{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import allel\n",
    "import json"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "pd.set_option('display.max_columns', None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "refgene = pd.read_csv('data/refseq-20200109.txt', sep='\\t')\n",
    "refgene['length'] = refgene['txEnd'] - refgene['txStart']\n",
    "refgene = refgene.sort_values('length', ascending=False)\n",
    "refgene = refgene.drop_duplicates('name2', keep='first')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>#bin</th>\n",
       "      <th>name</th>\n",
       "      <th>chrom</th>\n",
       "      <th>strand</th>\n",
       "      <th>txStart</th>\n",
       "      <th>txEnd</th>\n",
       "      <th>cdsStart</th>\n",
       "      <th>cdsEnd</th>\n",
       "      <th>exonCount</th>\n",
       "      <th>exonStarts</th>\n",
       "      <th>exonEnds</th>\n",
       "      <th>score</th>\n",
       "      <th>name2</th>\n",
       "      <th>cdsStartStat</th>\n",
       "      <th>cdsEndStat</th>\n",
       "      <th>exonFrames</th>\n",
       "      <th>length</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>21052</th>\n",
       "      <td>26</td>\n",
       "      <td>NM_014141.5</td>\n",
       "      <td>chr7</td>\n",
       "      <td>+</td>\n",
       "      <td>145813452</td>\n",
       "      <td>148118090</td>\n",
       "      <td>145813968</td>\n",
       "      <td>148112708</td>\n",
       "      <td>24</td>\n",
       "      <td>145813452,146471362,146536802,146740998,146805...</td>\n",
       "      <td>145814065,146471473,146536996,146741146,146805...</td>\n",
       "      <td>0</td>\n",
       "      <td>CNTNAP2</td>\n",
       "      <td>cmpl</td>\n",
       "      <td>cmpl</td>\n",
       "      <td>0,1,1,0,1,1,0,0,1,1,2,1,1,1,2,1,1,1,1,1,0,1,1,1,</td>\n",
       "      <td>2304638</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25593</th>\n",
       "      <td>1</td>\n",
       "      <td>NM_002839.3</td>\n",
       "      <td>chr9</td>\n",
       "      <td>-</td>\n",
       "      <td>8314245</td>\n",
       "      <td>10612723</td>\n",
       "      <td>8317873</td>\n",
       "      <td>8733843</td>\n",
       "      <td>46</td>\n",
       "      <td>8314245,8319830,8331581,8338921,8340342,834108...</td>\n",
       "      <td>8317942,8319966,8331736,8339047,8340469,834126...</td>\n",
       "      <td>0</td>\n",
       "      <td>PTPRD</td>\n",
       "      <td>cmpl</td>\n",
       "      <td>cmpl</td>\n",
       "      <td>0,2,0,0,2,0,2,0,0,1,0,1,2,0,0,2,0,1,1,0,0,1,1,...</td>\n",
       "      <td>2298478</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27890</th>\n",
       "      <td>12</td>\n",
       "      <td>NM_000109.3</td>\n",
       "      <td>chrX</td>\n",
       "      <td>-</td>\n",
       "      <td>31137344</td>\n",
       "      <td>33357726</td>\n",
       "      <td>31140035</td>\n",
       "      <td>33357382</td>\n",
       "      <td>79</td>\n",
       "      <td>31137344,31144758,31152218,31164407,31165391,3...</td>\n",
       "      <td>31140047,31144790,31152311,31164531,31165635,3...</td>\n",
       "      <td>0</td>\n",
       "      <td>DMD</td>\n",
       "      <td>cmpl</td>\n",
       "      <td>cmpl</td>\n",
       "      <td>0,1,1,0,2,2,2,2,2,0,2,0,1,2,1,1,2,1,0,0,1,0,2,...</td>\n",
       "      <td>2220382</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33217</th>\n",
       "      <td>2</td>\n",
       "      <td>NM_001142699.1</td>\n",
       "      <td>chr11</td>\n",
       "      <td>-</td>\n",
       "      <td>83166054</td>\n",
       "      <td>85338314</td>\n",
       "      <td>83170860</td>\n",
       "      <td>85309740</td>\n",
       "      <td>28</td>\n",
       "      <td>83166054,83173044,83177750,83180243,83182668,8...</td>\n",
       "      <td>83170967,83173136,83177860,83180416,83182770,8...</td>\n",
       "      <td>0</td>\n",
       "      <td>DLG2</td>\n",
       "      <td>cmpl</td>\n",
       "      <td>cmpl</td>\n",
       "      <td>1,2,0,1,1,1,0,2,2,1,0,2,2,2,1,0,1,2,0,0,0,0,0,...</td>\n",
       "      <td>2172260</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23572</th>\n",
       "      <td>9</td>\n",
       "      <td>NM_033225.5</td>\n",
       "      <td>chr8</td>\n",
       "      <td>-</td>\n",
       "      <td>2792874</td>\n",
       "      <td>4852328</td>\n",
       "      <td>2796106</td>\n",
       "      <td>4851938</td>\n",
       "      <td>70</td>\n",
       "      <td>2792874,2799993,2806820,2807752,2808635,281174...</td>\n",
       "      <td>2796266,2800126,2806908,2807865,2808797,281179...</td>\n",
       "      <td>0</td>\n",
       "      <td>CSMD1</td>\n",
       "      <td>cmpl</td>\n",
       "      <td>cmpl</td>\n",
       "      <td>2,1,0,1,1,1,1,0,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,...</td>\n",
       "      <td>2059454</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       #bin            name  chrom strand    txStart      txEnd   cdsStart  \\\n",
       "21052    26     NM_014141.5   chr7      +  145813452  148118090  145813968   \n",
       "25593     1     NM_002839.3   chr9      -    8314245   10612723    8317873   \n",
       "27890    12     NM_000109.3   chrX      -   31137344   33357726   31140035   \n",
       "33217     2  NM_001142699.1  chr11      -   83166054   85338314   83170860   \n",
       "23572     9     NM_033225.5   chr8      -    2792874    4852328    2796106   \n",
       "\n",
       "          cdsEnd  exonCount  \\\n",
       "21052  148112708         24   \n",
       "25593    8733843         46   \n",
       "27890   33357382         79   \n",
       "33217   85309740         28   \n",
       "23572    4851938         70   \n",
       "\n",
       "                                              exonStarts  \\\n",
       "21052  145813452,146471362,146536802,146740998,146805...   \n",
       "25593  8314245,8319830,8331581,8338921,8340342,834108...   \n",
       "27890  31137344,31144758,31152218,31164407,31165391,3...   \n",
       "33217  83166054,83173044,83177750,83180243,83182668,8...   \n",
       "23572  2792874,2799993,2806820,2807752,2808635,281174...   \n",
       "\n",
       "                                                exonEnds  score    name2  \\\n",
       "21052  145814065,146471473,146536996,146741146,146805...      0  CNTNAP2   \n",
       "25593  8317942,8319966,8331736,8339047,8340469,834126...      0    PTPRD   \n",
       "27890  31140047,31144790,31152311,31164531,31165635,3...      0      DMD   \n",
       "33217  83170967,83173136,83177860,83180416,83182770,8...      0     DLG2   \n",
       "23572  2796266,2800126,2806908,2807865,2808797,281179...      0    CSMD1   \n",
       "\n",
       "      cdsStartStat cdsEndStat  \\\n",
       "21052         cmpl       cmpl   \n",
       "25593         cmpl       cmpl   \n",
       "27890         cmpl       cmpl   \n",
       "33217         cmpl       cmpl   \n",
       "23572         cmpl       cmpl   \n",
       "\n",
       "                                              exonFrames   length  \n",
       "21052   0,1,1,0,1,1,0,0,1,1,2,1,1,1,2,1,1,1,1,1,0,1,1,1,  2304638  \n",
       "25593  0,2,0,0,2,0,2,0,0,1,0,1,2,0,0,2,0,1,1,0,0,1,1,...  2298478  \n",
       "27890  0,1,1,0,2,2,2,2,2,0,2,0,1,2,1,1,2,1,0,0,1,0,2,...  2220382  \n",
       "33217  1,2,0,1,1,1,0,2,2,1,0,2,2,2,1,0,1,2,0,0,0,0,0,...  2172260  \n",
       "23572  2,1,0,1,1,1,1,0,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,...  2059454  "
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "refgene.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_info = pd.read_csv('data/Homo_sapiens.gene_info.gz', sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>#tax_id</th>\n",
       "      <th>GeneID</th>\n",
       "      <th>Symbol</th>\n",
       "      <th>LocusTag</th>\n",
       "      <th>Synonyms</th>\n",
       "      <th>dbXrefs</th>\n",
       "      <th>chromosome</th>\n",
       "      <th>map_location</th>\n",
       "      <th>description</th>\n",
       "      <th>type_of_gene</th>\n",
       "      <th>Symbol_from_nomenclature_authority</th>\n",
       "      <th>Full_name_from_nomenclature_authority</th>\n",
       "      <th>Nomenclature_status</th>\n",
       "      <th>Other_designations</th>\n",
       "      <th>Modification_date</th>\n",
       "      <th>Feature_type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>9606</td>\n",
       "      <td>1</td>\n",
       "      <td>A1BG</td>\n",
       "      <td>-</td>\n",
       "      <td>A1B|ABG|GAB|HYST2477</td>\n",
       "      <td>MIM:138670|HGNC:HGNC:5|Ensembl:ENSG00000121410</td>\n",
       "      <td>19</td>\n",
       "      <td>19q13.43</td>\n",
       "      <td>alpha-1-B glycoprotein</td>\n",
       "      <td>protein-coding</td>\n",
       "      <td>A1BG</td>\n",
       "      <td>alpha-1-B glycoprotein</td>\n",
       "      <td>O</td>\n",
       "      <td>alpha-1B-glycoprotein|HEL-S-163pA|epididymis s...</td>\n",
       "      <td>20191220</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>9606</td>\n",
       "      <td>2</td>\n",
       "      <td>A2M</td>\n",
       "      <td>-</td>\n",
       "      <td>A2MD|CPAMD5|FWP007|S863-7</td>\n",
       "      <td>MIM:103950|HGNC:HGNC:7|Ensembl:ENSG00000175899</td>\n",
       "      <td>12</td>\n",
       "      <td>12p13.31</td>\n",
       "      <td>alpha-2-macroglobulin</td>\n",
       "      <td>protein-coding</td>\n",
       "      <td>A2M</td>\n",
       "      <td>alpha-2-macroglobulin</td>\n",
       "      <td>O</td>\n",
       "      <td>alpha-2-macroglobulin|C3 and PZP-like alpha-2-...</td>\n",
       "      <td>20200113</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>9606</td>\n",
       "      <td>3</td>\n",
       "      <td>A2MP1</td>\n",
       "      <td>-</td>\n",
       "      <td>A2MP</td>\n",
       "      <td>HGNC:HGNC:8|Ensembl:ENSG00000256069</td>\n",
       "      <td>12</td>\n",
       "      <td>12p13.31</td>\n",
       "      <td>alpha-2-macroglobulin pseudogene 1</td>\n",
       "      <td>pseudo</td>\n",
       "      <td>A2MP1</td>\n",
       "      <td>alpha-2-macroglobulin pseudogene 1</td>\n",
       "      <td>O</td>\n",
       "      <td>pregnancy-zone protein pseudogene</td>\n",
       "      <td>20191221</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>9606</td>\n",
       "      <td>9</td>\n",
       "      <td>NAT1</td>\n",
       "      <td>-</td>\n",
       "      <td>AAC1|MNAT|NAT-1|NATI</td>\n",
       "      <td>MIM:108345|HGNC:HGNC:7645|Ensembl:ENSG00000171428</td>\n",
       "      <td>8</td>\n",
       "      <td>8p22</td>\n",
       "      <td>N-acetyltransferase 1</td>\n",
       "      <td>protein-coding</td>\n",
       "      <td>NAT1</td>\n",
       "      <td>N-acetyltransferase 1</td>\n",
       "      <td>O</td>\n",
       "      <td>arylamine N-acetyltransferase 1|N-acetyltransf...</td>\n",
       "      <td>20200113</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>9606</td>\n",
       "      <td>10</td>\n",
       "      <td>NAT2</td>\n",
       "      <td>-</td>\n",
       "      <td>AAC2|NAT-2|PNAT</td>\n",
       "      <td>MIM:612182|HGNC:HGNC:7646|Ensembl:ENSG00000156006</td>\n",
       "      <td>8</td>\n",
       "      <td>8p22</td>\n",
       "      <td>N-acetyltransferase 2</td>\n",
       "      <td>protein-coding</td>\n",
       "      <td>NAT2</td>\n",
       "      <td>N-acetyltransferase 2</td>\n",
       "      <td>O</td>\n",
       "      <td>arylamine N-acetyltransferase 2|N-acetyltransf...</td>\n",
       "      <td>20200121</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   #tax_id  GeneID Symbol LocusTag                   Synonyms  \\\n",
       "0     9606       1   A1BG        -       A1B|ABG|GAB|HYST2477   \n",
       "1     9606       2    A2M        -  A2MD|CPAMD5|FWP007|S863-7   \n",
       "2     9606       3  A2MP1        -                       A2MP   \n",
       "3     9606       9   NAT1        -       AAC1|MNAT|NAT-1|NATI   \n",
       "4     9606      10   NAT2        -            AAC2|NAT-2|PNAT   \n",
       "\n",
       "                                             dbXrefs chromosome map_location  \\\n",
       "0     MIM:138670|HGNC:HGNC:5|Ensembl:ENSG00000121410         19     19q13.43   \n",
       "1     MIM:103950|HGNC:HGNC:7|Ensembl:ENSG00000175899         12     12p13.31   \n",
       "2                HGNC:HGNC:8|Ensembl:ENSG00000256069         12     12p13.31   \n",
       "3  MIM:108345|HGNC:HGNC:7645|Ensembl:ENSG00000171428          8         8p22   \n",
       "4  MIM:612182|HGNC:HGNC:7646|Ensembl:ENSG00000156006          8         8p22   \n",
       "\n",
       "                          description    type_of_gene  \\\n",
       "0              alpha-1-B glycoprotein  protein-coding   \n",
       "1               alpha-2-macroglobulin  protein-coding   \n",
       "2  alpha-2-macroglobulin pseudogene 1          pseudo   \n",
       "3               N-acetyltransferase 1  protein-coding   \n",
       "4               N-acetyltransferase 2  protein-coding   \n",
       "\n",
       "  Symbol_from_nomenclature_authority Full_name_from_nomenclature_authority  \\\n",
       "0                               A1BG                alpha-1-B glycoprotein   \n",
       "1                                A2M                 alpha-2-macroglobulin   \n",
       "2                              A2MP1    alpha-2-macroglobulin pseudogene 1   \n",
       "3                               NAT1                 N-acetyltransferase 1   \n",
       "4                               NAT2                 N-acetyltransferase 2   \n",
       "\n",
       "  Nomenclature_status                                 Other_designations  \\\n",
       "0                   O  alpha-1B-glycoprotein|HEL-S-163pA|epididymis s...   \n",
       "1                   O  alpha-2-macroglobulin|C3 and PZP-like alpha-2-...   \n",
       "2                   O                  pregnancy-zone protein pseudogene   \n",
       "3                   O  arylamine N-acetyltransferase 1|N-acetyltransf...   \n",
       "4                   O  arylamine N-acetyltransferase 2|N-acetyltransf...   \n",
       "\n",
       "   Modification_date Feature_type  \n",
       "0           20191220            -  \n",
       "1           20200113            -  \n",
       "2           20191221            -  \n",
       "3           20200113            -  \n",
       "4           20200121            -  "
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_info.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "refgene_info = refgene.merge(gene_info, left_on='name2', right_on='Symbol')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>#bin</th>\n",
       "      <th>name</th>\n",
       "      <th>chrom</th>\n",
       "      <th>strand</th>\n",
       "      <th>txStart</th>\n",
       "      <th>txEnd</th>\n",
       "      <th>cdsStart</th>\n",
       "      <th>cdsEnd</th>\n",
       "      <th>exonCount</th>\n",
       "      <th>exonStarts</th>\n",
       "      <th>exonEnds</th>\n",
       "      <th>score</th>\n",
       "      <th>name2</th>\n",
       "      <th>cdsStartStat</th>\n",
       "      <th>cdsEndStat</th>\n",
       "      <th>exonFrames</th>\n",
       "      <th>length</th>\n",
       "      <th>#tax_id</th>\n",
       "      <th>GeneID</th>\n",
       "      <th>Symbol</th>\n",
       "      <th>LocusTag</th>\n",
       "      <th>Synonyms</th>\n",
       "      <th>dbXrefs</th>\n",
       "      <th>chromosome</th>\n",
       "      <th>map_location</th>\n",
       "      <th>description</th>\n",
       "      <th>type_of_gene</th>\n",
       "      <th>Symbol_from_nomenclature_authority</th>\n",
       "      <th>Full_name_from_nomenclature_authority</th>\n",
       "      <th>Nomenclature_status</th>\n",
       "      <th>Other_designations</th>\n",
       "      <th>Modification_date</th>\n",
       "      <th>Feature_type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>26</td>\n",
       "      <td>NM_014141.5</td>\n",
       "      <td>chr7</td>\n",
       "      <td>+</td>\n",
       "      <td>145813452</td>\n",
       "      <td>148118090</td>\n",
       "      <td>145813968</td>\n",
       "      <td>148112708</td>\n",
       "      <td>24</td>\n",
       "      <td>145813452,146471362,146536802,146740998,146805...</td>\n",
       "      <td>145814065,146471473,146536996,146741146,146805...</td>\n",
       "      <td>0</td>\n",
       "      <td>CNTNAP2</td>\n",
       "      <td>cmpl</td>\n",
       "      <td>cmpl</td>\n",
       "      <td>0,1,1,0,1,1,0,0,1,1,2,1,1,1,2,1,1,1,1,1,0,1,1,1,</td>\n",
       "      <td>2304638</td>\n",
       "      <td>9606</td>\n",
       "      <td>26047</td>\n",
       "      <td>CNTNAP2</td>\n",
       "      <td>-</td>\n",
       "      <td>AUTS15|CASPR2|CDFE|NRXN4|PTHSL1</td>\n",
       "      <td>MIM:604569|HGNC:HGNC:13830|Ensembl:ENSG0000017...</td>\n",
       "      <td>7</td>\n",
       "      <td>7q35-q36.1</td>\n",
       "      <td>contactin associated protein 2</td>\n",
       "      <td>protein-coding</td>\n",
       "      <td>CNTNAP2</td>\n",
       "      <td>contactin associated protein 2</td>\n",
       "      <td>O</td>\n",
       "      <td>contactin-associated protein-like 2|cell recog...</td>\n",
       "      <td>20200105</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>NM_002839.3</td>\n",
       "      <td>chr9</td>\n",
       "      <td>-</td>\n",
       "      <td>8314245</td>\n",
       "      <td>10612723</td>\n",
       "      <td>8317873</td>\n",
       "      <td>8733843</td>\n",
       "      <td>46</td>\n",
       "      <td>8314245,8319830,8331581,8338921,8340342,834108...</td>\n",
       "      <td>8317942,8319966,8331736,8339047,8340469,834126...</td>\n",
       "      <td>0</td>\n",
       "      <td>PTPRD</td>\n",
       "      <td>cmpl</td>\n",
       "      <td>cmpl</td>\n",
       "      <td>0,2,0,0,2,0,2,0,0,1,0,1,2,0,0,2,0,1,1,0,0,1,1,...</td>\n",
       "      <td>2298478</td>\n",
       "      <td>9606</td>\n",
       "      <td>5789</td>\n",
       "      <td>PTPRD</td>\n",
       "      <td>-</td>\n",
       "      <td>HPTP|HPTPD|HPTPDELTA|PTPD|R-PTP-delta|RPTPDELTA</td>\n",
       "      <td>MIM:601598|HGNC:HGNC:9668|Ensembl:ENSG00000153707</td>\n",
       "      <td>9</td>\n",
       "      <td>9p24.1-p23</td>\n",
       "      <td>protein tyrosine phosphatase receptor type D</td>\n",
       "      <td>protein-coding</td>\n",
       "      <td>PTPRD</td>\n",
       "      <td>protein tyrosine phosphatase receptor type D</td>\n",
       "      <td>O</td>\n",
       "      <td>receptor-type tyrosine-protein phosphatase del...</td>\n",
       "      <td>20200128</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>12</td>\n",
       "      <td>NM_000109.3</td>\n",
       "      <td>chrX</td>\n",
       "      <td>-</td>\n",
       "      <td>31137344</td>\n",
       "      <td>33357726</td>\n",
       "      <td>31140035</td>\n",
       "      <td>33357382</td>\n",
       "      <td>79</td>\n",
       "      <td>31137344,31144758,31152218,31164407,31165391,3...</td>\n",
       "      <td>31140047,31144790,31152311,31164531,31165635,3...</td>\n",
       "      <td>0</td>\n",
       "      <td>DMD</td>\n",
       "      <td>cmpl</td>\n",
       "      <td>cmpl</td>\n",
       "      <td>0,1,1,0,2,2,2,2,2,0,2,0,1,2,1,1,2,1,0,0,1,0,2,...</td>\n",
       "      <td>2220382</td>\n",
       "      <td>9606</td>\n",
       "      <td>1756</td>\n",
       "      <td>DMD</td>\n",
       "      <td>-</td>\n",
       "      <td>BMD|CMD3B|DXS142|DXS164|DXS206|DXS230|DXS239|D...</td>\n",
       "      <td>MIM:300377|HGNC:HGNC:2928|Ensembl:ENSG00000198947</td>\n",
       "      <td>X</td>\n",
       "      <td>Xp21.2-p21.1</td>\n",
       "      <td>dystrophin</td>\n",
       "      <td>protein-coding</td>\n",
       "      <td>DMD</td>\n",
       "      <td>dystrophin</td>\n",
       "      <td>O</td>\n",
       "      <td>dystrophin</td>\n",
       "      <td>20200105</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>2</td>\n",
       "      <td>NM_001142699.1</td>\n",
       "      <td>chr11</td>\n",
       "      <td>-</td>\n",
       "      <td>83166054</td>\n",
       "      <td>85338314</td>\n",
       "      <td>83170860</td>\n",
       "      <td>85309740</td>\n",
       "      <td>28</td>\n",
       "      <td>83166054,83173044,83177750,83180243,83182668,8...</td>\n",
       "      <td>83170967,83173136,83177860,83180416,83182770,8...</td>\n",
       "      <td>0</td>\n",
       "      <td>DLG2</td>\n",
       "      <td>cmpl</td>\n",
       "      <td>cmpl</td>\n",
       "      <td>1,2,0,1,1,1,0,2,2,1,0,2,2,2,1,0,1,2,0,0,0,0,0,...</td>\n",
       "      <td>2172260</td>\n",
       "      <td>9606</td>\n",
       "      <td>1740</td>\n",
       "      <td>DLG2</td>\n",
       "      <td>-</td>\n",
       "      <td>PPP1R58|PSD-93|PSD93|chapsyn-110</td>\n",
       "      <td>MIM:603583|HGNC:HGNC:2901|Ensembl:ENSG00000150672</td>\n",
       "      <td>11</td>\n",
       "      <td>11q14.1</td>\n",
       "      <td>discs large MAGUK scaffold protein 2</td>\n",
       "      <td>protein-coding</td>\n",
       "      <td>DLG2</td>\n",
       "      <td>discs large MAGUK scaffold protein 2</td>\n",
       "      <td>O</td>\n",
       "      <td>disks large homolog 2|DLG2 protein isoform|cha...</td>\n",
       "      <td>20200124</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>9</td>\n",
       "      <td>NM_033225.5</td>\n",
       "      <td>chr8</td>\n",
       "      <td>-</td>\n",
       "      <td>2792874</td>\n",
       "      <td>4852328</td>\n",
       "      <td>2796106</td>\n",
       "      <td>4851938</td>\n",
       "      <td>70</td>\n",
       "      <td>2792874,2799993,2806820,2807752,2808635,281174...</td>\n",
       "      <td>2796266,2800126,2806908,2807865,2808797,281179...</td>\n",
       "      <td>0</td>\n",
       "      <td>CSMD1</td>\n",
       "      <td>cmpl</td>\n",
       "      <td>cmpl</td>\n",
       "      <td>2,1,0,1,1,1,1,0,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,...</td>\n",
       "      <td>2059454</td>\n",
       "      <td>9606</td>\n",
       "      <td>64478</td>\n",
       "      <td>CSMD1</td>\n",
       "      <td>-</td>\n",
       "      <td>PPP1R24</td>\n",
       "      <td>MIM:608397|HGNC:HGNC:14026|Ensembl:ENSG0000018...</td>\n",
       "      <td>8</td>\n",
       "      <td>8p23.2</td>\n",
       "      <td>CUB and Sushi multiple domains 1</td>\n",
       "      <td>protein-coding</td>\n",
       "      <td>CSMD1</td>\n",
       "      <td>CUB and Sushi multiple domains 1</td>\n",
       "      <td>O</td>\n",
       "      <td>CUB and sushi domain-containing protein 1|prot...</td>\n",
       "      <td>20200127</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   #bin            name  chrom strand    txStart      txEnd   cdsStart  \\\n",
       "0    26     NM_014141.5   chr7      +  145813452  148118090  145813968   \n",
       "1     1     NM_002839.3   chr9      -    8314245   10612723    8317873   \n",
       "2    12     NM_000109.3   chrX      -   31137344   33357726   31140035   \n",
       "3     2  NM_001142699.1  chr11      -   83166054   85338314   83170860   \n",
       "4     9     NM_033225.5   chr8      -    2792874    4852328    2796106   \n",
       "\n",
       "      cdsEnd  exonCount                                         exonStarts  \\\n",
       "0  148112708         24  145813452,146471362,146536802,146740998,146805...   \n",
       "1    8733843         46  8314245,8319830,8331581,8338921,8340342,834108...   \n",
       "2   33357382         79  31137344,31144758,31152218,31164407,31165391,3...   \n",
       "3   85309740         28  83166054,83173044,83177750,83180243,83182668,8...   \n",
       "4    4851938         70  2792874,2799993,2806820,2807752,2808635,281174...   \n",
       "\n",
       "                                            exonEnds  score    name2  \\\n",
       "0  145814065,146471473,146536996,146741146,146805...      0  CNTNAP2   \n",
       "1  8317942,8319966,8331736,8339047,8340469,834126...      0    PTPRD   \n",
       "2  31140047,31144790,31152311,31164531,31165635,3...      0      DMD   \n",
       "3  83170967,83173136,83177860,83180416,83182770,8...      0     DLG2   \n",
       "4  2796266,2800126,2806908,2807865,2808797,281179...      0    CSMD1   \n",
       "\n",
       "  cdsStartStat cdsEndStat                                         exonFrames  \\\n",
       "0         cmpl       cmpl   0,1,1,0,1,1,0,0,1,1,2,1,1,1,2,1,1,1,1,1,0,1,1,1,   \n",
       "1         cmpl       cmpl  0,2,0,0,2,0,2,0,0,1,0,1,2,0,0,2,0,1,1,0,0,1,1,...   \n",
       "2         cmpl       cmpl  0,1,1,0,2,2,2,2,2,0,2,0,1,2,1,1,2,1,0,0,1,0,2,...   \n",
       "3         cmpl       cmpl  1,2,0,1,1,1,0,2,2,1,0,2,2,2,1,0,1,2,0,0,0,0,0,...   \n",
       "4         cmpl       cmpl  2,1,0,1,1,1,1,0,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,...   \n",
       "\n",
       "    length  #tax_id  GeneID   Symbol LocusTag  \\\n",
       "0  2304638     9606   26047  CNTNAP2        -   \n",
       "1  2298478     9606    5789    PTPRD        -   \n",
       "2  2220382     9606    1756      DMD        -   \n",
       "3  2172260     9606    1740     DLG2        -   \n",
       "4  2059454     9606   64478    CSMD1        -   \n",
       "\n",
       "                                            Synonyms  \\\n",
       "0                    AUTS15|CASPR2|CDFE|NRXN4|PTHSL1   \n",
       "1    HPTP|HPTPD|HPTPDELTA|PTPD|R-PTP-delta|RPTPDELTA   \n",
       "2  BMD|CMD3B|DXS142|DXS164|DXS206|DXS230|DXS239|D...   \n",
       "3                   PPP1R58|PSD-93|PSD93|chapsyn-110   \n",
       "4                                            PPP1R24   \n",
       "\n",
       "                                             dbXrefs chromosome  map_location  \\\n",
       "0  MIM:604569|HGNC:HGNC:13830|Ensembl:ENSG0000017...          7    7q35-q36.1   \n",
       "1  MIM:601598|HGNC:HGNC:9668|Ensembl:ENSG00000153707          9    9p24.1-p23   \n",
       "2  MIM:300377|HGNC:HGNC:2928|Ensembl:ENSG00000198947          X  Xp21.2-p21.1   \n",
       "3  MIM:603583|HGNC:HGNC:2901|Ensembl:ENSG00000150672         11       11q14.1   \n",
       "4  MIM:608397|HGNC:HGNC:14026|Ensembl:ENSG0000018...          8        8p23.2   \n",
       "\n",
       "                                    description    type_of_gene  \\\n",
       "0                contactin associated protein 2  protein-coding   \n",
       "1  protein tyrosine phosphatase receptor type D  protein-coding   \n",
       "2                                    dystrophin  protein-coding   \n",
       "3          discs large MAGUK scaffold protein 2  protein-coding   \n",
       "4              CUB and Sushi multiple domains 1  protein-coding   \n",
       "\n",
       "  Symbol_from_nomenclature_authority  \\\n",
       "0                            CNTNAP2   \n",
       "1                              PTPRD   \n",
       "2                                DMD   \n",
       "3                               DLG2   \n",
       "4                              CSMD1   \n",
       "\n",
       "          Full_name_from_nomenclature_authority Nomenclature_status  \\\n",
       "0                contactin associated protein 2                   O   \n",
       "1  protein tyrosine phosphatase receptor type D                   O   \n",
       "2                                    dystrophin                   O   \n",
       "3          discs large MAGUK scaffold protein 2                   O   \n",
       "4              CUB and Sushi multiple domains 1                   O   \n",
       "\n",
       "                                  Other_designations  Modification_date  \\\n",
       "0  contactin-associated protein-like 2|cell recog...           20200105   \n",
       "1  receptor-type tyrosine-protein phosphatase del...           20200128   \n",
       "2                                         dystrophin           20200105   \n",
       "3  disks large homolog 2|DLG2 protein isoform|cha...           20200124   \n",
       "4  CUB and sushi domain-containing protein 1|prot...           20200127   \n",
       "\n",
       "  Feature_type  \n",
       "0            -  \n",
       "1            -  \n",
       "2            -  \n",
       "3            -  \n",
       "4            -  "
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "refgene_info.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "gene = refgene_info.loc[\n",
    "    refgene_info['type_of_gene'] == 'protein-coding',\n",
    "    ['chrom', 'txStart', 'txEnd', 'GeneID', 'name2', 'name', 'strand']\n",
    "].sort_values(['chrom', 'txStart', 'txEnd'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>txStart</th>\n",
       "      <th>txEnd</th>\n",
       "      <th>GeneID</th>\n",
       "      <th>name2</th>\n",
       "      <th>name</th>\n",
       "      <th>strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>22981</th>\n",
       "      <td>chr1</td>\n",
       "      <td>69090</td>\n",
       "      <td>70008</td>\n",
       "      <td>79501</td>\n",
       "      <td>OR4F5</td>\n",
       "      <td>NM_001005484.1</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22876</th>\n",
       "      <td>chr1</td>\n",
       "      <td>367658</td>\n",
       "      <td>368597</td>\n",
       "      <td>729759</td>\n",
       "      <td>OR4F29</td>\n",
       "      <td>NM_001005221.2</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22845</th>\n",
       "      <td>chr1</td>\n",
       "      <td>621095</td>\n",
       "      <td>622034</td>\n",
       "      <td>81399</td>\n",
       "      <td>OR4F16</td>\n",
       "      <td>NM_001005277.1</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12034</th>\n",
       "      <td>chr1</td>\n",
       "      <td>861120</td>\n",
       "      <td>879961</td>\n",
       "      <td>148398</td>\n",
       "      <td>SAMD11</td>\n",
       "      <td>NM_152486.2</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13343</th>\n",
       "      <td>chr1</td>\n",
       "      <td>879582</td>\n",
       "      <td>894679</td>\n",
       "      <td>26155</td>\n",
       "      <td>NOC2L</td>\n",
       "      <td>NM_015658.3</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      chrom  txStart   txEnd  GeneID   name2            name strand\n",
       "22981  chr1    69090   70008   79501   OR4F5  NM_001005484.1      +\n",
       "22876  chr1   367658  368597  729759  OR4F29  NM_001005221.2      +\n",
       "22845  chr1   621095  622034   81399  OR4F16  NM_001005277.1      -\n",
       "12034  chr1   861120  879961  148398  SAMD11     NM_152486.2      +\n",
       "13343  chr1   879582  894679   26155   NOC2L     NM_015658.3      -"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "gene.rename(columns={\n",
    "    'chrom': '#chrom', 'txStart': 'start', 'txEnd': 'end',\n",
    "    'GeneID': 'gene_id', 'name2': 'symbol', 'name': 'transcript'\n",
    "}).to_csv('data/gene-20200109.sorted.bed', index=False, sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "!bgzip -cf data/gene-20200109.sorted.bed > ../data/gene-20200109.sorted.bed.gz"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "!tabix -fp bed ../data/gene-20200109.sorted.bed.gz"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open('data/omim-gene.list') as f:\n",
    "    omim_gene = f.read().split('\\n')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "omim_gene = gene[gene['name2'].isin(omim_gene)].copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>txStart</th>\n",
       "      <th>txEnd</th>\n",
       "      <th>GeneID</th>\n",
       "      <th>name2</th>\n",
       "      <th>name</th>\n",
       "      <th>strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>22498</th>\n",
       "      <td>chr1</td>\n",
       "      <td>948846</td>\n",
       "      <td>949920</td>\n",
       "      <td>9636</td>\n",
       "      <td>ISG15</td>\n",
       "      <td>NM_005101.3</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8184</th>\n",
       "      <td>chr1</td>\n",
       "      <td>955502</td>\n",
       "      <td>991499</td>\n",
       "      <td>375790</td>\n",
       "      <td>AGRN</td>\n",
       "      <td>NM_198576.3</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20527</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1146705</td>\n",
       "      <td>1149548</td>\n",
       "      <td>7293</td>\n",
       "      <td>TNFRSF4</td>\n",
       "      <td>NM_003327.3</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20581</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1167628</td>\n",
       "      <td>1170421</td>\n",
       "      <td>126792</td>\n",
       "      <td>B3GALT6</td>\n",
       "      <td>NM_080605.3</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13882</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1270657</td>\n",
       "      <td>1284492</td>\n",
       "      <td>1855</td>\n",
       "      <td>DVL1</td>\n",
       "      <td>NM_004421.2</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      chrom  txStart    txEnd  GeneID    name2         name strand\n",
       "22498  chr1   948846   949920    9636    ISG15  NM_005101.3      +\n",
       "8184   chr1   955502   991499  375790     AGRN  NM_198576.3      +\n",
       "20527  chr1  1146705  1149548    7293  TNFRSF4  NM_003327.3      -\n",
       "20581  chr1  1167628  1170421  126792  B3GALT6  NM_080605.3      +\n",
       "13882  chr1  1270657  1284492    1855     DVL1  NM_004421.2      -"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "omim_gene.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "omim_gene.rename(columns={\n",
    "    'chrom': '#chrom', 'txStart': 'start', 'txEnd': 'end',\n",
    "    'GeneID': 'gene_id', 'name2': 'symbol', 'name': 'transcript'\n",
    "}).to_csv('data/omim-gene-20200109.sorted.bed', index=False, sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "!bgzip -cf data/omim-gene-20200109.sorted.bed > ../data/omim-gene-20200109.sorted.bed.gz"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "!tabix -fp bed ../data/omim-gene-20200109.sorted.bed.gz"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "curation_gene = pd.read_csv('data/ClinGen_gene_curation_list_GRCh37.tsv', sep='\\t', dtype=str, skiprows=5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>#Gene Symbol</th>\n",
       "      <th>Gene ID</th>\n",
       "      <th>cytoBand</th>\n",
       "      <th>Genomic Location</th>\n",
       "      <th>Haploinsufficiency Score</th>\n",
       "      <th>Haploinsufficiency Description</th>\n",
       "      <th>Haploinsufficiency PMID1</th>\n",
       "      <th>Haploinsufficiency PMID2</th>\n",
       "      <th>Haploinsufficiency PMID3</th>\n",
       "      <th>Triplosensitivity Score</th>\n",
       "      <th>Triplosensitivity Description</th>\n",
       "      <th>Triplosensitivity PMID1</th>\n",
       "      <th>Triplosensitivity PMID2</th>\n",
       "      <th>Triplosensitivity PMID3</th>\n",
       "      <th>Date Last Evaluated</th>\n",
       "      <th>Loss phenotype OMIM ID</th>\n",
       "      <th>Triplosensitive phenotype OMIM ID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A4GALT</td>\n",
       "      <td>53947</td>\n",
       "      <td>22q13.2</td>\n",
       "      <td>chr22:43088121-43117307</td>\n",
       "      <td>30</td>\n",
       "      <td>Gene associated with autosomal recessive pheno...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>No evidence available</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2014-12-11</td>\n",
       "      <td>111400</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AAGAB</td>\n",
       "      <td>79719</td>\n",
       "      <td>15q23</td>\n",
       "      <td>chr15:67493013-67547536</td>\n",
       "      <td>3</td>\n",
       "      <td>Sufficient evidence for dosage pathogenicity</td>\n",
       "      <td>23064416</td>\n",
       "      <td>23000146</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>No evidence available</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2013-02-28</td>\n",
       "      <td>148600</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AARS</td>\n",
       "      <td>16</td>\n",
       "      <td>16q22.1</td>\n",
       "      <td>chr16:70286297-70323412</td>\n",
       "      <td>0</td>\n",
       "      <td>No evidence available</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>No evidence available</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2018-01-11</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AARS2</td>\n",
       "      <td>57505</td>\n",
       "      <td>6p21.1</td>\n",
       "      <td>chr6:44266463-44281063</td>\n",
       "      <td>30</td>\n",
       "      <td>Gene associated with autosomal recessive pheno...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Not yet evaluated</td>\n",
       "      <td>Not yet evaluated</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2016-08-22</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AASS</td>\n",
       "      <td>10157</td>\n",
       "      <td>7q31.32</td>\n",
       "      <td>chr7:121713598-121784344</td>\n",
       "      <td>30</td>\n",
       "      <td>Gene associated with autosomal recessive pheno...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Not yet evaluated</td>\n",
       "      <td>Not yet evaluated</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2016-08-22</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  #Gene Symbol Gene ID cytoBand          Genomic Location  \\\n",
       "0       A4GALT   53947  22q13.2   chr22:43088121-43117307   \n",
       "1        AAGAB   79719    15q23   chr15:67493013-67547536   \n",
       "2         AARS      16  16q22.1   chr16:70286297-70323412   \n",
       "3        AARS2   57505   6p21.1    chr6:44266463-44281063   \n",
       "4         AASS   10157  7q31.32  chr7:121713598-121784344   \n",
       "\n",
       "  Haploinsufficiency Score                     Haploinsufficiency Description  \\\n",
       "0                       30  Gene associated with autosomal recessive pheno...   \n",
       "1                        3       Sufficient evidence for dosage pathogenicity   \n",
       "2                        0                              No evidence available   \n",
       "3                       30  Gene associated with autosomal recessive pheno...   \n",
       "4                       30  Gene associated with autosomal recessive pheno...   \n",
       "\n",
       "  Haploinsufficiency PMID1 Haploinsufficiency PMID2 Haploinsufficiency PMID3  \\\n",
       "0                      NaN                      NaN                      NaN   \n",
       "1                 23064416                 23000146                      NaN   \n",
       "2                      NaN                      NaN                      NaN   \n",
       "3                      NaN                      NaN                      NaN   \n",
       "4                      NaN                      NaN                      NaN   \n",
       "\n",
       "  Triplosensitivity Score Triplosensitivity Description  \\\n",
       "0                       0         No evidence available   \n",
       "1                       0         No evidence available   \n",
       "2                       0         No evidence available   \n",
       "3       Not yet evaluated             Not yet evaluated   \n",
       "4       Not yet evaluated             Not yet evaluated   \n",
       "\n",
       "  Triplosensitivity PMID1 Triplosensitivity PMID2 Triplosensitivity PMID3  \\\n",
       "0                     NaN                     NaN                     NaN   \n",
       "1                     NaN                     NaN                     NaN   \n",
       "2                     NaN                     NaN                     NaN   \n",
       "3                     NaN                     NaN                     NaN   \n",
       "4                     NaN                     NaN                     NaN   \n",
       "\n",
       "  Date Last Evaluated Loss phenotype OMIM ID Triplosensitive phenotype OMIM ID  \n",
       "0          2014-12-11                 111400                               NaN  \n",
       "1          2013-02-28                 148600                               NaN  \n",
       "2          2018-01-11                    NaN                               NaN  \n",
       "3          2016-08-22                    NaN                               NaN  \n",
       "4          2016-08-22                    NaN                               NaN  "
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "curation_gene.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "hi_genes = set(\n",
    "    curation_gene.loc[\n",
    "        curation_gene['Haploinsufficiency Score'] == '3', '#Gene Symbol'\n",
    "    ]\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "hi_gene = gene[gene['name2'].isin(hi_genes)].copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>txStart</th>\n",
       "      <th>txEnd</th>\n",
       "      <th>GeneID</th>\n",
       "      <th>name2</th>\n",
       "      <th>name</th>\n",
       "      <th>strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>59</th>\n",
       "      <td>chr1</td>\n",
       "      <td>6845383</td>\n",
       "      <td>7829766</td>\n",
       "      <td>23261</td>\n",
       "      <td>CAMTA1</td>\n",
       "      <td>NM_015215.3</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8264</th>\n",
       "      <td>chr1</td>\n",
       "      <td>17345216</td>\n",
       "      <td>17380665</td>\n",
       "      <td>6390</td>\n",
       "      <td>SDHB</td>\n",
       "      <td>NM_003000.2</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3896</th>\n",
       "      <td>chr1</td>\n",
       "      <td>27022521</td>\n",
       "      <td>27108601</td>\n",
       "      <td>8289</td>\n",
       "      <td>ARID1A</td>\n",
       "      <td>NM_139135.2</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4768</th>\n",
       "      <td>chr1</td>\n",
       "      <td>27860755</td>\n",
       "      <td>27930005</td>\n",
       "      <td>27245</td>\n",
       "      <td>AHDC1</td>\n",
       "      <td>NM_001029882.3</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8525</th>\n",
       "      <td>chr1</td>\n",
       "      <td>43391045</td>\n",
       "      <td>43424847</td>\n",
       "      <td>6513</td>\n",
       "      <td>SLC2A1</td>\n",
       "      <td>NM_006516.2</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     chrom   txStart     txEnd  GeneID   name2            name strand\n",
       "59    chr1   6845383   7829766   23261  CAMTA1     NM_015215.3      +\n",
       "8264  chr1  17345216  17380665    6390    SDHB     NM_003000.2      -\n",
       "3896  chr1  27022521  27108601    8289  ARID1A     NM_139135.2      +\n",
       "4768  chr1  27860755  27930005   27245   AHDC1  NM_001029882.3      -\n",
       "8525  chr1  43391045  43424847    6513  SLC2A1     NM_006516.2      -"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "hi_gene.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "hi_gene.rename(columns={\n",
    "    'chrom': '#chrom', 'txStart': 'start', 'txEnd': 'end',\n",
    "    'GeneID': 'gene_id', 'name2': 'symbol', 'name': 'transcript'\n",
    "}).to_csv('data/hi-gene-20200109.sorted.bed', index=False, sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "!bgzip -cf data/hi-gene-20200109.sorted.bed > ../data/hi-gene-20200109.sorted.bed.gz"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [],
   "source": [
    "!tabix -fp bed ../data/hi-gene-20200109.sorted.bed.gz"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [],
   "source": [
    "hi_cds = refgene_info.loc[\n",
    "    refgene_info['name2'].isin(hi_genes), ['chrom', 'cdsStart', 'cdsEnd', 'GeneID', 'name2', 'name']\n",
    "].copy()\n",
    "hi_cds = hi_cds.sort_values(['chrom', 'cdsStart', 'cdsEnd'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>cdsStart</th>\n",
       "      <th>cdsEnd</th>\n",
       "      <th>GeneID</th>\n",
       "      <th>name2</th>\n",
       "      <th>name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>59</th>\n",
       "      <td>chr1</td>\n",
       "      <td>6845590</td>\n",
       "      <td>7826551</td>\n",
       "      <td>23261</td>\n",
       "      <td>CAMTA1</td>\n",
       "      <td>NM_015215.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8264</th>\n",
       "      <td>chr1</td>\n",
       "      <td>17345375</td>\n",
       "      <td>17380514</td>\n",
       "      <td>6390</td>\n",
       "      <td>SDHB</td>\n",
       "      <td>NM_003000.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3896</th>\n",
       "      <td>chr1</td>\n",
       "      <td>27022894</td>\n",
       "      <td>27107247</td>\n",
       "      <td>8289</td>\n",
       "      <td>ARID1A</td>\n",
       "      <td>NM_139135.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4768</th>\n",
       "      <td>chr1</td>\n",
       "      <td>27873814</td>\n",
       "      <td>27878626</td>\n",
       "      <td>27245</td>\n",
       "      <td>AHDC1</td>\n",
       "      <td>NM_001029882.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8525</th>\n",
       "      <td>chr1</td>\n",
       "      <td>43392711</td>\n",
       "      <td>43424322</td>\n",
       "      <td>6513</td>\n",
       "      <td>SLC2A1</td>\n",
       "      <td>NM_006516.2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     chrom  cdsStart    cdsEnd  GeneID   name2            name\n",
       "59    chr1   6845590   7826551   23261  CAMTA1     NM_015215.3\n",
       "8264  chr1  17345375  17380514    6390    SDHB     NM_003000.2\n",
       "3896  chr1  27022894  27107247    8289  ARID1A     NM_139135.2\n",
       "4768  chr1  27873814  27878626   27245   AHDC1  NM_001029882.3\n",
       "8525  chr1  43392711  43424322    6513  SLC2A1     NM_006516.2"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "hi_cds.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [],
   "source": [
    "hi_cds.rename(columns={\n",
    "    'chrom': '#chrom', 'cdsStart': 'start', 'cdsEnd': 'end',\n",
    "    'GeneID': 'gene_id', 'name2': 'symbol', 'name': 'transcript'\n",
    "}).to_csv('data/hi-cds-20200109.sorted.bed', index=False, sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [],
   "source": [
    "!bgzip -cf data/hi-cds-20200109.sorted.bed > ../data/hi-cds-20200109.sorted.bed.gz"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [],
   "source": [
    "!tabix -fp bed ../data/hi-cds-20200109.sorted.bed.gz"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [],
   "source": [
    "hi_exons = refgene_info.loc[\n",
    "    refgene_info['name2'].isin(hi_genes), ['chrom', 'exonStarts', 'exonEnds', 'GeneID', 'name2', 'name', 'strand']\n",
    "].copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>exonStarts</th>\n",
       "      <th>exonEnds</th>\n",
       "      <th>GeneID</th>\n",
       "      <th>name2</th>\n",
       "      <th>name</th>\n",
       "      <th>strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chrX</td>\n",
       "      <td>31137344,31144758,31152218,31164407,31165391,3...</td>\n",
       "      <td>31140047,31144790,31152311,31164531,31165635,3...</td>\n",
       "      <td>1756</td>\n",
       "      <td>DMD</td>\n",
       "      <td>NM_000109.3</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>chrX</td>\n",
       "      <td>28605680,28807436,29301054,29414374,29417271,2...</td>\n",
       "      <td>28606164,28807542,29301334,29414561,29417425,2...</td>\n",
       "      <td>11141</td>\n",
       "      <td>IL1RAPL1</td>\n",
       "      <td>NM_014271.3</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>chr7</td>\n",
       "      <td>69063904,69364271,69583117,69599521,69900737,7...</td>\n",
       "      <td>69064948,69364484,69583219,69599557,69900767,7...</td>\n",
       "      <td>26053</td>\n",
       "      <td>AUTS2</td>\n",
       "      <td>NM_015570.3</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46</th>\n",
       "      <td>chr2</td>\n",
       "      <td>50145642,50170850,50280408,50282092,50318460,5...</td>\n",
       "      <td>50149389,50170929,50280728,50282182,50318632,5...</td>\n",
       "      <td>9378</td>\n",
       "      <td>NRXN1</td>\n",
       "      <td>NM_001330094.1</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>55</th>\n",
       "      <td>chr12</td>\n",
       "      <td>23682437,23689386,23696144,23699249,23716191,2...</td>\n",
       "      <td>23687456,23689603,23696318,23699358,23716337,2...</td>\n",
       "      <td>6660</td>\n",
       "      <td>SOX5</td>\n",
       "      <td>NM_152989.4</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    chrom                                         exonStarts  \\\n",
       "2    chrX  31137344,31144758,31152218,31164407,31165391,3...   \n",
       "21   chrX  28605680,28807436,29301054,29414374,29417271,2...   \n",
       "33   chr7  69063904,69364271,69583117,69599521,69900737,7...   \n",
       "46   chr2  50145642,50170850,50280408,50282092,50318460,5...   \n",
       "55  chr12  23682437,23689386,23696144,23699249,23716191,2...   \n",
       "\n",
       "                                             exonEnds  GeneID     name2  \\\n",
       "2   31140047,31144790,31152311,31164531,31165635,3...    1756       DMD   \n",
       "21  28606164,28807542,29301334,29414561,29417425,2...   11141  IL1RAPL1   \n",
       "33  69064948,69364484,69583219,69599557,69900767,7...   26053     AUTS2   \n",
       "46  50149389,50170929,50280728,50282182,50318632,5...    9378     NRXN1   \n",
       "55  23687456,23689603,23696318,23699358,23716337,2...    6660      SOX5   \n",
       "\n",
       "              name strand  \n",
       "2      NM_000109.3      -  \n",
       "21     NM_014271.3      +  \n",
       "33     NM_015570.3      +  \n",
       "46  NM_001330094.1      -  \n",
       "55     NM_152989.4      -  "
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "hi_exons.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [],
   "source": [
    "hi_exons['exonStarts'] = hi_exons['exonStarts'].str.replace(r',$', '')\n",
    "hi_exons['exonEnds'] = hi_exons['exonEnds'].str.replace(r',$', '')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>exonStarts</th>\n",
       "      <th>exonEnds</th>\n",
       "      <th>GeneID</th>\n",
       "      <th>name2</th>\n",
       "      <th>name</th>\n",
       "      <th>strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chrX</td>\n",
       "      <td>31137344,31144758,31152218,31164407,31165391,3...</td>\n",
       "      <td>31140047,31144790,31152311,31164531,31165635,3...</td>\n",
       "      <td>1756</td>\n",
       "      <td>DMD</td>\n",
       "      <td>NM_000109.3</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>chrX</td>\n",
       "      <td>28605680,28807436,29301054,29414374,29417271,2...</td>\n",
       "      <td>28606164,28807542,29301334,29414561,29417425,2...</td>\n",
       "      <td>11141</td>\n",
       "      <td>IL1RAPL1</td>\n",
       "      <td>NM_014271.3</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>chr7</td>\n",
       "      <td>69063904,69364271,69583117,69599521,69900737,7...</td>\n",
       "      <td>69064948,69364484,69583219,69599557,69900767,7...</td>\n",
       "      <td>26053</td>\n",
       "      <td>AUTS2</td>\n",
       "      <td>NM_015570.3</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46</th>\n",
       "      <td>chr2</td>\n",
       "      <td>50145642,50170850,50280408,50282092,50318460,5...</td>\n",
       "      <td>50149389,50170929,50280728,50282182,50318632,5...</td>\n",
       "      <td>9378</td>\n",
       "      <td>NRXN1</td>\n",
       "      <td>NM_001330094.1</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>55</th>\n",
       "      <td>chr12</td>\n",
       "      <td>23682437,23689386,23696144,23699249,23716191,2...</td>\n",
       "      <td>23687456,23689603,23696318,23699358,23716337,2...</td>\n",
       "      <td>6660</td>\n",
       "      <td>SOX5</td>\n",
       "      <td>NM_152989.4</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    chrom                                         exonStarts  \\\n",
       "2    chrX  31137344,31144758,31152218,31164407,31165391,3...   \n",
       "21   chrX  28605680,28807436,29301054,29414374,29417271,2...   \n",
       "33   chr7  69063904,69364271,69583117,69599521,69900737,7...   \n",
       "46   chr2  50145642,50170850,50280408,50282092,50318460,5...   \n",
       "55  chr12  23682437,23689386,23696144,23699249,23716191,2...   \n",
       "\n",
       "                                             exonEnds  GeneID     name2  \\\n",
       "2   31140047,31144790,31152311,31164531,31165635,3...    1756       DMD   \n",
       "21  28606164,28807542,29301334,29414561,29417425,2...   11141  IL1RAPL1   \n",
       "33  69064948,69364484,69583219,69599557,69900767,7...   26053     AUTS2   \n",
       "46  50149389,50170929,50280728,50282182,50318632,5...    9378     NRXN1   \n",
       "55  23687456,23689603,23696318,23699358,23716337,2...    6660      SOX5   \n",
       "\n",
       "              name strand  \n",
       "2      NM_000109.3      -  \n",
       "21     NM_014271.3      +  \n",
       "33     NM_015570.3      +  \n",
       "46  NM_001330094.1      -  \n",
       "55     NM_152989.4      -  "
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "hi_exons.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [],
   "source": [
    "start = hi_exons['exonStarts'].str.split(',').apply(pd.Series).stack().reset_index()\n",
    "start = start.rename(columns={'level_0': 'row', 0: 'start'})[['row', 'start']]\n",
    "start['start'] = start['start'].astype(int)\n",
    "end = hi_exons['exonEnds'].str.split(',').apply(pd.Series).stack().reset_index()\n",
    "end = end.rename(columns={0: 'end'})['end'].astype(int)\n",
    "position = start.join(end)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>row</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>2</td>\n",
       "      <td>31137344</td>\n",
       "      <td>31140047</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2</td>\n",
       "      <td>31144758</td>\n",
       "      <td>31144790</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>31152218</td>\n",
       "      <td>31152311</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>2</td>\n",
       "      <td>31164407</td>\n",
       "      <td>31164531</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>2</td>\n",
       "      <td>31165391</td>\n",
       "      <td>31165635</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   row     start       end\n",
       "0    2  31137344  31140047\n",
       "1    2  31144758  31144790\n",
       "2    2  31152218  31152311\n",
       "3    2  31164407  31164531\n",
       "4    2  31165391  31165635"
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "position.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [],
   "source": [
    "exon = position.merge(\n",
    "    hi_exons[['chrom', 'GeneID', 'name2', 'name', 'strand']], how='left', left_on='row', right_index=True\n",
    ")\n",
    "exon = exon.sort_values(['chrom', 'start', 'end'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>row</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>chrom</th>\n",
       "      <th>GeneID</th>\n",
       "      <th>name2</th>\n",
       "      <th>name</th>\n",
       "      <th>strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>149</th>\n",
       "      <td>59</td>\n",
       "      <td>6845383</td>\n",
       "      <td>6845635</td>\n",
       "      <td>chr1</td>\n",
       "      <td>23261</td>\n",
       "      <td>CAMTA1</td>\n",
       "      <td>NM_015215.3</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>150</th>\n",
       "      <td>59</td>\n",
       "      <td>6880240</td>\n",
       "      <td>6880310</td>\n",
       "      <td>chr1</td>\n",
       "      <td>23261</td>\n",
       "      <td>CAMTA1</td>\n",
       "      <td>NM_015215.3</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>151</th>\n",
       "      <td>59</td>\n",
       "      <td>6885151</td>\n",
       "      <td>6885270</td>\n",
       "      <td>chr1</td>\n",
       "      <td>23261</td>\n",
       "      <td>CAMTA1</td>\n",
       "      <td>NM_015215.3</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>152</th>\n",
       "      <td>59</td>\n",
       "      <td>7151363</td>\n",
       "      <td>7151431</td>\n",
       "      <td>chr1</td>\n",
       "      <td>23261</td>\n",
       "      <td>CAMTA1</td>\n",
       "      <td>NM_015215.3</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>153</th>\n",
       "      <td>59</td>\n",
       "      <td>7309550</td>\n",
       "      <td>7309686</td>\n",
       "      <td>chr1</td>\n",
       "      <td>23261</td>\n",
       "      <td>CAMTA1</td>\n",
       "      <td>NM_015215.3</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     row    start      end chrom  GeneID   name2         name strand\n",
       "149   59  6845383  6845635  chr1   23261  CAMTA1  NM_015215.3      +\n",
       "150   59  6880240  6880310  chr1   23261  CAMTA1  NM_015215.3      +\n",
       "151   59  6885151  6885270  chr1   23261  CAMTA1  NM_015215.3      +\n",
       "152   59  7151363  7151431  chr1   23261  CAMTA1  NM_015215.3      +\n",
       "153   59  7309550  7309686  chr1   23261  CAMTA1  NM_015215.3      +"
      ]
     },
     "execution_count": 41,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "exon.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [],
   "source": [
    "exon['+'] = exon.groupby(['name2', 'name'])['start'].rank('first', ascending=True).astype(int)\n",
    "exon['-'] = exon.groupby(['name2', 'name'])['start'].rank('first', ascending=False).astype(int)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>row</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>chrom</th>\n",
       "      <th>GeneID</th>\n",
       "      <th>name2</th>\n",
       "      <th>name</th>\n",
       "      <th>strand</th>\n",
       "      <th>+</th>\n",
       "      <th>-</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>149</th>\n",
       "      <td>59</td>\n",
       "      <td>6845383</td>\n",
       "      <td>6845635</td>\n",
       "      <td>chr1</td>\n",
       "      <td>23261</td>\n",
       "      <td>CAMTA1</td>\n",
       "      <td>NM_015215.3</td>\n",
       "      <td>+</td>\n",
       "      <td>1</td>\n",
       "      <td>23</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>150</th>\n",
       "      <td>59</td>\n",
       "      <td>6880240</td>\n",
       "      <td>6880310</td>\n",
       "      <td>chr1</td>\n",
       "      <td>23261</td>\n",
       "      <td>CAMTA1</td>\n",
       "      <td>NM_015215.3</td>\n",
       "      <td>+</td>\n",
       "      <td>2</td>\n",
       "      <td>22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>151</th>\n",
       "      <td>59</td>\n",
       "      <td>6885151</td>\n",
       "      <td>6885270</td>\n",
       "      <td>chr1</td>\n",
       "      <td>23261</td>\n",
       "      <td>CAMTA1</td>\n",
       "      <td>NM_015215.3</td>\n",
       "      <td>+</td>\n",
       "      <td>3</td>\n",
       "      <td>21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>152</th>\n",
       "      <td>59</td>\n",
       "      <td>7151363</td>\n",
       "      <td>7151431</td>\n",
       "      <td>chr1</td>\n",
       "      <td>23261</td>\n",
       "      <td>CAMTA1</td>\n",
       "      <td>NM_015215.3</td>\n",
       "      <td>+</td>\n",
       "      <td>4</td>\n",
       "      <td>20</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>153</th>\n",
       "      <td>59</td>\n",
       "      <td>7309550</td>\n",
       "      <td>7309686</td>\n",
       "      <td>chr1</td>\n",
       "      <td>23261</td>\n",
       "      <td>CAMTA1</td>\n",
       "      <td>NM_015215.3</td>\n",
       "      <td>+</td>\n",
       "      <td>5</td>\n",
       "      <td>19</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     row    start      end chrom  GeneID   name2         name strand  +   -\n",
       "149   59  6845383  6845635  chr1   23261  CAMTA1  NM_015215.3      +  1  23\n",
       "150   59  6880240  6880310  chr1   23261  CAMTA1  NM_015215.3      +  2  22\n",
       "151   59  6885151  6885270  chr1   23261  CAMTA1  NM_015215.3      +  3  21\n",
       "152   59  7151363  7151431  chr1   23261  CAMTA1  NM_015215.3      +  4  20\n",
       "153   59  7309550  7309686  chr1   23261  CAMTA1  NM_015215.3      +  5  19"
      ]
     },
     "execution_count": 43,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "exon.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>row</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>chrom</th>\n",
       "      <th>GeneID</th>\n",
       "      <th>name2</th>\n",
       "      <th>name</th>\n",
       "      <th>strand</th>\n",
       "      <th>+</th>\n",
       "      <th>-</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1511</th>\n",
       "      <td>1511</td>\n",
       "      <td>154227753</td>\n",
       "      <td>154227875</td>\n",
       "      <td>chrX</td>\n",
       "      <td>2157</td>\n",
       "      <td>F8</td>\n",
       "      <td>NM_000132.3</td>\n",
       "      <td>-</td>\n",
       "      <td>25</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1512</th>\n",
       "      <td>1511</td>\n",
       "      <td>154250684</td>\n",
       "      <td>154250998</td>\n",
       "      <td>chrX</td>\n",
       "      <td>2157</td>\n",
       "      <td>F8</td>\n",
       "      <td>NM_000132.3</td>\n",
       "      <td>-</td>\n",
       "      <td>26</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4621</th>\n",
       "      <td>17777</td>\n",
       "      <td>154487518</td>\n",
       "      <td>154490514</td>\n",
       "      <td>chrX</td>\n",
       "      <td>116442</td>\n",
       "      <td>RAB39B</td>\n",
       "      <td>NM_171998.3</td>\n",
       "      <td>-</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4622</th>\n",
       "      <td>17777</td>\n",
       "      <td>154493358</td>\n",
       "      <td>154493874</td>\n",
       "      <td>chrX</td>\n",
       "      <td>116442</td>\n",
       "      <td>RAB39B</td>\n",
       "      <td>NM_171998.3</td>\n",
       "      <td>-</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4685</th>\n",
       "      <td>23010</td>\n",
       "      <td>2654895</td>\n",
       "      <td>2655782</td>\n",
       "      <td>chrY</td>\n",
       "      <td>6736</td>\n",
       "      <td>SRY</td>\n",
       "      <td>NM_003140.2</td>\n",
       "      <td>-</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        row      start        end chrom  GeneID   name2         name strand  \\\n",
       "1511   1511  154227753  154227875  chrX    2157      F8  NM_000132.3      -   \n",
       "1512   1511  154250684  154250998  chrX    2157      F8  NM_000132.3      -   \n",
       "4621  17777  154487518  154490514  chrX  116442  RAB39B  NM_171998.3      -   \n",
       "4622  17777  154493358  154493874  chrX  116442  RAB39B  NM_171998.3      -   \n",
       "4685  23010    2654895    2655782  chrY    6736     SRY  NM_003140.2      -   \n",
       "\n",
       "       +  -  \n",
       "1511  25  2  \n",
       "1512  26  1  \n",
       "4621   1  2  \n",
       "4622   2  1  \n",
       "4685   1  1  "
      ]
     },
     "execution_count": 44,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "exon.tail()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [],
   "source": [
    "exon['exon'] = pd.concat([exon.loc[exon['strand'] == '+', '+'], exon.loc[exon['strand'] == '-', '-']])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>row</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>chrom</th>\n",
       "      <th>GeneID</th>\n",
       "      <th>name2</th>\n",
       "      <th>name</th>\n",
       "      <th>strand</th>\n",
       "      <th>+</th>\n",
       "      <th>-</th>\n",
       "      <th>exon</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>149</th>\n",
       "      <td>59</td>\n",
       "      <td>6845383</td>\n",
       "      <td>6845635</td>\n",
       "      <td>chr1</td>\n",
       "      <td>23261</td>\n",
       "      <td>CAMTA1</td>\n",
       "      <td>NM_015215.3</td>\n",
       "      <td>+</td>\n",
       "      <td>1</td>\n",
       "      <td>23</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>150</th>\n",
       "      <td>59</td>\n",
       "      <td>6880240</td>\n",
       "      <td>6880310</td>\n",
       "      <td>chr1</td>\n",
       "      <td>23261</td>\n",
       "      <td>CAMTA1</td>\n",
       "      <td>NM_015215.3</td>\n",
       "      <td>+</td>\n",
       "      <td>2</td>\n",
       "      <td>22</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>151</th>\n",
       "      <td>59</td>\n",
       "      <td>6885151</td>\n",
       "      <td>6885270</td>\n",
       "      <td>chr1</td>\n",
       "      <td>23261</td>\n",
       "      <td>CAMTA1</td>\n",
       "      <td>NM_015215.3</td>\n",
       "      <td>+</td>\n",
       "      <td>3</td>\n",
       "      <td>21</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>152</th>\n",
       "      <td>59</td>\n",
       "      <td>7151363</td>\n",
       "      <td>7151431</td>\n",
       "      <td>chr1</td>\n",
       "      <td>23261</td>\n",
       "      <td>CAMTA1</td>\n",
       "      <td>NM_015215.3</td>\n",
       "      <td>+</td>\n",
       "      <td>4</td>\n",
       "      <td>20</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>153</th>\n",
       "      <td>59</td>\n",
       "      <td>7309550</td>\n",
       "      <td>7309686</td>\n",
       "      <td>chr1</td>\n",
       "      <td>23261</td>\n",
       "      <td>CAMTA1</td>\n",
       "      <td>NM_015215.3</td>\n",
       "      <td>+</td>\n",
       "      <td>5</td>\n",
       "      <td>19</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     row    start      end chrom  GeneID   name2         name strand  +   -  \\\n",
       "149   59  6845383  6845635  chr1   23261  CAMTA1  NM_015215.3      +  1  23   \n",
       "150   59  6880240  6880310  chr1   23261  CAMTA1  NM_015215.3      +  2  22   \n",
       "151   59  6885151  6885270  chr1   23261  CAMTA1  NM_015215.3      +  3  21   \n",
       "152   59  7151363  7151431  chr1   23261  CAMTA1  NM_015215.3      +  4  20   \n",
       "153   59  7309550  7309686  chr1   23261  CAMTA1  NM_015215.3      +  5  19   \n",
       "\n",
       "     exon  \n",
       "149     1  \n",
       "150     2  \n",
       "151     3  \n",
       "152     4  \n",
       "153     5  "
      ]
     },
     "execution_count": 46,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "exon.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>row</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>chrom</th>\n",
       "      <th>GeneID</th>\n",
       "      <th>name2</th>\n",
       "      <th>name</th>\n",
       "      <th>strand</th>\n",
       "      <th>+</th>\n",
       "      <th>-</th>\n",
       "      <th>exon</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1511</th>\n",
       "      <td>1511</td>\n",
       "      <td>154227753</td>\n",
       "      <td>154227875</td>\n",
       "      <td>chrX</td>\n",
       "      <td>2157</td>\n",
       "      <td>F8</td>\n",
       "      <td>NM_000132.3</td>\n",
       "      <td>-</td>\n",
       "      <td>25</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1512</th>\n",
       "      <td>1511</td>\n",
       "      <td>154250684</td>\n",
       "      <td>154250998</td>\n",
       "      <td>chrX</td>\n",
       "      <td>2157</td>\n",
       "      <td>F8</td>\n",
       "      <td>NM_000132.3</td>\n",
       "      <td>-</td>\n",
       "      <td>26</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4621</th>\n",
       "      <td>17777</td>\n",
       "      <td>154487518</td>\n",
       "      <td>154490514</td>\n",
       "      <td>chrX</td>\n",
       "      <td>116442</td>\n",
       "      <td>RAB39B</td>\n",
       "      <td>NM_171998.3</td>\n",
       "      <td>-</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4622</th>\n",
       "      <td>17777</td>\n",
       "      <td>154493358</td>\n",
       "      <td>154493874</td>\n",
       "      <td>chrX</td>\n",
       "      <td>116442</td>\n",
       "      <td>RAB39B</td>\n",
       "      <td>NM_171998.3</td>\n",
       "      <td>-</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4685</th>\n",
       "      <td>23010</td>\n",
       "      <td>2654895</td>\n",
       "      <td>2655782</td>\n",
       "      <td>chrY</td>\n",
       "      <td>6736</td>\n",
       "      <td>SRY</td>\n",
       "      <td>NM_003140.2</td>\n",
       "      <td>-</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        row      start        end chrom  GeneID   name2         name strand  \\\n",
       "1511   1511  154227753  154227875  chrX    2157      F8  NM_000132.3      -   \n",
       "1512   1511  154250684  154250998  chrX    2157      F8  NM_000132.3      -   \n",
       "4621  17777  154487518  154490514  chrX  116442  RAB39B  NM_171998.3      -   \n",
       "4622  17777  154493358  154493874  chrX  116442  RAB39B  NM_171998.3      -   \n",
       "4685  23010    2654895    2655782  chrY    6736     SRY  NM_003140.2      -   \n",
       "\n",
       "       +  -  exon  \n",
       "1511  25  2     2  \n",
       "1512  26  1     1  \n",
       "4621   1  2     2  \n",
       "4622   2  1     1  \n",
       "4685   1  1     1  "
      ]
     },
     "execution_count": 47,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "exon.tail()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [],
   "source": [
    "exon['last_exon'] = exon.groupby(['name2', 'name'])['exon'].transform('max') == exon['exon']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>row</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>chrom</th>\n",
       "      <th>GeneID</th>\n",
       "      <th>name2</th>\n",
       "      <th>name</th>\n",
       "      <th>strand</th>\n",
       "      <th>+</th>\n",
       "      <th>-</th>\n",
       "      <th>exon</th>\n",
       "      <th>last_exon</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>149</th>\n",
       "      <td>59</td>\n",
       "      <td>6845383</td>\n",
       "      <td>6845635</td>\n",
       "      <td>chr1</td>\n",
       "      <td>23261</td>\n",
       "      <td>CAMTA1</td>\n",
       "      <td>NM_015215.3</td>\n",
       "      <td>+</td>\n",
       "      <td>1</td>\n",
       "      <td>23</td>\n",
       "      <td>1</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>150</th>\n",
       "      <td>59</td>\n",
       "      <td>6880240</td>\n",
       "      <td>6880310</td>\n",
       "      <td>chr1</td>\n",
       "      <td>23261</td>\n",
       "      <td>CAMTA1</td>\n",
       "      <td>NM_015215.3</td>\n",
       "      <td>+</td>\n",
       "      <td>2</td>\n",
       "      <td>22</td>\n",
       "      <td>2</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>151</th>\n",
       "      <td>59</td>\n",
       "      <td>6885151</td>\n",
       "      <td>6885270</td>\n",
       "      <td>chr1</td>\n",
       "      <td>23261</td>\n",
       "      <td>CAMTA1</td>\n",
       "      <td>NM_015215.3</td>\n",
       "      <td>+</td>\n",
       "      <td>3</td>\n",
       "      <td>21</td>\n",
       "      <td>3</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>152</th>\n",
       "      <td>59</td>\n",
       "      <td>7151363</td>\n",
       "      <td>7151431</td>\n",
       "      <td>chr1</td>\n",
       "      <td>23261</td>\n",
       "      <td>CAMTA1</td>\n",
       "      <td>NM_015215.3</td>\n",
       "      <td>+</td>\n",
       "      <td>4</td>\n",
       "      <td>20</td>\n",
       "      <td>4</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>153</th>\n",
       "      <td>59</td>\n",
       "      <td>7309550</td>\n",
       "      <td>7309686</td>\n",
       "      <td>chr1</td>\n",
       "      <td>23261</td>\n",
       "      <td>CAMTA1</td>\n",
       "      <td>NM_015215.3</td>\n",
       "      <td>+</td>\n",
       "      <td>5</td>\n",
       "      <td>19</td>\n",
       "      <td>5</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     row    start      end chrom  GeneID   name2         name strand  +   -  \\\n",
       "149   59  6845383  6845635  chr1   23261  CAMTA1  NM_015215.3      +  1  23   \n",
       "150   59  6880240  6880310  chr1   23261  CAMTA1  NM_015215.3      +  2  22   \n",
       "151   59  6885151  6885270  chr1   23261  CAMTA1  NM_015215.3      +  3  21   \n",
       "152   59  7151363  7151431  chr1   23261  CAMTA1  NM_015215.3      +  4  20   \n",
       "153   59  7309550  7309686  chr1   23261  CAMTA1  NM_015215.3      +  5  19   \n",
       "\n",
       "     exon  last_exon  \n",
       "149     1      False  \n",
       "150     2      False  \n",
       "151     3      False  \n",
       "152     4      False  \n",
       "153     5      False  "
      ]
     },
     "execution_count": 49,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "exon.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>row</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>chrom</th>\n",
       "      <th>GeneID</th>\n",
       "      <th>name2</th>\n",
       "      <th>name</th>\n",
       "      <th>strand</th>\n",
       "      <th>+</th>\n",
       "      <th>-</th>\n",
       "      <th>exon</th>\n",
       "      <th>last_exon</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1511</th>\n",
       "      <td>1511</td>\n",
       "      <td>154227753</td>\n",
       "      <td>154227875</td>\n",
       "      <td>chrX</td>\n",
       "      <td>2157</td>\n",
       "      <td>F8</td>\n",
       "      <td>NM_000132.3</td>\n",
       "      <td>-</td>\n",
       "      <td>25</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1512</th>\n",
       "      <td>1511</td>\n",
       "      <td>154250684</td>\n",
       "      <td>154250998</td>\n",
       "      <td>chrX</td>\n",
       "      <td>2157</td>\n",
       "      <td>F8</td>\n",
       "      <td>NM_000132.3</td>\n",
       "      <td>-</td>\n",
       "      <td>26</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4621</th>\n",
       "      <td>17777</td>\n",
       "      <td>154487518</td>\n",
       "      <td>154490514</td>\n",
       "      <td>chrX</td>\n",
       "      <td>116442</td>\n",
       "      <td>RAB39B</td>\n",
       "      <td>NM_171998.3</td>\n",
       "      <td>-</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4622</th>\n",
       "      <td>17777</td>\n",
       "      <td>154493358</td>\n",
       "      <td>154493874</td>\n",
       "      <td>chrX</td>\n",
       "      <td>116442</td>\n",
       "      <td>RAB39B</td>\n",
       "      <td>NM_171998.3</td>\n",
       "      <td>-</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4685</th>\n",
       "      <td>23010</td>\n",
       "      <td>2654895</td>\n",
       "      <td>2655782</td>\n",
       "      <td>chrY</td>\n",
       "      <td>6736</td>\n",
       "      <td>SRY</td>\n",
       "      <td>NM_003140.2</td>\n",
       "      <td>-</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        row      start        end chrom  GeneID   name2         name strand  \\\n",
       "1511   1511  154227753  154227875  chrX    2157      F8  NM_000132.3      -   \n",
       "1512   1511  154250684  154250998  chrX    2157      F8  NM_000132.3      -   \n",
       "4621  17777  154487518  154490514  chrX  116442  RAB39B  NM_171998.3      -   \n",
       "4622  17777  154493358  154493874  chrX  116442  RAB39B  NM_171998.3      -   \n",
       "4685  23010    2654895    2655782  chrY    6736     SRY  NM_003140.2      -   \n",
       "\n",
       "       +  -  exon  last_exon  \n",
       "1511  25  2     2      False  \n",
       "1512  26  1     1      False  \n",
       "4621   1  2     2       True  \n",
       "4622   2  1     1      False  \n",
       "4685   1  1     1       True  "
      ]
     },
     "execution_count": 50,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "exon.tail()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [],
   "source": [
    "exon = exon[\n",
    "    ['chrom', 'start', 'end', 'GeneID', 'name2', 'name', 'exon', 'last_exon']\n",
    "].sort_values(['chrom', 'start', 'end'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>GeneID</th>\n",
       "      <th>name2</th>\n",
       "      <th>name</th>\n",
       "      <th>exon</th>\n",
       "      <th>last_exon</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>149</th>\n",
       "      <td>chr1</td>\n",
       "      <td>6845383</td>\n",
       "      <td>6845635</td>\n",
       "      <td>23261</td>\n",
       "      <td>CAMTA1</td>\n",
       "      <td>NM_015215.3</td>\n",
       "      <td>1</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>150</th>\n",
       "      <td>chr1</td>\n",
       "      <td>6880240</td>\n",
       "      <td>6880310</td>\n",
       "      <td>23261</td>\n",
       "      <td>CAMTA1</td>\n",
       "      <td>NM_015215.3</td>\n",
       "      <td>2</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>151</th>\n",
       "      <td>chr1</td>\n",
       "      <td>6885151</td>\n",
       "      <td>6885270</td>\n",
       "      <td>23261</td>\n",
       "      <td>CAMTA1</td>\n",
       "      <td>NM_015215.3</td>\n",
       "      <td>3</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>152</th>\n",
       "      <td>chr1</td>\n",
       "      <td>7151363</td>\n",
       "      <td>7151431</td>\n",
       "      <td>23261</td>\n",
       "      <td>CAMTA1</td>\n",
       "      <td>NM_015215.3</td>\n",
       "      <td>4</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>153</th>\n",
       "      <td>chr1</td>\n",
       "      <td>7309550</td>\n",
       "      <td>7309686</td>\n",
       "      <td>23261</td>\n",
       "      <td>CAMTA1</td>\n",
       "      <td>NM_015215.3</td>\n",
       "      <td>5</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    chrom    start      end  GeneID   name2         name  exon  last_exon\n",
       "149  chr1  6845383  6845635   23261  CAMTA1  NM_015215.3     1      False\n",
       "150  chr1  6880240  6880310   23261  CAMTA1  NM_015215.3     2      False\n",
       "151  chr1  6885151  6885270   23261  CAMTA1  NM_015215.3     3      False\n",
       "152  chr1  7151363  7151431   23261  CAMTA1  NM_015215.3     4      False\n",
       "153  chr1  7309550  7309686   23261  CAMTA1  NM_015215.3     5      False"
      ]
     },
     "execution_count": 52,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "exon.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [],
   "source": [
    "exon.rename(columns={\n",
    "    'chrom': '#chrom',\n",
    "    'GeneID': 'gene_id', 'name2': 'symbol', 'name': 'transcript'\n",
    "}).to_csv('data/hi-exon-20200109.sorted.bed', index=False, sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [],
   "source": [
    "!bgzip -cf data/hi-exon-20200109.sorted.bed > ../data/hi-exon-20200109.sorted.bed.gz"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {},
   "outputs": [],
   "source": [
    "!tabix -fp bed ../data/hi-exon-20200109.sorted.bed.gz"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {},
   "outputs": [],
   "source": [
    "last_exon = exon[exon['last_exon']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>GeneID</th>\n",
       "      <th>name2</th>\n",
       "      <th>name</th>\n",
       "      <th>exon</th>\n",
       "      <th>last_exon</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>171</th>\n",
       "      <td>chr1</td>\n",
       "      <td>7826518</td>\n",
       "      <td>7829766</td>\n",
       "      <td>23261</td>\n",
       "      <td>CAMTA1</td>\n",
       "      <td>NM_015215.3</td>\n",
       "      <td>23</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3726</th>\n",
       "      <td>chr1</td>\n",
       "      <td>17345216</td>\n",
       "      <td>17345453</td>\n",
       "      <td>6390</td>\n",
       "      <td>SDHB</td>\n",
       "      <td>NM_003000.2</td>\n",
       "      <td>8</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2508</th>\n",
       "      <td>chr1</td>\n",
       "      <td>27105513</td>\n",
       "      <td>27108601</td>\n",
       "      <td>8289</td>\n",
       "      <td>ARID1A</td>\n",
       "      <td>NM_139135.2</td>\n",
       "      <td>20</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2798</th>\n",
       "      <td>chr1</td>\n",
       "      <td>27860755</td>\n",
       "      <td>27861427</td>\n",
       "      <td>27245</td>\n",
       "      <td>AHDC1</td>\n",
       "      <td>NM_001029882.3</td>\n",
       "      <td>7</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3802</th>\n",
       "      <td>chr1</td>\n",
       "      <td>43391045</td>\n",
       "      <td>43392912</td>\n",
       "      <td>6513</td>\n",
       "      <td>SLC2A1</td>\n",
       "      <td>NM_006516.2</td>\n",
       "      <td>10</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     chrom     start       end  GeneID   name2            name  exon  \\\n",
       "171   chr1   7826518   7829766   23261  CAMTA1     NM_015215.3    23   \n",
       "3726  chr1  17345216  17345453    6390    SDHB     NM_003000.2     8   \n",
       "2508  chr1  27105513  27108601    8289  ARID1A     NM_139135.2    20   \n",
       "2798  chr1  27860755  27861427   27245   AHDC1  NM_001029882.3     7   \n",
       "3802  chr1  43391045  43392912    6513  SLC2A1     NM_006516.2    10   \n",
       "\n",
       "      last_exon  \n",
       "171        True  \n",
       "3726       True  \n",
       "2508       True  \n",
       "2798       True  \n",
       "3802       True  "
      ]
     },
     "execution_count": 57,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "last_exon.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {},
   "outputs": [],
   "source": [
    "last_exon_region = last_exon['chrom'] + ':' + last_exon['start'].astype(str) + '-' + last_exon['end'].astype(str)\n",
    "last_exon_region = last_exon_region.str.replace('chr', '')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "171       1:7826518-7829766\n",
       "3726    1:17345216-17345453\n",
       "2508    1:27105513-27108601\n",
       "2798    1:27860755-27861427\n",
       "3802    1:43391045-43392912\n",
       "dtype: object"
      ]
     },
     "execution_count": 59,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "last_exon_region.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {},
   "outputs": [],
   "source": [
    "need_fields = [\n",
    "    'variants/CHROM', 'variants/POS', 'variants/REF', 'variants/ALT',\n",
    "    'variants/AF_ESP', 'variants/AF_EXAC', 'variants/AF_TGP', 'variants/CLNSIG'\n",
    "]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open('data/clinvar_pathogenic_20191125.sorted.vcf', 'w') as f:\n",
    "    headers = allel.read_vcf_headers('data/clinvar_20191125.vcf.gz')\n",
    "    f.write(''.join(headers.headers))\n",
    "    \n",
    "    for region in last_exon_region:\n",
    "        fields, samples, headers, it = allel.iter_vcf_chunks(\n",
    "            'data/clinvar_20191125.vcf.gz', fields=need_fields, alt_number=1, region=region\n",
    "        )\n",
    "        for variants, *_ in it:\n",
    "            esp_filter = np.isnan(variants['variants/AF_ESP'])\n",
    "            esp_filter[~esp_filter] |= variants['variants/AF_ESP'][~esp_filter] < 0.01\n",
    "\n",
    "            exac_filter = np.isnan(variants['variants/AF_EXAC'])\n",
    "            exac_filter[~exac_filter] |= variants['variants/AF_EXAC'][~exac_filter] < 0.01\n",
    "\n",
    "            tgp_filter = np.isnan(variants['variants/AF_TGP'])\n",
    "            tgp_filter[~tgp_filter] |= variants['variants/AF_TGP'][~tgp_filter] < 0.01\n",
    "\n",
    "            pathogenic_filter = np.isin(\n",
    "                variants['variants/CLNSIG'], ['Likely_pathogenic', 'Pathogenic', 'Pathogenic/Likely_pathogenic']\n",
    "            )\n",
    "\n",
    "            af_filter = esp_filter & exac_filter & tgp_filter & pathogenic_filter\n",
    "\n",
    "            filtered_variants = {k: v[af_filter] for k, v in variants.items()}\n",
    "\n",
    "            filtered_variants['variants/CHROM'] = 'chr' + filtered_variants['variants/CHROM']\n",
    "\n",
    "            names, callset = allel.normalize_callset(filtered_variants)\n",
    "            allel.write_vcf_data(f, names, callset, None, {field: np.nan for field in need_fields})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "sed: 1: \"data/clinvar_pathogenic ...\": extra characters at the end of d command\r\n"
     ]
    }
   ],
   "source": [
    "!sed -i 's/AF_.\\+=nan;//' data/clinvar_pathogenic_20191125.sorted.vcf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "metadata": {},
   "outputs": [],
   "source": [
    "!bgzip -cf data/clinvar_pathogenic_20191125.sorted.vcf > ../data/clinvar-pathogenic-20191125.sorted.vcf.gz"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "metadata": {},
   "outputs": [],
   "source": [
    "!tabix -fp vcf ../data/clinvar-pathogenic-20191125.sorted.vcf.gz"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "metadata": {},
   "outputs": [],
   "source": [
    "uhi_genes = set(\n",
    "    curation_gene.loc[curation_gene['Haploinsufficiency Score'] == '40', '#Gene Symbol']\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "metadata": {},
   "outputs": [],
   "source": [
    "uhi_gene = gene[gene['name2'].isin(uhi_genes)].copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>txStart</th>\n",
       "      <th>txEnd</th>\n",
       "      <th>GeneID</th>\n",
       "      <th>name2</th>\n",
       "      <th>name</th>\n",
       "      <th>strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>10291</th>\n",
       "      <td>chr1</td>\n",
       "      <td>55505148</td>\n",
       "      <td>55530526</td>\n",
       "      <td>255738</td>\n",
       "      <td>PCSK9</td>\n",
       "      <td>NM_174936.3</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18375</th>\n",
       "      <td>chr13</td>\n",
       "      <td>20761601</td>\n",
       "      <td>20767114</td>\n",
       "      <td>2706</td>\n",
       "      <td>GJB2</td>\n",
       "      <td>NM_004004.5</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       chrom   txStart     txEnd  GeneID  name2         name strand\n",
       "10291   chr1  55505148  55530526  255738  PCSK9  NM_174936.3      +\n",
       "18375  chr13  20761601  20767114    2706   GJB2  NM_004004.5      -"
      ]
     },
     "execution_count": 67,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "uhi_gene.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "metadata": {},
   "outputs": [],
   "source": [
    "uhi_gene.rename(columns={\n",
    "    'chrom': '#chrom', 'txStart': 'start', 'txEnd': 'end',\n",
    "    'GeneID': 'gene_id', 'name2': 'symbol', 'name': 'transcript'\n",
    "}).to_csv('data/uhi-gene-20200109.sorted.bed', index=False, sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "metadata": {},
   "outputs": [],
   "source": [
    "!bgzip -cf data/uhi-gene-20200109.sorted.bed > ../data/uhi-gene-20200109.sorted.bed.gz"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "metadata": {},
   "outputs": [],
   "source": [
    "!tabix -fp bed ../data/uhi-gene-20200109.sorted.bed.gz"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "metadata": {},
   "outputs": [],
   "source": [
    "region = pd.read_csv('data/ClinGen_region_curation_list_GRCh37.tsv', sep='\\t', skiprows=5, dtype=str)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>#ISCA ID</th>\n",
       "      <th>ISCA Region Name</th>\n",
       "      <th>cytoBand</th>\n",
       "      <th>Genomic Location</th>\n",
       "      <th>Haploinsufficiency Score</th>\n",
       "      <th>Haploinsufficiency Description</th>\n",
       "      <th>Haploinsufficiency PMID1</th>\n",
       "      <th>Haploinsufficiency PMID2</th>\n",
       "      <th>Haploinsufficiency PMID3</th>\n",
       "      <th>Triplosensitivity Score</th>\n",
       "      <th>Triplosensitivity Description</th>\n",
       "      <th>Triplosensitivity PMID1</th>\n",
       "      <th>Triplosensitivity PMID2</th>\n",
       "      <th>Triplosensitivity PMID3</th>\n",
       "      <th>Date Last Evaluated</th>\n",
       "      <th>Loss phenotype OMIM ID</th>\n",
       "      <th>Triplosensitive phenotype OMIM ID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ISCA-46300</td>\n",
       "      <td>15q24 recurrent region (C-D) (includes SIN3A)</td>\n",
       "      <td>NaN</td>\n",
       "      <td>chr15: 75631787-75972909</td>\n",
       "      <td>3</td>\n",
       "      <td>Sufficient evidence for dosage pathogenicity</td>\n",
       "      <td>27399968</td>\n",
       "      <td>22180641</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>No evidence available</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2020-01-13</td>\n",
       "      <td>613406</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ISCA-46299</td>\n",
       "      <td>Xp11.22 region (includes HUWE1)</td>\n",
       "      <td>Xp11.22</td>\n",
       "      <td>chrX:53363456-53793054</td>\n",
       "      <td>0</td>\n",
       "      <td>No evidence available</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3</td>\n",
       "      <td>Sufficient evidence for dosage pathogenicity</td>\n",
       "      <td>22840365</td>\n",
       "      <td>20655035</td>\n",
       "      <td>26692240</td>\n",
       "      <td>2018-11-19</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ISCA-46298</td>\n",
       "      <td>13q12.12 recurrent region (includes SACS and S...</td>\n",
       "      <td>13q12.12</td>\n",
       "      <td>chr13:23544582-24881361</td>\n",
       "      <td>30</td>\n",
       "      <td>Gene associated with autosomal recessive pheno...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>No evidence available</td>\n",
       "      <td>23239789</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2019-11-03</td>\n",
       "      <td>270550,253700</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ISCA-46297</td>\n",
       "      <td>16p12.2 recurrent region (includes OTOA) (dist...</td>\n",
       "      <td>16p12.2</td>\n",
       "      <td>chr16:21570113-21740423</td>\n",
       "      <td>30</td>\n",
       "      <td>Gene associated with autosomal recessive pheno...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>40</td>\n",
       "      <td>Dosage sensitivity unlikely</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2019-11-03</td>\n",
       "      <td>607039</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ISCA-46296</td>\n",
       "      <td>15q24 recurrent region (A-C)</td>\n",
       "      <td>15q24.1-q24.2</td>\n",
       "      <td>chr15:72963715-75508312</td>\n",
       "      <td>3</td>\n",
       "      <td>Sufficient evidence for dosage pathogenicity</td>\n",
       "      <td>22180641</td>\n",
       "      <td>19921647</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>Little evidence for dosage pathogenicity</td>\n",
       "      <td>19557438</td>\n",
       "      <td>20860070</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2020-01-03</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     #ISCA ID                                   ISCA Region Name  \\\n",
       "0  ISCA-46300      15q24 recurrent region (C-D) (includes SIN3A)   \n",
       "1  ISCA-46299                    Xp11.22 region (includes HUWE1)   \n",
       "2  ISCA-46298  13q12.12 recurrent region (includes SACS and S...   \n",
       "3  ISCA-46297  16p12.2 recurrent region (includes OTOA) (dist...   \n",
       "4  ISCA-46296                       15q24 recurrent region (A-C)   \n",
       "\n",
       "        cytoBand          Genomic Location Haploinsufficiency Score  \\\n",
       "0            NaN  chr15: 75631787-75972909                        3   \n",
       "1        Xp11.22    chrX:53363456-53793054                        0   \n",
       "2       13q12.12   chr13:23544582-24881361                       30   \n",
       "3        16p12.2   chr16:21570113-21740423                       30   \n",
       "4  15q24.1-q24.2   chr15:72963715-75508312                        3   \n",
       "\n",
       "                      Haploinsufficiency Description Haploinsufficiency PMID1  \\\n",
       "0       Sufficient evidence for dosage pathogenicity                 27399968   \n",
       "1                              No evidence available                      NaN   \n",
       "2  Gene associated with autosomal recessive pheno...                      NaN   \n",
       "3  Gene associated with autosomal recessive pheno...                      NaN   \n",
       "4       Sufficient evidence for dosage pathogenicity                 22180641   \n",
       "\n",
       "  Haploinsufficiency PMID2 Haploinsufficiency PMID3 Triplosensitivity Score  \\\n",
       "0                 22180641                      NaN                       0   \n",
       "1                      NaN                      NaN                       3   \n",
       "2                      NaN                      NaN                       0   \n",
       "3                      NaN                      NaN                      40   \n",
       "4                 19921647                      NaN                       1   \n",
       "\n",
       "                  Triplosensitivity Description Triplosensitivity PMID1  \\\n",
       "0                         No evidence available                     NaN   \n",
       "1  Sufficient evidence for dosage pathogenicity                22840365   \n",
       "2                         No evidence available                23239789   \n",
       "3                   Dosage sensitivity unlikely                     NaN   \n",
       "4      Little evidence for dosage pathogenicity                19557438   \n",
       "\n",
       "  Triplosensitivity PMID2 Triplosensitivity PMID3 Date Last Evaluated  \\\n",
       "0                     NaN                     NaN          2020-01-13   \n",
       "1                20655035                26692240          2018-11-19   \n",
       "2                     NaN                     NaN          2019-11-03   \n",
       "3                     NaN                     NaN          2019-11-03   \n",
       "4                20860070                     NaN          2020-01-03   \n",
       "\n",
       "  Loss phenotype OMIM ID Triplosensitive phenotype OMIM ID  \n",
       "0                 613406                               NaN  \n",
       "1                    NaN                               NaN  \n",
       "2          270550,253700                               NaN  \n",
       "3                 607039                               NaN  \n",
       "4                    NaN                               NaN  "
      ]
     },
     "execution_count": 72,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "region.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "metadata": {},
   "outputs": [],
   "source": [
    "position = region['Genomic Location'].str.extract(r'(?P<chrom>chr\\w+)\\s*:\\s*(?P<start>\\d+)\\s*-\\s*(?P<end>\\d+)')\n",
    "position['start'] = position['start'].astype(int)\n",
    "position['end'] = position['end'].astype(int)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr15</td>\n",
       "      <td>75631787</td>\n",
       "      <td>75972909</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chrX</td>\n",
       "      <td>53363456</td>\n",
       "      <td>53793054</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr13</td>\n",
       "      <td>23544582</td>\n",
       "      <td>24881361</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr16</td>\n",
       "      <td>21570113</td>\n",
       "      <td>21740423</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr15</td>\n",
       "      <td>72963715</td>\n",
       "      <td>75508312</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   chrom     start       end\n",
       "0  chr15  75631787  75972909\n",
       "1   chrX  53363456  53793054\n",
       "2  chr13  23544582  24881361\n",
       "3  chr16  21570113  21740423\n",
       "4  chr15  72963715  75508312"
      ]
     },
     "execution_count": 74,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "position.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "metadata": {},
   "outputs": [],
   "source": [
    "region_pos = region.merge(position, how='left', left_index=True, right_index=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>#ISCA ID</th>\n",
       "      <th>ISCA Region Name</th>\n",
       "      <th>cytoBand</th>\n",
       "      <th>Genomic Location</th>\n",
       "      <th>Haploinsufficiency Score</th>\n",
       "      <th>Haploinsufficiency Description</th>\n",
       "      <th>Haploinsufficiency PMID1</th>\n",
       "      <th>Haploinsufficiency PMID2</th>\n",
       "      <th>Haploinsufficiency PMID3</th>\n",
       "      <th>Triplosensitivity Score</th>\n",
       "      <th>Triplosensitivity Description</th>\n",
       "      <th>Triplosensitivity PMID1</th>\n",
       "      <th>Triplosensitivity PMID2</th>\n",
       "      <th>Triplosensitivity PMID3</th>\n",
       "      <th>Date Last Evaluated</th>\n",
       "      <th>Loss phenotype OMIM ID</th>\n",
       "      <th>Triplosensitive phenotype OMIM ID</th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ISCA-46300</td>\n",
       "      <td>15q24 recurrent region (C-D) (includes SIN3A)</td>\n",
       "      <td>NaN</td>\n",
       "      <td>chr15: 75631787-75972909</td>\n",
       "      <td>3</td>\n",
       "      <td>Sufficient evidence for dosage pathogenicity</td>\n",
       "      <td>27399968</td>\n",
       "      <td>22180641</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>No evidence available</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2020-01-13</td>\n",
       "      <td>613406</td>\n",
       "      <td>NaN</td>\n",
       "      <td>chr15</td>\n",
       "      <td>75631787</td>\n",
       "      <td>75972909</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ISCA-46299</td>\n",
       "      <td>Xp11.22 region (includes HUWE1)</td>\n",
       "      <td>Xp11.22</td>\n",
       "      <td>chrX:53363456-53793054</td>\n",
       "      <td>0</td>\n",
       "      <td>No evidence available</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3</td>\n",
       "      <td>Sufficient evidence for dosage pathogenicity</td>\n",
       "      <td>22840365</td>\n",
       "      <td>20655035</td>\n",
       "      <td>26692240</td>\n",
       "      <td>2018-11-19</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>chrX</td>\n",
       "      <td>53363456</td>\n",
       "      <td>53793054</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ISCA-46298</td>\n",
       "      <td>13q12.12 recurrent region (includes SACS and S...</td>\n",
       "      <td>13q12.12</td>\n",
       "      <td>chr13:23544582-24881361</td>\n",
       "      <td>30</td>\n",
       "      <td>Gene associated with autosomal recessive pheno...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>No evidence available</td>\n",
       "      <td>23239789</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2019-11-03</td>\n",
       "      <td>270550,253700</td>\n",
       "      <td>NaN</td>\n",
       "      <td>chr13</td>\n",
       "      <td>23544582</td>\n",
       "      <td>24881361</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ISCA-46297</td>\n",
       "      <td>16p12.2 recurrent region (includes OTOA) (dist...</td>\n",
       "      <td>16p12.2</td>\n",
       "      <td>chr16:21570113-21740423</td>\n",
       "      <td>30</td>\n",
       "      <td>Gene associated with autosomal recessive pheno...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>40</td>\n",
       "      <td>Dosage sensitivity unlikely</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2019-11-03</td>\n",
       "      <td>607039</td>\n",
       "      <td>NaN</td>\n",
       "      <td>chr16</td>\n",
       "      <td>21570113</td>\n",
       "      <td>21740423</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ISCA-46296</td>\n",
       "      <td>15q24 recurrent region (A-C)</td>\n",
       "      <td>15q24.1-q24.2</td>\n",
       "      <td>chr15:72963715-75508312</td>\n",
       "      <td>3</td>\n",
       "      <td>Sufficient evidence for dosage pathogenicity</td>\n",
       "      <td>22180641</td>\n",
       "      <td>19921647</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>Little evidence for dosage pathogenicity</td>\n",
       "      <td>19557438</td>\n",
       "      <td>20860070</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2020-01-03</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>chr15</td>\n",
       "      <td>72963715</td>\n",
       "      <td>75508312</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     #ISCA ID                                   ISCA Region Name  \\\n",
       "0  ISCA-46300      15q24 recurrent region (C-D) (includes SIN3A)   \n",
       "1  ISCA-46299                    Xp11.22 region (includes HUWE1)   \n",
       "2  ISCA-46298  13q12.12 recurrent region (includes SACS and S...   \n",
       "3  ISCA-46297  16p12.2 recurrent region (includes OTOA) (dist...   \n",
       "4  ISCA-46296                       15q24 recurrent region (A-C)   \n",
       "\n",
       "        cytoBand          Genomic Location Haploinsufficiency Score  \\\n",
       "0            NaN  chr15: 75631787-75972909                        3   \n",
       "1        Xp11.22    chrX:53363456-53793054                        0   \n",
       "2       13q12.12   chr13:23544582-24881361                       30   \n",
       "3        16p12.2   chr16:21570113-21740423                       30   \n",
       "4  15q24.1-q24.2   chr15:72963715-75508312                        3   \n",
       "\n",
       "                      Haploinsufficiency Description Haploinsufficiency PMID1  \\\n",
       "0       Sufficient evidence for dosage pathogenicity                 27399968   \n",
       "1                              No evidence available                      NaN   \n",
       "2  Gene associated with autosomal recessive pheno...                      NaN   \n",
       "3  Gene associated with autosomal recessive pheno...                      NaN   \n",
       "4       Sufficient evidence for dosage pathogenicity                 22180641   \n",
       "\n",
       "  Haploinsufficiency PMID2 Haploinsufficiency PMID3 Triplosensitivity Score  \\\n",
       "0                 22180641                      NaN                       0   \n",
       "1                      NaN                      NaN                       3   \n",
       "2                      NaN                      NaN                       0   \n",
       "3                      NaN                      NaN                      40   \n",
       "4                 19921647                      NaN                       1   \n",
       "\n",
       "                  Triplosensitivity Description Triplosensitivity PMID1  \\\n",
       "0                         No evidence available                     NaN   \n",
       "1  Sufficient evidence for dosage pathogenicity                22840365   \n",
       "2                         No evidence available                23239789   \n",
       "3                   Dosage sensitivity unlikely                     NaN   \n",
       "4      Little evidence for dosage pathogenicity                19557438   \n",
       "\n",
       "  Triplosensitivity PMID2 Triplosensitivity PMID3 Date Last Evaluated  \\\n",
       "0                     NaN                     NaN          2020-01-13   \n",
       "1                20655035                26692240          2018-11-19   \n",
       "2                     NaN                     NaN          2019-11-03   \n",
       "3                     NaN                     NaN          2019-11-03   \n",
       "4                20860070                     NaN          2020-01-03   \n",
       "\n",
       "  Loss phenotype OMIM ID Triplosensitive phenotype OMIM ID  chrom     start  \\\n",
       "0                 613406                               NaN  chr15  75631787   \n",
       "1                    NaN                               NaN   chrX  53363456   \n",
       "2          270550,253700                               NaN  chr13  23544582   \n",
       "3                 607039                               NaN  chr16  21570113   \n",
       "4                    NaN                               NaN  chr15  72963715   \n",
       "\n",
       "        end  \n",
       "0  75972909  \n",
       "1  53793054  \n",
       "2  24881361  \n",
       "3  21740423  \n",
       "4  75508312  "
      ]
     },
     "execution_count": 76,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "region_pos.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "metadata": {},
   "outputs": [],
   "source": [
    "hi_region = region_pos.loc[\n",
    "    region_pos['Haploinsufficiency Score'] == '3',\n",
    "    ['chrom', 'start', 'end', '#ISCA ID', 'ISCA Region Name']\n",
    "].sort_values(['chrom', 'start', 'end'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>#ISCA ID</th>\n",
       "      <th>ISCA Region Name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>43</th>\n",
       "      <td>chr1</td>\n",
       "      <td>834083</td>\n",
       "      <td>6289973</td>\n",
       "      <td>ISCA-37434</td>\n",
       "      <td>1p36 terminal region (includes GABRD)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>52</th>\n",
       "      <td>chr1</td>\n",
       "      <td>146577486</td>\n",
       "      <td>147394506</td>\n",
       "      <td>ISCA-37421</td>\n",
       "      <td>1q21.1 recurrent region (BP3-BP4, distal) (inc...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>chr1</td>\n",
       "      <td>243287730</td>\n",
       "      <td>245318287</td>\n",
       "      <td>ISCA-37493</td>\n",
       "      <td>1q43q44 terminal region (includes AKT3)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>51</th>\n",
       "      <td>chr10</td>\n",
       "      <td>81682843</td>\n",
       "      <td>88739388</td>\n",
       "      <td>ISCA-37424</td>\n",
       "      <td>10q22.3q23.2 recurrent region (LCR-3/4-flanked...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>63</th>\n",
       "      <td>chr11</td>\n",
       "      <td>31803509</td>\n",
       "      <td>32510988</td>\n",
       "      <td>ISCA-37401</td>\n",
       "      <td>11p13 (WAGR syndrome) region</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    chrom      start        end    #ISCA ID  \\\n",
       "43   chr1     834083    6289973  ISCA-37434   \n",
       "52   chr1  146577486  147394506  ISCA-37421   \n",
       "17   chr1  243287730  245318287  ISCA-37493   \n",
       "51  chr10   81682843   88739388  ISCA-37424   \n",
       "63  chr11   31803509   32510988  ISCA-37401   \n",
       "\n",
       "                                     ISCA Region Name  \n",
       "43              1p36 terminal region (includes GABRD)  \n",
       "52  1q21.1 recurrent region (BP3-BP4, distal) (inc...  \n",
       "17            1q43q44 terminal region (includes AKT3)  \n",
       "51  10q22.3q23.2 recurrent region (LCR-3/4-flanked...  \n",
       "63                       11p13 (WAGR syndrome) region  "
      ]
     },
     "execution_count": 78,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "hi_region.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "metadata": {},
   "outputs": [],
   "source": [
    "hi_region['omim_genes'] = hi_region.apply(\n",
    "    lambda row: ','.join(omim_gene.loc[\n",
    "        (omim_gene['chrom'] == row['chrom']) & (omim_gene['txStart'] >= row['start']) & (omim_gene['txEnd'] <= row['end']), 'name2'\n",
    "    ]),\n",
    "    axis=1\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>#ISCA ID</th>\n",
       "      <th>ISCA Region Name</th>\n",
       "      <th>omim_genes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>43</th>\n",
       "      <td>chr1</td>\n",
       "      <td>834083</td>\n",
       "      <td>6289973</td>\n",
       "      <td>ISCA-37434</td>\n",
       "      <td>1p36 terminal region (includes GABRD)</td>\n",
       "      <td>ISG15,AGRN,TNFRSF4,B3GALT6,DVL1,ATAD3A,TMEM240...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>52</th>\n",
       "      <td>chr1</td>\n",
       "      <td>146577486</td>\n",
       "      <td>147394506</td>\n",
       "      <td>ISCA-37421</td>\n",
       "      <td>1q21.1 recurrent region (BP3-BP4, distal) (inc...</td>\n",
       "      <td>GJA5,GJA8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>chr1</td>\n",
       "      <td>243287730</td>\n",
       "      <td>245318287</td>\n",
       "      <td>ISCA-37493</td>\n",
       "      <td>1q43q44 terminal region (includes AKT3)</td>\n",
       "      <td>SDCCAG8,AKT3,ZBTB18,COX20,HNRNPU</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>51</th>\n",
       "      <td>chr10</td>\n",
       "      <td>81682843</td>\n",
       "      <td>88739388</td>\n",
       "      <td>ISCA-37424</td>\n",
       "      <td>10q22.3q23.2 recurrent region (LCR-3/4-flanked...</td>\n",
       "      <td>ANXA11,MAT1A,CDHR1,RGR,LDB3,BMPR1A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>63</th>\n",
       "      <td>chr11</td>\n",
       "      <td>31803509</td>\n",
       "      <td>32510988</td>\n",
       "      <td>ISCA-37401</td>\n",
       "      <td>11p13 (WAGR syndrome) region</td>\n",
       "      <td>PAX6,WT1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    chrom      start        end    #ISCA ID  \\\n",
       "43   chr1     834083    6289973  ISCA-37434   \n",
       "52   chr1  146577486  147394506  ISCA-37421   \n",
       "17   chr1  243287730  245318287  ISCA-37493   \n",
       "51  chr10   81682843   88739388  ISCA-37424   \n",
       "63  chr11   31803509   32510988  ISCA-37401   \n",
       "\n",
       "                                     ISCA Region Name  \\\n",
       "43              1p36 terminal region (includes GABRD)   \n",
       "52  1q21.1 recurrent region (BP3-BP4, distal) (inc...   \n",
       "17            1q43q44 terminal region (includes AKT3)   \n",
       "51  10q22.3q23.2 recurrent region (LCR-3/4-flanked...   \n",
       "63                       11p13 (WAGR syndrome) region   \n",
       "\n",
       "                                           omim_genes  \n",
       "43  ISG15,AGRN,TNFRSF4,B3GALT6,DVL1,ATAD3A,TMEM240...  \n",
       "52                                          GJA5,GJA8  \n",
       "17                   SDCCAG8,AKT3,ZBTB18,COX20,HNRNPU  \n",
       "51                 ANXA11,MAT1A,CDHR1,RGR,LDB3,BMPR1A  \n",
       "63                                           PAX6,WT1  "
      ]
     },
     "execution_count": 80,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "hi_region.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "metadata": {},
   "outputs": [],
   "source": [
    "hi_region.rename(columns={\n",
    "    'chrom': '#chrom',\n",
    "    '#ISCA ID': 'isca_id', 'ISCA Region Name': 'name'\n",
    "}).to_csv('data/hi-region-20200109.sorted.bed', sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "metadata": {},
   "outputs": [],
   "source": [
    "!bgzip -cf data/hi-region-20200109.sorted.bed > ../data/hi-region-20200109.sorted.bed.gz"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "metadata": {},
   "outputs": [],
   "source": [
    "!tabix -fp bed ../data/hi-region-20200109.sorted.bed.gz"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "metadata": {},
   "outputs": [],
   "source": [
    "uhi_region = region_pos.loc[\n",
    "    region_pos['Haploinsufficiency Score'] == '40',\n",
    "    ['chrom', 'start', 'end', '#ISCA ID', 'ISCA Region Name']\n",
    "].sort_values(['chrom', 'start', 'end'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>#ISCA ID</th>\n",
       "      <th>ISCA Region Name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>chr1</td>\n",
       "      <td>148867551</td>\n",
       "      <td>149768855</td>\n",
       "      <td>ISCA-37469</td>\n",
       "      <td>1q21.2 region (polymorphic region)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>chr14</td>\n",
       "      <td>22111109</td>\n",
       "      <td>23021097</td>\n",
       "      <td>ISCA-37476</td>\n",
       "      <td>14q11.2 region (TCRA region)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>chr14</td>\n",
       "      <td>106050000</td>\n",
       "      <td>107289540</td>\n",
       "      <td>ISCA-37477</td>\n",
       "      <td>1 copy: 14q telomere; 3 copies: 14q telomere</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>chr16</td>\n",
       "      <td>34202088</td>\n",
       "      <td>35147508</td>\n",
       "      <td>ISCA-37481</td>\n",
       "      <td>3 copies: 16p centromere</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>chr19</td>\n",
       "      <td>43242796</td>\n",
       "      <td>43741310</td>\n",
       "      <td>ISCA-37483</td>\n",
       "      <td>19q13.3 region (PSG gene cluster)</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    chrom      start        end    #ISCA ID  \\\n",
       "29   chr1  148867551  149768855  ISCA-37469   \n",
       "25  chr14   22111109   23021097  ISCA-37476   \n",
       "24  chr14  106050000  107289540  ISCA-37477   \n",
       "21  chr16   34202088   35147508  ISCA-37481   \n",
       "20  chr19   43242796   43741310  ISCA-37483   \n",
       "\n",
       "                                ISCA Region Name  \n",
       "29            1q21.2 region (polymorphic region)  \n",
       "25                  14q11.2 region (TCRA region)  \n",
       "24  1 copy: 14q telomere; 3 copies: 14q telomere  \n",
       "21                      3 copies: 16p centromere  \n",
       "20             19q13.3 region (PSG gene cluster)  "
      ]
     },
     "execution_count": 85,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "uhi_region.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "metadata": {},
   "outputs": [],
   "source": [
    "uhi_region['genes'] = uhi_region.apply(\n",
    "    lambda row: ','.join(gene.loc[\n",
    "        (gene['chrom'] == row['chrom']) & (gene['txStart'] >= row['start']) & (gene['txEnd'] <= row['end']), 'name2'\n",
    "    ]),\n",
    "    axis=1\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 87,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>#ISCA ID</th>\n",
       "      <th>ISCA Region Name</th>\n",
       "      <th>genes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>chr1</td>\n",
       "      <td>148867551</td>\n",
       "      <td>149768855</td>\n",
       "      <td>ISCA-37469</td>\n",
       "      <td>1q21.2 region (polymorphic region)</td>\n",
       "      <td>FAM72C,PPIAL4C,FCGR1A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>chr14</td>\n",
       "      <td>22111109</td>\n",
       "      <td>23021097</td>\n",
       "      <td>ISCA-37476</td>\n",
       "      <td>14q11.2 region (TCRA region)</td>\n",
       "      <td>OR4E2,OR4E1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>chr14</td>\n",
       "      <td>106050000</td>\n",
       "      <td>107289540</td>\n",
       "      <td>ISCA-37477</td>\n",
       "      <td>1 copy: 14q telomere; 3 copies: 14q telomere</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>chr16</td>\n",
       "      <td>34202088</td>\n",
       "      <td>35147508</td>\n",
       "      <td>ISCA-37481</td>\n",
       "      <td>3 copies: 16p centromere</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>chr19</td>\n",
       "      <td>43242796</td>\n",
       "      <td>43741310</td>\n",
       "      <td>ISCA-37483</td>\n",
       "      <td>19q13.3 region (PSG gene cluster)</td>\n",
       "      <td>PSG8,PSG1,PSG6,PSG7,PSG11,PSG2,PSG5,PSG4</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    chrom      start        end    #ISCA ID  \\\n",
       "29   chr1  148867551  149768855  ISCA-37469   \n",
       "25  chr14   22111109   23021097  ISCA-37476   \n",
       "24  chr14  106050000  107289540  ISCA-37477   \n",
       "21  chr16   34202088   35147508  ISCA-37481   \n",
       "20  chr19   43242796   43741310  ISCA-37483   \n",
       "\n",
       "                                ISCA Region Name  \\\n",
       "29            1q21.2 region (polymorphic region)   \n",
       "25                  14q11.2 region (TCRA region)   \n",
       "24  1 copy: 14q telomere; 3 copies: 14q telomere   \n",
       "21                      3 copies: 16p centromere   \n",
       "20             19q13.3 region (PSG gene cluster)   \n",
       "\n",
       "                                       genes  \n",
       "29                     FAM72C,PPIAL4C,FCGR1A  \n",
       "25                               OR4E2,OR4E1  \n",
       "24                                            \n",
       "21                                            \n",
       "20  PSG8,PSG1,PSG6,PSG7,PSG11,PSG2,PSG5,PSG4  "
      ]
     },
     "execution_count": 87,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "uhi_region.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 88,
   "metadata": {},
   "outputs": [],
   "source": [
    "uhi_region.rename(columns={\n",
    "    'chrom': '#chrom',\n",
    "    '#ISCA ID': 'isca_id', 'ISCA Region Name': 'name'\n",
    "}).to_csv('data/uhi-region-20200109.sorted.bed', sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 89,
   "metadata": {},
   "outputs": [],
   "source": [
    "!bgzip -cf data/uhi-region-20200109.sorted.bed > ../data/uhi-region-20200109.sorted.bed.gz"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 90,
   "metadata": {},
   "outputs": [],
   "source": [
    "!tabix -fp bed ../data/uhi-region-20200109.sorted.bed.gz"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 91,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open('data/gene-table.json') as f:\n",
    "    data = json.load(f)\n",
    "    decipher = pd.DataFrame.from_records(data, columns=['name', 'hi_score', 'p_li']).rename(columns={'name': 'symbol'})\n",
    "    decipher = decipher.merge(gene, left_on='symbol', right_on='name2')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 92,
   "metadata": {},
   "outputs": [],
   "source": [
    "decipher = decipher.loc[\n",
    "    (decipher['p_li'] >= 0.9) & (decipher['hi_score'] < 0.1),\n",
    "    ['chrom', 'txStart', 'txEnd', 'GeneID', 'name2', 'name', 'p_li', 'hi_score']\n",
    "].sort_values(['chrom', 'txStart', 'txEnd'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 93,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>txStart</th>\n",
       "      <th>txEnd</th>\n",
       "      <th>GeneID</th>\n",
       "      <th>name2</th>\n",
       "      <th>name</th>\n",
       "      <th>p_li</th>\n",
       "      <th>hi_score</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2112</th>\n",
       "      <td>chr1</td>\n",
       "      <td>2439969</td>\n",
       "      <td>2458067</td>\n",
       "      <td>55229</td>\n",
       "      <td>PANK4</td>\n",
       "      <td>NM_018216.2</td>\n",
       "      <td>0.945333</td>\n",
       "      <td>0.098966</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2587</th>\n",
       "      <td>chr1</td>\n",
       "      <td>155290639</td>\n",
       "      <td>155300909</td>\n",
       "      <td>23623</td>\n",
       "      <td>RUSC1</td>\n",
       "      <td>NM_001105204.1</td>\n",
       "      <td>0.949429</td>\n",
       "      <td>0.035285</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2705</th>\n",
       "      <td>chr1</td>\n",
       "      <td>232533710</td>\n",
       "      <td>232651354</td>\n",
       "      <td>57568</td>\n",
       "      <td>SIPA1L2</td>\n",
       "      <td>NM_020808.4</td>\n",
       "      <td>0.999220</td>\n",
       "      <td>0.079950</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1608</th>\n",
       "      <td>chr10</td>\n",
       "      <td>852853</td>\n",
       "      <td>977645</td>\n",
       "      <td>23185</td>\n",
       "      <td>LARP4B</td>\n",
       "      <td>NM_015155.2</td>\n",
       "      <td>0.999886</td>\n",
       "      <td>0.071477</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>199</th>\n",
       "      <td>chr10</td>\n",
       "      <td>24872537</td>\n",
       "      <td>25012597</td>\n",
       "      <td>57584</td>\n",
       "      <td>ARHGAP21</td>\n",
       "      <td>NM_020824.3</td>\n",
       "      <td>0.999947</td>\n",
       "      <td>0.073197</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      chrom    txStart      txEnd  GeneID     name2            name      p_li  \\\n",
       "2112   chr1    2439969    2458067   55229     PANK4     NM_018216.2  0.945333   \n",
       "2587   chr1  155290639  155300909   23623     RUSC1  NM_001105204.1  0.949429   \n",
       "2705   chr1  232533710  232651354   57568   SIPA1L2     NM_020808.4  0.999220   \n",
       "1608  chr10     852853     977645   23185    LARP4B     NM_015155.2  0.999886   \n",
       "199   chr10   24872537   25012597   57584  ARHGAP21     NM_020824.3  0.999947   \n",
       "\n",
       "      hi_score  \n",
       "2112  0.098966  \n",
       "2587  0.035285  \n",
       "2705  0.079950  \n",
       "1608  0.071477  \n",
       "199   0.073197  "
      ]
     },
     "execution_count": 93,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "decipher.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 94,
   "metadata": {},
   "outputs": [],
   "source": [
    "decipher.rename(columns={\n",
    "    'chrom': '#chrom', 'txStart': 'start', 'txEnd': 'end',\n",
    "    'GeneID': 'gene_id', 'name2': 'symbol', 'name': 'transcript'\n",
    "}).to_csv('data/decipher-gene-20200109.sorted.bed', sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 95,
   "metadata": {},
   "outputs": [],
   "source": [
    "!bgzip -cf data/decipher-gene-20200109.sorted.bed > ../data/decipher-gene-20200109.sorted.bed.gz"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 96,
   "metadata": {},
   "outputs": [],
   "source": [
    "!tabix -fp bed ../data/decipher-gene-20200109.sorted.bed.gz"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 97,
   "metadata": {},
   "outputs": [],
   "source": [
    "ts_genes = set(\n",
    "    curation_gene.loc[\n",
    "        curation_gene['Triplosensitivity Score'] == '3', '#Gene Symbol'\n",
    "    ]\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 98,
   "metadata": {},
   "outputs": [],
   "source": [
    "ts_gene = gene[gene['name2'].isin(ts_genes)].copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 99,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>txStart</th>\n",
       "      <th>txEnd</th>\n",
       "      <th>GeneID</th>\n",
       "      <th>name2</th>\n",
       "      <th>name</th>\n",
       "      <th>strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>5423</th>\n",
       "      <td>chr5</td>\n",
       "      <td>126112314</td>\n",
       "      <td>126172712</td>\n",
       "      <td>4001</td>\n",
       "      <td>LMNB1</td>\n",
       "      <td>NR_134488.1</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12962</th>\n",
       "      <td>chrX</td>\n",
       "      <td>103031433</td>\n",
       "      <td>103047548</td>\n",
       "      <td>5354</td>\n",
       "      <td>PLP1</td>\n",
       "      <td>NM_001128834.2</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4411</th>\n",
       "      <td>chrX</td>\n",
       "      <td>153287263</td>\n",
       "      <td>153363188</td>\n",
       "      <td>4204</td>\n",
       "      <td>MECP2</td>\n",
       "      <td>NM_004992.3</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      chrom    txStart      txEnd  GeneID  name2            name strand\n",
       "5423   chr5  126112314  126172712    4001  LMNB1     NR_134488.1      +\n",
       "12962  chrX  103031433  103047548    5354   PLP1  NM_001128834.2      +\n",
       "4411   chrX  153287263  153363188    4204  MECP2     NM_004992.3      -"
      ]
     },
     "execution_count": 99,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ts_gene.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 100,
   "metadata": {},
   "outputs": [],
   "source": [
    "ts_gene.rename(columns={\n",
    "    'chrom': '#chrom', 'txStart': 'start', 'txEnd': 'end',\n",
    "    'GeneID': 'gene_id', 'name2': 'symbol', 'name': 'transcript'\n",
    "}).to_csv('data/ts-gene-20200109.sorted.bed', index=False, sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 101,
   "metadata": {},
   "outputs": [],
   "source": [
    "!bgzip -cf data/ts-gene-20200109.sorted.bed > ../data/ts-gene-20200109.sorted.bed.gz"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 102,
   "metadata": {},
   "outputs": [],
   "source": [
    "!tabix -fp bed ../data/ts-gene-20200109.sorted.bed.gz"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 103,
   "metadata": {},
   "outputs": [],
   "source": [
    "uts_genes = set(\n",
    "    curation_gene.loc[curation_gene['Triplosensitivity Score'] == '40', '#Gene Symbol']\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 104,
   "metadata": {},
   "outputs": [],
   "source": [
    "uts_gene = gene[gene['name2'].isin(uts_genes)].copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 105,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>txStart</th>\n",
       "      <th>txEnd</th>\n",
       "      <th>GeneID</th>\n",
       "      <th>name2</th>\n",
       "      <th>name</th>\n",
       "      <th>strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>21975</th>\n",
       "      <td>chrX</td>\n",
       "      <td>6451658</td>\n",
       "      <td>6453159</td>\n",
       "      <td>51481</td>\n",
       "      <td>VCX3A</td>\n",
       "      <td>NM_016379.3</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      chrom  txStart    txEnd  GeneID  name2         name strand\n",
       "21975  chrX  6451658  6453159   51481  VCX3A  NM_016379.3      -"
      ]
     },
     "execution_count": 105,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "uts_gene.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 106,
   "metadata": {},
   "outputs": [],
   "source": [
    "uts_gene.rename(columns={\n",
    "    'chrom': '#chrom', 'txStart': 'start', 'txEnd': 'end',\n",
    "    'GeneID': 'gene_id', 'name2': 'symbol', 'name': 'transcript'\n",
    "}).to_csv('data/uts-gene-20200109.sorted.bed', index=False, sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 107,
   "metadata": {},
   "outputs": [],
   "source": [
    "!bgzip -cf data/uts-gene-20200109.sorted.bed > ../data/uts-gene-20200109.sorted.bed.gz"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 108,
   "metadata": {},
   "outputs": [],
   "source": [
    "!tabix -fp bed ../data/uts-gene-20200109.sorted.bed.gz"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 109,
   "metadata": {},
   "outputs": [],
   "source": [
    "ts_region = region_pos.loc[\n",
    "    region_pos['Triplosensitivity Score'] == '3',\n",
    "    ['chrom', 'start', 'end', '#ISCA ID', 'ISCA Region Name']\n",
    "].sort_values(['chrom', 'start', 'end'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 110,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>#ISCA ID</th>\n",
       "      <th>ISCA Region Name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>52</th>\n",
       "      <td>chr1</td>\n",
       "      <td>146577486</td>\n",
       "      <td>147394506</td>\n",
       "      <td>ISCA-37421</td>\n",
       "      <td>1q21.1 recurrent region (BP3-BP4, distal) (inc...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>62</th>\n",
       "      <td>chr15</td>\n",
       "      <td>22832519</td>\n",
       "      <td>28379874</td>\n",
       "      <td>ISCA-37404</td>\n",
       "      <td>15q11q13 recurrent (PWS/AS) region (BP1-BP3, C...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>chr15</td>\n",
       "      <td>23747996</td>\n",
       "      <td>28379874</td>\n",
       "      <td>ISCA-37478</td>\n",
       "      <td>15q11q13 recurrent (PWS/AS) region (BP2-BP3, C...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>64</th>\n",
       "      <td>chr16</td>\n",
       "      <td>29649997</td>\n",
       "      <td>30199852</td>\n",
       "      <td>ISCA-37400</td>\n",
       "      <td>16p11.2 recurrent region (proximal, BP4-BP5) (...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>47</th>\n",
       "      <td>chr17</td>\n",
       "      <td>1247833</td>\n",
       "      <td>2588909</td>\n",
       "      <td>ISCA-37430</td>\n",
       "      <td>17p13.3 (Miller-Dieker syndrome) region (inclu...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    chrom      start        end    #ISCA ID  \\\n",
       "52   chr1  146577486  147394506  ISCA-37421   \n",
       "62  chr15   22832519   28379874  ISCA-37404   \n",
       "23  chr15   23747996   28379874  ISCA-37478   \n",
       "64  chr16   29649997   30199852  ISCA-37400   \n",
       "47  chr17    1247833    2588909  ISCA-37430   \n",
       "\n",
       "                                     ISCA Region Name  \n",
       "52  1q21.1 recurrent region (BP3-BP4, distal) (inc...  \n",
       "62  15q11q13 recurrent (PWS/AS) region (BP1-BP3, C...  \n",
       "23  15q11q13 recurrent (PWS/AS) region (BP2-BP3, C...  \n",
       "64  16p11.2 recurrent region (proximal, BP4-BP5) (...  \n",
       "47  17p13.3 (Miller-Dieker syndrome) region (inclu...  "
      ]
     },
     "execution_count": 110,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ts_region.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 111,
   "metadata": {},
   "outputs": [],
   "source": [
    "ts_region['omim_genes'] = ts_region.apply(\n",
    "    lambda row: ','.join(omim_gene.loc[\n",
    "        (omim_gene['chrom'] == row['chrom']) & (omim_gene['txStart'] >= row['start']) & (omim_gene['txEnd'] <= row['end']), 'name2'\n",
    "    ]),\n",
    "    axis=1\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 112,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>#ISCA ID</th>\n",
       "      <th>ISCA Region Name</th>\n",
       "      <th>omim_genes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>52</th>\n",
       "      <td>chr1</td>\n",
       "      <td>146577486</td>\n",
       "      <td>147394506</td>\n",
       "      <td>ISCA-37421</td>\n",
       "      <td>1q21.1 recurrent region (BP3-BP4, distal) (inc...</td>\n",
       "      <td>GJA5,GJA8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>62</th>\n",
       "      <td>chr15</td>\n",
       "      <td>22832519</td>\n",
       "      <td>28379874</td>\n",
       "      <td>ISCA-37404</td>\n",
       "      <td>15q11q13 recurrent (PWS/AS) region (BP1-BP3, C...</td>\n",
       "      <td>NIPA1,MKRN3,MAGEL2,NDN,SNRPN,UBE3A,GABRB3,GABR...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>chr15</td>\n",
       "      <td>23747996</td>\n",
       "      <td>28379874</td>\n",
       "      <td>ISCA-37478</td>\n",
       "      <td>15q11q13 recurrent (PWS/AS) region (BP2-BP3, C...</td>\n",
       "      <td>MKRN3,MAGEL2,NDN,SNRPN,UBE3A,GABRB3,GABRA5,OCA2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>64</th>\n",
       "      <td>chr16</td>\n",
       "      <td>29649997</td>\n",
       "      <td>30199852</td>\n",
       "      <td>ISCA-37400</td>\n",
       "      <td>16p11.2 recurrent region (proximal, BP4-BP5) (...</td>\n",
       "      <td>KIF22,PRRT2,ALDOA,TBX6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>47</th>\n",
       "      <td>chr17</td>\n",
       "      <td>1247833</td>\n",
       "      <td>2588909</td>\n",
       "      <td>ISCA-37430</td>\n",
       "      <td>17p13.3 (Miller-Dieker syndrome) region (inclu...</td>\n",
       "      <td>INPP5K,PRPF8,WDR81,SERPINF2,SERPINF1,DPH1,PAFA...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    chrom      start        end    #ISCA ID  \\\n",
       "52   chr1  146577486  147394506  ISCA-37421   \n",
       "62  chr15   22832519   28379874  ISCA-37404   \n",
       "23  chr15   23747996   28379874  ISCA-37478   \n",
       "64  chr16   29649997   30199852  ISCA-37400   \n",
       "47  chr17    1247833    2588909  ISCA-37430   \n",
       "\n",
       "                                     ISCA Region Name  \\\n",
       "52  1q21.1 recurrent region (BP3-BP4, distal) (inc...   \n",
       "62  15q11q13 recurrent (PWS/AS) region (BP1-BP3, C...   \n",
       "23  15q11q13 recurrent (PWS/AS) region (BP2-BP3, C...   \n",
       "64  16p11.2 recurrent region (proximal, BP4-BP5) (...   \n",
       "47  17p13.3 (Miller-Dieker syndrome) region (inclu...   \n",
       "\n",
       "                                           omim_genes  \n",
       "52                                          GJA5,GJA8  \n",
       "62  NIPA1,MKRN3,MAGEL2,NDN,SNRPN,UBE3A,GABRB3,GABR...  \n",
       "23    MKRN3,MAGEL2,NDN,SNRPN,UBE3A,GABRB3,GABRA5,OCA2  \n",
       "64                             KIF22,PRRT2,ALDOA,TBX6  \n",
       "47  INPP5K,PRPF8,WDR81,SERPINF2,SERPINF1,DPH1,PAFA...  "
      ]
     },
     "execution_count": 112,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ts_region.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 113,
   "metadata": {},
   "outputs": [],
   "source": [
    "ts_region.rename(columns={\n",
    "    'chrom': '#chrom',\n",
    "    '#ISCA ID': 'isca_id', 'ISCA Region Name': 'name'\n",
    "}).to_csv('data/ts-region-20200109.sorted.bed', sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 114,
   "metadata": {},
   "outputs": [],
   "source": [
    "!bgzip -cf data/ts-region-20200109.sorted.bed > ../data/ts-region-20200109.sorted.bed.gz"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 115,
   "metadata": {},
   "outputs": [],
   "source": [
    "!tabix -fp bed ../data/ts-region-20200109.sorted.bed.gz"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 116,
   "metadata": {},
   "outputs": [],
   "source": [
    "uts_region = region_pos.loc[\n",
    "    region_pos['Triplosensitivity Score'] == '40',\n",
    "    ['chrom', 'start', 'end', '#ISCA ID', 'ISCA Region Name']\n",
    "].sort_values(['chrom', 'start', 'end'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 117,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>#ISCA ID</th>\n",
       "      <th>ISCA Region Name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>chr1</td>\n",
       "      <td>148867551</td>\n",
       "      <td>149768855</td>\n",
       "      <td>ISCA-37469</td>\n",
       "      <td>1q21.2 region (polymorphic region)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>chr14</td>\n",
       "      <td>22111109</td>\n",
       "      <td>23021097</td>\n",
       "      <td>ISCA-37476</td>\n",
       "      <td>14q11.2 region (TCRA region)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>chr14</td>\n",
       "      <td>106050000</td>\n",
       "      <td>107289540</td>\n",
       "      <td>ISCA-37477</td>\n",
       "      <td>1 copy: 14q telomere; 3 copies: 14q telomere</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>chr15</td>\n",
       "      <td>22832519</td>\n",
       "      <td>23090897</td>\n",
       "      <td>ISCA-37448</td>\n",
       "      <td>15q11.2 recurrent region (BP1-BP2) (includes N...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>chr15</td>\n",
       "      <td>32019621</td>\n",
       "      <td>32445405</td>\n",
       "      <td>ISCA-46295</td>\n",
       "      <td>15q13.3 recurrent region (D-CHRNA7 to BP5) (in...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    chrom      start        end    #ISCA ID  \\\n",
       "29   chr1  148867551  149768855  ISCA-37469   \n",
       "25  chr14   22111109   23021097  ISCA-37476   \n",
       "24  chr14  106050000  107289540  ISCA-37477   \n",
       "33  chr15   22832519   23090897  ISCA-37448   \n",
       "5   chr15   32019621   32445405  ISCA-46295   \n",
       "\n",
       "                                     ISCA Region Name  \n",
       "29                 1q21.2 region (polymorphic region)  \n",
       "25                       14q11.2 region (TCRA region)  \n",
       "24       1 copy: 14q telomere; 3 copies: 14q telomere  \n",
       "33  15q11.2 recurrent region (BP1-BP2) (includes N...  \n",
       "5   15q13.3 recurrent region (D-CHRNA7 to BP5) (in...  "
      ]
     },
     "execution_count": 117,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "uts_region.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 118,
   "metadata": {},
   "outputs": [],
   "source": [
    "uts_region['genes'] = uts_region.apply(\n",
    "    lambda row: ','.join(gene.loc[\n",
    "        (gene['chrom'] == row['chrom']) & (gene['txStart'] >= row['start']) & (gene['txEnd'] <= row['end']), 'name2'\n",
    "    ]),\n",
    "    axis=1\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 119,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>#ISCA ID</th>\n",
       "      <th>ISCA Region Name</th>\n",
       "      <th>genes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>chr1</td>\n",
       "      <td>148867551</td>\n",
       "      <td>149768855</td>\n",
       "      <td>ISCA-37469</td>\n",
       "      <td>1q21.2 region (polymorphic region)</td>\n",
       "      <td>FAM72C,PPIAL4C,FCGR1A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>chr14</td>\n",
       "      <td>22111109</td>\n",
       "      <td>23021097</td>\n",
       "      <td>ISCA-37476</td>\n",
       "      <td>14q11.2 region (TCRA region)</td>\n",
       "      <td>OR4E2,OR4E1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>chr14</td>\n",
       "      <td>106050000</td>\n",
       "      <td>107289540</td>\n",
       "      <td>ISCA-37477</td>\n",
       "      <td>1 copy: 14q telomere; 3 copies: 14q telomere</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>chr15</td>\n",
       "      <td>22832519</td>\n",
       "      <td>23090897</td>\n",
       "      <td>ISCA-37448</td>\n",
       "      <td>15q11.2 recurrent region (BP1-BP2) (includes N...</td>\n",
       "      <td>TUBGCP5,CYFIP1,NIPA2,NIPA1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>chr15</td>\n",
       "      <td>32019621</td>\n",
       "      <td>32445405</td>\n",
       "      <td>ISCA-46295</td>\n",
       "      <td>15q13.3 recurrent region (D-CHRNA7 to BP5) (in...</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    chrom      start        end    #ISCA ID  \\\n",
       "29   chr1  148867551  149768855  ISCA-37469   \n",
       "25  chr14   22111109   23021097  ISCA-37476   \n",
       "24  chr14  106050000  107289540  ISCA-37477   \n",
       "33  chr15   22832519   23090897  ISCA-37448   \n",
       "5   chr15   32019621   32445405  ISCA-46295   \n",
       "\n",
       "                                     ISCA Region Name  \\\n",
       "29                 1q21.2 region (polymorphic region)   \n",
       "25                       14q11.2 region (TCRA region)   \n",
       "24       1 copy: 14q telomere; 3 copies: 14q telomere   \n",
       "33  15q11.2 recurrent region (BP1-BP2) (includes N...   \n",
       "5   15q13.3 recurrent region (D-CHRNA7 to BP5) (in...   \n",
       "\n",
       "                         genes  \n",
       "29       FAM72C,PPIAL4C,FCGR1A  \n",
       "25                 OR4E2,OR4E1  \n",
       "24                              \n",
       "33  TUBGCP5,CYFIP1,NIPA2,NIPA1  \n",
       "5                               "
      ]
     },
     "execution_count": 119,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "uts_region.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 120,
   "metadata": {},
   "outputs": [],
   "source": [
    "uts_region.rename(columns={\n",
    "    'chrom': '#chrom',\n",
    "    '#ISCA ID': 'isca_id', 'ISCA Region Name': 'name'\n",
    "}).to_csv('data/uts-region-20200109.sorted.bed', sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 121,
   "metadata": {},
   "outputs": [],
   "source": [
    "!bgzip -cf data/uts-region-20200109.sorted.bed > ../data/uts-region-20200109.sorted.bed.gz"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 122,
   "metadata": {},
   "outputs": [],
   "source": [
    "!tabix -fp bed ../data/uts-region-20200109.sorted.bed.gz"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "jupyter",
   "language": "python",
   "name": "jupyter"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
