{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "from multiprocessing import Pool\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import allel\n",
    "import pysam\n",
    "import json\n",
    "import os\n",
    "from datetime import datetime\n",
    "pd.set_option('display.max_columns', None)\n",
    "from urllib import request\n",
    "from tqdm import tqdm\n",
    "from autocnv import settings\n",
    "from gtfparse import read_gtf\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "work_dir = settings.BASE_DIR\n",
    "raw_data_dir = os.path.join(work_dir, 'raw_data')\n",
    "result_data_dir = os.path.join(work_dir, 'data')\n",
    "if not os.path.exists(raw_data_dir):\n",
    "    os.mkdir(raw_data_dir)\n",
    "if not os.path.exists(result_data_dir):\n",
    "    os.mkdir(result_data_dir)\n",
    "\n",
    "# version_date = datetime.today().strftime('%Y-%m-%d')\n",
    "version_date = '2021-10-28'\n",
    "version_date_dir = os.path.join(raw_data_dir, version_date)\n",
    "if not os.path.exists(version_date_dir):\n",
    "    os.mkdir(version_date_dir)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "pycharm": {
     "name": "#%%\n"
    }
   },
   "outputs": [],
   "source": [
    "exon_url = 'https://hgdownload.soe.ucsc.edu/goldenPath/hg19/bigZips/genes/hg19.refGene.gtf.gz'\n",
    "cyto_band_url = 'https://hgdownload.soe.ucsc.edu/goldenPath/hg19/database/cytoBand.txt.gz'\n",
    "gene_info_url = 'https://ftp.ncbi.nih.gov/refseq/H_sapiens/Homo_sapiens.gene_info.gz'\n",
    "ref_gene_url = 'https://hgdownload.soe.ucsc.edu/goldenPath/hg19/database/refGene.txt.gz'\n",
    "clingen_gene_curation_url = 'https://ftp.clinicalgenome.org/ClinGen_gene_curation_list_GRCh37.tsv'\n",
    "clingene_region_curation_url = 'https://ftp.clinicalgenome.org/ClinGen_region_curation_list_GRCh37.tsv'\n",
    "clinvar_url = 'https://ftp.ncbi.nlm.nih.gov/pub/clinvar/vcf_GRCh37/clinvar.vcf.gz'\n",
    "hi_pred_url = 'https://decipher.sanger.ac.uk/files/downloads/HI_Predictions_Version3.bed.gz'\n",
    "gnomad_lof_url = 'https://datasetgnomad.blob.core.windows.net/dataset/release/2.1.1/constraint/gnomad.v2.1.1.lof_metrics.by_gene.txt.bgz'\n",
    "gnomad_control_only_url = 'https://datasetgnomad.blob.core.windows.net/dataset/papers/2019-sv/gnomad_v2.1_sv.controls_only.sites.bed.gz'\n",
    "hgnc_gene_fam_url = 'ftp://ftp.ebi.ac.uk/pub/databases/genenames/hgnc/csv/genefamily_db_tables/family.csv'\n",
    "# ori file\n",
    "exon_ori_file = os.path.join(version_date_dir, 'hg19.refGene.gtf.gz ')\n",
    "cyto_band_ori_file = os.path.join(version_date_dir, 'cytoBand.txt.gz')\n",
    "gene_info_ori_file = os.path.join(version_date_dir, 'Homo_sapiens.gene_info.gz')\n",
    "ref_gene_ori_file = os.path.join(version_date_dir, 'refGene.txt.gz')\n",
    "clingen_gene_ori_file = os.path.join(version_date_dir, 'clingen_gene_hg19.tsv')\n",
    "clingen_region_ori_file = os.path.join(version_date_dir, 'clingen_region_hg19.tsv')\n",
    "clinvar_ori_vcf_file = os.path.join(version_date_dir, 'clinvar.vcf.gz')\n",
    "hi_pred_ori_file = os.path.join(version_date_dir, 'HI_Predictions_Version3.bed.gz')\n",
    "gnomad_lof_ori_file = os.path.join(version_date_dir, 'gnomad.v2.1.1.lof_metrics.by_gene.txt.bgz')\n",
    "gnomad_control_ori_file = os.path.join(version_date_dir, 'gnomad_v2.1_sv.controls_only.sites.bed.gz')\n",
    "# prep an external omim gene list\n",
    "omim_gene_list_file = os.path.join(raw_data_dir, 'gene-list-key-lte3.xlsx')\n",
    "dgv_ori_file = os.path.join(raw_data_dir, 'DGV.GS.March2016.50percent.GainLossSep.Final.hg19.gff3')\n",
    "# result file\n",
    "\n",
    "gene_file = os.path.join(result_data_dir, 'gene.sorted.bed')\n",
    "omim_gene_file = os.path.join(result_data_dir, 'omim-gene.sorted.bed')\n",
    "clinvar_file = os.path.join(result_data_dir, 'clinvar-pathogenic.sorted.vcf')\n",
    "decipher_gene_file = os.path.join(result_data_dir, 'decipher-gene.sorted.bed')\n",
    "dgv_gain_file = os.path.join(result_data_dir, 'dgv-gain.sorted.bed')\n",
    "dgv_loss_file = os.path.join(result_data_dir, 'dgv-loss.sorted.bed')\n",
    "func_region_file = os.path.join(result_data_dir, 'func-region.sorted')\n",
    "gnomad_del_file = os.path.join(result_data_dir, 'gnomad-del.sorted.bed')\n",
    "gnomad_dup_file = os.path.join(result_data_dir, 'gnomad-dup.sorted.bed')\n",
    "hi_cds_file = os.path.join(result_data_dir, 'hi-cds.sorted.bed')\n",
    "hi_exon_file = os.path.join(result_data_dir, 'hi-exon.sorted.bed')\n",
    "hi_gene_file = os.path.join(result_data_dir, 'hi-gene.sorted.bed')\n",
    "hi_region_file = os.path.join(result_data_dir, 'hi-region.sorted.bed')\n",
    "ts_gene_file = os.path.join(result_data_dir, 'ts-gene.sorted.bed')\n",
    "ts_region_file = os.path.join(result_data_dir, 'ts-region.sorted.bed')\n",
    "uhi_gene_file = os.path.join(result_data_dir, 'uhi-gene.sorted.bed')\n",
    "uhi_region_file = os.path.join(result_data_dir, 'uhi-region.sorted.bed')\n",
    "uts_gene_file = os.path.join(result_data_dir, 'uts-gene.sorted.bed')\n",
    "uts_region_file = os.path.join(result_data_dir, 'uts-region.sorted.bed')\n",
    "hgnc_gene_fam_file = os.path.join(version_date_dir, 'family.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "pycharm": {
     "name": "#%% Download required data\n"
    },
    "tags": []
   },
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'os' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001B[0;31m---------------------------------------------------------------------------\u001B[0m",
      "\u001B[0;31mNameError\u001B[0m                                 Traceback (most recent call last)",
      "\u001B[0;32m<ipython-input-1-1aace11afd31>\u001B[0m in \u001B[0;36m<module>\u001B[0;34m\u001B[0m\n\u001B[1;32m      1\u001B[0m \u001B[0;31m# Download require data\u001B[0m\u001B[0;34m\u001B[0m\u001B[0;34m\u001B[0m\u001B[0;34m\u001B[0m\u001B[0m\n\u001B[1;32m      2\u001B[0m \u001B[0;34m\u001B[0m\u001B[0m\n\u001B[0;32m----> 3\u001B[0;31m \u001B[0;32mif\u001B[0m \u001B[0;32mnot\u001B[0m \u001B[0mos\u001B[0m\u001B[0;34m.\u001B[0m\u001B[0mpath\u001B[0m\u001B[0;34m.\u001B[0m\u001B[0mexists\u001B[0m\u001B[0;34m(\u001B[0m\u001B[0mexon_ori_file\u001B[0m\u001B[0;34m)\u001B[0m\u001B[0;34m:\u001B[0m\u001B[0;34m\u001B[0m\u001B[0;34m\u001B[0m\u001B[0m\n\u001B[0m\u001B[1;32m      4\u001B[0m     \u001B[0mprint\u001B[0m\u001B[0;34m(\u001B[0m\u001B[0;34mf'downloading gene exon file'\u001B[0m\u001B[0;34m)\u001B[0m\u001B[0;34m\u001B[0m\u001B[0;34m\u001B[0m\u001B[0m\n\u001B[1;32m      5\u001B[0m \u001B[0;34m\u001B[0m\u001B[0m\n",
      "\u001B[0;31mNameError\u001B[0m: name 'os' is not defined"
     ]
    }
   ],
   "source": [
    "# Download require data\n",
    "\n",
    "if not os.path.exists(exon_ori_file):\n",
    "    print(f'downloading gene exon file')\n",
    "    request.urlretrieve(exon_url, exon_ori_file)\n",
    "    print('done')\n",
    "\n",
    "if not os.path.exists(cyto_band_ori_file):\n",
    "    print(f'downloading cytoband conf file to {cyto_band_ori_file}')\n",
    "    request.urlretrieve(cyto_band_url, cyto_band_ori_file)\n",
    "    print('done!')\n",
    "\n",
    "if not os.path.exists(gene_info_ori_file):\n",
    "    print(f'downloading gene info to {gene_info_ori_file}')\n",
    "    request.urlretrieve(gene_info_url, gene_info_ori_file)\n",
    "    print('done!')\n",
    "\n",
    "if not os.path.exists(ref_gene_ori_file):\n",
    "    print(f'downloading ref gene to {ref_gene_ori_file}')\n",
    "    request.urlretrieve(ref_gene_url, ref_gene_ori_file)\n",
    "    print('done!')\n",
    "\n",
    "if not os.path.exists(clingen_gene_ori_file):\n",
    "    print(f'downloading clingen gene list to {clingen_gene_ori_file}')\n",
    "    request.urlretrieve(clingen_gene_curation_url, clingen_gene_ori_file)\n",
    "    print('done!')\n",
    "\n",
    "if not os.path.exists(clingen_region_ori_file):\n",
    "    print(f'downloading clingen region file to {clingen_region_ori_file}')\n",
    "    request.urlretrieve(clingene_region_curation_url, clingen_region_ori_file)\n",
    "    print('done!')\n",
    "\n",
    "if not os.path.exists(clinvar_ori_vcf_file):\n",
    "    print(f'downloading clingen region file to {clinvar_ori_vcf_file}')\n",
    "    request.urlretrieve(clinvar_url, clinvar_ori_vcf_file)\n",
    "    print('done!')\n",
    "\n",
    "if not os.path.exists(hi_pred_ori_file):\n",
    "    print(f'downloading hi prediction file to {hi_pred_ori_file}')\n",
    "    request.urlretrieve(hi_pred_url, hi_pred_ori_file)\n",
    "    print('done!')\n",
    "if not os.path.exists(gnomad_lof_ori_file):\n",
    "    print(f'downloading pLoF file from GnomAD to {gnomad_lof_ori_file}')\n",
    "    request.urlretrieve(gnomad_lof_url, gnomad_lof_ori_file)\n",
    "    print('done!')\n",
    "if not os.path.exists(gnomad_control_ori_file):\n",
    "    print(f'downloading gnomad control only file to {gnomad_control_ori_file}')\n",
    "    request.urlretrieve(gnomad_control_only_url, gnomad_control_ori_file)\n",
    "    print('done!')\n",
    "\n",
    "if not os.path.exists(hgnc_gene_fam_file):\n",
    "    print(f'downloading gene family file to {hgnc_gene_fam_file}')\n",
    "    request.urlretrieve(hgnc_gene_fam_url, hgnc_gene_fam_file)\n",
    "    print('done!')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {
    "pycharm": {
     "name": "#%%\n"
    }
   },
   "outputs": [],
   "source": [
    "cyto_band_df = pd.read_csv(\n",
    "    cyto_band_ori_file, sep='\\t',\n",
    "    names=['#chrom', 'start', 'end', 'name', 'gieStain']\n",
    ")\n",
    "cyto_band_df.to_csv(settings.CYTO_BAND_FILE.strip('.gz'), sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "outputs": [],
   "source": [
    "!bgzip -cf {settings.CYTO_BAND_FILE.strip(\".gz\")} > {settings.CYTO_BAND_FILE}\n",
    "!tabix -fp bed {settings.CYTO_BAND_FILE}\n",
    "!rm {settings.CYTO_BAND_FILE.strip(\".gz\")}\n"
   ],
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%%\n"
    }
   }
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>bin</th>\n",
       "      <th>name</th>\n",
       "      <th>chrom</th>\n",
       "      <th>strand</th>\n",
       "      <th>txStart</th>\n",
       "      <th>txEnd</th>\n",
       "      <th>cdsStart</th>\n",
       "      <th>cdsEnd</th>\n",
       "      <th>exonCount</th>\n",
       "      <th>exonStarts</th>\n",
       "      <th>exonEnds</th>\n",
       "      <th>score</th>\n",
       "      <th>name2</th>\n",
       "      <th>cdsStartStat</th>\n",
       "      <th>cdsEndStat</th>\n",
       "      <th>ExonFrames</th>\n",
       "      <th>length</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>10573</th>\n",
       "      <td>26</td>\n",
       "      <td>NM_001351365</td>\n",
       "      <td>chr1</td>\n",
       "      <td>+</td>\n",
       "      <td>144146811</td>\n",
       "      <td>146467744</td>\n",
       "      <td>144146846</td>\n",
       "      <td>146466121</td>\n",
       "      <td>93</td>\n",
       "      <td>144146811,144148789,144149726,144150981,144151...</td>\n",
       "      <td>144147021,144148892,144149941,144151054,144151...</td>\n",
       "      <td>0</td>\n",
       "      <td>NBPF19</td>\n",
       "      <td>cmpl</td>\n",
       "      <td>cmpl</td>\n",
       "      <td>0,1,2,1,2,1,2,1,2,1,2,1,2,1,2,1,2,1,2,1,2,1,2,...</td>\n",
       "      <td>2319275</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10459</th>\n",
       "      <td>26</td>\n",
       "      <td>NM_001278267</td>\n",
       "      <td>chr1</td>\n",
       "      <td>+</td>\n",
       "      <td>144146810</td>\n",
       "      <td>146467744</td>\n",
       "      <td>144158383</td>\n",
       "      <td>146466121</td>\n",
       "      <td>131</td>\n",
       "      <td>144146810,144148789,144149726,144150981,144151...</td>\n",
       "      <td>144147021,144148892,144149941,144151054,144151...</td>\n",
       "      <td>0</td>\n",
       "      <td>NBPF20</td>\n",
       "      <td>cmpl</td>\n",
       "      <td>cmpl</td>\n",
       "      <td>-1,-1,-1,-1,-1,-1,-1,-1,0,2,1,2,1,2,1,2,1,2,1,...</td>\n",
       "      <td>2307738</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18617</th>\n",
       "      <td>12</td>\n",
       "      <td>NM_000109</td>\n",
       "      <td>chrX</td>\n",
       "      <td>-</td>\n",
       "      <td>31137340</td>\n",
       "      <td>33357505</td>\n",
       "      <td>31140035</td>\n",
       "      <td>33357382</td>\n",
       "      <td>79</td>\n",
       "      <td>31137340,31144758,31152218,31164407,31165391,3...</td>\n",
       "      <td>31140047,31144790,31152311,31164531,31165635,3...</td>\n",
       "      <td>0</td>\n",
       "      <td>DMD</td>\n",
       "      <td>cmpl</td>\n",
       "      <td>cmpl</td>\n",
       "      <td>0,1,1,0,2,2,2,2,2,0,2,0,1,2,1,1,2,1,0,0,1,0,2,...</td>\n",
       "      <td>2217347</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29979</th>\n",
       "      <td>2</td>\n",
       "      <td>NM_001351274</td>\n",
       "      <td>chr11</td>\n",
       "      <td>-</td>\n",
       "      <td>83166055</td>\n",
       "      <td>85339417</td>\n",
       "      <td>83170860</td>\n",
       "      <td>85337661</td>\n",
       "      <td>27</td>\n",
       "      <td>83166055,83173044,83177750,83180243,83182668,8...</td>\n",
       "      <td>83170967,83173136,83177860,83180416,83182770,8...</td>\n",
       "      <td>0</td>\n",
       "      <td>DLG2</td>\n",
       "      <td>cmpl</td>\n",
       "      <td>cmpl</td>\n",
       "      <td>1,2,0,1,1,1,0,2,2,1,0,2,2,2,1,0,1,2,0,0,0,0,0,...</td>\n",
       "      <td>2166801</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>68033</th>\n",
       "      <td>9</td>\n",
       "      <td>NM_033225</td>\n",
       "      <td>chr8</td>\n",
       "      <td>-</td>\n",
       "      <td>2792882</td>\n",
       "      <td>4852436</td>\n",
       "      <td>2796106</td>\n",
       "      <td>4851938</td>\n",
       "      <td>70</td>\n",
       "      <td>2792882,2799993,2806820,2807752,2808635,281174...</td>\n",
       "      <td>2796266,2800126,2806908,2807865,2808797,281179...</td>\n",
       "      <td>0</td>\n",
       "      <td>CSMD1</td>\n",
       "      <td>cmpl</td>\n",
       "      <td>cmpl</td>\n",
       "      <td>2,1,0,1,1,1,1,0,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,...</td>\n",
       "      <td>2055832</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       bin          name  chrom strand    txStart      txEnd   cdsStart  \\\n",
       "10573   26  NM_001351365   chr1      +  144146811  146467744  144146846   \n",
       "10459   26  NM_001278267   chr1      +  144146810  146467744  144158383   \n",
       "18617   12     NM_000109   chrX      -   31137340   33357505   31140035   \n",
       "29979    2  NM_001351274  chr11      -   83166055   85339417   83170860   \n",
       "68033    9     NM_033225   chr8      -    2792882    4852436    2796106   \n",
       "\n",
       "          cdsEnd  exonCount  \\\n",
       "10573  146466121         93   \n",
       "10459  146466121        131   \n",
       "18617   33357382         79   \n",
       "29979   85337661         27   \n",
       "68033    4851938         70   \n",
       "\n",
       "                                              exonStarts  \\\n",
       "10573  144146811,144148789,144149726,144150981,144151...   \n",
       "10459  144146810,144148789,144149726,144150981,144151...   \n",
       "18617  31137340,31144758,31152218,31164407,31165391,3...   \n",
       "29979  83166055,83173044,83177750,83180243,83182668,8...   \n",
       "68033  2792882,2799993,2806820,2807752,2808635,281174...   \n",
       "\n",
       "                                                exonEnds  score   name2  \\\n",
       "10573  144147021,144148892,144149941,144151054,144151...      0  NBPF19   \n",
       "10459  144147021,144148892,144149941,144151054,144151...      0  NBPF20   \n",
       "18617  31140047,31144790,31152311,31164531,31165635,3...      0     DMD   \n",
       "29979  83170967,83173136,83177860,83180416,83182770,8...      0    DLG2   \n",
       "68033  2796266,2800126,2806908,2807865,2808797,281179...      0   CSMD1   \n",
       "\n",
       "      cdsStartStat cdsEndStat  \\\n",
       "10573         cmpl       cmpl   \n",
       "10459         cmpl       cmpl   \n",
       "18617         cmpl       cmpl   \n",
       "29979         cmpl       cmpl   \n",
       "68033         cmpl       cmpl   \n",
       "\n",
       "                                              ExonFrames   length  \n",
       "10573  0,1,2,1,2,1,2,1,2,1,2,1,2,1,2,1,2,1,2,1,2,1,2,...  2319275  \n",
       "10459  -1,-1,-1,-1,-1,-1,-1,-1,0,2,1,2,1,2,1,2,1,2,1,...  2307738  \n",
       "18617  0,1,1,0,2,2,2,2,2,0,2,0,1,2,1,1,2,1,0,0,1,0,2,...  2217347  \n",
       "29979  1,2,0,1,1,1,0,2,2,1,0,2,2,2,1,0,1,2,0,0,0,0,0,...  2166801  \n",
       "68033  2,1,0,1,1,1,1,0,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,...  2055832  "
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cols = [\n",
    "    'bin', 'name', 'chrom', 'strand', 'txStart', 'txEnd', 'cdsStart', 'cdsEnd', 'exonCount', 'exonStarts', 'exonEnds',\n",
    "    'score', 'name2', 'cdsStartStat', 'cdsEndStat', 'ExonFrames']\n",
    "refgene = pd.read_csv(ref_gene_ori_file, sep='\\t', names=cols)\n",
    "refgene = refgene[~refgene['chrom'].str.match(r'.*fix$')]\n",
    "refgene['length'] = refgene['cdsEnd'] - refgene['cdsStart']\n",
    "refgene = refgene.sort_values('length', ascending=False)\n",
    "refgene = refgene.drop_duplicates('name2', keep='first')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "refgene.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>#tax_id</th>\n",
       "      <th>GeneID</th>\n",
       "      <th>Symbol</th>\n",
       "      <th>LocusTag</th>\n",
       "      <th>Synonyms</th>\n",
       "      <th>dbXrefs</th>\n",
       "      <th>chromosome</th>\n",
       "      <th>map_location</th>\n",
       "      <th>description</th>\n",
       "      <th>type_of_gene</th>\n",
       "      <th>Symbol_from_nomenclature_authority</th>\n",
       "      <th>Full_name_from_nomenclature_authority</th>\n",
       "      <th>Nomenclature_status</th>\n",
       "      <th>Other_designations</th>\n",
       "      <th>Modification_date</th>\n",
       "      <th>Feature_type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>9606</td>\n",
       "      <td>1</td>\n",
       "      <td>A1BG</td>\n",
       "      <td>-</td>\n",
       "      <td>A1B|ABG|GAB|HYST2477</td>\n",
       "      <td>MIM:138670|HGNC:HGNC:5|Ensembl:ENSG00000121410</td>\n",
       "      <td>19</td>\n",
       "      <td>19q13.43</td>\n",
       "      <td>alpha-1-B glycoprotein</td>\n",
       "      <td>protein-coding</td>\n",
       "      <td>A1BG</td>\n",
       "      <td>alpha-1-B glycoprotein</td>\n",
       "      <td>O</td>\n",
       "      <td>alpha-1B-glycoprotein|HEL-S-163pA|epididymis s...</td>\n",
       "      <td>20210708</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>9606</td>\n",
       "      <td>2</td>\n",
       "      <td>A2M</td>\n",
       "      <td>-</td>\n",
       "      <td>A2MD|CPAMD5|FWP007|S863-7</td>\n",
       "      <td>MIM:103950|HGNC:HGNC:7|Ensembl:ENSG00000175899</td>\n",
       "      <td>12</td>\n",
       "      <td>12p13.31</td>\n",
       "      <td>alpha-2-macroglobulin</td>\n",
       "      <td>protein-coding</td>\n",
       "      <td>A2M</td>\n",
       "      <td>alpha-2-macroglobulin</td>\n",
       "      <td>O</td>\n",
       "      <td>alpha-2-macroglobulin|C3 and PZP-like alpha-2-...</td>\n",
       "      <td>20211009</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>9606</td>\n",
       "      <td>3</td>\n",
       "      <td>A2MP1</td>\n",
       "      <td>-</td>\n",
       "      <td>A2MP</td>\n",
       "      <td>HGNC:HGNC:8|Ensembl:ENSG00000256069</td>\n",
       "      <td>12</td>\n",
       "      <td>12p13.31</td>\n",
       "      <td>alpha-2-macroglobulin pseudogene 1</td>\n",
       "      <td>pseudo</td>\n",
       "      <td>A2MP1</td>\n",
       "      <td>alpha-2-macroglobulin pseudogene 1</td>\n",
       "      <td>O</td>\n",
       "      <td>pregnancy-zone protein pseudogene</td>\n",
       "      <td>20210611</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>9606</td>\n",
       "      <td>9</td>\n",
       "      <td>NAT1</td>\n",
       "      <td>-</td>\n",
       "      <td>AAC1|MNAT|NAT-1|NATI</td>\n",
       "      <td>MIM:108345|HGNC:HGNC:7645|Ensembl:ENSG00000171428</td>\n",
       "      <td>8</td>\n",
       "      <td>8p22</td>\n",
       "      <td>N-acetyltransferase 1</td>\n",
       "      <td>protein-coding</td>\n",
       "      <td>NAT1</td>\n",
       "      <td>N-acetyltransferase 1</td>\n",
       "      <td>O</td>\n",
       "      <td>arylamine N-acetyltransferase 1|N-acetyltransf...</td>\n",
       "      <td>20210926</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>9606</td>\n",
       "      <td>10</td>\n",
       "      <td>NAT2</td>\n",
       "      <td>-</td>\n",
       "      <td>AAC2|NAT-2|PNAT</td>\n",
       "      <td>MIM:612182|HGNC:HGNC:7646|Ensembl:ENSG00000156006</td>\n",
       "      <td>8</td>\n",
       "      <td>8p22</td>\n",
       "      <td>N-acetyltransferase 2</td>\n",
       "      <td>protein-coding</td>\n",
       "      <td>NAT2</td>\n",
       "      <td>N-acetyltransferase 2</td>\n",
       "      <td>O</td>\n",
       "      <td>arylamine N-acetyltransferase 2|N-acetyltransf...</td>\n",
       "      <td>20210926</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   #tax_id  GeneID Symbol LocusTag                   Synonyms  \\\n",
       "0     9606       1   A1BG        -       A1B|ABG|GAB|HYST2477   \n",
       "1     9606       2    A2M        -  A2MD|CPAMD5|FWP007|S863-7   \n",
       "2     9606       3  A2MP1        -                       A2MP   \n",
       "3     9606       9   NAT1        -       AAC1|MNAT|NAT-1|NATI   \n",
       "4     9606      10   NAT2        -            AAC2|NAT-2|PNAT   \n",
       "\n",
       "                                             dbXrefs chromosome map_location  \\\n",
       "0     MIM:138670|HGNC:HGNC:5|Ensembl:ENSG00000121410         19     19q13.43   \n",
       "1     MIM:103950|HGNC:HGNC:7|Ensembl:ENSG00000175899         12     12p13.31   \n",
       "2                HGNC:HGNC:8|Ensembl:ENSG00000256069         12     12p13.31   \n",
       "3  MIM:108345|HGNC:HGNC:7645|Ensembl:ENSG00000171428          8         8p22   \n",
       "4  MIM:612182|HGNC:HGNC:7646|Ensembl:ENSG00000156006          8         8p22   \n",
       "\n",
       "                          description    type_of_gene  \\\n",
       "0              alpha-1-B glycoprotein  protein-coding   \n",
       "1               alpha-2-macroglobulin  protein-coding   \n",
       "2  alpha-2-macroglobulin pseudogene 1          pseudo   \n",
       "3               N-acetyltransferase 1  protein-coding   \n",
       "4               N-acetyltransferase 2  protein-coding   \n",
       "\n",
       "  Symbol_from_nomenclature_authority Full_name_from_nomenclature_authority  \\\n",
       "0                               A1BG                alpha-1-B glycoprotein   \n",
       "1                                A2M                 alpha-2-macroglobulin   \n",
       "2                              A2MP1    alpha-2-macroglobulin pseudogene 1   \n",
       "3                               NAT1                 N-acetyltransferase 1   \n",
       "4                               NAT2                 N-acetyltransferase 2   \n",
       "\n",
       "  Nomenclature_status                                 Other_designations  \\\n",
       "0                   O  alpha-1B-glycoprotein|HEL-S-163pA|epididymis s...   \n",
       "1                   O  alpha-2-macroglobulin|C3 and PZP-like alpha-2-...   \n",
       "2                   O                  pregnancy-zone protein pseudogene   \n",
       "3                   O  arylamine N-acetyltransferase 1|N-acetyltransf...   \n",
       "4                   O  arylamine N-acetyltransferase 2|N-acetyltransf...   \n",
       "\n",
       "   Modification_date Feature_type  \n",
       "0           20210708            -  \n",
       "1           20211009            -  \n",
       "2           20210611            -  \n",
       "3           20210926            -  \n",
       "4           20210926            -  "
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_info = pd.read_csv(gene_info_ori_file, sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_info.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>bin</th>\n",
       "      <th>name</th>\n",
       "      <th>chrom</th>\n",
       "      <th>strand</th>\n",
       "      <th>txStart</th>\n",
       "      <th>txEnd</th>\n",
       "      <th>cdsStart</th>\n",
       "      <th>cdsEnd</th>\n",
       "      <th>exonCount</th>\n",
       "      <th>exonStarts</th>\n",
       "      <th>exonEnds</th>\n",
       "      <th>score</th>\n",
       "      <th>name2</th>\n",
       "      <th>cdsStartStat</th>\n",
       "      <th>cdsEndStat</th>\n",
       "      <th>ExonFrames</th>\n",
       "      <th>length</th>\n",
       "      <th>#tax_id</th>\n",
       "      <th>GeneID</th>\n",
       "      <th>Symbol</th>\n",
       "      <th>LocusTag</th>\n",
       "      <th>Synonyms</th>\n",
       "      <th>dbXrefs</th>\n",
       "      <th>chromosome</th>\n",
       "      <th>map_location</th>\n",
       "      <th>description</th>\n",
       "      <th>type_of_gene</th>\n",
       "      <th>Symbol_from_nomenclature_authority</th>\n",
       "      <th>Full_name_from_nomenclature_authority</th>\n",
       "      <th>Nomenclature_status</th>\n",
       "      <th>Other_designations</th>\n",
       "      <th>Modification_date</th>\n",
       "      <th>Feature_type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>26</td>\n",
       "      <td>NM_001351365</td>\n",
       "      <td>chr1</td>\n",
       "      <td>+</td>\n",
       "      <td>144146811</td>\n",
       "      <td>146467744</td>\n",
       "      <td>144146846</td>\n",
       "      <td>146466121</td>\n",
       "      <td>93</td>\n",
       "      <td>144146811,144148789,144149726,144150981,144151...</td>\n",
       "      <td>144147021,144148892,144149941,144151054,144151...</td>\n",
       "      <td>0</td>\n",
       "      <td>NBPF19</td>\n",
       "      <td>cmpl</td>\n",
       "      <td>cmpl</td>\n",
       "      <td>0,1,2,1,2,1,2,1,2,1,2,1,2,1,2,1,2,1,2,1,2,1,2,...</td>\n",
       "      <td>2319275</td>\n",
       "      <td>9606</td>\n",
       "      <td>101060226</td>\n",
       "      <td>NBPF19</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>MIM:614006|HGNC:HGNC:31999|Ensembl:ENSG0000027...</td>\n",
       "      <td>1</td>\n",
       "      <td>1q21.2</td>\n",
       "      <td>NBPF member 19</td>\n",
       "      <td>protein-coding</td>\n",
       "      <td>NBPF19</td>\n",
       "      <td>NBPF member 19</td>\n",
       "      <td>O</td>\n",
       "      <td>neuroblastoma breakpoint family member 19</td>\n",
       "      <td>20210912</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>26</td>\n",
       "      <td>NM_001278267</td>\n",
       "      <td>chr1</td>\n",
       "      <td>+</td>\n",
       "      <td>144146810</td>\n",
       "      <td>146467744</td>\n",
       "      <td>144158383</td>\n",
       "      <td>146466121</td>\n",
       "      <td>131</td>\n",
       "      <td>144146810,144148789,144149726,144150981,144151...</td>\n",
       "      <td>144147021,144148892,144149941,144151054,144151...</td>\n",
       "      <td>0</td>\n",
       "      <td>NBPF20</td>\n",
       "      <td>cmpl</td>\n",
       "      <td>cmpl</td>\n",
       "      <td>-1,-1,-1,-1,-1,-1,-1,-1,0,2,1,2,1,2,1,2,1,2,1,...</td>\n",
       "      <td>2307738</td>\n",
       "      <td>9606</td>\n",
       "      <td>100288142</td>\n",
       "      <td>NBPF20</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>MIM:614007|HGNC:HGNC:32000|Ensembl:ENSG0000016...</td>\n",
       "      <td>1</td>\n",
       "      <td>1q21.1</td>\n",
       "      <td>NBPF member 20</td>\n",
       "      <td>protein-coding</td>\n",
       "      <td>NBPF20</td>\n",
       "      <td>NBPF member 20</td>\n",
       "      <td>O</td>\n",
       "      <td>neuroblastoma breakpoint family member 20</td>\n",
       "      <td>20210611</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>12</td>\n",
       "      <td>NM_000109</td>\n",
       "      <td>chrX</td>\n",
       "      <td>-</td>\n",
       "      <td>31137340</td>\n",
       "      <td>33357505</td>\n",
       "      <td>31140035</td>\n",
       "      <td>33357382</td>\n",
       "      <td>79</td>\n",
       "      <td>31137340,31144758,31152218,31164407,31165391,3...</td>\n",
       "      <td>31140047,31144790,31152311,31164531,31165635,3...</td>\n",
       "      <td>0</td>\n",
       "      <td>DMD</td>\n",
       "      <td>cmpl</td>\n",
       "      <td>cmpl</td>\n",
       "      <td>0,1,1,0,2,2,2,2,2,0,2,0,1,2,1,1,2,1,0,0,1,0,2,...</td>\n",
       "      <td>2217347</td>\n",
       "      <td>9606</td>\n",
       "      <td>1756</td>\n",
       "      <td>DMD</td>\n",
       "      <td>-</td>\n",
       "      <td>BMD|CMD3B|DXS142|DXS164|DXS206|DXS230|DXS239|D...</td>\n",
       "      <td>MIM:300377|HGNC:HGNC:2928|Ensembl:ENSG00000198947</td>\n",
       "      <td>X</td>\n",
       "      <td>Xp21.2-p21.1</td>\n",
       "      <td>dystrophin</td>\n",
       "      <td>protein-coding</td>\n",
       "      <td>DMD</td>\n",
       "      <td>dystrophin</td>\n",
       "      <td>O</td>\n",
       "      <td>dystrophin|truncated dystrophin</td>\n",
       "      <td>20211026</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>2</td>\n",
       "      <td>NM_001351274</td>\n",
       "      <td>chr11</td>\n",
       "      <td>-</td>\n",
       "      <td>83166055</td>\n",
       "      <td>85339417</td>\n",
       "      <td>83170860</td>\n",
       "      <td>85337661</td>\n",
       "      <td>27</td>\n",
       "      <td>83166055,83173044,83177750,83180243,83182668,8...</td>\n",
       "      <td>83170967,83173136,83177860,83180416,83182770,8...</td>\n",
       "      <td>0</td>\n",
       "      <td>DLG2</td>\n",
       "      <td>cmpl</td>\n",
       "      <td>cmpl</td>\n",
       "      <td>1,2,0,1,1,1,0,2,2,1,0,2,2,2,1,0,1,2,0,0,0,0,0,...</td>\n",
       "      <td>2166801</td>\n",
       "      <td>9606</td>\n",
       "      <td>1740</td>\n",
       "      <td>DLG2</td>\n",
       "      <td>-</td>\n",
       "      <td>PPP1R58|PSD-93|PSD93|chapsyn-110</td>\n",
       "      <td>MIM:603583|HGNC:HGNC:2901|Ensembl:ENSG00000150672</td>\n",
       "      <td>11</td>\n",
       "      <td>11q14.1</td>\n",
       "      <td>discs large MAGUK scaffold protein 2</td>\n",
       "      <td>protein-coding</td>\n",
       "      <td>DLG2</td>\n",
       "      <td>discs large MAGUK scaffold protein 2</td>\n",
       "      <td>O</td>\n",
       "      <td>disks large homolog 2|channel-associated prote...</td>\n",
       "      <td>20211016</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>9</td>\n",
       "      <td>NM_033225</td>\n",
       "      <td>chr8</td>\n",
       "      <td>-</td>\n",
       "      <td>2792882</td>\n",
       "      <td>4852436</td>\n",
       "      <td>2796106</td>\n",
       "      <td>4851938</td>\n",
       "      <td>70</td>\n",
       "      <td>2792882,2799993,2806820,2807752,2808635,281174...</td>\n",
       "      <td>2796266,2800126,2806908,2807865,2808797,281179...</td>\n",
       "      <td>0</td>\n",
       "      <td>CSMD1</td>\n",
       "      <td>cmpl</td>\n",
       "      <td>cmpl</td>\n",
       "      <td>2,1,0,1,1,1,1,0,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,...</td>\n",
       "      <td>2055832</td>\n",
       "      <td>9606</td>\n",
       "      <td>64478</td>\n",
       "      <td>CSMD1</td>\n",
       "      <td>-</td>\n",
       "      <td>PPP1R24</td>\n",
       "      <td>MIM:608397|HGNC:HGNC:14026|Ensembl:ENSG0000018...</td>\n",
       "      <td>8</td>\n",
       "      <td>8p23.2</td>\n",
       "      <td>CUB and Sushi multiple domains 1</td>\n",
       "      <td>protein-coding</td>\n",
       "      <td>CSMD1</td>\n",
       "      <td>CUB and Sushi multiple domains 1</td>\n",
       "      <td>O</td>\n",
       "      <td>CUB and sushi domain-containing protein 1|prot...</td>\n",
       "      <td>20210919</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   bin          name  chrom strand    txStart      txEnd   cdsStart  \\\n",
       "0   26  NM_001351365   chr1      +  144146811  146467744  144146846   \n",
       "1   26  NM_001278267   chr1      +  144146810  146467744  144158383   \n",
       "2   12     NM_000109   chrX      -   31137340   33357505   31140035   \n",
       "3    2  NM_001351274  chr11      -   83166055   85339417   83170860   \n",
       "4    9     NM_033225   chr8      -    2792882    4852436    2796106   \n",
       "\n",
       "      cdsEnd  exonCount                                         exonStarts  \\\n",
       "0  146466121         93  144146811,144148789,144149726,144150981,144151...   \n",
       "1  146466121        131  144146810,144148789,144149726,144150981,144151...   \n",
       "2   33357382         79  31137340,31144758,31152218,31164407,31165391,3...   \n",
       "3   85337661         27  83166055,83173044,83177750,83180243,83182668,8...   \n",
       "4    4851938         70  2792882,2799993,2806820,2807752,2808635,281174...   \n",
       "\n",
       "                                            exonEnds  score   name2  \\\n",
       "0  144147021,144148892,144149941,144151054,144151...      0  NBPF19   \n",
       "1  144147021,144148892,144149941,144151054,144151...      0  NBPF20   \n",
       "2  31140047,31144790,31152311,31164531,31165635,3...      0     DMD   \n",
       "3  83170967,83173136,83177860,83180416,83182770,8...      0    DLG2   \n",
       "4  2796266,2800126,2806908,2807865,2808797,281179...      0   CSMD1   \n",
       "\n",
       "  cdsStartStat cdsEndStat                                         ExonFrames  \\\n",
       "0         cmpl       cmpl  0,1,2,1,2,1,2,1,2,1,2,1,2,1,2,1,2,1,2,1,2,1,2,...   \n",
       "1         cmpl       cmpl  -1,-1,-1,-1,-1,-1,-1,-1,0,2,1,2,1,2,1,2,1,2,1,...   \n",
       "2         cmpl       cmpl  0,1,1,0,2,2,2,2,2,0,2,0,1,2,1,1,2,1,0,0,1,0,2,...   \n",
       "3         cmpl       cmpl  1,2,0,1,1,1,0,2,2,1,0,2,2,2,1,0,1,2,0,0,0,0,0,...   \n",
       "4         cmpl       cmpl  2,1,0,1,1,1,1,0,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,...   \n",
       "\n",
       "    length  #tax_id     GeneID  Symbol LocusTag  \\\n",
       "0  2319275     9606  101060226  NBPF19        -   \n",
       "1  2307738     9606  100288142  NBPF20        -   \n",
       "2  2217347     9606       1756     DMD        -   \n",
       "3  2166801     9606       1740    DLG2        -   \n",
       "4  2055832     9606      64478   CSMD1        -   \n",
       "\n",
       "                                            Synonyms  \\\n",
       "0                                                  -   \n",
       "1                                                  -   \n",
       "2  BMD|CMD3B|DXS142|DXS164|DXS206|DXS230|DXS239|D...   \n",
       "3                   PPP1R58|PSD-93|PSD93|chapsyn-110   \n",
       "4                                            PPP1R24   \n",
       "\n",
       "                                             dbXrefs chromosome  map_location  \\\n",
       "0  MIM:614006|HGNC:HGNC:31999|Ensembl:ENSG0000027...          1        1q21.2   \n",
       "1  MIM:614007|HGNC:HGNC:32000|Ensembl:ENSG0000016...          1        1q21.1   \n",
       "2  MIM:300377|HGNC:HGNC:2928|Ensembl:ENSG00000198947          X  Xp21.2-p21.1   \n",
       "3  MIM:603583|HGNC:HGNC:2901|Ensembl:ENSG00000150672         11       11q14.1   \n",
       "4  MIM:608397|HGNC:HGNC:14026|Ensembl:ENSG0000018...          8        8p23.2   \n",
       "\n",
       "                            description    type_of_gene  \\\n",
       "0                        NBPF member 19  protein-coding   \n",
       "1                        NBPF member 20  protein-coding   \n",
       "2                            dystrophin  protein-coding   \n",
       "3  discs large MAGUK scaffold protein 2  protein-coding   \n",
       "4      CUB and Sushi multiple domains 1  protein-coding   \n",
       "\n",
       "  Symbol_from_nomenclature_authority Full_name_from_nomenclature_authority  \\\n",
       "0                             NBPF19                        NBPF member 19   \n",
       "1                             NBPF20                        NBPF member 20   \n",
       "2                                DMD                            dystrophin   \n",
       "3                               DLG2  discs large MAGUK scaffold protein 2   \n",
       "4                              CSMD1      CUB and Sushi multiple domains 1   \n",
       "\n",
       "  Nomenclature_status                                 Other_designations  \\\n",
       "0                   O          neuroblastoma breakpoint family member 19   \n",
       "1                   O          neuroblastoma breakpoint family member 20   \n",
       "2                   O                    dystrophin|truncated dystrophin   \n",
       "3                   O  disks large homolog 2|channel-associated prote...   \n",
       "4                   O  CUB and sushi domain-containing protein 1|prot...   \n",
       "\n",
       "   Modification_date Feature_type  \n",
       "0           20210912            -  \n",
       "1           20210611            -  \n",
       "2           20211026            -  \n",
       "3           20211016            -  \n",
       "4           20210919            -  "
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "refgene_info = refgene.merge(gene_info, left_on='name2', right_on='Symbol')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {
    "pycharm": {
     "name": "#%%\n"
    }
   },
   "outputs": [],
   "source": [
    "refgene_info.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_fam = pd.read_csv(hgnc_gene_fam_file, sep=',')\n",
    "gene_fam = gene_fam[gene_fam['typical_gene'].notna()].rename(columns={'id': 'fam'})\n",
    "gene_fam_grp = gene_fam.groupby('typical_gene').agg({'fam': set})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>txStart</th>\n",
       "      <th>txEnd</th>\n",
       "      <th>GeneID</th>\n",
       "      <th>name2</th>\n",
       "      <th>name</th>\n",
       "      <th>strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>18372</th>\n",
       "      <td>chr1</td>\n",
       "      <td>69090</td>\n",
       "      <td>70008</td>\n",
       "      <td>79501</td>\n",
       "      <td>OR4F5</td>\n",
       "      <td>NM_001005484</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18188</th>\n",
       "      <td>chr1</td>\n",
       "      <td>367658</td>\n",
       "      <td>368597</td>\n",
       "      <td>729759</td>\n",
       "      <td>OR4F29</td>\n",
       "      <td>NM_001005221</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18185</th>\n",
       "      <td>chr1</td>\n",
       "      <td>621095</td>\n",
       "      <td>622034</td>\n",
       "      <td>81399</td>\n",
       "      <td>OR4F16</td>\n",
       "      <td>NM_001005277</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9220</th>\n",
       "      <td>chr1</td>\n",
       "      <td>859302</td>\n",
       "      <td>879954</td>\n",
       "      <td>148398</td>\n",
       "      <td>SAMD11</td>\n",
       "      <td>NM_001385641</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10563</th>\n",
       "      <td>chr1</td>\n",
       "      <td>879582</td>\n",
       "      <td>894636</td>\n",
       "      <td>26155</td>\n",
       "      <td>NOC2L</td>\n",
       "      <td>NM_015658</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      chrom  txStart   txEnd  GeneID   name2          name strand\n",
       "18372  chr1    69090   70008   79501   OR4F5  NM_001005484      +\n",
       "18188  chr1   367658  368597  729759  OR4F29  NM_001005221      +\n",
       "18185  chr1   621095  622034   81399  OR4F16  NM_001005277      -\n",
       "9220   chr1   859302  879954  148398  SAMD11  NM_001385641      +\n",
       "10563  chr1   879582  894636   26155   NOC2L     NM_015658      -"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene = refgene_info.loc[\n",
    "    refgene_info['type_of_gene'] == 'protein-coding',\n",
    "    ['chrom', 'txStart', 'txEnd', 'GeneID', 'name2', 'name', 'strand']\n",
    "].sort_values(['chrom', 'txStart', 'txEnd'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "gene.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "gene.rename(columns={\n",
    "    'chrom': '#chrom', 'txStart': 'start', 'txEnd': 'end',\n",
    "    'GeneID': 'gene_id', 'name2': 'symbol', 'name': 'transcript'\n",
    "}).to_csv(gene_file, index=False, sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "!bgzip -cf {gene_file} > {gene_file}.gz"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "!tabix -fp bed {gene_file}.gz"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "!rm {gene_file}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "raw_omim_df = pd.read_excel(omim_gene_list_file)\n",
    "omim_gene = set(raw_omim_df['gene_symbol'].dropna())\n",
    "# with open(omim_gene_list_file) as f:\n",
    "#     omim_gene = set(f.read().split('\\n'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>txStart</th>\n",
       "      <th>txEnd</th>\n",
       "      <th>GeneID</th>\n",
       "      <th>name2</th>\n",
       "      <th>name</th>\n",
       "      <th>strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>18397</th>\n",
       "      <td>chr1</td>\n",
       "      <td>948876</td>\n",
       "      <td>949919</td>\n",
       "      <td>9636</td>\n",
       "      <td>ISG15</td>\n",
       "      <td>NM_005101</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6532</th>\n",
       "      <td>chr1</td>\n",
       "      <td>955499</td>\n",
       "      <td>991494</td>\n",
       "      <td>375790</td>\n",
       "      <td>AGRN</td>\n",
       "      <td>NM_001305275</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16101</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1146719</td>\n",
       "      <td>1149533</td>\n",
       "      <td>7293</td>\n",
       "      <td>TNFRSF4</td>\n",
       "      <td>NM_003327</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17883</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1167616</td>\n",
       "      <td>1170420</td>\n",
       "      <td>126792</td>\n",
       "      <td>B3GALT6</td>\n",
       "      <td>NM_080605</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11098</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1270657</td>\n",
       "      <td>1284798</td>\n",
       "      <td>1855</td>\n",
       "      <td>DVL1</td>\n",
       "      <td>NM_004421</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      chrom  txStart    txEnd  GeneID    name2          name strand\n",
       "18397  chr1   948876   949919    9636    ISG15     NM_005101      +\n",
       "6532   chr1   955499   991494  375790     AGRN  NM_001305275      +\n",
       "16101  chr1  1146719  1149533    7293  TNFRSF4     NM_003327      -\n",
       "17883  chr1  1167616  1170420  126792  B3GALT6     NM_080605      +\n",
       "11098  chr1  1270657  1284798    1855     DVL1     NM_004421      -"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "omim_gene = refgene_info.loc[\n",
    "    refgene_info['name2'].isin(omim_gene),\n",
    "    ['chrom', 'txStart', 'txEnd', 'GeneID', 'name2', 'name', 'strand']\n",
    "].sort_values(['chrom', 'txStart', 'txEnd'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "omim_gene.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [],
   "source": [
    "omim_gene.rename(columns={\n",
    "    'chrom': '#chrom', 'txStart': 'start', 'txEnd': 'end',\n",
    "    'GeneID': 'gene_id', 'name2': 'symbol', 'name': 'transcript'\n",
    "}).to_csv(omim_gene_file, index=False, sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [],
   "source": [
    "!bgzip -cf {omim_gene_file} > {omim_gene_file}.gz"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "outputs": [],
   "source": [
    "!tabix -fp bed {omim_gene_file}.gz\n",
    "!rm {omim_gene_file}"
   ],
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%%\n"
    }
   }
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>#Gene Symbol</th>\n",
       "      <th>Gene ID</th>\n",
       "      <th>cytoBand</th>\n",
       "      <th>Genomic Location</th>\n",
       "      <th>Haploinsufficiency Score</th>\n",
       "      <th>Haploinsufficiency Description</th>\n",
       "      <th>Haploinsufficiency PMID1</th>\n",
       "      <th>Haploinsufficiency PMID2</th>\n",
       "      <th>Haploinsufficiency PMID3</th>\n",
       "      <th>Haploinsufficiency PMID4</th>\n",
       "      <th>Haploinsufficiency PMID5</th>\n",
       "      <th>Haploinsufficiency PMID6</th>\n",
       "      <th>Triplosensitivity Score</th>\n",
       "      <th>Triplosensitivity Description</th>\n",
       "      <th>Triplosensitivity PMID1</th>\n",
       "      <th>Triplosensitivity PMID2</th>\n",
       "      <th>Triplosensitivity PMID3</th>\n",
       "      <th>Triplosensitivity PMID4</th>\n",
       "      <th>Triplosensitivity PMID5</th>\n",
       "      <th>Triplosensitivity PMID6</th>\n",
       "      <th>Date Last Evaluated</th>\n",
       "      <th>Loss phenotype OMIM ID</th>\n",
       "      <th>Triplosensitive phenotype OMIM ID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A4GALT</td>\n",
       "      <td>53947</td>\n",
       "      <td>22q13.2</td>\n",
       "      <td>chr22:43088121-43117307</td>\n",
       "      <td>30</td>\n",
       "      <td>Gene associated with autosomal recessive pheno...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>No evidence available</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2014-12-11</td>\n",
       "      <td>111400</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AAGAB</td>\n",
       "      <td>79719</td>\n",
       "      <td>15q23</td>\n",
       "      <td>chr15:67493013-67547536</td>\n",
       "      <td>3</td>\n",
       "      <td>Sufficient evidence for dosage pathogenicity</td>\n",
       "      <td>23064416</td>\n",
       "      <td>23000146</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>No evidence available</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2013-02-28</td>\n",
       "      <td>148600</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AARS1</td>\n",
       "      <td>16</td>\n",
       "      <td>16q22.1</td>\n",
       "      <td>chr16:70286297-70323412</td>\n",
       "      <td>0</td>\n",
       "      <td>No evidence available</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>No evidence available</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2018-01-11</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AARS2</td>\n",
       "      <td>57505</td>\n",
       "      <td>6p21.1</td>\n",
       "      <td>chr6:44266463-44281063</td>\n",
       "      <td>30</td>\n",
       "      <td>Gene associated with autosomal recessive pheno...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Not yet evaluated</td>\n",
       "      <td>Not yet evaluated</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2016-08-22</td>\n",
       "      <td>615889</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AASS</td>\n",
       "      <td>10157</td>\n",
       "      <td>7q31.32</td>\n",
       "      <td>chr7:121713598-121784344</td>\n",
       "      <td>30</td>\n",
       "      <td>Gene associated with autosomal recessive pheno...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Not yet evaluated</td>\n",
       "      <td>Not yet evaluated</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2016-08-22</td>\n",
       "      <td>238700</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  #Gene Symbol Gene ID cytoBand          Genomic Location  \\\n",
       "0       A4GALT   53947  22q13.2   chr22:43088121-43117307   \n",
       "1        AAGAB   79719    15q23   chr15:67493013-67547536   \n",
       "2        AARS1      16  16q22.1   chr16:70286297-70323412   \n",
       "3        AARS2   57505   6p21.1    chr6:44266463-44281063   \n",
       "4         AASS   10157  7q31.32  chr7:121713598-121784344   \n",
       "\n",
       "  Haploinsufficiency Score                     Haploinsufficiency Description  \\\n",
       "0                       30  Gene associated with autosomal recessive pheno...   \n",
       "1                        3       Sufficient evidence for dosage pathogenicity   \n",
       "2                        0                              No evidence available   \n",
       "3                       30  Gene associated with autosomal recessive pheno...   \n",
       "4                       30  Gene associated with autosomal recessive pheno...   \n",
       "\n",
       "  Haploinsufficiency PMID1 Haploinsufficiency PMID2 Haploinsufficiency PMID3  \\\n",
       "0                      NaN                      NaN                      NaN   \n",
       "1                 23064416                 23000146                      NaN   \n",
       "2                      NaN                      NaN                      NaN   \n",
       "3                      NaN                      NaN                      NaN   \n",
       "4                      NaN                      NaN                      NaN   \n",
       "\n",
       "  Haploinsufficiency PMID4 Haploinsufficiency PMID5 Haploinsufficiency PMID6  \\\n",
       "0                      NaN                      NaN                      NaN   \n",
       "1                      NaN                      NaN                      NaN   \n",
       "2                      NaN                      NaN                      NaN   \n",
       "3                      NaN                      NaN                      NaN   \n",
       "4                      NaN                      NaN                      NaN   \n",
       "\n",
       "  Triplosensitivity Score Triplosensitivity Description  \\\n",
       "0                       0         No evidence available   \n",
       "1                       0         No evidence available   \n",
       "2                       0         No evidence available   \n",
       "3       Not yet evaluated             Not yet evaluated   \n",
       "4       Not yet evaluated             Not yet evaluated   \n",
       "\n",
       "  Triplosensitivity PMID1 Triplosensitivity PMID2 Triplosensitivity PMID3  \\\n",
       "0                     NaN                     NaN                     NaN   \n",
       "1                     NaN                     NaN                     NaN   \n",
       "2                     NaN                     NaN                     NaN   \n",
       "3                     NaN                     NaN                     NaN   \n",
       "4                     NaN                     NaN                     NaN   \n",
       "\n",
       "  Triplosensitivity PMID4 Triplosensitivity PMID5 Triplosensitivity PMID6  \\\n",
       "0                     NaN                     NaN                     NaN   \n",
       "1                     NaN                     NaN                     NaN   \n",
       "2                     NaN                     NaN                     NaN   \n",
       "3                     NaN                     NaN                     NaN   \n",
       "4                     NaN                     NaN                     NaN   \n",
       "\n",
       "  Date Last Evaluated Loss phenotype OMIM ID Triplosensitive phenotype OMIM ID  \n",
       "0          2014-12-11                 111400                               NaN  \n",
       "1          2013-02-28                 148600                               NaN  \n",
       "2          2018-01-11                    NaN                               NaN  \n",
       "3          2016-08-22                 615889                               NaN  \n",
       "4          2016-08-22                 238700                               NaN  "
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "curation_gene = pd.read_csv(clingen_gene_ori_file, sep='\\t', dtype=str, skiprows=5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [],
   "source": [
    "hi_genes = set(\n",
    "    curation_gene.loc[\n",
    "        curation_gene['Haploinsufficiency Score'] == '3', '#Gene Symbol'\n",
    "    ]\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [],
   "source": [
    "hi_gene = refgene_info.loc[\n",
    "    refgene_info['name2'].isin(hi_genes),\n",
    "    ['chrom', 'txStart', 'txEnd', 'GeneID', 'name2', 'name', 'strand']\n",
    "].sort_values(['chrom', 'txStart', 'txEnd'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>txStart</th>\n",
       "      <th>txEnd</th>\n",
       "      <th>GeneID</th>\n",
       "      <th>name2</th>\n",
       "      <th>name</th>\n",
       "      <th>strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>51</th>\n",
       "      <td>chr1</td>\n",
       "      <td>6845513</td>\n",
       "      <td>7829766</td>\n",
       "      <td>23261</td>\n",
       "      <td>CAMTA1</td>\n",
       "      <td>NM_001349609</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6499</th>\n",
       "      <td>chr1</td>\n",
       "      <td>17345220</td>\n",
       "      <td>17380527</td>\n",
       "      <td>6390</td>\n",
       "      <td>SDHB</td>\n",
       "      <td>NM_003000</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3068</th>\n",
       "      <td>chr1</td>\n",
       "      <td>27022505</td>\n",
       "      <td>27108595</td>\n",
       "      <td>8289</td>\n",
       "      <td>ARID1A</td>\n",
       "      <td>NM_139135</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14667</th>\n",
       "      <td>chr1</td>\n",
       "      <td>27860755</td>\n",
       "      <td>27930655</td>\n",
       "      <td>27245</td>\n",
       "      <td>AHDC1</td>\n",
       "      <td>NM_001371928</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7013</th>\n",
       "      <td>chr1</td>\n",
       "      <td>43391025</td>\n",
       "      <td>43424539</td>\n",
       "      <td>6513</td>\n",
       "      <td>SLC2A1</td>\n",
       "      <td>NM_006516</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      chrom   txStart     txEnd  GeneID   name2          name strand\n",
       "51     chr1   6845513   7829766   23261  CAMTA1  NM_001349609      +\n",
       "6499   chr1  17345220  17380527    6390    SDHB     NM_003000      -\n",
       "3068   chr1  27022505  27108595    8289  ARID1A     NM_139135      +\n",
       "14667  chr1  27860755  27930655   27245   AHDC1  NM_001371928      -\n",
       "7013   chr1  43391025  43424539    6513  SLC2A1     NM_006516      -"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "hi_gene.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [],
   "source": [
    "hi_gene.rename(columns={\n",
    "    'chrom': '#chrom', 'txStart': 'start', 'txEnd': 'end',\n",
    "    'GeneID': 'gene_id', 'name2': 'symbol', 'name': 'transcript'\n",
    "}).to_csv(hi_gene_file, index=False, sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [],
   "source": [
    "!bgzip -cf {hi_gene_file} > {hi_gene_file}.gz"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [],
   "source": [
    "!tabix -fp bed {hi_gene_file}.gz\n",
    "!rm {hi_gene_file}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [],
   "source": [
    "hi_cds = refgene_info.loc[\n",
    "    (refgene_info['name2'].isin(hi_genes)) & (refgene_info['length'] != 0),\n",
    "    ['chrom', 'cdsStart', 'cdsEnd', 'GeneID', 'name2', 'name']\n",
    "].sort_values(['chrom', 'cdsStart', 'cdsEnd'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>cdsStart</th>\n",
       "      <th>cdsEnd</th>\n",
       "      <th>GeneID</th>\n",
       "      <th>name2</th>\n",
       "      <th>name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>51</th>\n",
       "      <td>chr1</td>\n",
       "      <td>6845590</td>\n",
       "      <td>7826582</td>\n",
       "      <td>23261</td>\n",
       "      <td>CAMTA1</td>\n",
       "      <td>NM_001349609</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6499</th>\n",
       "      <td>chr1</td>\n",
       "      <td>17345375</td>\n",
       "      <td>17380514</td>\n",
       "      <td>6390</td>\n",
       "      <td>SDHB</td>\n",
       "      <td>NM_003000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3068</th>\n",
       "      <td>chr1</td>\n",
       "      <td>27022894</td>\n",
       "      <td>27107247</td>\n",
       "      <td>8289</td>\n",
       "      <td>ARID1A</td>\n",
       "      <td>NM_139135</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14667</th>\n",
       "      <td>chr1</td>\n",
       "      <td>27873814</td>\n",
       "      <td>27878626</td>\n",
       "      <td>27245</td>\n",
       "      <td>AHDC1</td>\n",
       "      <td>NM_001371928</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7013</th>\n",
       "      <td>chr1</td>\n",
       "      <td>43392711</td>\n",
       "      <td>43424322</td>\n",
       "      <td>6513</td>\n",
       "      <td>SLC2A1</td>\n",
       "      <td>NM_006516</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      chrom  cdsStart    cdsEnd  GeneID   name2          name\n",
       "51     chr1   6845590   7826582   23261  CAMTA1  NM_001349609\n",
       "6499   chr1  17345375  17380514    6390    SDHB     NM_003000\n",
       "3068   chr1  27022894  27107247    8289  ARID1A     NM_139135\n",
       "14667  chr1  27873814  27878626   27245   AHDC1  NM_001371928\n",
       "7013   chr1  43392711  43424322    6513  SLC2A1     NM_006516"
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "hi_cds.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [],
   "source": [
    "hi_cds.rename(columns={\n",
    "    'chrom': '#chrom', 'cdsStart': 'start', 'cdsEnd': 'end',\n",
    "    'GeneID': 'gene_id', 'name2': 'symbol', 'name': 'transcript'\n",
    "}).to_csv(hi_cds_file, index=False, sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "!bgzip -cf {hi_cds_file} > {hi_cds_file}.gz"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [],
   "source": [
    "!tabix -fp bed {hi_cds_file}.gz\n",
    "!rm {hi_cds_file}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [],
   "source": [
    "hi_exons = refgene_info.loc[\n",
    "    refgene_info['name2'].isin(hi_genes), ['chrom', 'exonStarts', 'exonEnds', 'GeneID', 'name2', 'name', 'strand']\n",
    "].copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>exonStarts</th>\n",
       "      <th>exonEnds</th>\n",
       "      <th>GeneID</th>\n",
       "      <th>name2</th>\n",
       "      <th>name</th>\n",
       "      <th>strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chrX</td>\n",
       "      <td>31137340,31144758,31152218,31164407,31165391,3...</td>\n",
       "      <td>31140047,31144790,31152311,31164531,31165635,3...</td>\n",
       "      <td>1756</td>\n",
       "      <td>DMD</td>\n",
       "      <td>NM_000109</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>chr7</td>\n",
       "      <td>69063460,69364271,69583117,69599521,69900737,7...</td>\n",
       "      <td>69064948,69364484,69583219,69599557,69900767,7...</td>\n",
       "      <td>26053</td>\n",
       "      <td>AUTS2</td>\n",
       "      <td>NM_015570</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>chrX</td>\n",
       "      <td>28605562,28807436,29301054,29414374,29417271,2...</td>\n",
       "      <td>28606164,28807542,29301334,29414561,29417425,2...</td>\n",
       "      <td>11141</td>\n",
       "      <td>IL1RAPL1</td>\n",
       "      <td>NM_014271</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44</th>\n",
       "      <td>chr2</td>\n",
       "      <td>50145642,50170841,50280408,50282092,50318460,5...</td>\n",
       "      <td>50149389,50170929,50280728,50282182,50318632,5...</td>\n",
       "      <td>9378</td>\n",
       "      <td>NRXN1</td>\n",
       "      <td>NM_001330084</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>51</th>\n",
       "      <td>chr1</td>\n",
       "      <td>6845513,6880240,6885151,7151363,7309550,752788...</td>\n",
       "      <td>6845635,6880310,6885270,7151431,7309686,752796...</td>\n",
       "      <td>23261</td>\n",
       "      <td>CAMTA1</td>\n",
       "      <td>NM_001349609</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   chrom                                         exonStarts  \\\n",
       "2   chrX  31137340,31144758,31152218,31164407,31165391,3...   \n",
       "26  chr7  69063460,69364271,69583117,69599521,69900737,7...   \n",
       "31  chrX  28605562,28807436,29301054,29414374,29417271,2...   \n",
       "44  chr2  50145642,50170841,50280408,50282092,50318460,5...   \n",
       "51  chr1  6845513,6880240,6885151,7151363,7309550,752788...   \n",
       "\n",
       "                                             exonEnds  GeneID     name2  \\\n",
       "2   31140047,31144790,31152311,31164531,31165635,3...    1756       DMD   \n",
       "26  69064948,69364484,69583219,69599557,69900767,7...   26053     AUTS2   \n",
       "31  28606164,28807542,29301334,29414561,29417425,2...   11141  IL1RAPL1   \n",
       "44  50149389,50170929,50280728,50282182,50318632,5...    9378     NRXN1   \n",
       "51  6845635,6880310,6885270,7151431,7309686,752796...   23261    CAMTA1   \n",
       "\n",
       "            name strand  \n",
       "2      NM_000109      -  \n",
       "26     NM_015570      +  \n",
       "31     NM_014271      +  \n",
       "44  NM_001330084      -  \n",
       "51  NM_001349609      +  "
      ]
     },
     "execution_count": 43,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "hi_exons.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [],
   "source": [
    "hi_exons['exonStarts'] = hi_exons['exonStarts'].str.replace(r',$', '')\n",
    "hi_exons['exonEnds'] = hi_exons['exonEnds'].str.replace(r',$', '')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>exonStarts</th>\n",
       "      <th>exonEnds</th>\n",
       "      <th>GeneID</th>\n",
       "      <th>name2</th>\n",
       "      <th>name</th>\n",
       "      <th>strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chrX</td>\n",
       "      <td>31137340,31144758,31152218,31164407,31165391,3...</td>\n",
       "      <td>31140047,31144790,31152311,31164531,31165635,3...</td>\n",
       "      <td>1756</td>\n",
       "      <td>DMD</td>\n",
       "      <td>NM_000109</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>chr7</td>\n",
       "      <td>69063460,69364271,69583117,69599521,69900737,7...</td>\n",
       "      <td>69064948,69364484,69583219,69599557,69900767,7...</td>\n",
       "      <td>26053</td>\n",
       "      <td>AUTS2</td>\n",
       "      <td>NM_015570</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>chrX</td>\n",
       "      <td>28605562,28807436,29301054,29414374,29417271,2...</td>\n",
       "      <td>28606164,28807542,29301334,29414561,29417425,2...</td>\n",
       "      <td>11141</td>\n",
       "      <td>IL1RAPL1</td>\n",
       "      <td>NM_014271</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44</th>\n",
       "      <td>chr2</td>\n",
       "      <td>50145642,50170841,50280408,50282092,50318460,5...</td>\n",
       "      <td>50149389,50170929,50280728,50282182,50318632,5...</td>\n",
       "      <td>9378</td>\n",
       "      <td>NRXN1</td>\n",
       "      <td>NM_001330084</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>51</th>\n",
       "      <td>chr1</td>\n",
       "      <td>6845513,6880240,6885151,7151363,7309550,752788...</td>\n",
       "      <td>6845635,6880310,6885270,7151431,7309686,752796...</td>\n",
       "      <td>23261</td>\n",
       "      <td>CAMTA1</td>\n",
       "      <td>NM_001349609</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   chrom                                         exonStarts  \\\n",
       "2   chrX  31137340,31144758,31152218,31164407,31165391,3...   \n",
       "26  chr7  69063460,69364271,69583117,69599521,69900737,7...   \n",
       "31  chrX  28605562,28807436,29301054,29414374,29417271,2...   \n",
       "44  chr2  50145642,50170841,50280408,50282092,50318460,5...   \n",
       "51  chr1  6845513,6880240,6885151,7151363,7309550,752788...   \n",
       "\n",
       "                                             exonEnds  GeneID     name2  \\\n",
       "2   31140047,31144790,31152311,31164531,31165635,3...    1756       DMD   \n",
       "26  69064948,69364484,69583219,69599557,69900767,7...   26053     AUTS2   \n",
       "31  28606164,28807542,29301334,29414561,29417425,2...   11141  IL1RAPL1   \n",
       "44  50149389,50170929,50280728,50282182,50318632,5...    9378     NRXN1   \n",
       "51  6845635,6880310,6885270,7151431,7309686,752796...   23261    CAMTA1   \n",
       "\n",
       "            name strand  \n",
       "2      NM_000109      -  \n",
       "26     NM_015570      +  \n",
       "31     NM_014271      +  \n",
       "44  NM_001330084      -  \n",
       "51  NM_001349609      +  "
      ]
     },
     "execution_count": 45,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "hi_exons.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [],
   "source": [
    "start = hi_exons['exonStarts'].str.split(',').apply(pd.Series).stack().reset_index()\n",
    "start = start.rename(columns={'level_0': 'row', 0: 'start'})[['row', 'start']]\n",
    "start['start'] = start['start'].astype(int)\n",
    "end = hi_exons['exonEnds'].str.split(',').apply(pd.Series).stack().reset_index()\n",
    "end = end.rename(columns={0: 'end'})['end'].astype(int)\n",
    "position = start.join(end)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>row</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>2</td>\n",
       "      <td>31137340</td>\n",
       "      <td>31140047</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2</td>\n",
       "      <td>31144758</td>\n",
       "      <td>31144790</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>31152218</td>\n",
       "      <td>31152311</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>2</td>\n",
       "      <td>31164407</td>\n",
       "      <td>31164531</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>2</td>\n",
       "      <td>31165391</td>\n",
       "      <td>31165635</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   row     start       end\n",
       "0    2  31137340  31140047\n",
       "1    2  31144758  31144790\n",
       "2    2  31152218  31152311\n",
       "3    2  31164407  31164531\n",
       "4    2  31165391  31165635"
      ]
     },
     "execution_count": 47,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "position.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [],
   "source": [
    "exon = position.merge(\n",
    "    hi_exons[['chrom', 'GeneID', 'name2', 'name', 'strand']], how='left', left_on='row', right_index=True\n",
    ")\n",
    "exon = exon.sort_values(['chrom', 'start', 'end'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>row</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>chrom</th>\n",
       "      <th>GeneID</th>\n",
       "      <th>name2</th>\n",
       "      <th>name</th>\n",
       "      <th>strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>128</th>\n",
       "      <td>51</td>\n",
       "      <td>6845513</td>\n",
       "      <td>6845635</td>\n",
       "      <td>chr1</td>\n",
       "      <td>23261</td>\n",
       "      <td>CAMTA1</td>\n",
       "      <td>NM_001349609</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>129</th>\n",
       "      <td>51</td>\n",
       "      <td>6880240</td>\n",
       "      <td>6880310</td>\n",
       "      <td>chr1</td>\n",
       "      <td>23261</td>\n",
       "      <td>CAMTA1</td>\n",
       "      <td>NM_001349609</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>130</th>\n",
       "      <td>51</td>\n",
       "      <td>6885151</td>\n",
       "      <td>6885270</td>\n",
       "      <td>chr1</td>\n",
       "      <td>23261</td>\n",
       "      <td>CAMTA1</td>\n",
       "      <td>NM_001349609</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>131</th>\n",
       "      <td>51</td>\n",
       "      <td>7151363</td>\n",
       "      <td>7151431</td>\n",
       "      <td>chr1</td>\n",
       "      <td>23261</td>\n",
       "      <td>CAMTA1</td>\n",
       "      <td>NM_001349609</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>132</th>\n",
       "      <td>51</td>\n",
       "      <td>7309550</td>\n",
       "      <td>7309686</td>\n",
       "      <td>chr1</td>\n",
       "      <td>23261</td>\n",
       "      <td>CAMTA1</td>\n",
       "      <td>NM_001349609</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     row    start      end chrom  GeneID   name2          name strand\n",
       "128   51  6845513  6845635  chr1   23261  CAMTA1  NM_001349609      +\n",
       "129   51  6880240  6880310  chr1   23261  CAMTA1  NM_001349609      +\n",
       "130   51  6885151  6885270  chr1   23261  CAMTA1  NM_001349609      +\n",
       "131   51  7151363  7151431  chr1   23261  CAMTA1  NM_001349609      +\n",
       "132   51  7309550  7309686  chr1   23261  CAMTA1  NM_001349609      +"
      ]
     },
     "execution_count": 49,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "exon.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [],
   "source": [
    "exon['+'] = exon.groupby(['name2', 'name'])['start'].rank('first', ascending=True).astype(int)\n",
    "exon['-'] = exon.groupby(['name2', 'name'])['start'].rank('first', ascending=False).astype(int)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>row</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>chrom</th>\n",
       "      <th>GeneID</th>\n",
       "      <th>name2</th>\n",
       "      <th>name</th>\n",
       "      <th>strand</th>\n",
       "      <th>+</th>\n",
       "      <th>-</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>128</th>\n",
       "      <td>51</td>\n",
       "      <td>6845513</td>\n",
       "      <td>6845635</td>\n",
       "      <td>chr1</td>\n",
       "      <td>23261</td>\n",
       "      <td>CAMTA1</td>\n",
       "      <td>NM_001349609</td>\n",
       "      <td>+</td>\n",
       "      <td>1</td>\n",
       "      <td>23</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>129</th>\n",
       "      <td>51</td>\n",
       "      <td>6880240</td>\n",
       "      <td>6880310</td>\n",
       "      <td>chr1</td>\n",
       "      <td>23261</td>\n",
       "      <td>CAMTA1</td>\n",
       "      <td>NM_001349609</td>\n",
       "      <td>+</td>\n",
       "      <td>2</td>\n",
       "      <td>22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>130</th>\n",
       "      <td>51</td>\n",
       "      <td>6885151</td>\n",
       "      <td>6885270</td>\n",
       "      <td>chr1</td>\n",
       "      <td>23261</td>\n",
       "      <td>CAMTA1</td>\n",
       "      <td>NM_001349609</td>\n",
       "      <td>+</td>\n",
       "      <td>3</td>\n",
       "      <td>21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>131</th>\n",
       "      <td>51</td>\n",
       "      <td>7151363</td>\n",
       "      <td>7151431</td>\n",
       "      <td>chr1</td>\n",
       "      <td>23261</td>\n",
       "      <td>CAMTA1</td>\n",
       "      <td>NM_001349609</td>\n",
       "      <td>+</td>\n",
       "      <td>4</td>\n",
       "      <td>20</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>132</th>\n",
       "      <td>51</td>\n",
       "      <td>7309550</td>\n",
       "      <td>7309686</td>\n",
       "      <td>chr1</td>\n",
       "      <td>23261</td>\n",
       "      <td>CAMTA1</td>\n",
       "      <td>NM_001349609</td>\n",
       "      <td>+</td>\n",
       "      <td>5</td>\n",
       "      <td>19</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     row    start      end chrom  GeneID   name2          name strand  +   -\n",
       "128   51  6845513  6845635  chr1   23261  CAMTA1  NM_001349609      +  1  23\n",
       "129   51  6880240  6880310  chr1   23261  CAMTA1  NM_001349609      +  2  22\n",
       "130   51  6885151  6885270  chr1   23261  CAMTA1  NM_001349609      +  3  21\n",
       "131   51  7151363  7151431  chr1   23261  CAMTA1  NM_001349609      +  4  20\n",
       "132   51  7309550  7309686  chr1   23261  CAMTA1  NM_001349609      +  5  19"
      ]
     },
     "execution_count": 51,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "exon.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>row</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>chrom</th>\n",
       "      <th>GeneID</th>\n",
       "      <th>name2</th>\n",
       "      <th>name</th>\n",
       "      <th>strand</th>\n",
       "      <th>+</th>\n",
       "      <th>-</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1520</th>\n",
       "      <td>1195</td>\n",
       "      <td>154227753</td>\n",
       "      <td>154227875</td>\n",
       "      <td>chrX</td>\n",
       "      <td>2157</td>\n",
       "      <td>F8</td>\n",
       "      <td>NM_000132</td>\n",
       "      <td>-</td>\n",
       "      <td>25</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1521</th>\n",
       "      <td>1195</td>\n",
       "      <td>154250684</td>\n",
       "      <td>154250998</td>\n",
       "      <td>chrX</td>\n",
       "      <td>2157</td>\n",
       "      <td>F8</td>\n",
       "      <td>NM_000132</td>\n",
       "      <td>-</td>\n",
       "      <td>26</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5081</th>\n",
       "      <td>15481</td>\n",
       "      <td>154487519</td>\n",
       "      <td>154490514</td>\n",
       "      <td>chrX</td>\n",
       "      <td>116442</td>\n",
       "      <td>RAB39B</td>\n",
       "      <td>NM_171998</td>\n",
       "      <td>-</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5082</th>\n",
       "      <td>15481</td>\n",
       "      <td>154493358</td>\n",
       "      <td>154493776</td>\n",
       "      <td>chrX</td>\n",
       "      <td>116442</td>\n",
       "      <td>RAB39B</td>\n",
       "      <td>NM_171998</td>\n",
       "      <td>-</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5130</th>\n",
       "      <td>18757</td>\n",
       "      <td>2654895</td>\n",
       "      <td>2655723</td>\n",
       "      <td>chrY</td>\n",
       "      <td>6736</td>\n",
       "      <td>SRY</td>\n",
       "      <td>NM_003140</td>\n",
       "      <td>-</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        row      start        end chrom  GeneID   name2       name strand   +  \\\n",
       "1520   1195  154227753  154227875  chrX    2157      F8  NM_000132      -  25   \n",
       "1521   1195  154250684  154250998  chrX    2157      F8  NM_000132      -  26   \n",
       "5081  15481  154487519  154490514  chrX  116442  RAB39B  NM_171998      -   1   \n",
       "5082  15481  154493358  154493776  chrX  116442  RAB39B  NM_171998      -   2   \n",
       "5130  18757    2654895    2655723  chrY    6736     SRY  NM_003140      -   1   \n",
       "\n",
       "      -  \n",
       "1520  2  \n",
       "1521  1  \n",
       "5081  2  \n",
       "5082  1  \n",
       "5130  1  "
      ]
     },
     "execution_count": 52,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "exon.tail()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [],
   "source": [
    "exon['exon'] = pd.concat([exon.loc[exon['strand'] == '+', '+'], exon.loc[exon['strand'] == '-', '-']])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>row</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>chrom</th>\n",
       "      <th>GeneID</th>\n",
       "      <th>name2</th>\n",
       "      <th>name</th>\n",
       "      <th>strand</th>\n",
       "      <th>+</th>\n",
       "      <th>-</th>\n",
       "      <th>exon</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>128</th>\n",
       "      <td>51</td>\n",
       "      <td>6845513</td>\n",
       "      <td>6845635</td>\n",
       "      <td>chr1</td>\n",
       "      <td>23261</td>\n",
       "      <td>CAMTA1</td>\n",
       "      <td>NM_001349609</td>\n",
       "      <td>+</td>\n",
       "      <td>1</td>\n",
       "      <td>23</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>129</th>\n",
       "      <td>51</td>\n",
       "      <td>6880240</td>\n",
       "      <td>6880310</td>\n",
       "      <td>chr1</td>\n",
       "      <td>23261</td>\n",
       "      <td>CAMTA1</td>\n",
       "      <td>NM_001349609</td>\n",
       "      <td>+</td>\n",
       "      <td>2</td>\n",
       "      <td>22</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>130</th>\n",
       "      <td>51</td>\n",
       "      <td>6885151</td>\n",
       "      <td>6885270</td>\n",
       "      <td>chr1</td>\n",
       "      <td>23261</td>\n",
       "      <td>CAMTA1</td>\n",
       "      <td>NM_001349609</td>\n",
       "      <td>+</td>\n",
       "      <td>3</td>\n",
       "      <td>21</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>131</th>\n",
       "      <td>51</td>\n",
       "      <td>7151363</td>\n",
       "      <td>7151431</td>\n",
       "      <td>chr1</td>\n",
       "      <td>23261</td>\n",
       "      <td>CAMTA1</td>\n",
       "      <td>NM_001349609</td>\n",
       "      <td>+</td>\n",
       "      <td>4</td>\n",
       "      <td>20</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>132</th>\n",
       "      <td>51</td>\n",
       "      <td>7309550</td>\n",
       "      <td>7309686</td>\n",
       "      <td>chr1</td>\n",
       "      <td>23261</td>\n",
       "      <td>CAMTA1</td>\n",
       "      <td>NM_001349609</td>\n",
       "      <td>+</td>\n",
       "      <td>5</td>\n",
       "      <td>19</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     row    start      end chrom  GeneID   name2          name strand  +   -  \\\n",
       "128   51  6845513  6845635  chr1   23261  CAMTA1  NM_001349609      +  1  23   \n",
       "129   51  6880240  6880310  chr1   23261  CAMTA1  NM_001349609      +  2  22   \n",
       "130   51  6885151  6885270  chr1   23261  CAMTA1  NM_001349609      +  3  21   \n",
       "131   51  7151363  7151431  chr1   23261  CAMTA1  NM_001349609      +  4  20   \n",
       "132   51  7309550  7309686  chr1   23261  CAMTA1  NM_001349609      +  5  19   \n",
       "\n",
       "     exon  \n",
       "128     1  \n",
       "129     2  \n",
       "130     3  \n",
       "131     4  \n",
       "132     5  "
      ]
     },
     "execution_count": 54,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "exon.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>row</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>chrom</th>\n",
       "      <th>GeneID</th>\n",
       "      <th>name2</th>\n",
       "      <th>name</th>\n",
       "      <th>strand</th>\n",
       "      <th>+</th>\n",
       "      <th>-</th>\n",
       "      <th>exon</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1520</th>\n",
       "      <td>1195</td>\n",
       "      <td>154227753</td>\n",
       "      <td>154227875</td>\n",
       "      <td>chrX</td>\n",
       "      <td>2157</td>\n",
       "      <td>F8</td>\n",
       "      <td>NM_000132</td>\n",
       "      <td>-</td>\n",
       "      <td>25</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1521</th>\n",
       "      <td>1195</td>\n",
       "      <td>154250684</td>\n",
       "      <td>154250998</td>\n",
       "      <td>chrX</td>\n",
       "      <td>2157</td>\n",
       "      <td>F8</td>\n",
       "      <td>NM_000132</td>\n",
       "      <td>-</td>\n",
       "      <td>26</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5081</th>\n",
       "      <td>15481</td>\n",
       "      <td>154487519</td>\n",
       "      <td>154490514</td>\n",
       "      <td>chrX</td>\n",
       "      <td>116442</td>\n",
       "      <td>RAB39B</td>\n",
       "      <td>NM_171998</td>\n",
       "      <td>-</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5082</th>\n",
       "      <td>15481</td>\n",
       "      <td>154493358</td>\n",
       "      <td>154493776</td>\n",
       "      <td>chrX</td>\n",
       "      <td>116442</td>\n",
       "      <td>RAB39B</td>\n",
       "      <td>NM_171998</td>\n",
       "      <td>-</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5130</th>\n",
       "      <td>18757</td>\n",
       "      <td>2654895</td>\n",
       "      <td>2655723</td>\n",
       "      <td>chrY</td>\n",
       "      <td>6736</td>\n",
       "      <td>SRY</td>\n",
       "      <td>NM_003140</td>\n",
       "      <td>-</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        row      start        end chrom  GeneID   name2       name strand   +  \\\n",
       "1520   1195  154227753  154227875  chrX    2157      F8  NM_000132      -  25   \n",
       "1521   1195  154250684  154250998  chrX    2157      F8  NM_000132      -  26   \n",
       "5081  15481  154487519  154490514  chrX  116442  RAB39B  NM_171998      -   1   \n",
       "5082  15481  154493358  154493776  chrX  116442  RAB39B  NM_171998      -   2   \n",
       "5130  18757    2654895    2655723  chrY    6736     SRY  NM_003140      -   1   \n",
       "\n",
       "      -  exon  \n",
       "1520  2     2  \n",
       "1521  1     1  \n",
       "5081  2     2  \n",
       "5082  1     1  \n",
       "5130  1     1  "
      ]
     },
     "execution_count": 55,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "exon.tail()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {},
   "outputs": [],
   "source": [
    "exon['last_exon'] = exon.groupby(['name2', 'name'])['exon'].transform('max') == exon['exon']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>row</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>chrom</th>\n",
       "      <th>GeneID</th>\n",
       "      <th>name2</th>\n",
       "      <th>name</th>\n",
       "      <th>strand</th>\n",
       "      <th>+</th>\n",
       "      <th>-</th>\n",
       "      <th>exon</th>\n",
       "      <th>last_exon</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>128</th>\n",
       "      <td>51</td>\n",
       "      <td>6845513</td>\n",
       "      <td>6845635</td>\n",
       "      <td>chr1</td>\n",
       "      <td>23261</td>\n",
       "      <td>CAMTA1</td>\n",
       "      <td>NM_001349609</td>\n",
       "      <td>+</td>\n",
       "      <td>1</td>\n",
       "      <td>23</td>\n",
       "      <td>1</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>129</th>\n",
       "      <td>51</td>\n",
       "      <td>6880240</td>\n",
       "      <td>6880310</td>\n",
       "      <td>chr1</td>\n",
       "      <td>23261</td>\n",
       "      <td>CAMTA1</td>\n",
       "      <td>NM_001349609</td>\n",
       "      <td>+</td>\n",
       "      <td>2</td>\n",
       "      <td>22</td>\n",
       "      <td>2</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>130</th>\n",
       "      <td>51</td>\n",
       "      <td>6885151</td>\n",
       "      <td>6885270</td>\n",
       "      <td>chr1</td>\n",
       "      <td>23261</td>\n",
       "      <td>CAMTA1</td>\n",
       "      <td>NM_001349609</td>\n",
       "      <td>+</td>\n",
       "      <td>3</td>\n",
       "      <td>21</td>\n",
       "      <td>3</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>131</th>\n",
       "      <td>51</td>\n",
       "      <td>7151363</td>\n",
       "      <td>7151431</td>\n",
       "      <td>chr1</td>\n",
       "      <td>23261</td>\n",
       "      <td>CAMTA1</td>\n",
       "      <td>NM_001349609</td>\n",
       "      <td>+</td>\n",
       "      <td>4</td>\n",
       "      <td>20</td>\n",
       "      <td>4</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>132</th>\n",
       "      <td>51</td>\n",
       "      <td>7309550</td>\n",
       "      <td>7309686</td>\n",
       "      <td>chr1</td>\n",
       "      <td>23261</td>\n",
       "      <td>CAMTA1</td>\n",
       "      <td>NM_001349609</td>\n",
       "      <td>+</td>\n",
       "      <td>5</td>\n",
       "      <td>19</td>\n",
       "      <td>5</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     row    start      end chrom  GeneID   name2          name strand  +   -  \\\n",
       "128   51  6845513  6845635  chr1   23261  CAMTA1  NM_001349609      +  1  23   \n",
       "129   51  6880240  6880310  chr1   23261  CAMTA1  NM_001349609      +  2  22   \n",
       "130   51  6885151  6885270  chr1   23261  CAMTA1  NM_001349609      +  3  21   \n",
       "131   51  7151363  7151431  chr1   23261  CAMTA1  NM_001349609      +  4  20   \n",
       "132   51  7309550  7309686  chr1   23261  CAMTA1  NM_001349609      +  5  19   \n",
       "\n",
       "     exon  last_exon  \n",
       "128     1      False  \n",
       "129     2      False  \n",
       "130     3      False  \n",
       "131     4      False  \n",
       "132     5      False  "
      ]
     },
     "execution_count": 57,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "exon.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>row</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>chrom</th>\n",
       "      <th>GeneID</th>\n",
       "      <th>name2</th>\n",
       "      <th>name</th>\n",
       "      <th>strand</th>\n",
       "      <th>+</th>\n",
       "      <th>-</th>\n",
       "      <th>exon</th>\n",
       "      <th>last_exon</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1520</th>\n",
       "      <td>1195</td>\n",
       "      <td>154227753</td>\n",
       "      <td>154227875</td>\n",
       "      <td>chrX</td>\n",
       "      <td>2157</td>\n",
       "      <td>F8</td>\n",
       "      <td>NM_000132</td>\n",
       "      <td>-</td>\n",
       "      <td>25</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1521</th>\n",
       "      <td>1195</td>\n",
       "      <td>154250684</td>\n",
       "      <td>154250998</td>\n",
       "      <td>chrX</td>\n",
       "      <td>2157</td>\n",
       "      <td>F8</td>\n",
       "      <td>NM_000132</td>\n",
       "      <td>-</td>\n",
       "      <td>26</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5081</th>\n",
       "      <td>15481</td>\n",
       "      <td>154487519</td>\n",
       "      <td>154490514</td>\n",
       "      <td>chrX</td>\n",
       "      <td>116442</td>\n",
       "      <td>RAB39B</td>\n",
       "      <td>NM_171998</td>\n",
       "      <td>-</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5082</th>\n",
       "      <td>15481</td>\n",
       "      <td>154493358</td>\n",
       "      <td>154493776</td>\n",
       "      <td>chrX</td>\n",
       "      <td>116442</td>\n",
       "      <td>RAB39B</td>\n",
       "      <td>NM_171998</td>\n",
       "      <td>-</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5130</th>\n",
       "      <td>18757</td>\n",
       "      <td>2654895</td>\n",
       "      <td>2655723</td>\n",
       "      <td>chrY</td>\n",
       "      <td>6736</td>\n",
       "      <td>SRY</td>\n",
       "      <td>NM_003140</td>\n",
       "      <td>-</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        row      start        end chrom  GeneID   name2       name strand   +  \\\n",
       "1520   1195  154227753  154227875  chrX    2157      F8  NM_000132      -  25   \n",
       "1521   1195  154250684  154250998  chrX    2157      F8  NM_000132      -  26   \n",
       "5081  15481  154487519  154490514  chrX  116442  RAB39B  NM_171998      -   1   \n",
       "5082  15481  154493358  154493776  chrX  116442  RAB39B  NM_171998      -   2   \n",
       "5130  18757    2654895    2655723  chrY    6736     SRY  NM_003140      -   1   \n",
       "\n",
       "      -  exon  last_exon  \n",
       "1520  2     2      False  \n",
       "1521  1     1      False  \n",
       "5081  2     2       True  \n",
       "5082  1     1      False  \n",
       "5130  1     1       True  "
      ]
     },
     "execution_count": 58,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "exon.tail()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {},
   "outputs": [],
   "source": [
    "exon = exon[\n",
    "    ['chrom', 'start', 'end', 'GeneID', 'name2', 'name', 'exon', 'last_exon']\n",
    "].sort_values(['chrom', 'start', 'end'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>GeneID</th>\n",
       "      <th>name2</th>\n",
       "      <th>name</th>\n",
       "      <th>exon</th>\n",
       "      <th>last_exon</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>128</th>\n",
       "      <td>chr1</td>\n",
       "      <td>6845513</td>\n",
       "      <td>6845635</td>\n",
       "      <td>23261</td>\n",
       "      <td>CAMTA1</td>\n",
       "      <td>NM_001349609</td>\n",
       "      <td>1</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>129</th>\n",
       "      <td>chr1</td>\n",
       "      <td>6880240</td>\n",
       "      <td>6880310</td>\n",
       "      <td>23261</td>\n",
       "      <td>CAMTA1</td>\n",
       "      <td>NM_001349609</td>\n",
       "      <td>2</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>130</th>\n",
       "      <td>chr1</td>\n",
       "      <td>6885151</td>\n",
       "      <td>6885270</td>\n",
       "      <td>23261</td>\n",
       "      <td>CAMTA1</td>\n",
       "      <td>NM_001349609</td>\n",
       "      <td>3</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>131</th>\n",
       "      <td>chr1</td>\n",
       "      <td>7151363</td>\n",
       "      <td>7151431</td>\n",
       "      <td>23261</td>\n",
       "      <td>CAMTA1</td>\n",
       "      <td>NM_001349609</td>\n",
       "      <td>4</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>132</th>\n",
       "      <td>chr1</td>\n",
       "      <td>7309550</td>\n",
       "      <td>7309686</td>\n",
       "      <td>23261</td>\n",
       "      <td>CAMTA1</td>\n",
       "      <td>NM_001349609</td>\n",
       "      <td>5</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    chrom    start      end  GeneID   name2          name  exon  last_exon\n",
       "128  chr1  6845513  6845635   23261  CAMTA1  NM_001349609     1      False\n",
       "129  chr1  6880240  6880310   23261  CAMTA1  NM_001349609     2      False\n",
       "130  chr1  6885151  6885270   23261  CAMTA1  NM_001349609     3      False\n",
       "131  chr1  7151363  7151431   23261  CAMTA1  NM_001349609     4      False\n",
       "132  chr1  7309550  7309686   23261  CAMTA1  NM_001349609     5      False"
      ]
     },
     "execution_count": 60,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "exon.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {},
   "outputs": [],
   "source": [
    "exon.rename(columns={\n",
    "    'chrom': '#chrom',\n",
    "    'GeneID': 'gene_id', 'name2': 'symbol', 'name': 'transcript'\n",
    "}).to_csv(hi_exon_file, index=False, sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "metadata": {},
   "outputs": [],
   "source": [
    "!bgzip -cf {hi_exon_file} > {hi_exon_file}.gz"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "metadata": {},
   "outputs": [],
   "source": [
    "!tabix -fp bed {hi_exon_file}.gz\n",
    "!rm {hi_exon_file}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "metadata": {},
   "outputs": [],
   "source": [
    "last_exon = exon[exon['last_exon']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>GeneID</th>\n",
       "      <th>name2</th>\n",
       "      <th>name</th>\n",
       "      <th>exon</th>\n",
       "      <th>last_exon</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>150</th>\n",
       "      <td>chr1</td>\n",
       "      <td>7826518</td>\n",
       "      <td>7829766</td>\n",
       "      <td>23261</td>\n",
       "      <td>CAMTA1</td>\n",
       "      <td>NM_001349609</td>\n",
       "      <td>23</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3882</th>\n",
       "      <td>chr1</td>\n",
       "      <td>17345220</td>\n",
       "      <td>17345453</td>\n",
       "      <td>6390</td>\n",
       "      <td>SDHB</td>\n",
       "      <td>NM_003000</td>\n",
       "      <td>8</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2565</th>\n",
       "      <td>chr1</td>\n",
       "      <td>27105513</td>\n",
       "      <td>27108595</td>\n",
       "      <td>8289</td>\n",
       "      <td>ARID1A</td>\n",
       "      <td>NM_139135</td>\n",
       "      <td>20</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5033</th>\n",
       "      <td>chr1</td>\n",
       "      <td>27860755</td>\n",
       "      <td>27861427</td>\n",
       "      <td>27245</td>\n",
       "      <td>AHDC1</td>\n",
       "      <td>NM_001371928</td>\n",
       "      <td>9</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4072</th>\n",
       "      <td>chr1</td>\n",
       "      <td>43391025</td>\n",
       "      <td>43392912</td>\n",
       "      <td>6513</td>\n",
       "      <td>SLC2A1</td>\n",
       "      <td>NM_006516</td>\n",
       "      <td>10</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     chrom     start       end  GeneID   name2          name  exon  last_exon\n",
       "150   chr1   7826518   7829766   23261  CAMTA1  NM_001349609    23       True\n",
       "3882  chr1  17345220  17345453    6390    SDHB     NM_003000     8       True\n",
       "2565  chr1  27105513  27108595    8289  ARID1A     NM_139135    20       True\n",
       "5033  chr1  27860755  27861427   27245   AHDC1  NM_001371928     9       True\n",
       "4072  chr1  43391025  43392912    6513  SLC2A1     NM_006516    10       True"
      ]
     },
     "execution_count": 65,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "last_exon.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "metadata": {},
   "outputs": [],
   "source": [
    "last_exon_region = last_exon['chrom'] + ':' + last_exon['start'].astype(str) + '-' + last_exon['end'].astype(str)\n",
    "last_exon_region = last_exon_region.str.replace('chr', '')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "150       1:7826518-7829766\n",
       "3882    1:17345220-17345453\n",
       "2565    1:27105513-27108595\n",
       "5033    1:27860755-27861427\n",
       "4072    1:43391025-43392912\n",
       "dtype: object"
      ]
     },
     "execution_count": 67,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "last_exon_region.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "metadata": {},
   "outputs": [],
   "source": [
    "need_fields = [\n",
    "    'variants/CHROM', 'variants/POS', 'variants/REF', 'variants/ALT',\n",
    "    'variants/AF_ESP', 'variants/AF_EXAC', 'variants/AF_TGP', 'variants/CLNSIG'\n",
    "]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "metadata": {},
   "outputs": [],
   "source": [
    "!tabix -f {clinvar_ori_vcf_file}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "299it [00:00, 5543.95it/s]\n"
     ]
    }
   ],
   "source": [
    "\n",
    "with open(clinvar_file, 'w') as f:\n",
    "    headers = allel.read_vcf_headers(clinvar_ori_vcf_file)\n",
    "    f.write(''.join(headers.headers))\n",
    "    \n",
    "    def fetch_variants(region):\n",
    "        fields, samples, headers, it = allel.iter_vcf_chunks(\n",
    "            clinvar_ori_vcf_file, fields=need_fields, alt_number=1, region=region\n",
    "        )\n",
    "        for variants, *_ in it:\n",
    "            esp_filter = np.isnan(variants['variants/AF_ESP'])\n",
    "            esp_filter[~esp_filter] |= variants['variants/AF_ESP'][~esp_filter] < 0.01\n",
    "\n",
    "            exac_filter = np.isnan(variants['variants/AF_EXAC'])\n",
    "            exac_filter[~exac_filter] |= variants['variants/AF_EXAC'][~exac_filter] < 0.01\n",
    "\n",
    "            tgp_filter = np.isnan(variants['variants/AF_TGP'])\n",
    "            tgp_filter[~tgp_filter] |= variants['variants/AF_TGP'][~tgp_filter] < 0.01\n",
    "\n",
    "            pathogenic_filter = np.isin(\n",
    "                variants['variants/CLNSIG'], ['Likely_pathogenic', 'Pathogenic', 'Pathogenic/Likely_pathogenic']\n",
    "            )\n",
    "\n",
    "            af_filter = esp_filter & exac_filter & tgp_filter & pathogenic_filter\n",
    "\n",
    "            filtered_variants = {k: v[af_filter] for k, v in variants.items()}\n",
    "\n",
    "            filtered_variants['variants/CHROM'] = 'chr' + filtered_variants['variants/CHROM']\n",
    "\n",
    "            return allel.normalize_callset(filtered_variants)\n",
    "    \n",
    "    with Pool(processes=32) as pool:\n",
    "        variants = pool.map(fetch_variants, last_exon_region)\n",
    "    \n",
    "    for names, callset in tqdm(filter(lambda x: x is not None, variants)):\n",
    "        allel.write_vcf_data(f, names, callset, None, {field: np.nan for field in need_fields})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "sed: can't read s/AF_.\\+=nan;//: No such file or directory\n"
     ]
    }
   ],
   "source": [
    "!sed -i '' 's/AF_.\\+=nan;//' {clinvar_file}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "metadata": {},
   "outputs": [],
   "source": [
    "!bgzip -cf {clinvar_file} > {clinvar_file}.gz"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "metadata": {},
   "outputs": [],
   "source": [
    "!tabix -fp vcf {clinvar_file}.gz\n",
    "!rm {clinvar_file}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "metadata": {},
   "outputs": [],
   "source": [
    "uhi_genes = set(\n",
    "    curation_gene.loc[curation_gene['Haploinsufficiency Score'] == '40', '#Gene Symbol']\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "metadata": {},
   "outputs": [],
   "source": [
    "uhi_gene = refgene_info.loc[\n",
    "    refgene_info['name2'].isin(uhi_genes),\n",
    "    ['chrom', 'txStart', 'txEnd', 'GeneID', 'name2', 'name', 'strand']\n",
    "].sort_values(['chrom', 'txStart', 'txEnd'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>txStart</th>\n",
       "      <th>txEnd</th>\n",
       "      <th>GeneID</th>\n",
       "      <th>name2</th>\n",
       "      <th>name</th>\n",
       "      <th>strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>5019</th>\n",
       "      <td>chr1</td>\n",
       "      <td>32117864</td>\n",
       "      <td>32169618</td>\n",
       "      <td>1307</td>\n",
       "      <td>COL16A1</td>\n",
       "      <td>NM_001856</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17600</th>\n",
       "      <td>chr1</td>\n",
       "      <td>40974390</td>\n",
       "      <td>40981722</td>\n",
       "      <td>64789</td>\n",
       "      <td>EXO5</td>\n",
       "      <td>NM_001346956</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8348</th>\n",
       "      <td>chr1</td>\n",
       "      <td>55505220</td>\n",
       "      <td>55530525</td>\n",
       "      <td>255738</td>\n",
       "      <td>PCSK9</td>\n",
       "      <td>NM_174936</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8405</th>\n",
       "      <td>chr1</td>\n",
       "      <td>171154438</td>\n",
       "      <td>171181824</td>\n",
       "      <td>2327</td>\n",
       "      <td>FMO2</td>\n",
       "      <td>NM_001460</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9978</th>\n",
       "      <td>chr10</td>\n",
       "      <td>74870212</td>\n",
       "      <td>74891581</td>\n",
       "      <td>25961</td>\n",
       "      <td>NUDT13</td>\n",
       "      <td>NM_015901</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       chrom    txStart      txEnd  GeneID    name2          name strand\n",
       "5019    chr1   32117864   32169618    1307  COL16A1     NM_001856      -\n",
       "17600   chr1   40974390   40981722   64789     EXO5  NM_001346956      +\n",
       "8348    chr1   55505220   55530525  255738    PCSK9     NM_174936      +\n",
       "8405    chr1  171154438  171181824    2327     FMO2     NM_001460      +\n",
       "9978   chr10   74870212   74891581   25961   NUDT13     NM_015901      +"
      ]
     },
     "execution_count": 76,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "uhi_gene.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "metadata": {},
   "outputs": [],
   "source": [
    "uhi_gene.rename(columns={\n",
    "    'chrom': '#chrom', 'txStart': 'start', 'txEnd': 'end',\n",
    "    'GeneID': 'gene_id', 'name2': 'symbol', 'name': 'transcript'\n",
    "}).to_csv(uhi_gene_file, index=False, sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "metadata": {},
   "outputs": [],
   "source": [
    "!bgzip -cf {uhi_gene_file} > {uhi_gene_file}.gz"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "metadata": {},
   "outputs": [],
   "source": [
    "!tabix -fp bed {uhi_gene_file}.gz\n",
    "!rm {uhi_gene_file}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "metadata": {},
   "outputs": [],
   "source": [
    "region = pd.read_csv(clingen_region_ori_file, sep='\\t', skiprows=5, dtype=str)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>#ISCA ID</th>\n",
       "      <th>ISCA Region Name</th>\n",
       "      <th>cytoBand</th>\n",
       "      <th>Genomic Location</th>\n",
       "      <th>Haploinsufficiency Score</th>\n",
       "      <th>Haploinsufficiency Description</th>\n",
       "      <th>Haploinsufficiency PMID1</th>\n",
       "      <th>Haploinsufficiency PMID2</th>\n",
       "      <th>Haploinsufficiency PMID3</th>\n",
       "      <th>Haploinsufficiency PMID4</th>\n",
       "      <th>Haploinsufficiency PMID5</th>\n",
       "      <th>Haploinsufficiency PMID6</th>\n",
       "      <th>Triplosensitivity Score</th>\n",
       "      <th>Triplosensitivity Description</th>\n",
       "      <th>Triplosensitivity PMID1</th>\n",
       "      <th>Triplosensitivity PMID2</th>\n",
       "      <th>Triplosensitivity PMID3</th>\n",
       "      <th>Triplosensitivity PMID4</th>\n",
       "      <th>Triplosensitivity PMID5</th>\n",
       "      <th>Triplosensitivity PMID6</th>\n",
       "      <th>Date Last Evaluated</th>\n",
       "      <th>Loss phenotype OMIM ID</th>\n",
       "      <th>Triplosensitive phenotype OMIM ID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ISCA-46739</td>\n",
       "      <td>Yq11.223 population region (DGV_Gold_Standard_...</td>\n",
       "      <td>Yq11.223</td>\n",
       "      <td>chrY:25853189-26146959</td>\n",
       "      <td>40</td>\n",
       "      <td>Dosage sensitivity unlikely</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>No evidence available</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2021-08-03</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ISCA-46738</td>\n",
       "      <td>Yq11.223 population region (DGV_Gold_Standard_...</td>\n",
       "      <td>Yq11.223</td>\n",
       "      <td>chrY:24361866-24465171</td>\n",
       "      <td>40</td>\n",
       "      <td>Dosage sensitivity unlikely</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>No evidence available</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2021-08-03</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ISCA-46737</td>\n",
       "      <td>Xq27.3 population region (DGV_Gold_Standard_Ju...</td>\n",
       "      <td>Xq27.3</td>\n",
       "      <td>chrX:142601775-142800327</td>\n",
       "      <td>40</td>\n",
       "      <td>Dosage sensitivity unlikely</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>No evidence available</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2021-08-03</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ISCA-46736</td>\n",
       "      <td>Xq21.1 population region (DGV_Gold_Standard_Ju...</td>\n",
       "      <td>Xq21.1</td>\n",
       "      <td>chrX:76137628-76141022</td>\n",
       "      <td>40</td>\n",
       "      <td>Dosage sensitivity unlikely</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>No evidence available</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2021-08-03</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ISCA-46735</td>\n",
       "      <td>Xp11.23 population region (DGV_Gold_Standard_J...</td>\n",
       "      <td>Xp11.23</td>\n",
       "      <td>chrX:47880294-47952112</td>\n",
       "      <td>40</td>\n",
       "      <td>Dosage sensitivity unlikely</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>No evidence available</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2021-08-03</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     #ISCA ID                                   ISCA Region Name  cytoBand  \\\n",
       "0  ISCA-46739  Yq11.223 population region (DGV_Gold_Standard_...  Yq11.223   \n",
       "1  ISCA-46738  Yq11.223 population region (DGV_Gold_Standard_...  Yq11.223   \n",
       "2  ISCA-46737  Xq27.3 population region (DGV_Gold_Standard_Ju...    Xq27.3   \n",
       "3  ISCA-46736  Xq21.1 population region (DGV_Gold_Standard_Ju...    Xq21.1   \n",
       "4  ISCA-46735  Xp11.23 population region (DGV_Gold_Standard_J...   Xp11.23   \n",
       "\n",
       "           Genomic Location Haploinsufficiency Score  \\\n",
       "0    chrY:25853189-26146959                       40   \n",
       "1    chrY:24361866-24465171                       40   \n",
       "2  chrX:142601775-142800327                       40   \n",
       "3    chrX:76137628-76141022                       40   \n",
       "4    chrX:47880294-47952112                       40   \n",
       "\n",
       "  Haploinsufficiency Description Haploinsufficiency PMID1  \\\n",
       "0    Dosage sensitivity unlikely                      NaN   \n",
       "1    Dosage sensitivity unlikely                      NaN   \n",
       "2    Dosage sensitivity unlikely                      NaN   \n",
       "3    Dosage sensitivity unlikely                      NaN   \n",
       "4    Dosage sensitivity unlikely                      NaN   \n",
       "\n",
       "  Haploinsufficiency PMID2 Haploinsufficiency PMID3 Haploinsufficiency PMID4  \\\n",
       "0                      NaN                      NaN                      NaN   \n",
       "1                      NaN                      NaN                      NaN   \n",
       "2                      NaN                      NaN                      NaN   \n",
       "3                      NaN                      NaN                      NaN   \n",
       "4                      NaN                      NaN                      NaN   \n",
       "\n",
       "  Haploinsufficiency PMID5 Haploinsufficiency PMID6 Triplosensitivity Score  \\\n",
       "0                      NaN                      NaN                       0   \n",
       "1                      NaN                      NaN                       0   \n",
       "2                      NaN                      NaN                       0   \n",
       "3                      NaN                      NaN                       0   \n",
       "4                      NaN                      NaN                       0   \n",
       "\n",
       "  Triplosensitivity Description Triplosensitivity PMID1  \\\n",
       "0         No evidence available                     NaN   \n",
       "1         No evidence available                     NaN   \n",
       "2         No evidence available                     NaN   \n",
       "3         No evidence available                     NaN   \n",
       "4         No evidence available                     NaN   \n",
       "\n",
       "  Triplosensitivity PMID2 Triplosensitivity PMID3 Triplosensitivity PMID4  \\\n",
       "0                     NaN                     NaN                     NaN   \n",
       "1                     NaN                     NaN                     NaN   \n",
       "2                     NaN                     NaN                     NaN   \n",
       "3                     NaN                     NaN                     NaN   \n",
       "4                     NaN                     NaN                     NaN   \n",
       "\n",
       "  Triplosensitivity PMID5 Triplosensitivity PMID6 Date Last Evaluated  \\\n",
       "0                     NaN                     NaN          2021-08-03   \n",
       "1                     NaN                     NaN          2021-08-03   \n",
       "2                     NaN                     NaN          2021-08-03   \n",
       "3                     NaN                     NaN          2021-08-03   \n",
       "4                     NaN                     NaN          2021-08-03   \n",
       "\n",
       "  Loss phenotype OMIM ID Triplosensitive phenotype OMIM ID  \n",
       "0                    NaN                               NaN  \n",
       "1                    NaN                               NaN  \n",
       "2                    NaN                               NaN  \n",
       "3                    NaN                               NaN  \n",
       "4                    NaN                               NaN  "
      ]
     },
     "execution_count": 81,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "region.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "metadata": {},
   "outputs": [],
   "source": [
    "position = region['Genomic Location'].str.extract(r'(?P<chrom>chr\\w+)\\s*:\\s*(?P<start>\\d+)\\s*-\\s*(?P<end>\\d+)')\n",
    "position['start'] = position['start'].astype(int)\n",
    "position['end'] = position['end'].astype(int)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chrY</td>\n",
       "      <td>25853189</td>\n",
       "      <td>26146959</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chrY</td>\n",
       "      <td>24361866</td>\n",
       "      <td>24465171</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chrX</td>\n",
       "      <td>142601775</td>\n",
       "      <td>142800327</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chrX</td>\n",
       "      <td>76137628</td>\n",
       "      <td>76141022</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chrX</td>\n",
       "      <td>47880294</td>\n",
       "      <td>47952112</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  chrom      start        end\n",
       "0  chrY   25853189   26146959\n",
       "1  chrY   24361866   24465171\n",
       "2  chrX  142601775  142800327\n",
       "3  chrX   76137628   76141022\n",
       "4  chrX   47880294   47952112"
      ]
     },
     "execution_count": 83,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "position.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "metadata": {},
   "outputs": [],
   "source": [
    "region_pos = region.merge(position, how='left', left_index=True, right_index=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>#ISCA ID</th>\n",
       "      <th>ISCA Region Name</th>\n",
       "      <th>cytoBand</th>\n",
       "      <th>Genomic Location</th>\n",
       "      <th>Haploinsufficiency Score</th>\n",
       "      <th>Haploinsufficiency Description</th>\n",
       "      <th>Haploinsufficiency PMID1</th>\n",
       "      <th>Haploinsufficiency PMID2</th>\n",
       "      <th>Haploinsufficiency PMID3</th>\n",
       "      <th>Haploinsufficiency PMID4</th>\n",
       "      <th>Haploinsufficiency PMID5</th>\n",
       "      <th>Haploinsufficiency PMID6</th>\n",
       "      <th>Triplosensitivity Score</th>\n",
       "      <th>Triplosensitivity Description</th>\n",
       "      <th>Triplosensitivity PMID1</th>\n",
       "      <th>Triplosensitivity PMID2</th>\n",
       "      <th>Triplosensitivity PMID3</th>\n",
       "      <th>Triplosensitivity PMID4</th>\n",
       "      <th>Triplosensitivity PMID5</th>\n",
       "      <th>Triplosensitivity PMID6</th>\n",
       "      <th>Date Last Evaluated</th>\n",
       "      <th>Loss phenotype OMIM ID</th>\n",
       "      <th>Triplosensitive phenotype OMIM ID</th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ISCA-46739</td>\n",
       "      <td>Yq11.223 population region (DGV_Gold_Standard_...</td>\n",
       "      <td>Yq11.223</td>\n",
       "      <td>chrY:25853189-26146959</td>\n",
       "      <td>40</td>\n",
       "      <td>Dosage sensitivity unlikely</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>No evidence available</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2021-08-03</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>chrY</td>\n",
       "      <td>25853189</td>\n",
       "      <td>26146959</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ISCA-46738</td>\n",
       "      <td>Yq11.223 population region (DGV_Gold_Standard_...</td>\n",
       "      <td>Yq11.223</td>\n",
       "      <td>chrY:24361866-24465171</td>\n",
       "      <td>40</td>\n",
       "      <td>Dosage sensitivity unlikely</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>No evidence available</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2021-08-03</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>chrY</td>\n",
       "      <td>24361866</td>\n",
       "      <td>24465171</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ISCA-46737</td>\n",
       "      <td>Xq27.3 population region (DGV_Gold_Standard_Ju...</td>\n",
       "      <td>Xq27.3</td>\n",
       "      <td>chrX:142601775-142800327</td>\n",
       "      <td>40</td>\n",
       "      <td>Dosage sensitivity unlikely</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>No evidence available</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2021-08-03</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>chrX</td>\n",
       "      <td>142601775</td>\n",
       "      <td>142800327</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ISCA-46736</td>\n",
       "      <td>Xq21.1 population region (DGV_Gold_Standard_Ju...</td>\n",
       "      <td>Xq21.1</td>\n",
       "      <td>chrX:76137628-76141022</td>\n",
       "      <td>40</td>\n",
       "      <td>Dosage sensitivity unlikely</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>No evidence available</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2021-08-03</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>chrX</td>\n",
       "      <td>76137628</td>\n",
       "      <td>76141022</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ISCA-46735</td>\n",
       "      <td>Xp11.23 population region (DGV_Gold_Standard_J...</td>\n",
       "      <td>Xp11.23</td>\n",
       "      <td>chrX:47880294-47952112</td>\n",
       "      <td>40</td>\n",
       "      <td>Dosage sensitivity unlikely</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>No evidence available</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2021-08-03</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>chrX</td>\n",
       "      <td>47880294</td>\n",
       "      <td>47952112</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     #ISCA ID                                   ISCA Region Name  cytoBand  \\\n",
       "0  ISCA-46739  Yq11.223 population region (DGV_Gold_Standard_...  Yq11.223   \n",
       "1  ISCA-46738  Yq11.223 population region (DGV_Gold_Standard_...  Yq11.223   \n",
       "2  ISCA-46737  Xq27.3 population region (DGV_Gold_Standard_Ju...    Xq27.3   \n",
       "3  ISCA-46736  Xq21.1 population region (DGV_Gold_Standard_Ju...    Xq21.1   \n",
       "4  ISCA-46735  Xp11.23 population region (DGV_Gold_Standard_J...   Xp11.23   \n",
       "\n",
       "           Genomic Location Haploinsufficiency Score  \\\n",
       "0    chrY:25853189-26146959                       40   \n",
       "1    chrY:24361866-24465171                       40   \n",
       "2  chrX:142601775-142800327                       40   \n",
       "3    chrX:76137628-76141022                       40   \n",
       "4    chrX:47880294-47952112                       40   \n",
       "\n",
       "  Haploinsufficiency Description Haploinsufficiency PMID1  \\\n",
       "0    Dosage sensitivity unlikely                      NaN   \n",
       "1    Dosage sensitivity unlikely                      NaN   \n",
       "2    Dosage sensitivity unlikely                      NaN   \n",
       "3    Dosage sensitivity unlikely                      NaN   \n",
       "4    Dosage sensitivity unlikely                      NaN   \n",
       "\n",
       "  Haploinsufficiency PMID2 Haploinsufficiency PMID3 Haploinsufficiency PMID4  \\\n",
       "0                      NaN                      NaN                      NaN   \n",
       "1                      NaN                      NaN                      NaN   \n",
       "2                      NaN                      NaN                      NaN   \n",
       "3                      NaN                      NaN                      NaN   \n",
       "4                      NaN                      NaN                      NaN   \n",
       "\n",
       "  Haploinsufficiency PMID5 Haploinsufficiency PMID6 Triplosensitivity Score  \\\n",
       "0                      NaN                      NaN                       0   \n",
       "1                      NaN                      NaN                       0   \n",
       "2                      NaN                      NaN                       0   \n",
       "3                      NaN                      NaN                       0   \n",
       "4                      NaN                      NaN                       0   \n",
       "\n",
       "  Triplosensitivity Description Triplosensitivity PMID1  \\\n",
       "0         No evidence available                     NaN   \n",
       "1         No evidence available                     NaN   \n",
       "2         No evidence available                     NaN   \n",
       "3         No evidence available                     NaN   \n",
       "4         No evidence available                     NaN   \n",
       "\n",
       "  Triplosensitivity PMID2 Triplosensitivity PMID3 Triplosensitivity PMID4  \\\n",
       "0                     NaN                     NaN                     NaN   \n",
       "1                     NaN                     NaN                     NaN   \n",
       "2                     NaN                     NaN                     NaN   \n",
       "3                     NaN                     NaN                     NaN   \n",
       "4                     NaN                     NaN                     NaN   \n",
       "\n",
       "  Triplosensitivity PMID5 Triplosensitivity PMID6 Date Last Evaluated  \\\n",
       "0                     NaN                     NaN          2021-08-03   \n",
       "1                     NaN                     NaN          2021-08-03   \n",
       "2                     NaN                     NaN          2021-08-03   \n",
       "3                     NaN                     NaN          2021-08-03   \n",
       "4                     NaN                     NaN          2021-08-03   \n",
       "\n",
       "  Loss phenotype OMIM ID Triplosensitive phenotype OMIM ID chrom      start  \\\n",
       "0                    NaN                               NaN  chrY   25853189   \n",
       "1                    NaN                               NaN  chrY   24361866   \n",
       "2                    NaN                               NaN  chrX  142601775   \n",
       "3                    NaN                               NaN  chrX   76137628   \n",
       "4                    NaN                               NaN  chrX   47880294   \n",
       "\n",
       "         end  \n",
       "0   26146959  \n",
       "1   24465171  \n",
       "2  142800327  \n",
       "3   76141022  \n",
       "4   47952112  "
      ]
     },
     "execution_count": 85,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "region_pos.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "metadata": {},
   "outputs": [],
   "source": [
    "func_region = region_pos.loc[\n",
    "    (region_pos['Haploinsufficiency Score'].isin(['1', '2','3']))\n",
    "     | (region_pos['Triplosensitivity Score'].isin(['1', '2','3'])),\n",
    "     ['chrom', 'start', 'end', '#ISCA ID', 'ISCA Region Name']\n",
    "].sort_values(['chrom', 'start', 'end'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 87,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>#ISCA ID</th>\n",
       "      <th>ISCA Region Name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>477</th>\n",
       "      <td>chr1</td>\n",
       "      <td>834083</td>\n",
       "      <td>6289973</td>\n",
       "      <td>ISCA-37434</td>\n",
       "      <td>1p36 terminal region (includes GABRD)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>483</th>\n",
       "      <td>chr1</td>\n",
       "      <td>145386507</td>\n",
       "      <td>145748064</td>\n",
       "      <td>ISCA-37428</td>\n",
       "      <td>1q21.1 recurrent (TAR syndrome) region (proxim...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>486</th>\n",
       "      <td>chr1</td>\n",
       "      <td>146577486</td>\n",
       "      <td>147394506</td>\n",
       "      <td>ISCA-37421</td>\n",
       "      <td>1q21.1 recurrent region (BP3-BP4, distal) (inc...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>452</th>\n",
       "      <td>chr1</td>\n",
       "      <td>243287730</td>\n",
       "      <td>245318287</td>\n",
       "      <td>ISCA-37493</td>\n",
       "      <td>1q43q44 terminal region (includes AKT3)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>485</th>\n",
       "      <td>chr10</td>\n",
       "      <td>81682843</td>\n",
       "      <td>88739388</td>\n",
       "      <td>ISCA-37424</td>\n",
       "      <td>10q22.3q23.2 recurrent region (LCR 3-LCR 4) (I...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     chrom      start        end    #ISCA ID  \\\n",
       "477   chr1     834083    6289973  ISCA-37434   \n",
       "483   chr1  145386507  145748064  ISCA-37428   \n",
       "486   chr1  146577486  147394506  ISCA-37421   \n",
       "452   chr1  243287730  245318287  ISCA-37493   \n",
       "485  chr10   81682843   88739388  ISCA-37424   \n",
       "\n",
       "                                      ISCA Region Name  \n",
       "477              1p36 terminal region (includes GABRD)  \n",
       "483  1q21.1 recurrent (TAR syndrome) region (proxim...  \n",
       "486  1q21.1 recurrent region (BP3-BP4, distal) (inc...  \n",
       "452            1q43q44 terminal region (includes AKT3)  \n",
       "485  10q22.3q23.2 recurrent region (LCR 3-LCR 4) (I...  "
      ]
     },
     "execution_count": 87,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "func_region.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 88,
   "metadata": {},
   "outputs": [],
   "source": [
    "func_region.rename(columns={\n",
    "    'chrom': '#chrom',\n",
    "    '#ISCA ID': 'isca_id', 'ISCA Region Name': 'name'\n",
    "}).to_csv(func_region_file, sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 89,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "!bgzip -cf {func_region_file} > {func_region_file}.gz"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 90,
   "metadata": {},
   "outputs": [],
   "source": [
    "!tabix -fp bed {func_region_file}.gz\n",
    "!rm {func_region_file}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 91,
   "metadata": {},
   "outputs": [],
   "source": [
    "hi_region = region_pos.loc[\n",
    "    region_pos['Haploinsufficiency Score'] == '3',\n",
    "    ['chrom', 'start', 'end', '#ISCA ID', 'ISCA Region Name']\n",
    "].sort_values(['chrom', 'start', 'end'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 92,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>#ISCA ID</th>\n",
       "      <th>ISCA Region Name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>477</th>\n",
       "      <td>chr1</td>\n",
       "      <td>834083</td>\n",
       "      <td>6289973</td>\n",
       "      <td>ISCA-37434</td>\n",
       "      <td>1p36 terminal region (includes GABRD)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>486</th>\n",
       "      <td>chr1</td>\n",
       "      <td>146577486</td>\n",
       "      <td>147394506</td>\n",
       "      <td>ISCA-37421</td>\n",
       "      <td>1q21.1 recurrent region (BP3-BP4, distal) (inc...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>452</th>\n",
       "      <td>chr1</td>\n",
       "      <td>243287730</td>\n",
       "      <td>245318287</td>\n",
       "      <td>ISCA-37493</td>\n",
       "      <td>1q43q44 terminal region (includes AKT3)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>485</th>\n",
       "      <td>chr10</td>\n",
       "      <td>81682843</td>\n",
       "      <td>88739388</td>\n",
       "      <td>ISCA-37424</td>\n",
       "      <td>10q22.3q23.2 recurrent region (LCR 3-LCR 4) (I...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>497</th>\n",
       "      <td>chr11</td>\n",
       "      <td>31803509</td>\n",
       "      <td>32510988</td>\n",
       "      <td>ISCA-37401</td>\n",
       "      <td>11p13 (WAGR syndrome) region</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     chrom      start        end    #ISCA ID  \\\n",
       "477   chr1     834083    6289973  ISCA-37434   \n",
       "486   chr1  146577486  147394506  ISCA-37421   \n",
       "452   chr1  243287730  245318287  ISCA-37493   \n",
       "485  chr10   81682843   88739388  ISCA-37424   \n",
       "497  chr11   31803509   32510988  ISCA-37401   \n",
       "\n",
       "                                      ISCA Region Name  \n",
       "477              1p36 terminal region (includes GABRD)  \n",
       "486  1q21.1 recurrent region (BP3-BP4, distal) (inc...  \n",
       "452            1q43q44 terminal region (includes AKT3)  \n",
       "485  10q22.3q23.2 recurrent region (LCR 3-LCR 4) (I...  \n",
       "497                       11p13 (WAGR syndrome) region  "
      ]
     },
     "execution_count": 92,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "hi_region.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 93,
   "metadata": {},
   "outputs": [],
   "source": [
    "hi_region['omim_genes'] = hi_region.apply(\n",
    "    lambda row: ','.join(omim_gene.loc[\n",
    "        (omim_gene['chrom'] == row['chrom'])\n",
    "        & (omim_gene['txEnd'] >= row['start'])\n",
    "        & (omim_gene['txStart'] <= row['end']),\n",
    "        'name2'\n",
    "    ]),\n",
    "    axis=1\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 94,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>#ISCA ID</th>\n",
       "      <th>ISCA Region Name</th>\n",
       "      <th>omim_genes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>477</th>\n",
       "      <td>chr1</td>\n",
       "      <td>834083</td>\n",
       "      <td>6289973</td>\n",
       "      <td>ISCA-37434</td>\n",
       "      <td>1p36 terminal region (includes GABRD)</td>\n",
       "      <td>ISG15,AGRN,TNFRSF4,B3GALT6,DVL1,VWA1,ATAD3A,TM...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>486</th>\n",
       "      <td>chr1</td>\n",
       "      <td>146577486</td>\n",
       "      <td>147394506</td>\n",
       "      <td>ISCA-37421</td>\n",
       "      <td>1q21.1 recurrent region (BP3-BP4, distal) (inc...</td>\n",
       "      <td>GJA5,GJA8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>452</th>\n",
       "      <td>chr1</td>\n",
       "      <td>243287730</td>\n",
       "      <td>245318287</td>\n",
       "      <td>ISCA-37493</td>\n",
       "      <td>1q43q44 terminal region (includes AKT3)</td>\n",
       "      <td>SDCCAG8,AKT3,ZBTB18,COX20,HNRNPU</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>485</th>\n",
       "      <td>chr10</td>\n",
       "      <td>81682843</td>\n",
       "      <td>88739388</td>\n",
       "      <td>ISCA-37424</td>\n",
       "      <td>10q22.3q23.2 recurrent region (LCR 3-LCR 4) (I...</td>\n",
       "      <td>ANXA11,MAT1A,CDHR1,RGR,LDB3,BMPR1A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>497</th>\n",
       "      <td>chr11</td>\n",
       "      <td>31803509</td>\n",
       "      <td>32510988</td>\n",
       "      <td>ISCA-37401</td>\n",
       "      <td>11p13 (WAGR syndrome) region</td>\n",
       "      <td>ELP4,PAX6,WT1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     chrom      start        end    #ISCA ID  \\\n",
       "477   chr1     834083    6289973  ISCA-37434   \n",
       "486   chr1  146577486  147394506  ISCA-37421   \n",
       "452   chr1  243287730  245318287  ISCA-37493   \n",
       "485  chr10   81682843   88739388  ISCA-37424   \n",
       "497  chr11   31803509   32510988  ISCA-37401   \n",
       "\n",
       "                                      ISCA Region Name  \\\n",
       "477              1p36 terminal region (includes GABRD)   \n",
       "486  1q21.1 recurrent region (BP3-BP4, distal) (inc...   \n",
       "452            1q43q44 terminal region (includes AKT3)   \n",
       "485  10q22.3q23.2 recurrent region (LCR 3-LCR 4) (I...   \n",
       "497                       11p13 (WAGR syndrome) region   \n",
       "\n",
       "                                            omim_genes  \n",
       "477  ISG15,AGRN,TNFRSF4,B3GALT6,DVL1,VWA1,ATAD3A,TM...  \n",
       "486                                          GJA5,GJA8  \n",
       "452                   SDCCAG8,AKT3,ZBTB18,COX20,HNRNPU  \n",
       "485                 ANXA11,MAT1A,CDHR1,RGR,LDB3,BMPR1A  \n",
       "497                                      ELP4,PAX6,WT1  "
      ]
     },
     "execution_count": 94,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "hi_region.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 95,
   "metadata": {},
   "outputs": [],
   "source": [
    "hi_region.rename(columns={\n",
    "    'chrom': '#chrom',\n",
    "    '#ISCA ID': 'isca_id', 'ISCA Region Name': 'name'\n",
    "}).to_csv(hi_region_file, sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 96,
   "metadata": {},
   "outputs": [],
   "source": [
    "!bgzip -cf {hi_region_file} > {hi_region_file}.gz"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 97,
   "metadata": {},
   "outputs": [],
   "source": [
    "!tabix -fp bed {hi_region_file}.gz\n",
    "!rm {hi_region_file}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 98,
   "metadata": {},
   "outputs": [],
   "source": [
    "uhi_region = region_pos.loc[\n",
    "    region_pos['Haploinsufficiency Score'] == '40',\n",
    "    ['chrom', 'start', 'end', '#ISCA ID', 'ISCA Region Name']\n",
    "].sort_values(['chrom', 'start', 'end'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 99,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>#ISCA ID</th>\n",
       "      <th>ISCA Region Name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>429</th>\n",
       "      <td>chr1</td>\n",
       "      <td>363999</td>\n",
       "      <td>401750</td>\n",
       "      <td>ISCA-46306</td>\n",
       "      <td>1p36.33 population region (gnomAD-SV_v2.1_DEL_...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>428</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1640999</td>\n",
       "      <td>1647000</td>\n",
       "      <td>ISCA-46307</td>\n",
       "      <td>1p36.33 population region (gnomAD-SV_v2.1_DEL_...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>176</th>\n",
       "      <td>chr1</td>\n",
       "      <td>12886641</td>\n",
       "      <td>12942069</td>\n",
       "      <td>ISCA-46561</td>\n",
       "      <td>1p36.21 population region (DGV_Gold_Standard_J...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>426</th>\n",
       "      <td>chr1</td>\n",
       "      <td>12896424</td>\n",
       "      <td>12930275</td>\n",
       "      <td>ISCA-46309</td>\n",
       "      <td>1p36.21 population region (gnomAD-SV_v2.1_DEL_...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>175</th>\n",
       "      <td>chr1</td>\n",
       "      <td>12899395</td>\n",
       "      <td>12924283</td>\n",
       "      <td>ISCA-46562</td>\n",
       "      <td>1p36.21 population region (DGV_Gold_Standard_J...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    chrom     start       end    #ISCA ID  \\\n",
       "429  chr1    363999    401750  ISCA-46306   \n",
       "428  chr1   1640999   1647000  ISCA-46307   \n",
       "176  chr1  12886641  12942069  ISCA-46561   \n",
       "426  chr1  12896424  12930275  ISCA-46309   \n",
       "175  chr1  12899395  12924283  ISCA-46562   \n",
       "\n",
       "                                      ISCA Region Name  \n",
       "429  1p36.33 population region (gnomAD-SV_v2.1_DEL_...  \n",
       "428  1p36.33 population region (gnomAD-SV_v2.1_DEL_...  \n",
       "176  1p36.21 population region (DGV_Gold_Standard_J...  \n",
       "426  1p36.21 population region (gnomAD-SV_v2.1_DEL_...  \n",
       "175  1p36.21 population region (DGV_Gold_Standard_J...  "
      ]
     },
     "execution_count": 99,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "uhi_region.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 100,
   "metadata": {},
   "outputs": [],
   "source": [
    "uhi_region['genes'] = uhi_region.apply(\n",
    "    lambda row: ','.join(gene.loc[\n",
    "        (gene['chrom'] == row['chrom'])\n",
    "        & (gene['txEnd'] >= row['start'])\n",
    "        & (gene['txStart'] <= row['end']),\n",
    "        'name2'\n",
    "    ]),\n",
    "    axis=1\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 101,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>#ISCA ID</th>\n",
       "      <th>ISCA Region Name</th>\n",
       "      <th>genes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>429</th>\n",
       "      <td>chr1</td>\n",
       "      <td>363999</td>\n",
       "      <td>401750</td>\n",
       "      <td>ISCA-46306</td>\n",
       "      <td>1p36.33 population region (gnomAD-SV_v2.1_DEL_...</td>\n",
       "      <td>OR4F29</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>428</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1640999</td>\n",
       "      <td>1647000</td>\n",
       "      <td>ISCA-46307</td>\n",
       "      <td>1p36.33 population region (gnomAD-SV_v2.1_DEL_...</td>\n",
       "      <td>CDK11B,CDK11A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>176</th>\n",
       "      <td>chr1</td>\n",
       "      <td>12886641</td>\n",
       "      <td>12942069</td>\n",
       "      <td>ISCA-46561</td>\n",
       "      <td>1p36.21 population region (DGV_Gold_Standard_J...</td>\n",
       "      <td>PRAMEF11,HNRNPCL1,PRAMEF2,PRAMEF4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>426</th>\n",
       "      <td>chr1</td>\n",
       "      <td>12896424</td>\n",
       "      <td>12930275</td>\n",
       "      <td>ISCA-46309</td>\n",
       "      <td>1p36.21 population region (gnomAD-SV_v2.1_DEL_...</td>\n",
       "      <td>HNRNPCL1,PRAMEF2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>175</th>\n",
       "      <td>chr1</td>\n",
       "      <td>12899395</td>\n",
       "      <td>12924283</td>\n",
       "      <td>ISCA-46562</td>\n",
       "      <td>1p36.21 population region (DGV_Gold_Standard_J...</td>\n",
       "      <td>HNRNPCL1,PRAMEF2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    chrom     start       end    #ISCA ID  \\\n",
       "429  chr1    363999    401750  ISCA-46306   \n",
       "428  chr1   1640999   1647000  ISCA-46307   \n",
       "176  chr1  12886641  12942069  ISCA-46561   \n",
       "426  chr1  12896424  12930275  ISCA-46309   \n",
       "175  chr1  12899395  12924283  ISCA-46562   \n",
       "\n",
       "                                      ISCA Region Name  \\\n",
       "429  1p36.33 population region (gnomAD-SV_v2.1_DEL_...   \n",
       "428  1p36.33 population region (gnomAD-SV_v2.1_DEL_...   \n",
       "176  1p36.21 population region (DGV_Gold_Standard_J...   \n",
       "426  1p36.21 population region (gnomAD-SV_v2.1_DEL_...   \n",
       "175  1p36.21 population region (DGV_Gold_Standard_J...   \n",
       "\n",
       "                                 genes  \n",
       "429                             OR4F29  \n",
       "428                      CDK11B,CDK11A  \n",
       "176  PRAMEF11,HNRNPCL1,PRAMEF2,PRAMEF4  \n",
       "426                   HNRNPCL1,PRAMEF2  \n",
       "175                   HNRNPCL1,PRAMEF2  "
      ]
     },
     "execution_count": 101,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "uhi_region.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 102,
   "metadata": {},
   "outputs": [],
   "source": [
    "uhi_region.rename(columns={\n",
    "    'chrom': '#chrom',\n",
    "    '#ISCA ID': 'isca_id', 'ISCA Region Name': 'name'\n",
    "}).to_csv(uhi_region_file, sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 103,
   "metadata": {},
   "outputs": [],
   "source": [
    "!bgzip -cf {uhi_region_file} > {uhi_region_file}.gz"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 104,
   "metadata": {},
   "outputs": [],
   "source": [
    "!tabix -fp bed {uhi_region_file}.gz\n",
    "!rm {uhi_region_file}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 105,
   "metadata": {},
   "outputs": [],
   "source": [
    "decipher = pd.read_csv(hi_pred_ori_file, sep='\\t',skiprows=1, header=None, usecols=[3,])\n",
    "decipher = decipher[3].str.split('|', expand=True).rename(columns={0: 'symbol', 1: 'hi_score', 2: 'hi_index'})\n",
    "decipher['hi_index'] = decipher['hi_index'].str.replace('%', '').astype(float)\n",
    "decipher = decipher.merge(gene, left_on='symbol', right_on='name2')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 106,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>symbol</th>\n",
       "      <th>hi_score</th>\n",
       "      <th>hi_index</th>\n",
       "      <th>chrom</th>\n",
       "      <th>txStart</th>\n",
       "      <th>txEnd</th>\n",
       "      <th>GeneID</th>\n",
       "      <th>name2</th>\n",
       "      <th>name</th>\n",
       "      <th>strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ANXA2R</td>\n",
       "      <td>0.000047349</td>\n",
       "      <td>100.00</td>\n",
       "      <td>chr5</td>\n",
       "      <td>43039472</td>\n",
       "      <td>43043349</td>\n",
       "      <td>389289</td>\n",
       "      <td>ANXA2R</td>\n",
       "      <td>NM_001382352</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SCGB1D1</td>\n",
       "      <td>0.000054551</td>\n",
       "      <td>99.99</td>\n",
       "      <td>chr11</td>\n",
       "      <td>61957687</td>\n",
       "      <td>61961011</td>\n",
       "      <td>10648</td>\n",
       "      <td>SCGB1D1</td>\n",
       "      <td>NM_006552</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>IL31</td>\n",
       "      <td>0.000057228</td>\n",
       "      <td>99.99</td>\n",
       "      <td>chr12</td>\n",
       "      <td>122656575</td>\n",
       "      <td>122658768</td>\n",
       "      <td>386653</td>\n",
       "      <td>IL31</td>\n",
       "      <td>NM_001014336</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>BPIFA2</td>\n",
       "      <td>0.000061264</td>\n",
       "      <td>99.98</td>\n",
       "      <td>chr20</td>\n",
       "      <td>31749575</td>\n",
       "      <td>31769218</td>\n",
       "      <td>140683</td>\n",
       "      <td>BPIFA2</td>\n",
       "      <td>NM_001319164</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>MUCL1</td>\n",
       "      <td>0.000067779</td>\n",
       "      <td>99.98</td>\n",
       "      <td>chr12</td>\n",
       "      <td>55248299</td>\n",
       "      <td>55252171</td>\n",
       "      <td>118430</td>\n",
       "      <td>MUCL1</td>\n",
       "      <td>NM_058173</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    symbol     hi_score  hi_index  chrom    txStart      txEnd  GeneID  \\\n",
       "0   ANXA2R  0.000047349    100.00   chr5   43039472   43043349  389289   \n",
       "1  SCGB1D1  0.000054551     99.99  chr11   61957687   61961011   10648   \n",
       "2     IL31  0.000057228     99.99  chr12  122656575  122658768  386653   \n",
       "3   BPIFA2  0.000061264     99.98  chr20   31749575   31769218  140683   \n",
       "4    MUCL1  0.000067779     99.98  chr12   55248299   55252171  118430   \n",
       "\n",
       "     name2          name strand  \n",
       "0   ANXA2R  NM_001382352      -  \n",
       "1  SCGB1D1     NM_006552      +  \n",
       "2     IL31  NM_001014336      -  \n",
       "3   BPIFA2  NM_001319164      +  \n",
       "4    MUCL1     NM_058173      +  "
      ]
     },
     "execution_count": 106,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "decipher.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 107,
   "metadata": {},
   "outputs": [],
   "source": [
    "gnomad = pd.read_csv(gnomad_lof_ori_file, sep='\\t', index_col=0, compression='gzip')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 108,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>transcript</th>\n",
       "      <th>obs_mis</th>\n",
       "      <th>exp_mis</th>\n",
       "      <th>oe_mis</th>\n",
       "      <th>mu_mis</th>\n",
       "      <th>possible_mis</th>\n",
       "      <th>obs_mis_pphen</th>\n",
       "      <th>exp_mis_pphen</th>\n",
       "      <th>oe_mis_pphen</th>\n",
       "      <th>possible_mis_pphen</th>\n",
       "      <th>obs_syn</th>\n",
       "      <th>exp_syn</th>\n",
       "      <th>oe_syn</th>\n",
       "      <th>mu_syn</th>\n",
       "      <th>possible_syn</th>\n",
       "      <th>obs_lof</th>\n",
       "      <th>mu_lof</th>\n",
       "      <th>possible_lof</th>\n",
       "      <th>exp_lof</th>\n",
       "      <th>pLI</th>\n",
       "      <th>pNull</th>\n",
       "      <th>pRec</th>\n",
       "      <th>oe_lof</th>\n",
       "      <th>oe_syn_lower</th>\n",
       "      <th>oe_syn_upper</th>\n",
       "      <th>oe_mis_lower</th>\n",
       "      <th>oe_mis_upper</th>\n",
       "      <th>oe_lof_lower</th>\n",
       "      <th>oe_lof_upper</th>\n",
       "      <th>constraint_flag</th>\n",
       "      <th>syn_z</th>\n",
       "      <th>mis_z</th>\n",
       "      <th>lof_z</th>\n",
       "      <th>oe_lof_upper_rank</th>\n",
       "      <th>oe_lof_upper_bin</th>\n",
       "      <th>oe_lof_upper_bin_6</th>\n",
       "      <th>n_sites</th>\n",
       "      <th>classic_caf</th>\n",
       "      <th>max_af</th>\n",
       "      <th>no_lofs</th>\n",
       "      <th>obs_het_lof</th>\n",
       "      <th>obs_hom_lof</th>\n",
       "      <th>defined</th>\n",
       "      <th>p</th>\n",
       "      <th>exp_hom_lof</th>\n",
       "      <th>classic_caf_afr</th>\n",
       "      <th>classic_caf_amr</th>\n",
       "      <th>classic_caf_asj</th>\n",
       "      <th>classic_caf_eas</th>\n",
       "      <th>classic_caf_fin</th>\n",
       "      <th>classic_caf_nfe</th>\n",
       "      <th>classic_caf_oth</th>\n",
       "      <th>classic_caf_sas</th>\n",
       "      <th>p_afr</th>\n",
       "      <th>p_amr</th>\n",
       "      <th>p_asj</th>\n",
       "      <th>p_eas</th>\n",
       "      <th>p_fin</th>\n",
       "      <th>p_nfe</th>\n",
       "      <th>p_oth</th>\n",
       "      <th>p_sas</th>\n",
       "      <th>transcript_type</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>transcript_level</th>\n",
       "      <th>cds_length</th>\n",
       "      <th>num_coding_exons</th>\n",
       "      <th>gene_type</th>\n",
       "      <th>gene_length</th>\n",
       "      <th>exac_pLI</th>\n",
       "      <th>exac_obs_lof</th>\n",
       "      <th>exac_exp_lof</th>\n",
       "      <th>exac_oe_lof</th>\n",
       "      <th>brain_expression</th>\n",
       "      <th>chromosome</th>\n",
       "      <th>start_position</th>\n",
       "      <th>end_position</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>MED13</th>\n",
       "      <td>ENST00000397786</td>\n",
       "      <td>871</td>\n",
       "      <td>1117.80</td>\n",
       "      <td>0.77921</td>\n",
       "      <td>0.000056</td>\n",
       "      <td>14195</td>\n",
       "      <td>314.0</td>\n",
       "      <td>529.75</td>\n",
       "      <td>0.59273</td>\n",
       "      <td>6708.0</td>\n",
       "      <td>422</td>\n",
       "      <td>387.53</td>\n",
       "      <td>1.0890</td>\n",
       "      <td>0.000019</td>\n",
       "      <td>4248</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000005</td>\n",
       "      <td>1257.0</td>\n",
       "      <td>98.429</td>\n",
       "      <td>1.0</td>\n",
       "      <td>8.943600e-40</td>\n",
       "      <td>1.838300e-16</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.005</td>\n",
       "      <td>1.180</td>\n",
       "      <td>0.736</td>\n",
       "      <td>0.824</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.030</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-1.376500</td>\n",
       "      <td>2.6232</td>\n",
       "      <td>9.1935</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.000012</td>\n",
       "      <td>0.000008</td>\n",
       "      <td>124782.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>124785.0</td>\n",
       "      <td>0.000012</td>\n",
       "      <td>0.000018</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000093</td>\n",
       "      <td>0.000009</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000093</td>\n",
       "      <td>0.000009</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>ENSG00000108510</td>\n",
       "      <td>2</td>\n",
       "      <td>6522</td>\n",
       "      <td>30</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>122678</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>64.393</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>17</td>\n",
       "      <td>60019966</td>\n",
       "      <td>60142643</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NIPBL</th>\n",
       "      <td>ENST00000282516</td>\n",
       "      <td>846</td>\n",
       "      <td>1441.50</td>\n",
       "      <td>0.58688</td>\n",
       "      <td>0.000074</td>\n",
       "      <td>18540</td>\n",
       "      <td>158.0</td>\n",
       "      <td>543.10</td>\n",
       "      <td>0.29092</td>\n",
       "      <td>7135.0</td>\n",
       "      <td>496</td>\n",
       "      <td>495.01</td>\n",
       "      <td>1.0020</td>\n",
       "      <td>0.000025</td>\n",
       "      <td>5211</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.000009</td>\n",
       "      <td>1781.0</td>\n",
       "      <td>150.320</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.977300e-59</td>\n",
       "      <td>3.572400e-24</td>\n",
       "      <td>0.006653</td>\n",
       "      <td>0.930</td>\n",
       "      <td>1.079</td>\n",
       "      <td>0.554</td>\n",
       "      <td>0.621</td>\n",
       "      <td>0.001</td>\n",
       "      <td>0.032</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.035119</td>\n",
       "      <td>5.5737</td>\n",
       "      <td>11.2860</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.000012</td>\n",
       "      <td>0.000008</td>\n",
       "      <td>125693.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>125696.0</td>\n",
       "      <td>0.000012</td>\n",
       "      <td>0.000018</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000099</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000065</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000099</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000065</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>ENSG00000164190</td>\n",
       "      <td>2</td>\n",
       "      <td>8412</td>\n",
       "      <td>46</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>189655</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>110.570</td>\n",
       "      <td>0.009044</td>\n",
       "      <td>NaN</td>\n",
       "      <td>5</td>\n",
       "      <td>36876861</td>\n",
       "      <td>37066515</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SMC3</th>\n",
       "      <td>ENST00000361804</td>\n",
       "      <td>178</td>\n",
       "      <td>630.07</td>\n",
       "      <td>0.28251</td>\n",
       "      <td>0.000032</td>\n",
       "      <td>8109</td>\n",
       "      <td>21.0</td>\n",
       "      <td>182.52</td>\n",
       "      <td>0.11506</td>\n",
       "      <td>2197.0</td>\n",
       "      <td>215</td>\n",
       "      <td>203.25</td>\n",
       "      <td>1.0578</td>\n",
       "      <td>0.000010</td>\n",
       "      <td>2091</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000005</td>\n",
       "      <td>937.0</td>\n",
       "      <td>79.490</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.785300e-32</td>\n",
       "      <td>2.191400e-13</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.946</td>\n",
       "      <td>1.184</td>\n",
       "      <td>0.249</td>\n",
       "      <td>0.320</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.037</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.647760</td>\n",
       "      <td>6.3999</td>\n",
       "      <td>8.2618</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>0.000032</td>\n",
       "      <td>0.000004</td>\n",
       "      <td>125731.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>125739.0</td>\n",
       "      <td>0.000032</td>\n",
       "      <td>0.000127</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000099</td>\n",
       "      <td>0.000054</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000044</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000033</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000099</td>\n",
       "      <td>0.000054</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000044</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000033</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>ENSG00000108055</td>\n",
       "      <td>2</td>\n",
       "      <td>3651</td>\n",
       "      <td>29</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>36946</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>58.523</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>10</td>\n",
       "      <td>112327449</td>\n",
       "      <td>112364394</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CNOT1</th>\n",
       "      <td>ENST00000317147</td>\n",
       "      <td>561</td>\n",
       "      <td>1295.90</td>\n",
       "      <td>0.43290</td>\n",
       "      <td>0.000069</td>\n",
       "      <td>15670</td>\n",
       "      <td>51.0</td>\n",
       "      <td>290.68</td>\n",
       "      <td>0.17545</td>\n",
       "      <td>3560.0</td>\n",
       "      <td>470</td>\n",
       "      <td>456.03</td>\n",
       "      <td>1.0306</td>\n",
       "      <td>0.000024</td>\n",
       "      <td>4564</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.000007</td>\n",
       "      <td>1440.0</td>\n",
       "      <td>125.030</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.992400e-49</td>\n",
       "      <td>4.562900e-20</td>\n",
       "      <td>0.007998</td>\n",
       "      <td>0.955</td>\n",
       "      <td>1.112</td>\n",
       "      <td>0.403</td>\n",
       "      <td>0.464</td>\n",
       "      <td>0.002</td>\n",
       "      <td>0.038</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.514100</td>\n",
       "      <td>7.2546</td>\n",
       "      <td>10.2790</td>\n",
       "      <td>3.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>0.000020</td>\n",
       "      <td>0.000004</td>\n",
       "      <td>125740.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>125744.0</td>\n",
       "      <td>0.000016</td>\n",
       "      <td>0.000032</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000029</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000055</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000026</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000029</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000054</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000018</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>ENSG00000125107</td>\n",
       "      <td>2</td>\n",
       "      <td>7128</td>\n",
       "      <td>48</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>109936</td>\n",
       "      <td>1.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>90.130</td>\n",
       "      <td>0.033285</td>\n",
       "      <td>NaN</td>\n",
       "      <td>16</td>\n",
       "      <td>58553855</td>\n",
       "      <td>58663790</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RLF</th>\n",
       "      <td>ENST00000372771</td>\n",
       "      <td>669</td>\n",
       "      <td>972.87</td>\n",
       "      <td>0.68766</td>\n",
       "      <td>0.000047</td>\n",
       "      <td>12682</td>\n",
       "      <td>107.0</td>\n",
       "      <td>321.14</td>\n",
       "      <td>0.33319</td>\n",
       "      <td>4151.0</td>\n",
       "      <td>358</td>\n",
       "      <td>352.62</td>\n",
       "      <td>1.0153</td>\n",
       "      <td>0.000017</td>\n",
       "      <td>3482</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000004</td>\n",
       "      <td>1024.0</td>\n",
       "      <td>73.222</td>\n",
       "      <td>1.0</td>\n",
       "      <td>8.405500e-30</td>\n",
       "      <td>2.284200e-12</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.930</td>\n",
       "      <td>1.108</td>\n",
       "      <td>0.645</td>\n",
       "      <td>0.733</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.040</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.225180</td>\n",
       "      <td>3.4620</td>\n",
       "      <td>7.9294</td>\n",
       "      <td>4.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.000004</td>\n",
       "      <td>0.000004</td>\n",
       "      <td>125122.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>125123.0</td>\n",
       "      <td>0.000004</td>\n",
       "      <td>0.000002</td>\n",
       "      <td>0.000062</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000062</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>ENSG00000117000</td>\n",
       "      <td>2</td>\n",
       "      <td>5742</td>\n",
       "      <td>8</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>79549</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>43.607</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>40627045</td>\n",
       "      <td>40706593</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            transcript  obs_mis  exp_mis   oe_mis    mu_mis  possible_mis  \\\n",
       "gene                                                                        \n",
       "MED13  ENST00000397786      871  1117.80  0.77921  0.000056         14195   \n",
       "NIPBL  ENST00000282516      846  1441.50  0.58688  0.000074         18540   \n",
       "SMC3   ENST00000361804      178   630.07  0.28251  0.000032          8109   \n",
       "CNOT1  ENST00000317147      561  1295.90  0.43290  0.000069         15670   \n",
       "RLF    ENST00000372771      669   972.87  0.68766  0.000047         12682   \n",
       "\n",
       "       obs_mis_pphen  exp_mis_pphen  oe_mis_pphen  possible_mis_pphen  \\\n",
       "gene                                                                    \n",
       "MED13          314.0         529.75       0.59273              6708.0   \n",
       "NIPBL          158.0         543.10       0.29092              7135.0   \n",
       "SMC3            21.0         182.52       0.11506              2197.0   \n",
       "CNOT1           51.0         290.68       0.17545              3560.0   \n",
       "RLF            107.0         321.14       0.33319              4151.0   \n",
       "\n",
       "       obs_syn  exp_syn  oe_syn    mu_syn  possible_syn  obs_lof    mu_lof  \\\n",
       "gene                                                                         \n",
       "MED13      422   387.53  1.0890  0.000019          4248      0.0  0.000005   \n",
       "NIPBL      496   495.01  1.0020  0.000025          5211      1.0  0.000009   \n",
       "SMC3       215   203.25  1.0578  0.000010          2091      0.0  0.000005   \n",
       "CNOT1      470   456.03  1.0306  0.000024          4564      1.0  0.000007   \n",
       "RLF        358   352.62  1.0153  0.000017          3482      0.0  0.000004   \n",
       "\n",
       "       possible_lof  exp_lof  pLI         pNull          pRec    oe_lof  \\\n",
       "gene                                                                      \n",
       "MED13        1257.0   98.429  1.0  8.943600e-40  1.838300e-16  0.000000   \n",
       "NIPBL        1781.0  150.320  1.0  2.977300e-59  3.572400e-24  0.006653   \n",
       "SMC3          937.0   79.490  1.0  2.785300e-32  2.191400e-13  0.000000   \n",
       "CNOT1        1440.0  125.030  1.0  2.992400e-49  4.562900e-20  0.007998   \n",
       "RLF          1024.0   73.222  1.0  8.405500e-30  2.284200e-12  0.000000   \n",
       "\n",
       "       oe_syn_lower  oe_syn_upper  oe_mis_lower  oe_mis_upper  oe_lof_lower  \\\n",
       "gene                                                                          \n",
       "MED13         1.005         1.180         0.736         0.824         0.000   \n",
       "NIPBL         0.930         1.079         0.554         0.621         0.001   \n",
       "SMC3          0.946         1.184         0.249         0.320         0.000   \n",
       "CNOT1         0.955         1.112         0.403         0.464         0.002   \n",
       "RLF           0.930         1.108         0.645         0.733         0.000   \n",
       "\n",
       "       oe_lof_upper constraint_flag     syn_z   mis_z    lof_z  \\\n",
       "gene                                                             \n",
       "MED13         0.030             NaN -1.376500  2.6232   9.1935   \n",
       "NIPBL         0.032             NaN -0.035119  5.5737  11.2860   \n",
       "SMC3          0.037             NaN -0.647760  6.3999   8.2618   \n",
       "CNOT1         0.038             NaN -0.514100  7.2546  10.2790   \n",
       "RLF           0.040             NaN -0.225180  3.4620   7.9294   \n",
       "\n",
       "       oe_lof_upper_rank  oe_lof_upper_bin  oe_lof_upper_bin_6  n_sites  \\\n",
       "gene                                                                      \n",
       "MED13                0.0               0.0                 0.0      2.0   \n",
       "NIPBL                1.0               0.0                 0.0      2.0   \n",
       "SMC3                 2.0               0.0                 0.0      8.0   \n",
       "CNOT1                3.0               0.0                 0.0      5.0   \n",
       "RLF                  4.0               0.0                 0.0      1.0   \n",
       "\n",
       "       classic_caf    max_af   no_lofs  obs_het_lof  obs_hom_lof   defined  \\\n",
       "gene                                                                         \n",
       "MED13     0.000012  0.000008  124782.0          3.0          0.0  124785.0   \n",
       "NIPBL     0.000012  0.000008  125693.0          3.0          0.0  125696.0   \n",
       "SMC3      0.000032  0.000004  125731.0          8.0          0.0  125739.0   \n",
       "CNOT1     0.000020  0.000004  125740.0          4.0          0.0  125744.0   \n",
       "RLF       0.000004  0.000004  125122.0          1.0          0.0  125123.0   \n",
       "\n",
       "              p  exp_hom_lof  classic_caf_afr  classic_caf_amr  \\\n",
       "gene                                                             \n",
       "MED13  0.000012     0.000018         0.000000         0.000000   \n",
       "NIPBL  0.000012     0.000018         0.000000         0.000000   \n",
       "SMC3   0.000032     0.000127         0.000000         0.000000   \n",
       "CNOT1  0.000016     0.000032         0.000000         0.000029   \n",
       "RLF    0.000004     0.000002         0.000062         0.000000   \n",
       "\n",
       "       classic_caf_asj  classic_caf_eas  classic_caf_fin  classic_caf_nfe  \\\n",
       "gene                                                                        \n",
       "MED13         0.000000         0.000000         0.000093         0.000009   \n",
       "NIPBL         0.000099         0.000000         0.000000         0.000000   \n",
       "SMC3          0.000099         0.000054         0.000000         0.000044   \n",
       "CNOT1         0.000000         0.000055         0.000000         0.000026   \n",
       "RLF           0.000000         0.000000         0.000000         0.000000   \n",
       "\n",
       "       classic_caf_oth  classic_caf_sas     p_afr     p_amr     p_asj  \\\n",
       "gene                                                                    \n",
       "MED13              0.0         0.000000  0.000000  0.000000  0.000000   \n",
       "NIPBL              0.0         0.000065  0.000000  0.000000  0.000099   \n",
       "SMC3               0.0         0.000033  0.000000  0.000000  0.000099   \n",
       "CNOT1              0.0         0.000000  0.000000  0.000029  0.000000   \n",
       "RLF                0.0         0.000000  0.000062  0.000000  0.000000   \n",
       "\n",
       "          p_eas     p_fin     p_nfe  p_oth     p_sas transcript_type  \\\n",
       "gene                                                                   \n",
       "MED13  0.000000  0.000093  0.000009    0.0  0.000000  protein_coding   \n",
       "NIPBL  0.000000  0.000000  0.000000    0.0  0.000065  protein_coding   \n",
       "SMC3   0.000054  0.000000  0.000044    0.0  0.000033  protein_coding   \n",
       "CNOT1  0.000054  0.000000  0.000018    0.0  0.000000  protein_coding   \n",
       "RLF    0.000000  0.000000  0.000000    0.0  0.000000  protein_coding   \n",
       "\n",
       "               gene_id  transcript_level  cds_length  num_coding_exons  \\\n",
       "gene                                                                     \n",
       "MED13  ENSG00000108510                 2        6522                30   \n",
       "NIPBL  ENSG00000164190                 2        8412                46   \n",
       "SMC3   ENSG00000108055                 2        3651                29   \n",
       "CNOT1  ENSG00000125107                 2        7128                48   \n",
       "RLF    ENSG00000117000                 2        5742                 8   \n",
       "\n",
       "            gene_type  gene_length  exac_pLI  exac_obs_lof  exac_exp_lof  \\\n",
       "gene                                                                       \n",
       "MED13  protein_coding       122678       1.0           0.0        64.393   \n",
       "NIPBL  protein_coding       189655       1.0           1.0       110.570   \n",
       "SMC3   protein_coding        36946       1.0           0.0        58.523   \n",
       "CNOT1  protein_coding       109936       1.0           3.0        90.130   \n",
       "RLF    protein_coding        79549       1.0           0.0        43.607   \n",
       "\n",
       "       exac_oe_lof  brain_expression chromosome  start_position  end_position  \n",
       "gene                                                                           \n",
       "MED13     0.000000               NaN         17        60019966      60142643  \n",
       "NIPBL     0.009044               NaN          5        36876861      37066515  \n",
       "SMC3      0.000000               NaN         10       112327449     112364394  \n",
       "CNOT1     0.033285               NaN         16        58553855      58663790  \n",
       "RLF       0.000000               NaN          1        40627045      40706593  "
      ]
     },
     "execution_count": 108,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gnomad.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 109,
   "metadata": {},
   "outputs": [],
   "source": [
    "decipher = decipher.join(gnomad['pLI'], on='name2')\n",
    "decipher = decipher.join(gnomad['oe_lof_upper'], on='name2')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 110,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>symbol</th>\n",
       "      <th>hi_score</th>\n",
       "      <th>hi_index</th>\n",
       "      <th>chrom</th>\n",
       "      <th>txStart</th>\n",
       "      <th>txEnd</th>\n",
       "      <th>GeneID</th>\n",
       "      <th>name2</th>\n",
       "      <th>name</th>\n",
       "      <th>strand</th>\n",
       "      <th>pLI</th>\n",
       "      <th>oe_lof_upper</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ANXA2R</td>\n",
       "      <td>0.000047349</td>\n",
       "      <td>100.00</td>\n",
       "      <td>chr5</td>\n",
       "      <td>43039472</td>\n",
       "      <td>43043349</td>\n",
       "      <td>389289</td>\n",
       "      <td>ANXA2R</td>\n",
       "      <td>NM_001382352</td>\n",
       "      <td>-</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SCGB1D1</td>\n",
       "      <td>0.000054551</td>\n",
       "      <td>99.99</td>\n",
       "      <td>chr11</td>\n",
       "      <td>61957687</td>\n",
       "      <td>61961011</td>\n",
       "      <td>10648</td>\n",
       "      <td>SCGB1D1</td>\n",
       "      <td>NM_006552</td>\n",
       "      <td>+</td>\n",
       "      <td>1.294000e-07</td>\n",
       "      <td>1.968</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>IL31</td>\n",
       "      <td>0.000057228</td>\n",
       "      <td>99.99</td>\n",
       "      <td>chr12</td>\n",
       "      <td>122656575</td>\n",
       "      <td>122658768</td>\n",
       "      <td>386653</td>\n",
       "      <td>IL31</td>\n",
       "      <td>NM_001014336</td>\n",
       "      <td>-</td>\n",
       "      <td>1.144500e-02</td>\n",
       "      <td>1.715</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>BPIFA2</td>\n",
       "      <td>0.000061264</td>\n",
       "      <td>99.98</td>\n",
       "      <td>chr20</td>\n",
       "      <td>31749575</td>\n",
       "      <td>31769218</td>\n",
       "      <td>140683</td>\n",
       "      <td>BPIFA2</td>\n",
       "      <td>NM_001319164</td>\n",
       "      <td>+</td>\n",
       "      <td>8.012400e-07</td>\n",
       "      <td>1.495</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>MUCL1</td>\n",
       "      <td>0.000067779</td>\n",
       "      <td>99.98</td>\n",
       "      <td>chr12</td>\n",
       "      <td>55248299</td>\n",
       "      <td>55252171</td>\n",
       "      <td>118430</td>\n",
       "      <td>MUCL1</td>\n",
       "      <td>NM_058173</td>\n",
       "      <td>+</td>\n",
       "      <td>3.092300e-05</td>\n",
       "      <td>1.908</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    symbol     hi_score  hi_index  chrom    txStart      txEnd  GeneID  \\\n",
       "0   ANXA2R  0.000047349    100.00   chr5   43039472   43043349  389289   \n",
       "1  SCGB1D1  0.000054551     99.99  chr11   61957687   61961011   10648   \n",
       "2     IL31  0.000057228     99.99  chr12  122656575  122658768  386653   \n",
       "3   BPIFA2  0.000061264     99.98  chr20   31749575   31769218  140683   \n",
       "4    MUCL1  0.000067779     99.98  chr12   55248299   55252171  118430   \n",
       "\n",
       "     name2          name strand           pLI  oe_lof_upper  \n",
       "0   ANXA2R  NM_001382352      -           NaN           NaN  \n",
       "1  SCGB1D1     NM_006552      +  1.294000e-07         1.968  \n",
       "2     IL31  NM_001014336      -  1.144500e-02         1.715  \n",
       "3   BPIFA2  NM_001319164      +  8.012400e-07         1.495  \n",
       "4    MUCL1     NM_058173      +  3.092300e-05         1.908  "
      ]
     },
     "execution_count": 110,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "decipher.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 111,
   "metadata": {},
   "outputs": [],
   "source": [
    "decipher = decipher.loc[\n",
    "    (decipher['pLI'] >= 0.9) & (decipher['hi_index'] < 10) & (decipher['oe_lof_upper'] < 0.35),\n",
    "    ['chrom', 'txStart', 'txEnd', 'GeneID', 'name2', 'name', 'pLI', 'hi_score']\n",
    "].sort_values(['chrom', 'txStart', 'txEnd'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 112,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>txStart</th>\n",
       "      <th>txEnd</th>\n",
       "      <th>GeneID</th>\n",
       "      <th>name2</th>\n",
       "      <th>name</th>\n",
       "      <th>pLI</th>\n",
       "      <th>hi_score</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>17510</th>\n",
       "      <td>chr1</td>\n",
       "      <td>6845513</td>\n",
       "      <td>7829766</td>\n",
       "      <td>23261</td>\n",
       "      <td>CAMTA1</td>\n",
       "      <td>NM_001349609</td>\n",
       "      <td>1.00000</td>\n",
       "      <td>0.951338773</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17568</th>\n",
       "      <td>chr1</td>\n",
       "      <td>8412463</td>\n",
       "      <td>8877699</td>\n",
       "      <td>473</td>\n",
       "      <td>RERE</td>\n",
       "      <td>NM_001042681</td>\n",
       "      <td>1.00000</td>\n",
       "      <td>0.962236687</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17640</th>\n",
       "      <td>chr1</td>\n",
       "      <td>9751524</td>\n",
       "      <td>9788980</td>\n",
       "      <td>5293</td>\n",
       "      <td>PIK3CD</td>\n",
       "      <td>NM_001350235</td>\n",
       "      <td>0.99999</td>\n",
       "      <td>0.977549203</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16383</th>\n",
       "      <td>chr1</td>\n",
       "      <td>10270763</td>\n",
       "      <td>10441661</td>\n",
       "      <td>23095</td>\n",
       "      <td>KIF1B</td>\n",
       "      <td>NM_015074</td>\n",
       "      <td>1.00000</td>\n",
       "      <td>0.70541308</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17817</th>\n",
       "      <td>chr1</td>\n",
       "      <td>11166591</td>\n",
       "      <td>11322608</td>\n",
       "      <td>2475</td>\n",
       "      <td>MTOR</td>\n",
       "      <td>NM_004958</td>\n",
       "      <td>1.00000</td>\n",
       "      <td>0.997612705</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      chrom   txStart     txEnd  GeneID   name2          name      pLI  \\\n",
       "17510  chr1   6845513   7829766   23261  CAMTA1  NM_001349609  1.00000   \n",
       "17568  chr1   8412463   8877699     473    RERE  NM_001042681  1.00000   \n",
       "17640  chr1   9751524   9788980    5293  PIK3CD  NM_001350235  0.99999   \n",
       "16383  chr1  10270763  10441661   23095   KIF1B     NM_015074  1.00000   \n",
       "17817  chr1  11166591  11322608    2475    MTOR     NM_004958  1.00000   \n",
       "\n",
       "          hi_score  \n",
       "17510  0.951338773  \n",
       "17568  0.962236687  \n",
       "17640  0.977549203  \n",
       "16383   0.70541308  \n",
       "17817  0.997612705  "
      ]
     },
     "execution_count": 112,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "decipher.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 113,
   "metadata": {},
   "outputs": [],
   "source": [
    "decipher.rename(columns={\n",
    "    'chrom': '#chrom', 'txStart': 'start', 'txEnd': 'end',\n",
    "    'GeneID': 'gene_id', 'name2': 'symbol', 'name': 'transcript'\n",
    "}).to_csv(decipher_gene_file, sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 114,
   "metadata": {},
   "outputs": [],
   "source": [
    "!bgzip -cf {decipher_gene_file} > {decipher_gene_file}.gz"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 115,
   "metadata": {},
   "outputs": [],
   "source": [
    "!tabix -fp bed {decipher_gene_file}.gz\n",
    "!rm {decipher_gene_file}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 116,
   "metadata": {},
   "outputs": [],
   "source": [
    "ts_genes = set(\n",
    "    curation_gene.loc[\n",
    "        curation_gene['Triplosensitivity Score'] == '3', '#Gene Symbol'\n",
    "    ]\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 117,
   "metadata": {},
   "outputs": [],
   "source": [
    "ts_gene = refgene_info.loc[\n",
    "    refgene_info['name2'].isin(ts_genes),\n",
    "    ['chrom', 'txStart', 'txEnd', 'GeneID', 'name2', 'name', 'strand']\n",
    "].sort_values(['chrom', 'txStart', 'txEnd'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 118,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>txStart</th>\n",
       "      <th>txEnd</th>\n",
       "      <th>GeneID</th>\n",
       "      <th>name2</th>\n",
       "      <th>name</th>\n",
       "      <th>strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>4352</th>\n",
       "      <td>chr5</td>\n",
       "      <td>126112827</td>\n",
       "      <td>126172712</td>\n",
       "      <td>4001</td>\n",
       "      <td>LMNB1</td>\n",
       "      <td>NM_005573</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10862</th>\n",
       "      <td>chrX</td>\n",
       "      <td>103031780</td>\n",
       "      <td>103047547</td>\n",
       "      <td>5354</td>\n",
       "      <td>PLP1</td>\n",
       "      <td>NM_001305004</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      chrom    txStart      txEnd  GeneID  name2          name strand\n",
       "4352   chr5  126112827  126172712    4001  LMNB1     NM_005573      +\n",
       "10862  chrX  103031780  103047547    5354   PLP1  NM_001305004      +"
      ]
     },
     "execution_count": 118,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ts_gene.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 119,
   "metadata": {},
   "outputs": [],
   "source": [
    "ts_gene.rename(columns={\n",
    "    'chrom': '#chrom', 'txStart': 'start', 'txEnd': 'end',\n",
    "    'GeneID': 'gene_id', 'name2': 'symbol', 'name': 'transcript'\n",
    "}).to_csv(ts_gene_file, index=False, sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 120,
   "metadata": {},
   "outputs": [],
   "source": [
    "!bgzip -cf {ts_gene_file} > {ts_gene_file}.gz"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 121,
   "metadata": {},
   "outputs": [],
   "source": [
    "!tabix -fp bed {ts_gene_file}.gz\n",
    "!rm {ts_gene_file}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 122,
   "metadata": {},
   "outputs": [],
   "source": [
    "uts_genes = set(\n",
    "    curation_gene.loc[curation_gene['Triplosensitivity Score'] == '40', '#Gene Symbol']\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 123,
   "metadata": {},
   "outputs": [],
   "source": [
    "uts_gene = refgene_info.loc[\n",
    "    refgene_info['name2'].isin(uts_genes),\n",
    "    ['chrom', 'txStart', 'txEnd', 'GeneID', 'name2', 'name', 'strand']\n",
    "].sort_values(['chrom', 'txStart', 'txEnd'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 124,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>txStart</th>\n",
       "      <th>txEnd</th>\n",
       "      <th>GeneID</th>\n",
       "      <th>name2</th>\n",
       "      <th>name</th>\n",
       "      <th>strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>18575</th>\n",
       "      <td>chrX</td>\n",
       "      <td>6451658</td>\n",
       "      <td>6453159</td>\n",
       "      <td>51481</td>\n",
       "      <td>VCX3A</td>\n",
       "      <td>NM_016379</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      chrom  txStart    txEnd  GeneID  name2       name strand\n",
       "18575  chrX  6451658  6453159   51481  VCX3A  NM_016379      -"
      ]
     },
     "execution_count": 124,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "uts_gene.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 125,
   "metadata": {},
   "outputs": [],
   "source": [
    "uts_gene.rename(columns={\n",
    "    'chrom': '#chrom', 'txStart': 'start', 'txEnd': 'end',\n",
    "    'GeneID': 'gene_id', 'name2': 'symbol', 'name': 'transcript'\n",
    "}).to_csv(uts_gene_file, index=False, sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 126,
   "metadata": {},
   "outputs": [],
   "source": [
    "!bgzip -cf {uts_gene_file} > {uts_gene_file}.gz"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 127,
   "metadata": {},
   "outputs": [],
   "source": [
    "!tabix -fp bed {uts_gene_file}.gz\n",
    "!rm {uts_gene_file}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 128,
   "metadata": {},
   "outputs": [],
   "source": [
    "ts_region = region_pos.loc[\n",
    "    region_pos['Triplosensitivity Score'] == '3',\n",
    "    ['chrom', 'start', 'end', '#ISCA ID', 'ISCA Region Name']\n",
    "].sort_values(['chrom', 'start', 'end'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 129,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>#ISCA ID</th>\n",
       "      <th>ISCA Region Name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>486</th>\n",
       "      <td>chr1</td>\n",
       "      <td>146577486</td>\n",
       "      <td>147394506</td>\n",
       "      <td>ISCA-37421</td>\n",
       "      <td>1q21.1 recurrent region (BP3-BP4, distal) (inc...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>496</th>\n",
       "      <td>chr15</td>\n",
       "      <td>22832519</td>\n",
       "      <td>28379874</td>\n",
       "      <td>ISCA-37404</td>\n",
       "      <td>15q11.2q13 recurrent (PWS/AS) region (Class 1,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>457</th>\n",
       "      <td>chr15</td>\n",
       "      <td>23747996</td>\n",
       "      <td>28379874</td>\n",
       "      <td>ISCA-37478</td>\n",
       "      <td>15q11.2q13 recurrent (PWS/AS) region (Class 2,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>498</th>\n",
       "      <td>chr16</td>\n",
       "      <td>29649997</td>\n",
       "      <td>30199852</td>\n",
       "      <td>ISCA-37400</td>\n",
       "      <td>16p11.2 recurrent region (proximal, BP4-BP5) (...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>481</th>\n",
       "      <td>chr17</td>\n",
       "      <td>1247833</td>\n",
       "      <td>2588909</td>\n",
       "      <td>ISCA-37430</td>\n",
       "      <td>17p13.3 (Miller-Dieker syndrome) region (inclu...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     chrom      start        end    #ISCA ID  \\\n",
       "486   chr1  146577486  147394506  ISCA-37421   \n",
       "496  chr15   22832519   28379874  ISCA-37404   \n",
       "457  chr15   23747996   28379874  ISCA-37478   \n",
       "498  chr16   29649997   30199852  ISCA-37400   \n",
       "481  chr17    1247833    2588909  ISCA-37430   \n",
       "\n",
       "                                      ISCA Region Name  \n",
       "486  1q21.1 recurrent region (BP3-BP4, distal) (inc...  \n",
       "496  15q11.2q13 recurrent (PWS/AS) region (Class 1,...  \n",
       "457  15q11.2q13 recurrent (PWS/AS) region (Class 2,...  \n",
       "498  16p11.2 recurrent region (proximal, BP4-BP5) (...  \n",
       "481  17p13.3 (Miller-Dieker syndrome) region (inclu...  "
      ]
     },
     "execution_count": 129,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ts_region.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 130,
   "metadata": {},
   "outputs": [],
   "source": [
    "ts_region['omim_genes'] = ts_region.apply(\n",
    "    lambda row: ','.join(omim_gene.loc[\n",
    "        (omim_gene['chrom'] == row['chrom'])\n",
    "        & (omim_gene['txEnd'] >= row['start'])\n",
    "        & (omim_gene['txStart'] <= row['end']),\n",
    "        'name2'\n",
    "    ]),\n",
    "    axis=1\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 131,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>#ISCA ID</th>\n",
       "      <th>ISCA Region Name</th>\n",
       "      <th>omim_genes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>486</th>\n",
       "      <td>chr1</td>\n",
       "      <td>146577486</td>\n",
       "      <td>147394506</td>\n",
       "      <td>ISCA-37421</td>\n",
       "      <td>1q21.1 recurrent region (BP3-BP4, distal) (inc...</td>\n",
       "      <td>GJA5,GJA8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>496</th>\n",
       "      <td>chr15</td>\n",
       "      <td>22832519</td>\n",
       "      <td>28379874</td>\n",
       "      <td>ISCA-37404</td>\n",
       "      <td>15q11.2q13 recurrent (PWS/AS) region (Class 1,...</td>\n",
       "      <td>NIPA1,MKRN3,MAGEL2,NDN,SNRPN,UBE3A,GABRB3,GABR...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>457</th>\n",
       "      <td>chr15</td>\n",
       "      <td>23747996</td>\n",
       "      <td>28379874</td>\n",
       "      <td>ISCA-37478</td>\n",
       "      <td>15q11.2q13 recurrent (PWS/AS) region (Class 2,...</td>\n",
       "      <td>MKRN3,MAGEL2,NDN,SNRPN,UBE3A,GABRB3,GABRA5,OCA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>498</th>\n",
       "      <td>chr16</td>\n",
       "      <td>29649997</td>\n",
       "      <td>30199852</td>\n",
       "      <td>ISCA-37400</td>\n",
       "      <td>16p11.2 recurrent region (proximal, BP4-BP5) (...</td>\n",
       "      <td>KIF22,PRRT2,ALDOA,TBX6,CORO1A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>481</th>\n",
       "      <td>chr17</td>\n",
       "      <td>1247833</td>\n",
       "      <td>2588909</td>\n",
       "      <td>ISCA-37430</td>\n",
       "      <td>17p13.3 (Miller-Dieker syndrome) region (inclu...</td>\n",
       "      <td>INPP5K,PRPF8,WDR81,SERPINF2,SERPINF1,DPH1,PAFA...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     chrom      start        end    #ISCA ID  \\\n",
       "486   chr1  146577486  147394506  ISCA-37421   \n",
       "496  chr15   22832519   28379874  ISCA-37404   \n",
       "457  chr15   23747996   28379874  ISCA-37478   \n",
       "498  chr16   29649997   30199852  ISCA-37400   \n",
       "481  chr17    1247833    2588909  ISCA-37430   \n",
       "\n",
       "                                      ISCA Region Name  \\\n",
       "486  1q21.1 recurrent region (BP3-BP4, distal) (inc...   \n",
       "496  15q11.2q13 recurrent (PWS/AS) region (Class 1,...   \n",
       "457  15q11.2q13 recurrent (PWS/AS) region (Class 2,...   \n",
       "498  16p11.2 recurrent region (proximal, BP4-BP5) (...   \n",
       "481  17p13.3 (Miller-Dieker syndrome) region (inclu...   \n",
       "\n",
       "                                            omim_genes  \n",
       "486                                          GJA5,GJA8  \n",
       "496  NIPA1,MKRN3,MAGEL2,NDN,SNRPN,UBE3A,GABRB3,GABR...  \n",
       "457  MKRN3,MAGEL2,NDN,SNRPN,UBE3A,GABRB3,GABRA5,OCA...  \n",
       "498                      KIF22,PRRT2,ALDOA,TBX6,CORO1A  \n",
       "481  INPP5K,PRPF8,WDR81,SERPINF2,SERPINF1,DPH1,PAFA...  "
      ]
     },
     "execution_count": 131,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ts_region.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 132,
   "metadata": {},
   "outputs": [],
   "source": [
    "ts_region.rename(columns={\n",
    "    'chrom': '#chrom',\n",
    "    '#ISCA ID': 'isca_id', 'ISCA Region Name': 'name'\n",
    "}).to_csv(ts_region_file, sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 133,
   "metadata": {},
   "outputs": [],
   "source": [
    "!bgzip -cf {ts_region_file} > {ts_region_file}.gz"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 134,
   "metadata": {},
   "outputs": [],
   "source": [
    "!tabix -fp bed {ts_region_file}.gz\n",
    "!rm {ts_region_file}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 135,
   "metadata": {},
   "outputs": [],
   "source": [
    "uts_region = region_pos.loc[\n",
    "    region_pos['Triplosensitivity Score'] == '40',\n",
    "    ['chrom', 'start', 'end', '#ISCA ID', 'ISCA Region Name']\n",
    "].sort_values(['chrom', 'start', 'end'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 136,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>#ISCA ID</th>\n",
       "      <th>ISCA Region Name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>181</th>\n",
       "      <td>chr1</td>\n",
       "      <td>713039</td>\n",
       "      <td>756009</td>\n",
       "      <td>ISCA-46556</td>\n",
       "      <td>1p36.33 population region (DGV_Gold_Standard_J...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>182</th>\n",
       "      <td>chr1</td>\n",
       "      <td>713039</td>\n",
       "      <td>756009</td>\n",
       "      <td>ISCA-46554</td>\n",
       "      <td>1p36.33 population region ((DGV_Gold_Standard_...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>427</th>\n",
       "      <td>chr1</td>\n",
       "      <td>12894999</td>\n",
       "      <td>12940000</td>\n",
       "      <td>ISCA-46308</td>\n",
       "      <td>1p36.21 population region (gnomAD-SV_v2.1_DUP_...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>180</th>\n",
       "      <td>chr1</td>\n",
       "      <td>12897609</td>\n",
       "      <td>12922238</td>\n",
       "      <td>ISCA-46557</td>\n",
       "      <td>1p36.21 population region (DGV_Gold_Standard_J...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>414</th>\n",
       "      <td>chr1</td>\n",
       "      <td>104190999</td>\n",
       "      <td>104303200</td>\n",
       "      <td>ISCA-46321</td>\n",
       "      <td>1p21.1 population region (gnomAD-SV_v2.1_DUP_1...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    chrom      start        end    #ISCA ID  \\\n",
       "181  chr1     713039     756009  ISCA-46556   \n",
       "182  chr1     713039     756009  ISCA-46554   \n",
       "427  chr1   12894999   12940000  ISCA-46308   \n",
       "180  chr1   12897609   12922238  ISCA-46557   \n",
       "414  chr1  104190999  104303200  ISCA-46321   \n",
       "\n",
       "                                      ISCA Region Name  \n",
       "181  1p36.33 population region (DGV_Gold_Standard_J...  \n",
       "182  1p36.33 population region ((DGV_Gold_Standard_...  \n",
       "427  1p36.21 population region (gnomAD-SV_v2.1_DUP_...  \n",
       "180  1p36.21 population region (DGV_Gold_Standard_J...  \n",
       "414  1p21.1 population region (gnomAD-SV_v2.1_DUP_1...  "
      ]
     },
     "execution_count": 136,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "uts_region.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 137,
   "metadata": {},
   "outputs": [],
   "source": [
    "uts_region['genes'] = uts_region.apply(\n",
    "    lambda row: ','.join(gene.loc[\n",
    "        (gene['chrom'] == row['chrom'])\n",
    "        & (gene['txEnd'] >= row['start'])\n",
    "        & (gene['txStart'] <= row['end']),\n",
    "        'name2'\n",
    "    ]),\n",
    "    axis=1\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 138,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>#ISCA ID</th>\n",
       "      <th>ISCA Region Name</th>\n",
       "      <th>genes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>181</th>\n",
       "      <td>chr1</td>\n",
       "      <td>713039</td>\n",
       "      <td>756009</td>\n",
       "      <td>ISCA-46556</td>\n",
       "      <td>1p36.33 population region (DGV_Gold_Standard_J...</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>182</th>\n",
       "      <td>chr1</td>\n",
       "      <td>713039</td>\n",
       "      <td>756009</td>\n",
       "      <td>ISCA-46554</td>\n",
       "      <td>1p36.33 population region ((DGV_Gold_Standard_...</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>427</th>\n",
       "      <td>chr1</td>\n",
       "      <td>12894999</td>\n",
       "      <td>12940000</td>\n",
       "      <td>ISCA-46308</td>\n",
       "      <td>1p36.21 population region (gnomAD-SV_v2.1_DUP_...</td>\n",
       "      <td>HNRNPCL1,PRAMEF2,PRAMEF4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>180</th>\n",
       "      <td>chr1</td>\n",
       "      <td>12897609</td>\n",
       "      <td>12922238</td>\n",
       "      <td>ISCA-46557</td>\n",
       "      <td>1p36.21 population region (DGV_Gold_Standard_J...</td>\n",
       "      <td>HNRNPCL1,PRAMEF2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>414</th>\n",
       "      <td>chr1</td>\n",
       "      <td>104190999</td>\n",
       "      <td>104303200</td>\n",
       "      <td>ISCA-46321</td>\n",
       "      <td>1p21.1 population region (gnomAD-SV_v2.1_DUP_1...</td>\n",
       "      <td>AMY1C,AMY1A,AMY1B</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    chrom      start        end    #ISCA ID  \\\n",
       "181  chr1     713039     756009  ISCA-46556   \n",
       "182  chr1     713039     756009  ISCA-46554   \n",
       "427  chr1   12894999   12940000  ISCA-46308   \n",
       "180  chr1   12897609   12922238  ISCA-46557   \n",
       "414  chr1  104190999  104303200  ISCA-46321   \n",
       "\n",
       "                                      ISCA Region Name  \\\n",
       "181  1p36.33 population region (DGV_Gold_Standard_J...   \n",
       "182  1p36.33 population region ((DGV_Gold_Standard_...   \n",
       "427  1p36.21 population region (gnomAD-SV_v2.1_DUP_...   \n",
       "180  1p36.21 population region (DGV_Gold_Standard_J...   \n",
       "414  1p21.1 population region (gnomAD-SV_v2.1_DUP_1...   \n",
       "\n",
       "                        genes  \n",
       "181                            \n",
       "182                            \n",
       "427  HNRNPCL1,PRAMEF2,PRAMEF4  \n",
       "180          HNRNPCL1,PRAMEF2  \n",
       "414         AMY1C,AMY1A,AMY1B  "
      ]
     },
     "execution_count": 138,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "uts_region.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 139,
   "metadata": {},
   "outputs": [],
   "source": [
    "uts_region.rename(columns={\n",
    "    'chrom': '#chrom',\n",
    "    '#ISCA ID': 'isca_id', 'ISCA Region Name': 'name'\n",
    "}).to_csv(uts_region_file, sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 140,
   "metadata": {},
   "outputs": [],
   "source": [
    "!bgzip -cf {uts_region_file} > {uts_region_file}.gz"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 141,
   "metadata": {},
   "outputs": [],
   "source": [
    "!tabix -fp bed {uts_region_file}.gz\n",
    "!rm {uts_region_file}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 142,
   "metadata": {},
   "outputs": [],
   "source": [
    "dgv = pd.read_csv(\n",
    "    dgv_ori_file,\n",
    "    sep='\\t', names=['chrom', 'info'], usecols=[0, 8]\n",
    ").drop_duplicates('info')\n",
    "info = dgv['info'].str.extract(\n",
    "    r'ID=(?P<id>[^;]+).*variant_sub_type=(?P<type>[^;]+).*inner_start=(?P<start>[^;]+).*inner_end=(?P<end>[^;]+).*Frequency=(?P<freq>\\S+?)%;.*num_unique_samples_tested=(?P<sample>[^;]+)'\n",
    ").astype({'start': int, 'end': int, 'sample': int, 'freq': float})\n",
    "dgv = dgv.merge(info, left_index=True, right_index=True)\n",
    "dgv['af'] = dgv['freq'] / 100\n",
    "dgv = dgv[dgv['sample'] >= 1000].sort_values(['chrom', 'start', 'end'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 143,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>info</th>\n",
       "      <th>id</th>\n",
       "      <th>type</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>freq</th>\n",
       "      <th>sample</th>\n",
       "      <th>af</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>chr1</td>\n",
       "      <td>ID=gssvG9;Name=gssvG9;variant_type=CNV;variant...</td>\n",
       "      <td>gssvG9</td>\n",
       "      <td>Gain</td>\n",
       "      <td>49911</td>\n",
       "      <td>222421</td>\n",
       "      <td>1.13</td>\n",
       "      <td>1149</td>\n",
       "      <td>0.0113</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54</th>\n",
       "      <td>chr1</td>\n",
       "      <td>ID=gssvG28;Name=gssvG28;variant_type=CNV;varia...</td>\n",
       "      <td>gssvG28</td>\n",
       "      <td>Gain</td>\n",
       "      <td>60905</td>\n",
       "      <td>97505</td>\n",
       "      <td>0.61</td>\n",
       "      <td>2609</td>\n",
       "      <td>0.0061</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18651</th>\n",
       "      <td>chr1</td>\n",
       "      <td>ID=gssvL18;Name=gssvL18;variant_type=CNV;varia...</td>\n",
       "      <td>gssvL18</td>\n",
       "      <td>Loss</td>\n",
       "      <td>60905</td>\n",
       "      <td>97505</td>\n",
       "      <td>0.61</td>\n",
       "      <td>2609</td>\n",
       "      <td>0.0061</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18627</th>\n",
       "      <td>chr1</td>\n",
       "      <td>ID=gssvL4;Name=gssvL4;variant_type=CNV;variant...</td>\n",
       "      <td>gssvL4</td>\n",
       "      <td>Loss</td>\n",
       "      <td>61724</td>\n",
       "      <td>346583</td>\n",
       "      <td>0.30</td>\n",
       "      <td>12364</td>\n",
       "      <td>0.0030</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48</th>\n",
       "      <td>chr1</td>\n",
       "      <td>ID=gssvG24;Name=gssvG24;variant_type=CNV;varia...</td>\n",
       "      <td>gssvG24</td>\n",
       "      <td>Gain</td>\n",
       "      <td>125400</td>\n",
       "      <td>176984</td>\n",
       "      <td>0.73</td>\n",
       "      <td>1510</td>\n",
       "      <td>0.0073</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      chrom                                               info       id  type  \\\n",
       "12     chr1  ID=gssvG9;Name=gssvG9;variant_type=CNV;variant...   gssvG9  Gain   \n",
       "54     chr1  ID=gssvG28;Name=gssvG28;variant_type=CNV;varia...  gssvG28  Gain   \n",
       "18651  chr1  ID=gssvL18;Name=gssvL18;variant_type=CNV;varia...  gssvL18  Loss   \n",
       "18627  chr1  ID=gssvL4;Name=gssvL4;variant_type=CNV;variant...   gssvL4  Loss   \n",
       "48     chr1  ID=gssvG24;Name=gssvG24;variant_type=CNV;varia...  gssvG24  Gain   \n",
       "\n",
       "        start     end  freq  sample      af  \n",
       "12      49911  222421  1.13    1149  0.0113  \n",
       "54      60905   97505  0.61    2609  0.0061  \n",
       "18651   60905   97505  0.61    2609  0.0061  \n",
       "18627   61724  346583  0.30   12364  0.0030  \n",
       "48     125400  176984  0.73    1510  0.0073  "
      ]
     },
     "execution_count": 143,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dgv.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 144,
   "metadata": {},
   "outputs": [],
   "source": [
    "def fetch_gene(gene, chrom, start, end):\n",
    "    return ','.join(gene.loc[\n",
    "        (gene['chrom'] == chrom) & (gene['txEnd'] >= start) & (gene['txStart'] <= end), 'name2'\n",
    "    ])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 145,
   "metadata": {},
   "outputs": [],
   "source": [
    "with Pool(processes=7) as pool:\n",
    "    dgv['genes'] = pool.starmap(fetch_gene, (\n",
    "        (gene, row['chrom'], row['start'], row['end']) for _, row in dgv.iterrows()\n",
    "    ), chunksize=70)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 146,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>info</th>\n",
       "      <th>id</th>\n",
       "      <th>type</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>freq</th>\n",
       "      <th>sample</th>\n",
       "      <th>af</th>\n",
       "      <th>genes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>chr1</td>\n",
       "      <td>ID=gssvG9;Name=gssvG9;variant_type=CNV;variant...</td>\n",
       "      <td>gssvG9</td>\n",
       "      <td>Gain</td>\n",
       "      <td>49911</td>\n",
       "      <td>222421</td>\n",
       "      <td>1.13</td>\n",
       "      <td>1149</td>\n",
       "      <td>0.0113</td>\n",
       "      <td>OR4F5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54</th>\n",
       "      <td>chr1</td>\n",
       "      <td>ID=gssvG28;Name=gssvG28;variant_type=CNV;varia...</td>\n",
       "      <td>gssvG28</td>\n",
       "      <td>Gain</td>\n",
       "      <td>60905</td>\n",
       "      <td>97505</td>\n",
       "      <td>0.61</td>\n",
       "      <td>2609</td>\n",
       "      <td>0.0061</td>\n",
       "      <td>OR4F5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18651</th>\n",
       "      <td>chr1</td>\n",
       "      <td>ID=gssvL18;Name=gssvL18;variant_type=CNV;varia...</td>\n",
       "      <td>gssvL18</td>\n",
       "      <td>Loss</td>\n",
       "      <td>60905</td>\n",
       "      <td>97505</td>\n",
       "      <td>0.61</td>\n",
       "      <td>2609</td>\n",
       "      <td>0.0061</td>\n",
       "      <td>OR4F5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18627</th>\n",
       "      <td>chr1</td>\n",
       "      <td>ID=gssvL4;Name=gssvL4;variant_type=CNV;variant...</td>\n",
       "      <td>gssvL4</td>\n",
       "      <td>Loss</td>\n",
       "      <td>61724</td>\n",
       "      <td>346583</td>\n",
       "      <td>0.30</td>\n",
       "      <td>12364</td>\n",
       "      <td>0.0030</td>\n",
       "      <td>OR4F5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48</th>\n",
       "      <td>chr1</td>\n",
       "      <td>ID=gssvG24;Name=gssvG24;variant_type=CNV;varia...</td>\n",
       "      <td>gssvG24</td>\n",
       "      <td>Gain</td>\n",
       "      <td>125400</td>\n",
       "      <td>176984</td>\n",
       "      <td>0.73</td>\n",
       "      <td>1510</td>\n",
       "      <td>0.0073</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      chrom                                               info       id  type  \\\n",
       "12     chr1  ID=gssvG9;Name=gssvG9;variant_type=CNV;variant...   gssvG9  Gain   \n",
       "54     chr1  ID=gssvG28;Name=gssvG28;variant_type=CNV;varia...  gssvG28  Gain   \n",
       "18651  chr1  ID=gssvL18;Name=gssvL18;variant_type=CNV;varia...  gssvL18  Loss   \n",
       "18627  chr1  ID=gssvL4;Name=gssvL4;variant_type=CNV;variant...   gssvL4  Loss   \n",
       "48     chr1  ID=gssvG24;Name=gssvG24;variant_type=CNV;varia...  gssvG24  Gain   \n",
       "\n",
       "        start     end  freq  sample      af  genes  \n",
       "12      49911  222421  1.13    1149  0.0113  OR4F5  \n",
       "54      60905   97505  0.61    2609  0.0061  OR4F5  \n",
       "18651   60905   97505  0.61    2609  0.0061  OR4F5  \n",
       "18627   61724  346583  0.30   12364  0.0030  OR4F5  \n",
       "48     125400  176984  0.73    1510  0.0073         "
      ]
     },
     "execution_count": 146,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dgv.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 147,
   "metadata": {},
   "outputs": [],
   "source": [
    "dgv.loc[\n",
    "    dgv['type'] == 'Gain', ['chrom', 'start', 'end', 'id', 'genes', 'af']\n",
    "].rename(columns={'chrom': '#chrom'}).to_csv(dgv_gain_file, sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 148,
   "metadata": {},
   "outputs": [],
   "source": [
    "!bgzip -cf {dgv_gain_file} > {dgv_gain_file}.gz"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 149,
   "metadata": {},
   "outputs": [],
   "source": [
    "!tabix -fp bed {dgv_gain_file}.gz\n",
    "!rm {dgv_gain_file}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 150,
   "metadata": {},
   "outputs": [],
   "source": [
    "dgv.loc[\n",
    "    dgv['type'] == 'Loss', ['chrom', 'start', 'end', 'id', 'genes', 'af']\n",
    "].rename(columns={'chrom': '#chrom'}).to_csv(dgv_loss_file, sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 151,
   "metadata": {},
   "outputs": [],
   "source": [
    "!bgzip -cf {dgv_loss_file} > {dgv_loss_file}.gz"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 152,
   "metadata": {},
   "outputs": [],
   "source": [
    "!tabix -fp bed {dgv_loss_file}.gz\n",
    "!rm {dgv_loss_file}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 153,
   "metadata": {},
   "outputs": [],
   "source": [
    "gnomad = pd.read_csv(\n",
    "    gnomad_control_ori_file, sep='\\t',  dtype=str,\n",
    "    usecols=[0, 1, 2, 3, 4, 37, 38, 73, 74, 107, 108, 141, 142, 175, 176, 241]\n",
    ")\n",
    "gnomad = gnomad[\n",
    "    (gnomad['FILTER'] == 'PASS') & gnomad['svtype'].isin(['DEL', 'DUP'])\n",
    "]\n",
    "gnomad = gnomad[\n",
    "    (gnomad['N_BI_GENOS'].astype(int) >= 1000) |\n",
    "    (gnomad['AFR_N_BI_GENOS'].astype(int) >= 1000) |\n",
    "    (gnomad['AMR_N_BI_GENOS'].astype(int) >= 1000) |\n",
    "    (gnomad['EAS_N_BI_GENOS'].astype(int) >= 1000) |\n",
    "    (gnomad['EUR_N_BI_GENOS'].astype(int) >= 1000)\n",
    "]\n",
    "gnomad['#chrom'] = 'chr' + gnomad['#chrom']\n",
    "gnomad['start'] = gnomad['start'].astype(int)\n",
    "gnomad['end'] = gnomad['end'].astype(int)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 154,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>#chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>name</th>\n",
       "      <th>svtype</th>\n",
       "      <th>AF</th>\n",
       "      <th>N_BI_GENOS</th>\n",
       "      <th>AFR_AF</th>\n",
       "      <th>AFR_N_BI_GENOS</th>\n",
       "      <th>AMR_AF</th>\n",
       "      <th>AMR_N_BI_GENOS</th>\n",
       "      <th>EAS_AF</th>\n",
       "      <th>EAS_N_BI_GENOS</th>\n",
       "      <th>EUR_AF</th>\n",
       "      <th>EUR_N_BI_GENOS</th>\n",
       "      <th>FILTER</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>chr1</td>\n",
       "      <td>62398</td>\n",
       "      <td>62489</td>\n",
       "      <td>gnomAD-SV_v2.1_DEL_1_3</td>\n",
       "      <td>DEL</td>\n",
       "      <td>9.600000339560211e-05</td>\n",
       "      <td>5192</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2014</td>\n",
       "      <td>0.0</td>\n",
       "      <td>609</td>\n",
       "      <td>0.0005840000230818987</td>\n",
       "      <td>856</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1660</td>\n",
       "      <td>PASS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>chr1</td>\n",
       "      <td>66349</td>\n",
       "      <td>66427</td>\n",
       "      <td>gnomAD-SV_v2.1_DEL_1_4</td>\n",
       "      <td>DEL</td>\n",
       "      <td>0.0004830000107176602</td>\n",
       "      <td>5177</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2009</td>\n",
       "      <td>0.0024709999561309814</td>\n",
       "      <td>607</td>\n",
       "      <td>0.0</td>\n",
       "      <td>856</td>\n",
       "      <td>0.0006050000083632767</td>\n",
       "      <td>1652</td>\n",
       "      <td>PASS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>chr1</td>\n",
       "      <td>136999</td>\n",
       "      <td>155500</td>\n",
       "      <td>gnomAD-SV_v2.1_DUP_1_3</td>\n",
       "      <td>DUP</td>\n",
       "      <td>0.138032004237175</td>\n",
       "      <td>5122</td>\n",
       "      <td>0.10211800038814545</td>\n",
       "      <td>1983</td>\n",
       "      <td>0.1157890036702156</td>\n",
       "      <td>570</td>\n",
       "      <td>0.3907710015773773</td>\n",
       "      <td>856</td>\n",
       "      <td>0.05963899940252304</td>\n",
       "      <td>1660</td>\n",
       "      <td>PASS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>chr1</td>\n",
       "      <td>141474</td>\n",
       "      <td>155000</td>\n",
       "      <td>gnomAD-SV_v2.1_DUP_1_4</td>\n",
       "      <td>DUP</td>\n",
       "      <td>0.006382000166922808</td>\n",
       "      <td>4544</td>\n",
       "      <td>0.01448499970138073</td>\n",
       "      <td>1864</td>\n",
       "      <td>0.0</td>\n",
       "      <td>503</td>\n",
       "      <td>0.001905999961309135</td>\n",
       "      <td>787</td>\n",
       "      <td>0.00037200000951997936</td>\n",
       "      <td>1343</td>\n",
       "      <td>PASS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>chr1</td>\n",
       "      <td>319999</td>\n",
       "      <td>327650</td>\n",
       "      <td>gnomAD-SV_v2.1_DEL_1_9</td>\n",
       "      <td>DEL</td>\n",
       "      <td>0.1377120018005371</td>\n",
       "      <td>5192</td>\n",
       "      <td>0.29294899106025696</td>\n",
       "      <td>2014</td>\n",
       "      <td>0.020524999126791954</td>\n",
       "      <td>609</td>\n",
       "      <td>0.08235999941825867</td>\n",
       "      <td>856</td>\n",
       "      <td>0.022590000182390213</td>\n",
       "      <td>1660</td>\n",
       "      <td>PASS</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   #chrom   start     end                    name svtype  \\\n",
       "6    chr1   62398   62489  gnomAD-SV_v2.1_DEL_1_3    DEL   \n",
       "7    chr1   66349   66427  gnomAD-SV_v2.1_DEL_1_4    DEL   \n",
       "15   chr1  136999  155500  gnomAD-SV_v2.1_DUP_1_3    DUP   \n",
       "16   chr1  141474  155000  gnomAD-SV_v2.1_DUP_1_4    DUP   \n",
       "24   chr1  319999  327650  gnomAD-SV_v2.1_DEL_1_9    DEL   \n",
       "\n",
       "                       AF N_BI_GENOS               AFR_AF AFR_N_BI_GENOS  \\\n",
       "6   9.600000339560211e-05       5192                  0.0           2014   \n",
       "7   0.0004830000107176602       5177                  0.0           2009   \n",
       "15      0.138032004237175       5122  0.10211800038814545           1983   \n",
       "16   0.006382000166922808       4544  0.01448499970138073           1864   \n",
       "24     0.1377120018005371       5192  0.29294899106025696           2014   \n",
       "\n",
       "                   AMR_AF AMR_N_BI_GENOS                 EAS_AF  \\\n",
       "6                     0.0            609  0.0005840000230818987   \n",
       "7   0.0024709999561309814            607                    0.0   \n",
       "15     0.1157890036702156            570     0.3907710015773773   \n",
       "16                    0.0            503   0.001905999961309135   \n",
       "24   0.020524999126791954            609    0.08235999941825867   \n",
       "\n",
       "   EAS_N_BI_GENOS                  EUR_AF EUR_N_BI_GENOS FILTER  \n",
       "6             856                     0.0           1660   PASS  \n",
       "7             856   0.0006050000083632767           1652   PASS  \n",
       "15            856     0.05963899940252304           1660   PASS  \n",
       "16            787  0.00037200000951997936           1343   PASS  \n",
       "24            856    0.022590000182390213           1660   PASS  "
      ]
     },
     "execution_count": 154,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gnomad.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 155,
   "metadata": {},
   "outputs": [],
   "source": [
    "with Pool(processes=30) as pool:\n",
    "    gnomad['genes'] = pool.starmap(fetch_gene, (\n",
    "        (gene, row['#chrom'], row['start'], row['end']) for _, row in gnomad.iterrows()\n",
    "    ), chunksize=70)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 156,
   "metadata": {},
   "outputs": [],
   "source": [
    "# gnomad['genes'] = pool.starmap(fetch_gene, (\n",
    "#         (gene, row['#chrom'], row['start'], row['end']) for _, row in gnomad.iterrows()\n",
    "#     ), chunksize=70)\n",
    "gnomad['gene'] = gnomad.apply(lambda row: fetch_gene(gene, row['#chrom'], row['start'], row['end']), axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 157,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>#chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>name</th>\n",
       "      <th>svtype</th>\n",
       "      <th>AF</th>\n",
       "      <th>N_BI_GENOS</th>\n",
       "      <th>AFR_AF</th>\n",
       "      <th>AFR_N_BI_GENOS</th>\n",
       "      <th>AMR_AF</th>\n",
       "      <th>AMR_N_BI_GENOS</th>\n",
       "      <th>EAS_AF</th>\n",
       "      <th>EAS_N_BI_GENOS</th>\n",
       "      <th>EUR_AF</th>\n",
       "      <th>EUR_N_BI_GENOS</th>\n",
       "      <th>FILTER</th>\n",
       "      <th>genes</th>\n",
       "      <th>gene</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>chr1</td>\n",
       "      <td>62398</td>\n",
       "      <td>62489</td>\n",
       "      <td>gnomAD-SV_v2.1_DEL_1_3</td>\n",
       "      <td>DEL</td>\n",
       "      <td>9.600000339560211e-05</td>\n",
       "      <td>5192</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2014</td>\n",
       "      <td>0.0</td>\n",
       "      <td>609</td>\n",
       "      <td>0.0005840000230818987</td>\n",
       "      <td>856</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1660</td>\n",
       "      <td>PASS</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>chr1</td>\n",
       "      <td>66349</td>\n",
       "      <td>66427</td>\n",
       "      <td>gnomAD-SV_v2.1_DEL_1_4</td>\n",
       "      <td>DEL</td>\n",
       "      <td>0.0004830000107176602</td>\n",
       "      <td>5177</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2009</td>\n",
       "      <td>0.0024709999561309814</td>\n",
       "      <td>607</td>\n",
       "      <td>0.0</td>\n",
       "      <td>856</td>\n",
       "      <td>0.0006050000083632767</td>\n",
       "      <td>1652</td>\n",
       "      <td>PASS</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>chr1</td>\n",
       "      <td>136999</td>\n",
       "      <td>155500</td>\n",
       "      <td>gnomAD-SV_v2.1_DUP_1_3</td>\n",
       "      <td>DUP</td>\n",
       "      <td>0.138032004237175</td>\n",
       "      <td>5122</td>\n",
       "      <td>0.10211800038814545</td>\n",
       "      <td>1983</td>\n",
       "      <td>0.1157890036702156</td>\n",
       "      <td>570</td>\n",
       "      <td>0.3907710015773773</td>\n",
       "      <td>856</td>\n",
       "      <td>0.05963899940252304</td>\n",
       "      <td>1660</td>\n",
       "      <td>PASS</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>chr1</td>\n",
       "      <td>141474</td>\n",
       "      <td>155000</td>\n",
       "      <td>gnomAD-SV_v2.1_DUP_1_4</td>\n",
       "      <td>DUP</td>\n",
       "      <td>0.006382000166922808</td>\n",
       "      <td>4544</td>\n",
       "      <td>0.01448499970138073</td>\n",
       "      <td>1864</td>\n",
       "      <td>0.0</td>\n",
       "      <td>503</td>\n",
       "      <td>0.001905999961309135</td>\n",
       "      <td>787</td>\n",
       "      <td>0.00037200000951997936</td>\n",
       "      <td>1343</td>\n",
       "      <td>PASS</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>chr1</td>\n",
       "      <td>319999</td>\n",
       "      <td>327650</td>\n",
       "      <td>gnomAD-SV_v2.1_DEL_1_9</td>\n",
       "      <td>DEL</td>\n",
       "      <td>0.1377120018005371</td>\n",
       "      <td>5192</td>\n",
       "      <td>0.29294899106025696</td>\n",
       "      <td>2014</td>\n",
       "      <td>0.020524999126791954</td>\n",
       "      <td>609</td>\n",
       "      <td>0.08235999941825867</td>\n",
       "      <td>856</td>\n",
       "      <td>0.022590000182390213</td>\n",
       "      <td>1660</td>\n",
       "      <td>PASS</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   #chrom   start     end                    name svtype  \\\n",
       "6    chr1   62398   62489  gnomAD-SV_v2.1_DEL_1_3    DEL   \n",
       "7    chr1   66349   66427  gnomAD-SV_v2.1_DEL_1_4    DEL   \n",
       "15   chr1  136999  155500  gnomAD-SV_v2.1_DUP_1_3    DUP   \n",
       "16   chr1  141474  155000  gnomAD-SV_v2.1_DUP_1_4    DUP   \n",
       "24   chr1  319999  327650  gnomAD-SV_v2.1_DEL_1_9    DEL   \n",
       "\n",
       "                       AF N_BI_GENOS               AFR_AF AFR_N_BI_GENOS  \\\n",
       "6   9.600000339560211e-05       5192                  0.0           2014   \n",
       "7   0.0004830000107176602       5177                  0.0           2009   \n",
       "15      0.138032004237175       5122  0.10211800038814545           1983   \n",
       "16   0.006382000166922808       4544  0.01448499970138073           1864   \n",
       "24     0.1377120018005371       5192  0.29294899106025696           2014   \n",
       "\n",
       "                   AMR_AF AMR_N_BI_GENOS                 EAS_AF  \\\n",
       "6                     0.0            609  0.0005840000230818987   \n",
       "7   0.0024709999561309814            607                    0.0   \n",
       "15     0.1157890036702156            570     0.3907710015773773   \n",
       "16                    0.0            503   0.001905999961309135   \n",
       "24   0.020524999126791954            609    0.08235999941825867   \n",
       "\n",
       "   EAS_N_BI_GENOS                  EUR_AF EUR_N_BI_GENOS FILTER genes gene  \n",
       "6             856                     0.0           1660   PASS             \n",
       "7             856   0.0006050000083632767           1652   PASS             \n",
       "15            856     0.05963899940252304           1660   PASS             \n",
       "16            787  0.00037200000951997936           1343   PASS             \n",
       "24            856    0.022590000182390213           1660   PASS             "
      ]
     },
     "execution_count": 157,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gnomad.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 158,
   "metadata": {},
   "outputs": [],
   "source": [
    "gnomad[\n",
    "    gnomad['svtype'] == 'DEL'\n",
    "][['#chrom', 'start', 'end', 'name', 'genes', 'AF', 'AFR_AF', 'AMR_AF', 'EAS_AF', 'EUR_AF']].rename(columns={\n",
    "    'AF': 'af', 'AFR_AF': 'af_afr', 'AMR_AF': 'af_amr', 'EAS_AF': 'af_eas', 'EUR_AF': 'af_eur'\n",
    "}).to_csv(gnomad_del_file, sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 159,
   "metadata": {},
   "outputs": [],
   "source": [
    "!bgzip -cf {gnomad_del_file} > {gnomad_del_file}.gz"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 160,
   "metadata": {},
   "outputs": [],
   "source": [
    "!tabix -fp bed {gnomad_del_file}.gz\n",
    "!rm {gnomad_del_file}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 161,
   "metadata": {},
   "outputs": [],
   "source": [
    "gnomad.loc[\n",
    "    gnomad['svtype'] == 'DUP',\n",
    "    ['#chrom', 'start', 'end', 'name', 'genes', 'AF', 'AFR_AF', 'AMR_AF', 'EAS_AF', 'EUR_AF']\n",
    "].rename(columns={\n",
    "    'AF': 'af', 'AFR_AF': 'af_afr', 'AMR_AF': 'af_amr', 'EAS_AF': 'af_eas', 'EUR_AF': 'af_eur'\n",
    "}).to_csv(gnomad_dup_file, sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 162,
   "metadata": {
    "pycharm": {
     "name": "#%%\n"
    }
   },
   "outputs": [],
   "source": [
    "!bgzip -cf {gnomad_dup_file} > {gnomad_dup_file}.gz"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 163,
   "metadata": {
    "pycharm": {
     "name": "#%%\n"
    }
   },
   "outputs": [],
   "source": [
    "!tabix -fp bed {gnomad_dup_file}.gz\n",
    "!rm {gnomad_dup_file}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "pycharm": {
     "name": "#%%\n"
    }
   },
   "outputs": [],
   "source": [
    "refGene_df = read_gtf(exon_ori_file)\n",
    "refGene_df = refGene_df[refGene_df['feature'] == 'exon']\n",
    "gene_db = pd.read_csv(settings.GENE_DATABASE, sep='\\t')\n",
    "merge_df = gene_db.merge(refGene_df, left_on='transcript', right_on='transcript_id')[[\n",
    "    '#chrom', 'start_y', 'end_y', 'gene_id_x', 'symbol', 'exon_number', 'transcript'\n",
    "]]\n",
    "merge_df.columns = [x.strip('_x').strip('_y') for x in merge_df.columns]\n",
    "\n",
    "merge_df.to_csv(settings.GENE_EXON_DATABASE.strip('.gz'), sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "outputs": [],
   "source": [
    "!bgzip -cf {settings.GENE_EXON_DATABASE.strip('.gz')} > settings.GENE_EXON_DATABASE\n",
    "!tabix -fp bed {settings.GENE_EXON_DATABASE}\n",
    "!rm {settings.GENE_EXON_DATABASE.strip('.gz')}\n"
   ],
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%%\n"
    }
   }
  }
 ],
 "metadata": {
  "interpreter": {
   "hash": "a51fe8dc5e4b3f3bbe9cb07bc9df4f2d9831135c008ea9f53caf8f1c6eba4359"
  },
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}