{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "from multiprocessing import Pool\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import allel\n",
    "import pysam\n",
    "import json"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "pd.set_option('display.max_columns', None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "refgene = pd.read_csv('data/refseq-20200109.txt', sep='\\t')\n",
    "refgene['length'] = refgene['cdsEnd'] - refgene['cdsStart']\n",
    "refgene = refgene.sort_values('length', ascending=False)\n",
    "refgene = refgene.drop_duplicates('name2', keep='first')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>#bin</th>\n",
       "      <th>name</th>\n",
       "      <th>chrom</th>\n",
       "      <th>strand</th>\n",
       "      <th>txStart</th>\n",
       "      <th>txEnd</th>\n",
       "      <th>cdsStart</th>\n",
       "      <th>cdsEnd</th>\n",
       "      <th>exonCount</th>\n",
       "      <th>exonStarts</th>\n",
       "      <th>exonEnds</th>\n",
       "      <th>score</th>\n",
       "      <th>name2</th>\n",
       "      <th>cdsStartStat</th>\n",
       "      <th>cdsEndStat</th>\n",
       "      <th>exonFrames</th>\n",
       "      <th>length</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>21052</th>\n",
       "      <td>26</td>\n",
       "      <td>NM_014141.5</td>\n",
       "      <td>chr7</td>\n",
       "      <td>+</td>\n",
       "      <td>145813452</td>\n",
       "      <td>148118090</td>\n",
       "      <td>145813968</td>\n",
       "      <td>148112708</td>\n",
       "      <td>24</td>\n",
       "      <td>145813452,146471362,146536802,146740998,146805...</td>\n",
       "      <td>145814065,146471473,146536996,146741146,146805...</td>\n",
       "      <td>0</td>\n",
       "      <td>CNTNAP2</td>\n",
       "      <td>cmpl</td>\n",
       "      <td>cmpl</td>\n",
       "      <td>0,1,1,0,1,1,0,0,1,1,2,1,1,1,2,1,1,1,1,1,0,1,1,1,</td>\n",
       "      <td>2298740</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27890</th>\n",
       "      <td>12</td>\n",
       "      <td>NM_000109.3</td>\n",
       "      <td>chrX</td>\n",
       "      <td>-</td>\n",
       "      <td>31137344</td>\n",
       "      <td>33357726</td>\n",
       "      <td>31140035</td>\n",
       "      <td>33357382</td>\n",
       "      <td>79</td>\n",
       "      <td>31137344,31144758,31152218,31164407,31165391,3...</td>\n",
       "      <td>31140047,31144790,31152311,31164531,31165635,3...</td>\n",
       "      <td>0</td>\n",
       "      <td>DMD</td>\n",
       "      <td>cmpl</td>\n",
       "      <td>cmpl</td>\n",
       "      <td>0,1,1,0,2,2,2,2,2,0,2,0,1,2,1,1,2,1,0,0,1,0,2,...</td>\n",
       "      <td>2217347</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33217</th>\n",
       "      <td>2</td>\n",
       "      <td>NM_001142699.1</td>\n",
       "      <td>chr11</td>\n",
       "      <td>-</td>\n",
       "      <td>83166054</td>\n",
       "      <td>85338314</td>\n",
       "      <td>83170860</td>\n",
       "      <td>85309740</td>\n",
       "      <td>28</td>\n",
       "      <td>83166054,83173044,83177750,83180243,83182668,8...</td>\n",
       "      <td>83170967,83173136,83177860,83180416,83182770,8...</td>\n",
       "      <td>0</td>\n",
       "      <td>DLG2</td>\n",
       "      <td>cmpl</td>\n",
       "      <td>cmpl</td>\n",
       "      <td>1,2,0,1,1,1,0,2,2,1,0,2,2,2,1,0,1,2,0,0,0,0,0,...</td>\n",
       "      <td>2138880</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23572</th>\n",
       "      <td>9</td>\n",
       "      <td>NM_033225.5</td>\n",
       "      <td>chr8</td>\n",
       "      <td>-</td>\n",
       "      <td>2792874</td>\n",
       "      <td>4852328</td>\n",
       "      <td>2796106</td>\n",
       "      <td>4851938</td>\n",
       "      <td>70</td>\n",
       "      <td>2792874,2799993,2806820,2807752,2808635,281174...</td>\n",
       "      <td>2796266,2800126,2806908,2807865,2808797,281179...</td>\n",
       "      <td>0</td>\n",
       "      <td>CSMD1</td>\n",
       "      <td>cmpl</td>\n",
       "      <td>cmpl</td>\n",
       "      <td>2,1,0,1,1,1,1,0,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,...</td>\n",
       "      <td>2055832</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54718</th>\n",
       "      <td>10</td>\n",
       "      <td>NM_080676.5</td>\n",
       "      <td>chr20</td>\n",
       "      <td>+</td>\n",
       "      <td>13976145</td>\n",
       "      <td>16033842</td>\n",
       "      <td>13976409</td>\n",
       "      <td>16030521</td>\n",
       "      <td>17</td>\n",
       "      <td>13976145,13982933,14066266,14474124,14665488,1...</td>\n",
       "      <td>13976455,13983050,14066374,14474154,14665605,1...</td>\n",
       "      <td>0</td>\n",
       "      <td>MACROD2</td>\n",
       "      <td>cmpl</td>\n",
       "      <td>cmpl</td>\n",
       "      <td>0,1,1,1,1,1,0,1,0,1,1,1,1,1,1,1,1,</td>\n",
       "      <td>2054112</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       #bin            name  chrom strand    txStart      txEnd   cdsStart  \\\n",
       "21052    26     NM_014141.5   chr7      +  145813452  148118090  145813968   \n",
       "27890    12     NM_000109.3   chrX      -   31137344   33357726   31140035   \n",
       "33217     2  NM_001142699.1  chr11      -   83166054   85338314   83170860   \n",
       "23572     9     NM_033225.5   chr8      -    2792874    4852328    2796106   \n",
       "54718    10     NM_080676.5  chr20      +   13976145   16033842   13976409   \n",
       "\n",
       "          cdsEnd  exonCount  \\\n",
       "21052  148112708         24   \n",
       "27890   33357382         79   \n",
       "33217   85309740         28   \n",
       "23572    4851938         70   \n",
       "54718   16030521         17   \n",
       "\n",
       "                                              exonStarts  \\\n",
       "21052  145813452,146471362,146536802,146740998,146805...   \n",
       "27890  31137344,31144758,31152218,31164407,31165391,3...   \n",
       "33217  83166054,83173044,83177750,83180243,83182668,8...   \n",
       "23572  2792874,2799993,2806820,2807752,2808635,281174...   \n",
       "54718  13976145,13982933,14066266,14474124,14665488,1...   \n",
       "\n",
       "                                                exonEnds  score    name2  \\\n",
       "21052  145814065,146471473,146536996,146741146,146805...      0  CNTNAP2   \n",
       "27890  31140047,31144790,31152311,31164531,31165635,3...      0      DMD   \n",
       "33217  83170967,83173136,83177860,83180416,83182770,8...      0     DLG2   \n",
       "23572  2796266,2800126,2806908,2807865,2808797,281179...      0    CSMD1   \n",
       "54718  13976455,13983050,14066374,14474154,14665605,1...      0  MACROD2   \n",
       "\n",
       "      cdsStartStat cdsEndStat  \\\n",
       "21052         cmpl       cmpl   \n",
       "27890         cmpl       cmpl   \n",
       "33217         cmpl       cmpl   \n",
       "23572         cmpl       cmpl   \n",
       "54718         cmpl       cmpl   \n",
       "\n",
       "                                              exonFrames   length  \n",
       "21052   0,1,1,0,1,1,0,0,1,1,2,1,1,1,2,1,1,1,1,1,0,1,1,1,  2298740  \n",
       "27890  0,1,1,0,2,2,2,2,2,0,2,0,1,2,1,1,2,1,0,0,1,0,2,...  2217347  \n",
       "33217  1,2,0,1,1,1,0,2,2,1,0,2,2,2,1,0,1,2,0,0,0,0,0,...  2138880  \n",
       "23572  2,1,0,1,1,1,1,0,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,...  2055832  \n",
       "54718                 0,1,1,1,1,1,0,1,0,1,1,1,1,1,1,1,1,  2054112  "
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "refgene.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_info = pd.read_csv('data/Homo_sapiens.gene_info.gz', sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>#tax_id</th>\n",
       "      <th>GeneID</th>\n",
       "      <th>Symbol</th>\n",
       "      <th>LocusTag</th>\n",
       "      <th>Synonyms</th>\n",
       "      <th>dbXrefs</th>\n",
       "      <th>chromosome</th>\n",
       "      <th>map_location</th>\n",
       "      <th>description</th>\n",
       "      <th>type_of_gene</th>\n",
       "      <th>Symbol_from_nomenclature_authority</th>\n",
       "      <th>Full_name_from_nomenclature_authority</th>\n",
       "      <th>Nomenclature_status</th>\n",
       "      <th>Other_designations</th>\n",
       "      <th>Modification_date</th>\n",
       "      <th>Feature_type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>9606</td>\n",
       "      <td>1</td>\n",
       "      <td>A1BG</td>\n",
       "      <td>-</td>\n",
       "      <td>A1B|ABG|GAB|HYST2477</td>\n",
       "      <td>MIM:138670|HGNC:HGNC:5|Ensembl:ENSG00000121410</td>\n",
       "      <td>19</td>\n",
       "      <td>19q13.43</td>\n",
       "      <td>alpha-1-B glycoprotein</td>\n",
       "      <td>protein-coding</td>\n",
       "      <td>A1BG</td>\n",
       "      <td>alpha-1-B glycoprotein</td>\n",
       "      <td>O</td>\n",
       "      <td>alpha-1B-glycoprotein|HEL-S-163pA|epididymis s...</td>\n",
       "      <td>20191220</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>9606</td>\n",
       "      <td>2</td>\n",
       "      <td>A2M</td>\n",
       "      <td>-</td>\n",
       "      <td>A2MD|CPAMD5|FWP007|S863-7</td>\n",
       "      <td>MIM:103950|HGNC:HGNC:7|Ensembl:ENSG00000175899</td>\n",
       "      <td>12</td>\n",
       "      <td>12p13.31</td>\n",
       "      <td>alpha-2-macroglobulin</td>\n",
       "      <td>protein-coding</td>\n",
       "      <td>A2M</td>\n",
       "      <td>alpha-2-macroglobulin</td>\n",
       "      <td>O</td>\n",
       "      <td>alpha-2-macroglobulin|C3 and PZP-like alpha-2-...</td>\n",
       "      <td>20200113</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>9606</td>\n",
       "      <td>3</td>\n",
       "      <td>A2MP1</td>\n",
       "      <td>-</td>\n",
       "      <td>A2MP</td>\n",
       "      <td>HGNC:HGNC:8|Ensembl:ENSG00000256069</td>\n",
       "      <td>12</td>\n",
       "      <td>12p13.31</td>\n",
       "      <td>alpha-2-macroglobulin pseudogene 1</td>\n",
       "      <td>pseudo</td>\n",
       "      <td>A2MP1</td>\n",
       "      <td>alpha-2-macroglobulin pseudogene 1</td>\n",
       "      <td>O</td>\n",
       "      <td>pregnancy-zone protein pseudogene</td>\n",
       "      <td>20191221</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>9606</td>\n",
       "      <td>9</td>\n",
       "      <td>NAT1</td>\n",
       "      <td>-</td>\n",
       "      <td>AAC1|MNAT|NAT-1|NATI</td>\n",
       "      <td>MIM:108345|HGNC:HGNC:7645|Ensembl:ENSG00000171428</td>\n",
       "      <td>8</td>\n",
       "      <td>8p22</td>\n",
       "      <td>N-acetyltransferase 1</td>\n",
       "      <td>protein-coding</td>\n",
       "      <td>NAT1</td>\n",
       "      <td>N-acetyltransferase 1</td>\n",
       "      <td>O</td>\n",
       "      <td>arylamine N-acetyltransferase 1|N-acetyltransf...</td>\n",
       "      <td>20200209</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>9606</td>\n",
       "      <td>10</td>\n",
       "      <td>NAT2</td>\n",
       "      <td>-</td>\n",
       "      <td>AAC2|NAT-2|PNAT</td>\n",
       "      <td>MIM:612182|HGNC:HGNC:7646|Ensembl:ENSG00000156006</td>\n",
       "      <td>8</td>\n",
       "      <td>8p22</td>\n",
       "      <td>N-acetyltransferase 2</td>\n",
       "      <td>protein-coding</td>\n",
       "      <td>NAT2</td>\n",
       "      <td>N-acetyltransferase 2</td>\n",
       "      <td>O</td>\n",
       "      <td>arylamine N-acetyltransferase 2|N-acetyltransf...</td>\n",
       "      <td>20200121</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   #tax_id  GeneID Symbol LocusTag                   Synonyms  \\\n",
       "0     9606       1   A1BG        -       A1B|ABG|GAB|HYST2477   \n",
       "1     9606       2    A2M        -  A2MD|CPAMD5|FWP007|S863-7   \n",
       "2     9606       3  A2MP1        -                       A2MP   \n",
       "3     9606       9   NAT1        -       AAC1|MNAT|NAT-1|NATI   \n",
       "4     9606      10   NAT2        -            AAC2|NAT-2|PNAT   \n",
       "\n",
       "                                             dbXrefs chromosome map_location  \\\n",
       "0     MIM:138670|HGNC:HGNC:5|Ensembl:ENSG00000121410         19     19q13.43   \n",
       "1     MIM:103950|HGNC:HGNC:7|Ensembl:ENSG00000175899         12     12p13.31   \n",
       "2                HGNC:HGNC:8|Ensembl:ENSG00000256069         12     12p13.31   \n",
       "3  MIM:108345|HGNC:HGNC:7645|Ensembl:ENSG00000171428          8         8p22   \n",
       "4  MIM:612182|HGNC:HGNC:7646|Ensembl:ENSG00000156006          8         8p22   \n",
       "\n",
       "                          description    type_of_gene  \\\n",
       "0              alpha-1-B glycoprotein  protein-coding   \n",
       "1               alpha-2-macroglobulin  protein-coding   \n",
       "2  alpha-2-macroglobulin pseudogene 1          pseudo   \n",
       "3               N-acetyltransferase 1  protein-coding   \n",
       "4               N-acetyltransferase 2  protein-coding   \n",
       "\n",
       "  Symbol_from_nomenclature_authority Full_name_from_nomenclature_authority  \\\n",
       "0                               A1BG                alpha-1-B glycoprotein   \n",
       "1                                A2M                 alpha-2-macroglobulin   \n",
       "2                              A2MP1    alpha-2-macroglobulin pseudogene 1   \n",
       "3                               NAT1                 N-acetyltransferase 1   \n",
       "4                               NAT2                 N-acetyltransferase 2   \n",
       "\n",
       "  Nomenclature_status                                 Other_designations  \\\n",
       "0                   O  alpha-1B-glycoprotein|HEL-S-163pA|epididymis s...   \n",
       "1                   O  alpha-2-macroglobulin|C3 and PZP-like alpha-2-...   \n",
       "2                   O                  pregnancy-zone protein pseudogene   \n",
       "3                   O  arylamine N-acetyltransferase 1|N-acetyltransf...   \n",
       "4                   O  arylamine N-acetyltransferase 2|N-acetyltransf...   \n",
       "\n",
       "   Modification_date Feature_type  \n",
       "0           20191220            -  \n",
       "1           20200113            -  \n",
       "2           20191221            -  \n",
       "3           20200209            -  \n",
       "4           20200121            -  "
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_info.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "refgene_info = refgene.merge(gene_info, left_on='name2', right_on='Symbol')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>#bin</th>\n",
       "      <th>name</th>\n",
       "      <th>chrom</th>\n",
       "      <th>strand</th>\n",
       "      <th>txStart</th>\n",
       "      <th>txEnd</th>\n",
       "      <th>cdsStart</th>\n",
       "      <th>cdsEnd</th>\n",
       "      <th>exonCount</th>\n",
       "      <th>exonStarts</th>\n",
       "      <th>exonEnds</th>\n",
       "      <th>score</th>\n",
       "      <th>name2</th>\n",
       "      <th>cdsStartStat</th>\n",
       "      <th>cdsEndStat</th>\n",
       "      <th>exonFrames</th>\n",
       "      <th>length</th>\n",
       "      <th>#tax_id</th>\n",
       "      <th>GeneID</th>\n",
       "      <th>Symbol</th>\n",
       "      <th>LocusTag</th>\n",
       "      <th>Synonyms</th>\n",
       "      <th>dbXrefs</th>\n",
       "      <th>chromosome</th>\n",
       "      <th>map_location</th>\n",
       "      <th>description</th>\n",
       "      <th>type_of_gene</th>\n",
       "      <th>Symbol_from_nomenclature_authority</th>\n",
       "      <th>Full_name_from_nomenclature_authority</th>\n",
       "      <th>Nomenclature_status</th>\n",
       "      <th>Other_designations</th>\n",
       "      <th>Modification_date</th>\n",
       "      <th>Feature_type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>26</td>\n",
       "      <td>NM_014141.5</td>\n",
       "      <td>chr7</td>\n",
       "      <td>+</td>\n",
       "      <td>145813452</td>\n",
       "      <td>148118090</td>\n",
       "      <td>145813968</td>\n",
       "      <td>148112708</td>\n",
       "      <td>24</td>\n",
       "      <td>145813452,146471362,146536802,146740998,146805...</td>\n",
       "      <td>145814065,146471473,146536996,146741146,146805...</td>\n",
       "      <td>0</td>\n",
       "      <td>CNTNAP2</td>\n",
       "      <td>cmpl</td>\n",
       "      <td>cmpl</td>\n",
       "      <td>0,1,1,0,1,1,0,0,1,1,2,1,1,1,2,1,1,1,1,1,0,1,1,1,</td>\n",
       "      <td>2298740</td>\n",
       "      <td>9606</td>\n",
       "      <td>26047</td>\n",
       "      <td>CNTNAP2</td>\n",
       "      <td>-</td>\n",
       "      <td>AUTS15|CASPR2|CDFE|NRXN4|PTHSL1</td>\n",
       "      <td>MIM:604569|HGNC:HGNC:13830|Ensembl:ENSG0000017...</td>\n",
       "      <td>7</td>\n",
       "      <td>7q35-q36.1</td>\n",
       "      <td>contactin associated protein 2</td>\n",
       "      <td>protein-coding</td>\n",
       "      <td>CNTNAP2</td>\n",
       "      <td>contactin associated protein 2</td>\n",
       "      <td>O</td>\n",
       "      <td>contactin-associated protein-like 2|cell recog...</td>\n",
       "      <td>20200105</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>12</td>\n",
       "      <td>NM_000109.3</td>\n",
       "      <td>chrX</td>\n",
       "      <td>-</td>\n",
       "      <td>31137344</td>\n",
       "      <td>33357726</td>\n",
       "      <td>31140035</td>\n",
       "      <td>33357382</td>\n",
       "      <td>79</td>\n",
       "      <td>31137344,31144758,31152218,31164407,31165391,3...</td>\n",
       "      <td>31140047,31144790,31152311,31164531,31165635,3...</td>\n",
       "      <td>0</td>\n",
       "      <td>DMD</td>\n",
       "      <td>cmpl</td>\n",
       "      <td>cmpl</td>\n",
       "      <td>0,1,1,0,2,2,2,2,2,0,2,0,1,2,1,1,2,1,0,0,1,0,2,...</td>\n",
       "      <td>2217347</td>\n",
       "      <td>9606</td>\n",
       "      <td>1756</td>\n",
       "      <td>DMD</td>\n",
       "      <td>-</td>\n",
       "      <td>BMD|CMD3B|DXS142|DXS164|DXS206|DXS230|DXS239|D...</td>\n",
       "      <td>MIM:300377|HGNC:HGNC:2928|Ensembl:ENSG00000198947</td>\n",
       "      <td>X</td>\n",
       "      <td>Xp21.2-p21.1</td>\n",
       "      <td>dystrophin</td>\n",
       "      <td>protein-coding</td>\n",
       "      <td>DMD</td>\n",
       "      <td>dystrophin</td>\n",
       "      <td>O</td>\n",
       "      <td>dystrophin</td>\n",
       "      <td>20200210</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>NM_001142699.1</td>\n",
       "      <td>chr11</td>\n",
       "      <td>-</td>\n",
       "      <td>83166054</td>\n",
       "      <td>85338314</td>\n",
       "      <td>83170860</td>\n",
       "      <td>85309740</td>\n",
       "      <td>28</td>\n",
       "      <td>83166054,83173044,83177750,83180243,83182668,8...</td>\n",
       "      <td>83170967,83173136,83177860,83180416,83182770,8...</td>\n",
       "      <td>0</td>\n",
       "      <td>DLG2</td>\n",
       "      <td>cmpl</td>\n",
       "      <td>cmpl</td>\n",
       "      <td>1,2,0,1,1,1,0,2,2,1,0,2,2,2,1,0,1,2,0,0,0,0,0,...</td>\n",
       "      <td>2138880</td>\n",
       "      <td>9606</td>\n",
       "      <td>1740</td>\n",
       "      <td>DLG2</td>\n",
       "      <td>-</td>\n",
       "      <td>PPP1R58|PSD-93|PSD93|chapsyn-110</td>\n",
       "      <td>MIM:603583|HGNC:HGNC:2901|Ensembl:ENSG00000150672</td>\n",
       "      <td>11</td>\n",
       "      <td>11q14.1</td>\n",
       "      <td>discs large MAGUK scaffold protein 2</td>\n",
       "      <td>protein-coding</td>\n",
       "      <td>DLG2</td>\n",
       "      <td>discs large MAGUK scaffold protein 2</td>\n",
       "      <td>O</td>\n",
       "      <td>disks large homolog 2|DLG2 protein isoform|cha...</td>\n",
       "      <td>20200209</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>9</td>\n",
       "      <td>NM_033225.5</td>\n",
       "      <td>chr8</td>\n",
       "      <td>-</td>\n",
       "      <td>2792874</td>\n",
       "      <td>4852328</td>\n",
       "      <td>2796106</td>\n",
       "      <td>4851938</td>\n",
       "      <td>70</td>\n",
       "      <td>2792874,2799993,2806820,2807752,2808635,281174...</td>\n",
       "      <td>2796266,2800126,2806908,2807865,2808797,281179...</td>\n",
       "      <td>0</td>\n",
       "      <td>CSMD1</td>\n",
       "      <td>cmpl</td>\n",
       "      <td>cmpl</td>\n",
       "      <td>2,1,0,1,1,1,1,0,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,...</td>\n",
       "      <td>2055832</td>\n",
       "      <td>9606</td>\n",
       "      <td>64478</td>\n",
       "      <td>CSMD1</td>\n",
       "      <td>-</td>\n",
       "      <td>PPP1R24</td>\n",
       "      <td>MIM:608397|HGNC:HGNC:14026|Ensembl:ENSG0000018...</td>\n",
       "      <td>8</td>\n",
       "      <td>8p23.2</td>\n",
       "      <td>CUB and Sushi multiple domains 1</td>\n",
       "      <td>protein-coding</td>\n",
       "      <td>CSMD1</td>\n",
       "      <td>CUB and Sushi multiple domains 1</td>\n",
       "      <td>O</td>\n",
       "      <td>CUB and sushi domain-containing protein 1|prot...</td>\n",
       "      <td>20200209</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>10</td>\n",
       "      <td>NM_080676.5</td>\n",
       "      <td>chr20</td>\n",
       "      <td>+</td>\n",
       "      <td>13976145</td>\n",
       "      <td>16033842</td>\n",
       "      <td>13976409</td>\n",
       "      <td>16030521</td>\n",
       "      <td>17</td>\n",
       "      <td>13976145,13982933,14066266,14474124,14665488,1...</td>\n",
       "      <td>13976455,13983050,14066374,14474154,14665605,1...</td>\n",
       "      <td>0</td>\n",
       "      <td>MACROD2</td>\n",
       "      <td>cmpl</td>\n",
       "      <td>cmpl</td>\n",
       "      <td>0,1,1,1,1,1,0,1,0,1,1,1,1,1,1,1,1,</td>\n",
       "      <td>2054112</td>\n",
       "      <td>9606</td>\n",
       "      <td>140733</td>\n",
       "      <td>MACROD2</td>\n",
       "      <td>-</td>\n",
       "      <td>C20orf133|C2orf133</td>\n",
       "      <td>MIM:611567|HGNC:HGNC:16126|Ensembl:ENSG0000017...</td>\n",
       "      <td>20</td>\n",
       "      <td>20p12.1</td>\n",
       "      <td>mono-ADP ribosylhydrolase 2</td>\n",
       "      <td>protein-coding</td>\n",
       "      <td>MACROD2</td>\n",
       "      <td>mono-ADP ribosylhydrolase 2</td>\n",
       "      <td>O</td>\n",
       "      <td>ADP-ribose glycohydrolase MACROD2|MACRO domain...</td>\n",
       "      <td>20191221</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   #bin            name  chrom strand    txStart      txEnd   cdsStart  \\\n",
       "0    26     NM_014141.5   chr7      +  145813452  148118090  145813968   \n",
       "1    12     NM_000109.3   chrX      -   31137344   33357726   31140035   \n",
       "2     2  NM_001142699.1  chr11      -   83166054   85338314   83170860   \n",
       "3     9     NM_033225.5   chr8      -    2792874    4852328    2796106   \n",
       "4    10     NM_080676.5  chr20      +   13976145   16033842   13976409   \n",
       "\n",
       "      cdsEnd  exonCount                                         exonStarts  \\\n",
       "0  148112708         24  145813452,146471362,146536802,146740998,146805...   \n",
       "1   33357382         79  31137344,31144758,31152218,31164407,31165391,3...   \n",
       "2   85309740         28  83166054,83173044,83177750,83180243,83182668,8...   \n",
       "3    4851938         70  2792874,2799993,2806820,2807752,2808635,281174...   \n",
       "4   16030521         17  13976145,13982933,14066266,14474124,14665488,1...   \n",
       "\n",
       "                                            exonEnds  score    name2  \\\n",
       "0  145814065,146471473,146536996,146741146,146805...      0  CNTNAP2   \n",
       "1  31140047,31144790,31152311,31164531,31165635,3...      0      DMD   \n",
       "2  83170967,83173136,83177860,83180416,83182770,8...      0     DLG2   \n",
       "3  2796266,2800126,2806908,2807865,2808797,281179...      0    CSMD1   \n",
       "4  13976455,13983050,14066374,14474154,14665605,1...      0  MACROD2   \n",
       "\n",
       "  cdsStartStat cdsEndStat                                         exonFrames  \\\n",
       "0         cmpl       cmpl   0,1,1,0,1,1,0,0,1,1,2,1,1,1,2,1,1,1,1,1,0,1,1,1,   \n",
       "1         cmpl       cmpl  0,1,1,0,2,2,2,2,2,0,2,0,1,2,1,1,2,1,0,0,1,0,2,...   \n",
       "2         cmpl       cmpl  1,2,0,1,1,1,0,2,2,1,0,2,2,2,1,0,1,2,0,0,0,0,0,...   \n",
       "3         cmpl       cmpl  2,1,0,1,1,1,1,0,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,...   \n",
       "4         cmpl       cmpl                 0,1,1,1,1,1,0,1,0,1,1,1,1,1,1,1,1,   \n",
       "\n",
       "    length  #tax_id  GeneID   Symbol LocusTag  \\\n",
       "0  2298740     9606   26047  CNTNAP2        -   \n",
       "1  2217347     9606    1756      DMD        -   \n",
       "2  2138880     9606    1740     DLG2        -   \n",
       "3  2055832     9606   64478    CSMD1        -   \n",
       "4  2054112     9606  140733  MACROD2        -   \n",
       "\n",
       "                                            Synonyms  \\\n",
       "0                    AUTS15|CASPR2|CDFE|NRXN4|PTHSL1   \n",
       "1  BMD|CMD3B|DXS142|DXS164|DXS206|DXS230|DXS239|D...   \n",
       "2                   PPP1R58|PSD-93|PSD93|chapsyn-110   \n",
       "3                                            PPP1R24   \n",
       "4                                 C20orf133|C2orf133   \n",
       "\n",
       "                                             dbXrefs chromosome  map_location  \\\n",
       "0  MIM:604569|HGNC:HGNC:13830|Ensembl:ENSG0000017...          7    7q35-q36.1   \n",
       "1  MIM:300377|HGNC:HGNC:2928|Ensembl:ENSG00000198947          X  Xp21.2-p21.1   \n",
       "2  MIM:603583|HGNC:HGNC:2901|Ensembl:ENSG00000150672         11       11q14.1   \n",
       "3  MIM:608397|HGNC:HGNC:14026|Ensembl:ENSG0000018...          8        8p23.2   \n",
       "4  MIM:611567|HGNC:HGNC:16126|Ensembl:ENSG0000017...         20       20p12.1   \n",
       "\n",
       "                            description    type_of_gene  \\\n",
       "0        contactin associated protein 2  protein-coding   \n",
       "1                            dystrophin  protein-coding   \n",
       "2  discs large MAGUK scaffold protein 2  protein-coding   \n",
       "3      CUB and Sushi multiple domains 1  protein-coding   \n",
       "4           mono-ADP ribosylhydrolase 2  protein-coding   \n",
       "\n",
       "  Symbol_from_nomenclature_authority Full_name_from_nomenclature_authority  \\\n",
       "0                            CNTNAP2        contactin associated protein 2   \n",
       "1                                DMD                            dystrophin   \n",
       "2                               DLG2  discs large MAGUK scaffold protein 2   \n",
       "3                              CSMD1      CUB and Sushi multiple domains 1   \n",
       "4                            MACROD2           mono-ADP ribosylhydrolase 2   \n",
       "\n",
       "  Nomenclature_status                                 Other_designations  \\\n",
       "0                   O  contactin-associated protein-like 2|cell recog...   \n",
       "1                   O                                         dystrophin   \n",
       "2                   O  disks large homolog 2|DLG2 protein isoform|cha...   \n",
       "3                   O  CUB and sushi domain-containing protein 1|prot...   \n",
       "4                   O  ADP-ribose glycohydrolase MACROD2|MACRO domain...   \n",
       "\n",
       "   Modification_date Feature_type  \n",
       "0           20200105            -  \n",
       "1           20200210            -  \n",
       "2           20200209            -  \n",
       "3           20200209            -  \n",
       "4           20191221            -  "
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "refgene_info.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "gene = refgene_info.loc[\n",
    "    refgene_info['type_of_gene'] == 'protein-coding',\n",
    "    ['chrom', 'txStart', 'txEnd', 'GeneID', 'name2', 'name', 'strand']\n",
    "].sort_values(['chrom', 'txStart', 'txEnd'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>txStart</th>\n",
       "      <th>txEnd</th>\n",
       "      <th>GeneID</th>\n",
       "      <th>name2</th>\n",
       "      <th>name</th>\n",
       "      <th>strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>17803</th>\n",
       "      <td>chr1</td>\n",
       "      <td>69090</td>\n",
       "      <td>70008</td>\n",
       "      <td>79501</td>\n",
       "      <td>OR4F5</td>\n",
       "      <td>NM_001005484.1</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17672</th>\n",
       "      <td>chr1</td>\n",
       "      <td>367658</td>\n",
       "      <td>368597</td>\n",
       "      <td>729759</td>\n",
       "      <td>OR4F29</td>\n",
       "      <td>NM_001005221.2</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17664</th>\n",
       "      <td>chr1</td>\n",
       "      <td>621095</td>\n",
       "      <td>622034</td>\n",
       "      <td>81399</td>\n",
       "      <td>OR4F16</td>\n",
       "      <td>NM_001005277.1</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9248</th>\n",
       "      <td>chr1</td>\n",
       "      <td>861120</td>\n",
       "      <td>879961</td>\n",
       "      <td>148398</td>\n",
       "      <td>SAMD11</td>\n",
       "      <td>NM_152486.2</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10208</th>\n",
       "      <td>chr1</td>\n",
       "      <td>879582</td>\n",
       "      <td>894679</td>\n",
       "      <td>26155</td>\n",
       "      <td>NOC2L</td>\n",
       "      <td>NM_015658.3</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      chrom  txStart   txEnd  GeneID   name2            name strand\n",
       "17803  chr1    69090   70008   79501   OR4F5  NM_001005484.1      +\n",
       "17672  chr1   367658  368597  729759  OR4F29  NM_001005221.2      +\n",
       "17664  chr1   621095  622034   81399  OR4F16  NM_001005277.1      -\n",
       "9248   chr1   861120  879961  148398  SAMD11     NM_152486.2      +\n",
       "10208  chr1   879582  894679   26155   NOC2L     NM_015658.3      -"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "gene.rename(columns={\n",
    "    'chrom': '#chrom', 'txStart': 'start', 'txEnd': 'end',\n",
    "    'GeneID': 'gene_id', 'name2': 'symbol', 'name': 'transcript'\n",
    "}).to_csv('data/gene-20200109.sorted.bed', index=False, sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "!bgzip -cf data/gene-20200109.sorted.bed > ../data/gene-20200109.sorted.bed.gz"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "!tabix -fp bed ../data/gene-20200109.sorted.bed.gz"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open('data/omim-gene.list') as f:\n",
    "    omim_gene = set(f.read().split('\\n'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "omim_gene = refgene_info.loc[\n",
    "    refgene_info['name2'].isin(omim_gene),\n",
    "    ['chrom', 'txStart', 'txEnd', 'GeneID', 'name2', 'name', 'strand']\n",
    "].sort_values(['chrom', 'txStart', 'txEnd'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>txStart</th>\n",
       "      <th>txEnd</th>\n",
       "      <th>GeneID</th>\n",
       "      <th>name2</th>\n",
       "      <th>name</th>\n",
       "      <th>strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>17825</th>\n",
       "      <td>chr1</td>\n",
       "      <td>948846</td>\n",
       "      <td>949920</td>\n",
       "      <td>9636</td>\n",
       "      <td>ISG15</td>\n",
       "      <td>NM_005101.3</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6298</th>\n",
       "      <td>chr1</td>\n",
       "      <td>955502</td>\n",
       "      <td>991499</td>\n",
       "      <td>375790</td>\n",
       "      <td>AGRN</td>\n",
       "      <td>NM_001305275.1</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15599</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1146705</td>\n",
       "      <td>1149548</td>\n",
       "      <td>7293</td>\n",
       "      <td>TNFRSF4</td>\n",
       "      <td>NM_003327.3</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17342</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1167628</td>\n",
       "      <td>1170421</td>\n",
       "      <td>126792</td>\n",
       "      <td>B3GALT6</td>\n",
       "      <td>NM_080605.3</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10735</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1270657</td>\n",
       "      <td>1284492</td>\n",
       "      <td>1855</td>\n",
       "      <td>DVL1</td>\n",
       "      <td>NM_001330311.1</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      chrom  txStart    txEnd  GeneID    name2            name strand\n",
       "17825  chr1   948846   949920    9636    ISG15     NM_005101.3      +\n",
       "6298   chr1   955502   991499  375790     AGRN  NM_001305275.1      +\n",
       "15599  chr1  1146705  1149548    7293  TNFRSF4     NM_003327.3      -\n",
       "17342  chr1  1167628  1170421  126792  B3GALT6     NM_080605.3      +\n",
       "10735  chr1  1270657  1284492    1855     DVL1  NM_001330311.1      -"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "omim_gene.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "omim_gene.rename(columns={\n",
    "    'chrom': '#chrom', 'txStart': 'start', 'txEnd': 'end',\n",
    "    'GeneID': 'gene_id', 'name2': 'symbol', 'name': 'transcript'\n",
    "}).to_csv('data/omim-gene-20200109.sorted.bed', index=False, sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "!bgzip -cf data/omim-gene-20200109.sorted.bed > ../data/omim-gene-20200109.sorted.bed.gz"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "!tabix -fp bed ../data/omim-gene-20200109.sorted.bed.gz"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "curation_gene = pd.read_csv('data/ClinGen_gene_curation_list_GRCh37.tsv', sep='\\t', dtype=str, skiprows=5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>#Gene Symbol</th>\n",
       "      <th>Gene ID</th>\n",
       "      <th>cytoBand</th>\n",
       "      <th>Genomic Location</th>\n",
       "      <th>Haploinsufficiency Score</th>\n",
       "      <th>Haploinsufficiency Description</th>\n",
       "      <th>Haploinsufficiency PMID1</th>\n",
       "      <th>Haploinsufficiency PMID2</th>\n",
       "      <th>Haploinsufficiency PMID3</th>\n",
       "      <th>Triplosensitivity Score</th>\n",
       "      <th>Triplosensitivity Description</th>\n",
       "      <th>Triplosensitivity PMID1</th>\n",
       "      <th>Triplosensitivity PMID2</th>\n",
       "      <th>Triplosensitivity PMID3</th>\n",
       "      <th>Date Last Evaluated</th>\n",
       "      <th>Loss phenotype OMIM ID</th>\n",
       "      <th>Triplosensitive phenotype OMIM ID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A4GALT</td>\n",
       "      <td>53947</td>\n",
       "      <td>22q13.2</td>\n",
       "      <td>chr22:43088121-43117307</td>\n",
       "      <td>30</td>\n",
       "      <td>Gene associated with autosomal recessive pheno...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>No evidence available</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2014-12-11</td>\n",
       "      <td>111400</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AAGAB</td>\n",
       "      <td>79719</td>\n",
       "      <td>15q23</td>\n",
       "      <td>chr15:67493013-67547536</td>\n",
       "      <td>3</td>\n",
       "      <td>Sufficient evidence for dosage pathogenicity</td>\n",
       "      <td>23064416</td>\n",
       "      <td>23000146</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>No evidence available</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2013-02-28</td>\n",
       "      <td>148600</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AARS</td>\n",
       "      <td>16</td>\n",
       "      <td>16q22.1</td>\n",
       "      <td>chr16:70286297-70323412</td>\n",
       "      <td>0</td>\n",
       "      <td>No evidence available</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>No evidence available</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2018-01-11</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AARS2</td>\n",
       "      <td>57505</td>\n",
       "      <td>6p21.1</td>\n",
       "      <td>chr6:44266463-44281063</td>\n",
       "      <td>30</td>\n",
       "      <td>Gene associated with autosomal recessive pheno...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Not yet evaluated</td>\n",
       "      <td>Not yet evaluated</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2016-08-22</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AASS</td>\n",
       "      <td>10157</td>\n",
       "      <td>7q31.32</td>\n",
       "      <td>chr7:121713598-121784344</td>\n",
       "      <td>30</td>\n",
       "      <td>Gene associated with autosomal recessive pheno...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Not yet evaluated</td>\n",
       "      <td>Not yet evaluated</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2016-08-22</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  #Gene Symbol Gene ID cytoBand          Genomic Location  \\\n",
       "0       A4GALT   53947  22q13.2   chr22:43088121-43117307   \n",
       "1        AAGAB   79719    15q23   chr15:67493013-67547536   \n",
       "2         AARS      16  16q22.1   chr16:70286297-70323412   \n",
       "3        AARS2   57505   6p21.1    chr6:44266463-44281063   \n",
       "4         AASS   10157  7q31.32  chr7:121713598-121784344   \n",
       "\n",
       "  Haploinsufficiency Score                     Haploinsufficiency Description  \\\n",
       "0                       30  Gene associated with autosomal recessive pheno...   \n",
       "1                        3       Sufficient evidence for dosage pathogenicity   \n",
       "2                        0                              No evidence available   \n",
       "3                       30  Gene associated with autosomal recessive pheno...   \n",
       "4                       30  Gene associated with autosomal recessive pheno...   \n",
       "\n",
       "  Haploinsufficiency PMID1 Haploinsufficiency PMID2 Haploinsufficiency PMID3  \\\n",
       "0                      NaN                      NaN                      NaN   \n",
       "1                 23064416                 23000146                      NaN   \n",
       "2                      NaN                      NaN                      NaN   \n",
       "3                      NaN                      NaN                      NaN   \n",
       "4                      NaN                      NaN                      NaN   \n",
       "\n",
       "  Triplosensitivity Score Triplosensitivity Description  \\\n",
       "0                       0         No evidence available   \n",
       "1                       0         No evidence available   \n",
       "2                       0         No evidence available   \n",
       "3       Not yet evaluated             Not yet evaluated   \n",
       "4       Not yet evaluated             Not yet evaluated   \n",
       "\n",
       "  Triplosensitivity PMID1 Triplosensitivity PMID2 Triplosensitivity PMID3  \\\n",
       "0                     NaN                     NaN                     NaN   \n",
       "1                     NaN                     NaN                     NaN   \n",
       "2                     NaN                     NaN                     NaN   \n",
       "3                     NaN                     NaN                     NaN   \n",
       "4                     NaN                     NaN                     NaN   \n",
       "\n",
       "  Date Last Evaluated Loss phenotype OMIM ID Triplosensitive phenotype OMIM ID  \n",
       "0          2014-12-11                 111400                               NaN  \n",
       "1          2013-02-28                 148600                               NaN  \n",
       "2          2018-01-11                    NaN                               NaN  \n",
       "3          2016-08-22                    NaN                               NaN  \n",
       "4          2016-08-22                    NaN                               NaN  "
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "curation_gene.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "hi_genes = set(\n",
    "    curation_gene.loc[\n",
    "        curation_gene['Haploinsufficiency Score'] == '3', '#Gene Symbol'\n",
    "    ]\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "hi_gene = refgene_info.loc[\n",
    "    refgene_info['name2'].isin(hi_genes),\n",
    "    ['chrom', 'txStart', 'txEnd', 'GeneID', 'name2', 'name', 'strand']\n",
    "].sort_values(['chrom', 'txStart', 'txEnd'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>txStart</th>\n",
       "      <th>txEnd</th>\n",
       "      <th>GeneID</th>\n",
       "      <th>name2</th>\n",
       "      <th>name</th>\n",
       "      <th>strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>47</th>\n",
       "      <td>chr1</td>\n",
       "      <td>6845383</td>\n",
       "      <td>7829766</td>\n",
       "      <td>23261</td>\n",
       "      <td>CAMTA1</td>\n",
       "      <td>NM_015215.3</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6265</th>\n",
       "      <td>chr1</td>\n",
       "      <td>17345216</td>\n",
       "      <td>17380665</td>\n",
       "      <td>6390</td>\n",
       "      <td>SDHB</td>\n",
       "      <td>NM_003000.2</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2946</th>\n",
       "      <td>chr1</td>\n",
       "      <td>27022521</td>\n",
       "      <td>27108601</td>\n",
       "      <td>8289</td>\n",
       "      <td>ARID1A</td>\n",
       "      <td>NM_006015.4</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14219</th>\n",
       "      <td>chr1</td>\n",
       "      <td>27860755</td>\n",
       "      <td>27930005</td>\n",
       "      <td>27245</td>\n",
       "      <td>AHDC1</td>\n",
       "      <td>NM_001029882.3</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6768</th>\n",
       "      <td>chr1</td>\n",
       "      <td>43391045</td>\n",
       "      <td>43424847</td>\n",
       "      <td>6513</td>\n",
       "      <td>SLC2A1</td>\n",
       "      <td>NM_006516.2</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      chrom   txStart     txEnd  GeneID   name2            name strand\n",
       "47     chr1   6845383   7829766   23261  CAMTA1     NM_015215.3      +\n",
       "6265   chr1  17345216  17380665    6390    SDHB     NM_003000.2      -\n",
       "2946   chr1  27022521  27108601    8289  ARID1A     NM_006015.4      +\n",
       "14219  chr1  27860755  27930005   27245   AHDC1  NM_001029882.3      -\n",
       "6768   chr1  43391045  43424847    6513  SLC2A1     NM_006516.2      -"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "hi_gene.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "hi_gene.rename(columns={\n",
    "    'chrom': '#chrom', 'txStart': 'start', 'txEnd': 'end',\n",
    "    'GeneID': 'gene_id', 'name2': 'symbol', 'name': 'transcript'\n",
    "}).to_csv('data/hi-gene-20200109.sorted.bed', index=False, sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "!bgzip -cf data/hi-gene-20200109.sorted.bed > ../data/hi-gene-20200109.sorted.bed.gz"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [],
   "source": [
    "!tabix -fp bed ../data/hi-gene-20200109.sorted.bed.gz"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [],
   "source": [
    "hi_cds = refgene_info.loc[\n",
    "    (refgene_info['name2'].isin(hi_genes)) & (refgene_info['length'] != 0),\n",
    "    ['chrom', 'cdsStart', 'cdsEnd', 'GeneID', 'name2', 'name']\n",
    "].sort_values(['chrom', 'cdsStart', 'cdsEnd'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>cdsStart</th>\n",
       "      <th>cdsEnd</th>\n",
       "      <th>GeneID</th>\n",
       "      <th>name2</th>\n",
       "      <th>name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>47</th>\n",
       "      <td>chr1</td>\n",
       "      <td>6845590</td>\n",
       "      <td>7826551</td>\n",
       "      <td>23261</td>\n",
       "      <td>CAMTA1</td>\n",
       "      <td>NM_015215.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6265</th>\n",
       "      <td>chr1</td>\n",
       "      <td>17345375</td>\n",
       "      <td>17380514</td>\n",
       "      <td>6390</td>\n",
       "      <td>SDHB</td>\n",
       "      <td>NM_003000.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2946</th>\n",
       "      <td>chr1</td>\n",
       "      <td>27022894</td>\n",
       "      <td>27107247</td>\n",
       "      <td>8289</td>\n",
       "      <td>ARID1A</td>\n",
       "      <td>NM_006015.4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14219</th>\n",
       "      <td>chr1</td>\n",
       "      <td>27873814</td>\n",
       "      <td>27878626</td>\n",
       "      <td>27245</td>\n",
       "      <td>AHDC1</td>\n",
       "      <td>NM_001029882.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6768</th>\n",
       "      <td>chr1</td>\n",
       "      <td>43392711</td>\n",
       "      <td>43424322</td>\n",
       "      <td>6513</td>\n",
       "      <td>SLC2A1</td>\n",
       "      <td>NM_006516.2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      chrom  cdsStart    cdsEnd  GeneID   name2            name\n",
       "47     chr1   6845590   7826551   23261  CAMTA1     NM_015215.3\n",
       "6265   chr1  17345375  17380514    6390    SDHB     NM_003000.2\n",
       "2946   chr1  27022894  27107247    8289  ARID1A     NM_006015.4\n",
       "14219  chr1  27873814  27878626   27245   AHDC1  NM_001029882.3\n",
       "6768   chr1  43392711  43424322    6513  SLC2A1     NM_006516.2"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "hi_cds.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [],
   "source": [
    "hi_cds.rename(columns={\n",
    "    'chrom': '#chrom', 'cdsStart': 'start', 'cdsEnd': 'end',\n",
    "    'GeneID': 'gene_id', 'name2': 'symbol', 'name': 'transcript'\n",
    "}).to_csv('data/hi-cds-20200109.sorted.bed', index=False, sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [],
   "source": [
    "!bgzip -cf data/hi-cds-20200109.sorted.bed > ../data/hi-cds-20200109.sorted.bed.gz"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [],
   "source": [
    "!tabix -fp bed ../data/hi-cds-20200109.sorted.bed.gz"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [],
   "source": [
    "hi_exons = refgene_info.loc[\n",
    "    refgene_info['name2'].isin(hi_genes), ['chrom', 'exonStarts', 'exonEnds', 'GeneID', 'name2', 'name', 'strand']\n",
    "].copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>exonStarts</th>\n",
       "      <th>exonEnds</th>\n",
       "      <th>GeneID</th>\n",
       "      <th>name2</th>\n",
       "      <th>name</th>\n",
       "      <th>strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chrX</td>\n",
       "      <td>31137344,31144758,31152218,31164407,31165391,3...</td>\n",
       "      <td>31140047,31144790,31152311,31164531,31165635,3...</td>\n",
       "      <td>1756</td>\n",
       "      <td>DMD</td>\n",
       "      <td>NM_000109.3</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>chr7</td>\n",
       "      <td>69063904,69364271,69583117,69599521,69900737,7...</td>\n",
       "      <td>69064948,69364484,69583219,69599557,69900767,7...</td>\n",
       "      <td>26053</td>\n",
       "      <td>AUTS2</td>\n",
       "      <td>NM_015570.3</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>chrX</td>\n",
       "      <td>28605680,28807436,29301054,29414374,29417271,2...</td>\n",
       "      <td>28606164,28807542,29301334,29414561,29417425,2...</td>\n",
       "      <td>11141</td>\n",
       "      <td>IL1RAPL1</td>\n",
       "      <td>NM_014271.3</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>chr2</td>\n",
       "      <td>50145642,50170850,50280408,50318460,50463926,5...</td>\n",
       "      <td>50149389,50170929,50280728,50318632,50464108,5...</td>\n",
       "      <td>9378</td>\n",
       "      <td>NRXN1</td>\n",
       "      <td>NM_001330096.1</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>47</th>\n",
       "      <td>chr1</td>\n",
       "      <td>6845383,6880240,6885151,7151363,7309550,752788...</td>\n",
       "      <td>6845635,6880310,6885270,7151431,7309686,752796...</td>\n",
       "      <td>23261</td>\n",
       "      <td>CAMTA1</td>\n",
       "      <td>NM_015215.3</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   chrom                                         exonStarts  \\\n",
       "1   chrX  31137344,31144758,31152218,31164407,31165391,3...   \n",
       "23  chr7  69063904,69364271,69583117,69599521,69900737,7...   \n",
       "27  chrX  28605680,28807436,29301054,29414374,29417271,2...   \n",
       "40  chr2  50145642,50170850,50280408,50318460,50463926,5...   \n",
       "47  chr1  6845383,6880240,6885151,7151363,7309550,752788...   \n",
       "\n",
       "                                             exonEnds  GeneID     name2  \\\n",
       "1   31140047,31144790,31152311,31164531,31165635,3...    1756       DMD   \n",
       "23  69064948,69364484,69583219,69599557,69900767,7...   26053     AUTS2   \n",
       "27  28606164,28807542,29301334,29414561,29417425,2...   11141  IL1RAPL1   \n",
       "40  50149389,50170929,50280728,50318632,50464108,5...    9378     NRXN1   \n",
       "47  6845635,6880310,6885270,7151431,7309686,752796...   23261    CAMTA1   \n",
       "\n",
       "              name strand  \n",
       "1      NM_000109.3      -  \n",
       "23     NM_015570.3      +  \n",
       "27     NM_014271.3      +  \n",
       "40  NM_001330096.1      -  \n",
       "47     NM_015215.3      +  "
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "hi_exons.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [],
   "source": [
    "hi_exons['exonStarts'] = hi_exons['exonStarts'].str.replace(r',$', '')\n",
    "hi_exons['exonEnds'] = hi_exons['exonEnds'].str.replace(r',$', '')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>exonStarts</th>\n",
       "      <th>exonEnds</th>\n",
       "      <th>GeneID</th>\n",
       "      <th>name2</th>\n",
       "      <th>name</th>\n",
       "      <th>strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chrX</td>\n",
       "      <td>31137344,31144758,31152218,31164407,31165391,3...</td>\n",
       "      <td>31140047,31144790,31152311,31164531,31165635,3...</td>\n",
       "      <td>1756</td>\n",
       "      <td>DMD</td>\n",
       "      <td>NM_000109.3</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>chr7</td>\n",
       "      <td>69063904,69364271,69583117,69599521,69900737,7...</td>\n",
       "      <td>69064948,69364484,69583219,69599557,69900767,7...</td>\n",
       "      <td>26053</td>\n",
       "      <td>AUTS2</td>\n",
       "      <td>NM_015570.3</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>chrX</td>\n",
       "      <td>28605680,28807436,29301054,29414374,29417271,2...</td>\n",
       "      <td>28606164,28807542,29301334,29414561,29417425,2...</td>\n",
       "      <td>11141</td>\n",
       "      <td>IL1RAPL1</td>\n",
       "      <td>NM_014271.3</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>chr2</td>\n",
       "      <td>50145642,50170850,50280408,50318460,50463926,5...</td>\n",
       "      <td>50149389,50170929,50280728,50318632,50464108,5...</td>\n",
       "      <td>9378</td>\n",
       "      <td>NRXN1</td>\n",
       "      <td>NM_001330096.1</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>47</th>\n",
       "      <td>chr1</td>\n",
       "      <td>6845383,6880240,6885151,7151363,7309550,752788...</td>\n",
       "      <td>6845635,6880310,6885270,7151431,7309686,752796...</td>\n",
       "      <td>23261</td>\n",
       "      <td>CAMTA1</td>\n",
       "      <td>NM_015215.3</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   chrom                                         exonStarts  \\\n",
       "1   chrX  31137344,31144758,31152218,31164407,31165391,3...   \n",
       "23  chr7  69063904,69364271,69583117,69599521,69900737,7...   \n",
       "27  chrX  28605680,28807436,29301054,29414374,29417271,2...   \n",
       "40  chr2  50145642,50170850,50280408,50318460,50463926,5...   \n",
       "47  chr1  6845383,6880240,6885151,7151363,7309550,752788...   \n",
       "\n",
       "                                             exonEnds  GeneID     name2  \\\n",
       "1   31140047,31144790,31152311,31164531,31165635,3...    1756       DMD   \n",
       "23  69064948,69364484,69583219,69599557,69900767,7...   26053     AUTS2   \n",
       "27  28606164,28807542,29301334,29414561,29417425,2...   11141  IL1RAPL1   \n",
       "40  50149389,50170929,50280728,50318632,50464108,5...    9378     NRXN1   \n",
       "47  6845635,6880310,6885270,7151431,7309686,752796...   23261    CAMTA1   \n",
       "\n",
       "              name strand  \n",
       "1      NM_000109.3      -  \n",
       "23     NM_015570.3      +  \n",
       "27     NM_014271.3      +  \n",
       "40  NM_001330096.1      -  \n",
       "47     NM_015215.3      +  "
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "hi_exons.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [],
   "source": [
    "start = hi_exons['exonStarts'].str.split(',').apply(pd.Series).stack().reset_index()\n",
    "start = start.rename(columns={'level_0': 'row', 0: 'start'})[['row', 'start']]\n",
    "start['start'] = start['start'].astype(int)\n",
    "end = hi_exons['exonEnds'].str.split(',').apply(pd.Series).stack().reset_index()\n",
    "end = end.rename(columns={0: 'end'})['end'].astype(int)\n",
    "position = start.join(end)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>row</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>31137344</td>\n",
       "      <td>31140047</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>31144758</td>\n",
       "      <td>31144790</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>31152218</td>\n",
       "      <td>31152311</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>31164407</td>\n",
       "      <td>31164531</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>31165391</td>\n",
       "      <td>31165635</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   row     start       end\n",
       "0    1  31137344  31140047\n",
       "1    1  31144758  31144790\n",
       "2    1  31152218  31152311\n",
       "3    1  31164407  31164531\n",
       "4    1  31165391  31165635"
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "position.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [],
   "source": [
    "exon = position.merge(\n",
    "    hi_exons[['chrom', 'GeneID', 'name2', 'name', 'strand']], how='left', left_on='row', right_index=True\n",
    ")\n",
    "exon = exon.sort_values(['chrom', 'start', 'end'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>row</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>chrom</th>\n",
       "      <th>GeneID</th>\n",
       "      <th>name2</th>\n",
       "      <th>name</th>\n",
       "      <th>strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>126</th>\n",
       "      <td>47</td>\n",
       "      <td>6845383</td>\n",
       "      <td>6845635</td>\n",
       "      <td>chr1</td>\n",
       "      <td>23261</td>\n",
       "      <td>CAMTA1</td>\n",
       "      <td>NM_015215.3</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>127</th>\n",
       "      <td>47</td>\n",
       "      <td>6880240</td>\n",
       "      <td>6880310</td>\n",
       "      <td>chr1</td>\n",
       "      <td>23261</td>\n",
       "      <td>CAMTA1</td>\n",
       "      <td>NM_015215.3</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>128</th>\n",
       "      <td>47</td>\n",
       "      <td>6885151</td>\n",
       "      <td>6885270</td>\n",
       "      <td>chr1</td>\n",
       "      <td>23261</td>\n",
       "      <td>CAMTA1</td>\n",
       "      <td>NM_015215.3</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>129</th>\n",
       "      <td>47</td>\n",
       "      <td>7151363</td>\n",
       "      <td>7151431</td>\n",
       "      <td>chr1</td>\n",
       "      <td>23261</td>\n",
       "      <td>CAMTA1</td>\n",
       "      <td>NM_015215.3</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>130</th>\n",
       "      <td>47</td>\n",
       "      <td>7309550</td>\n",
       "      <td>7309686</td>\n",
       "      <td>chr1</td>\n",
       "      <td>23261</td>\n",
       "      <td>CAMTA1</td>\n",
       "      <td>NM_015215.3</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     row    start      end chrom  GeneID   name2         name strand\n",
       "126   47  6845383  6845635  chr1   23261  CAMTA1  NM_015215.3      +\n",
       "127   47  6880240  6880310  chr1   23261  CAMTA1  NM_015215.3      +\n",
       "128   47  6885151  6885270  chr1   23261  CAMTA1  NM_015215.3      +\n",
       "129   47  7151363  7151431  chr1   23261  CAMTA1  NM_015215.3      +\n",
       "130   47  7309550  7309686  chr1   23261  CAMTA1  NM_015215.3      +"
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "exon.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [],
   "source": [
    "exon['+'] = exon.groupby(['name2', 'name'])['start'].rank('first', ascending=True).astype(int)\n",
    "exon['-'] = exon.groupby(['name2', 'name'])['start'].rank('first', ascending=False).astype(int)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>row</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>chrom</th>\n",
       "      <th>GeneID</th>\n",
       "      <th>name2</th>\n",
       "      <th>name</th>\n",
       "      <th>strand</th>\n",
       "      <th>+</th>\n",
       "      <th>-</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>126</th>\n",
       "      <td>47</td>\n",
       "      <td>6845383</td>\n",
       "      <td>6845635</td>\n",
       "      <td>chr1</td>\n",
       "      <td>23261</td>\n",
       "      <td>CAMTA1</td>\n",
       "      <td>NM_015215.3</td>\n",
       "      <td>+</td>\n",
       "      <td>1</td>\n",
       "      <td>23</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>127</th>\n",
       "      <td>47</td>\n",
       "      <td>6880240</td>\n",
       "      <td>6880310</td>\n",
       "      <td>chr1</td>\n",
       "      <td>23261</td>\n",
       "      <td>CAMTA1</td>\n",
       "      <td>NM_015215.3</td>\n",
       "      <td>+</td>\n",
       "      <td>2</td>\n",
       "      <td>22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>128</th>\n",
       "      <td>47</td>\n",
       "      <td>6885151</td>\n",
       "      <td>6885270</td>\n",
       "      <td>chr1</td>\n",
       "      <td>23261</td>\n",
       "      <td>CAMTA1</td>\n",
       "      <td>NM_015215.3</td>\n",
       "      <td>+</td>\n",
       "      <td>3</td>\n",
       "      <td>21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>129</th>\n",
       "      <td>47</td>\n",
       "      <td>7151363</td>\n",
       "      <td>7151431</td>\n",
       "      <td>chr1</td>\n",
       "      <td>23261</td>\n",
       "      <td>CAMTA1</td>\n",
       "      <td>NM_015215.3</td>\n",
       "      <td>+</td>\n",
       "      <td>4</td>\n",
       "      <td>20</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>130</th>\n",
       "      <td>47</td>\n",
       "      <td>7309550</td>\n",
       "      <td>7309686</td>\n",
       "      <td>chr1</td>\n",
       "      <td>23261</td>\n",
       "      <td>CAMTA1</td>\n",
       "      <td>NM_015215.3</td>\n",
       "      <td>+</td>\n",
       "      <td>5</td>\n",
       "      <td>19</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     row    start      end chrom  GeneID   name2         name strand  +   -\n",
       "126   47  6845383  6845635  chr1   23261  CAMTA1  NM_015215.3      +  1  23\n",
       "127   47  6880240  6880310  chr1   23261  CAMTA1  NM_015215.3      +  2  22\n",
       "128   47  6885151  6885270  chr1   23261  CAMTA1  NM_015215.3      +  3  21\n",
       "129   47  7151363  7151431  chr1   23261  CAMTA1  NM_015215.3      +  4  20\n",
       "130   47  7309550  7309686  chr1   23261  CAMTA1  NM_015215.3      +  5  19"
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "exon.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>row</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>chrom</th>\n",
       "      <th>GeneID</th>\n",
       "      <th>name2</th>\n",
       "      <th>name</th>\n",
       "      <th>strand</th>\n",
       "      <th>+</th>\n",
       "      <th>-</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>3759</th>\n",
       "      <td>7339</td>\n",
       "      <td>545352</td>\n",
       "      <td>545561</td>\n",
       "      <td>chrY</td>\n",
       "      <td>6473</td>\n",
       "      <td>SHOX</td>\n",
       "      <td>NM_006883.2</td>\n",
       "      <td>+</td>\n",
       "      <td>3</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3760</th>\n",
       "      <td>7339</td>\n",
       "      <td>551555</td>\n",
       "      <td>551613</td>\n",
       "      <td>chrY</td>\n",
       "      <td>6473</td>\n",
       "      <td>SHOX</td>\n",
       "      <td>NM_006883.2</td>\n",
       "      <td>+</td>\n",
       "      <td>4</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3761</th>\n",
       "      <td>7339</td>\n",
       "      <td>551733</td>\n",
       "      <td>551822</td>\n",
       "      <td>chrY</td>\n",
       "      <td>6473</td>\n",
       "      <td>SHOX</td>\n",
       "      <td>NM_006883.2</td>\n",
       "      <td>+</td>\n",
       "      <td>5</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3762</th>\n",
       "      <td>7339</td>\n",
       "      <td>569519</td>\n",
       "      <td>570146</td>\n",
       "      <td>chrY</td>\n",
       "      <td>6473</td>\n",
       "      <td>SHOX</td>\n",
       "      <td>NM_006883.2</td>\n",
       "      <td>+</td>\n",
       "      <td>6</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4615</th>\n",
       "      <td>18161</td>\n",
       "      <td>2654895</td>\n",
       "      <td>2655782</td>\n",
       "      <td>chrY</td>\n",
       "      <td>6736</td>\n",
       "      <td>SRY</td>\n",
       "      <td>NM_003140.2</td>\n",
       "      <td>-</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        row    start      end chrom  GeneID name2         name strand  +  -\n",
       "3759   7339   545352   545561  chrY    6473  SHOX  NM_006883.2      +  3  4\n",
       "3760   7339   551555   551613  chrY    6473  SHOX  NM_006883.2      +  4  3\n",
       "3761   7339   551733   551822  chrY    6473  SHOX  NM_006883.2      +  5  2\n",
       "3762   7339   569519   570146  chrY    6473  SHOX  NM_006883.2      +  6  1\n",
       "4615  18161  2654895  2655782  chrY    6736   SRY  NM_003140.2      -  1  1"
      ]
     },
     "execution_count": 43,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "exon.tail()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [],
   "source": [
    "exon['exon'] = pd.concat([exon.loc[exon['strand'] == '+', '+'], exon.loc[exon['strand'] == '-', '-']])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>row</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>chrom</th>\n",
       "      <th>GeneID</th>\n",
       "      <th>name2</th>\n",
       "      <th>name</th>\n",
       "      <th>strand</th>\n",
       "      <th>+</th>\n",
       "      <th>-</th>\n",
       "      <th>exon</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>126</th>\n",
       "      <td>47</td>\n",
       "      <td>6845383</td>\n",
       "      <td>6845635</td>\n",
       "      <td>chr1</td>\n",
       "      <td>23261</td>\n",
       "      <td>CAMTA1</td>\n",
       "      <td>NM_015215.3</td>\n",
       "      <td>+</td>\n",
       "      <td>1</td>\n",
       "      <td>23</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>127</th>\n",
       "      <td>47</td>\n",
       "      <td>6880240</td>\n",
       "      <td>6880310</td>\n",
       "      <td>chr1</td>\n",
       "      <td>23261</td>\n",
       "      <td>CAMTA1</td>\n",
       "      <td>NM_015215.3</td>\n",
       "      <td>+</td>\n",
       "      <td>2</td>\n",
       "      <td>22</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>128</th>\n",
       "      <td>47</td>\n",
       "      <td>6885151</td>\n",
       "      <td>6885270</td>\n",
       "      <td>chr1</td>\n",
       "      <td>23261</td>\n",
       "      <td>CAMTA1</td>\n",
       "      <td>NM_015215.3</td>\n",
       "      <td>+</td>\n",
       "      <td>3</td>\n",
       "      <td>21</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>129</th>\n",
       "      <td>47</td>\n",
       "      <td>7151363</td>\n",
       "      <td>7151431</td>\n",
       "      <td>chr1</td>\n",
       "      <td>23261</td>\n",
       "      <td>CAMTA1</td>\n",
       "      <td>NM_015215.3</td>\n",
       "      <td>+</td>\n",
       "      <td>4</td>\n",
       "      <td>20</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>130</th>\n",
       "      <td>47</td>\n",
       "      <td>7309550</td>\n",
       "      <td>7309686</td>\n",
       "      <td>chr1</td>\n",
       "      <td>23261</td>\n",
       "      <td>CAMTA1</td>\n",
       "      <td>NM_015215.3</td>\n",
       "      <td>+</td>\n",
       "      <td>5</td>\n",
       "      <td>19</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     row    start      end chrom  GeneID   name2         name strand  +   -  \\\n",
       "126   47  6845383  6845635  chr1   23261  CAMTA1  NM_015215.3      +  1  23   \n",
       "127   47  6880240  6880310  chr1   23261  CAMTA1  NM_015215.3      +  2  22   \n",
       "128   47  6885151  6885270  chr1   23261  CAMTA1  NM_015215.3      +  3  21   \n",
       "129   47  7151363  7151431  chr1   23261  CAMTA1  NM_015215.3      +  4  20   \n",
       "130   47  7309550  7309686  chr1   23261  CAMTA1  NM_015215.3      +  5  19   \n",
       "\n",
       "     exon  \n",
       "126     1  \n",
       "127     2  \n",
       "128     3  \n",
       "129     4  \n",
       "130     5  "
      ]
     },
     "execution_count": 45,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "exon.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>row</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>chrom</th>\n",
       "      <th>GeneID</th>\n",
       "      <th>name2</th>\n",
       "      <th>name</th>\n",
       "      <th>strand</th>\n",
       "      <th>+</th>\n",
       "      <th>-</th>\n",
       "      <th>exon</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>3759</th>\n",
       "      <td>7339</td>\n",
       "      <td>545352</td>\n",
       "      <td>545561</td>\n",
       "      <td>chrY</td>\n",
       "      <td>6473</td>\n",
       "      <td>SHOX</td>\n",
       "      <td>NM_006883.2</td>\n",
       "      <td>+</td>\n",
       "      <td>3</td>\n",
       "      <td>4</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3760</th>\n",
       "      <td>7339</td>\n",
       "      <td>551555</td>\n",
       "      <td>551613</td>\n",
       "      <td>chrY</td>\n",
       "      <td>6473</td>\n",
       "      <td>SHOX</td>\n",
       "      <td>NM_006883.2</td>\n",
       "      <td>+</td>\n",
       "      <td>4</td>\n",
       "      <td>3</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3761</th>\n",
       "      <td>7339</td>\n",
       "      <td>551733</td>\n",
       "      <td>551822</td>\n",
       "      <td>chrY</td>\n",
       "      <td>6473</td>\n",
       "      <td>SHOX</td>\n",
       "      <td>NM_006883.2</td>\n",
       "      <td>+</td>\n",
       "      <td>5</td>\n",
       "      <td>2</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3762</th>\n",
       "      <td>7339</td>\n",
       "      <td>569519</td>\n",
       "      <td>570146</td>\n",
       "      <td>chrY</td>\n",
       "      <td>6473</td>\n",
       "      <td>SHOX</td>\n",
       "      <td>NM_006883.2</td>\n",
       "      <td>+</td>\n",
       "      <td>6</td>\n",
       "      <td>1</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4615</th>\n",
       "      <td>18161</td>\n",
       "      <td>2654895</td>\n",
       "      <td>2655782</td>\n",
       "      <td>chrY</td>\n",
       "      <td>6736</td>\n",
       "      <td>SRY</td>\n",
       "      <td>NM_003140.2</td>\n",
       "      <td>-</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        row    start      end chrom  GeneID name2         name strand  +  -  \\\n",
       "3759   7339   545352   545561  chrY    6473  SHOX  NM_006883.2      +  3  4   \n",
       "3760   7339   551555   551613  chrY    6473  SHOX  NM_006883.2      +  4  3   \n",
       "3761   7339   551733   551822  chrY    6473  SHOX  NM_006883.2      +  5  2   \n",
       "3762   7339   569519   570146  chrY    6473  SHOX  NM_006883.2      +  6  1   \n",
       "4615  18161  2654895  2655782  chrY    6736   SRY  NM_003140.2      -  1  1   \n",
       "\n",
       "      exon  \n",
       "3759     3  \n",
       "3760     4  \n",
       "3761     5  \n",
       "3762     6  \n",
       "4615     1  "
      ]
     },
     "execution_count": 46,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "exon.tail()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [],
   "source": [
    "exon['last_exon'] = exon.groupby(['name2', 'name'])['exon'].transform('max') == exon['exon']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>row</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>chrom</th>\n",
       "      <th>GeneID</th>\n",
       "      <th>name2</th>\n",
       "      <th>name</th>\n",
       "      <th>strand</th>\n",
       "      <th>+</th>\n",
       "      <th>-</th>\n",
       "      <th>exon</th>\n",
       "      <th>last_exon</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>126</th>\n",
       "      <td>47</td>\n",
       "      <td>6845383</td>\n",
       "      <td>6845635</td>\n",
       "      <td>chr1</td>\n",
       "      <td>23261</td>\n",
       "      <td>CAMTA1</td>\n",
       "      <td>NM_015215.3</td>\n",
       "      <td>+</td>\n",
       "      <td>1</td>\n",
       "      <td>23</td>\n",
       "      <td>1</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>127</th>\n",
       "      <td>47</td>\n",
       "      <td>6880240</td>\n",
       "      <td>6880310</td>\n",
       "      <td>chr1</td>\n",
       "      <td>23261</td>\n",
       "      <td>CAMTA1</td>\n",
       "      <td>NM_015215.3</td>\n",
       "      <td>+</td>\n",
       "      <td>2</td>\n",
       "      <td>22</td>\n",
       "      <td>2</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>128</th>\n",
       "      <td>47</td>\n",
       "      <td>6885151</td>\n",
       "      <td>6885270</td>\n",
       "      <td>chr1</td>\n",
       "      <td>23261</td>\n",
       "      <td>CAMTA1</td>\n",
       "      <td>NM_015215.3</td>\n",
       "      <td>+</td>\n",
       "      <td>3</td>\n",
       "      <td>21</td>\n",
       "      <td>3</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>129</th>\n",
       "      <td>47</td>\n",
       "      <td>7151363</td>\n",
       "      <td>7151431</td>\n",
       "      <td>chr1</td>\n",
       "      <td>23261</td>\n",
       "      <td>CAMTA1</td>\n",
       "      <td>NM_015215.3</td>\n",
       "      <td>+</td>\n",
       "      <td>4</td>\n",
       "      <td>20</td>\n",
       "      <td>4</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>130</th>\n",
       "      <td>47</td>\n",
       "      <td>7309550</td>\n",
       "      <td>7309686</td>\n",
       "      <td>chr1</td>\n",
       "      <td>23261</td>\n",
       "      <td>CAMTA1</td>\n",
       "      <td>NM_015215.3</td>\n",
       "      <td>+</td>\n",
       "      <td>5</td>\n",
       "      <td>19</td>\n",
       "      <td>5</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     row    start      end chrom  GeneID   name2         name strand  +   -  \\\n",
       "126   47  6845383  6845635  chr1   23261  CAMTA1  NM_015215.3      +  1  23   \n",
       "127   47  6880240  6880310  chr1   23261  CAMTA1  NM_015215.3      +  2  22   \n",
       "128   47  6885151  6885270  chr1   23261  CAMTA1  NM_015215.3      +  3  21   \n",
       "129   47  7151363  7151431  chr1   23261  CAMTA1  NM_015215.3      +  4  20   \n",
       "130   47  7309550  7309686  chr1   23261  CAMTA1  NM_015215.3      +  5  19   \n",
       "\n",
       "     exon  last_exon  \n",
       "126     1      False  \n",
       "127     2      False  \n",
       "128     3      False  \n",
       "129     4      False  \n",
       "130     5      False  "
      ]
     },
     "execution_count": 48,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "exon.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>row</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>chrom</th>\n",
       "      <th>GeneID</th>\n",
       "      <th>name2</th>\n",
       "      <th>name</th>\n",
       "      <th>strand</th>\n",
       "      <th>+</th>\n",
       "      <th>-</th>\n",
       "      <th>exon</th>\n",
       "      <th>last_exon</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>3759</th>\n",
       "      <td>7339</td>\n",
       "      <td>545352</td>\n",
       "      <td>545561</td>\n",
       "      <td>chrY</td>\n",
       "      <td>6473</td>\n",
       "      <td>SHOX</td>\n",
       "      <td>NM_006883.2</td>\n",
       "      <td>+</td>\n",
       "      <td>3</td>\n",
       "      <td>4</td>\n",
       "      <td>3</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3760</th>\n",
       "      <td>7339</td>\n",
       "      <td>551555</td>\n",
       "      <td>551613</td>\n",
       "      <td>chrY</td>\n",
       "      <td>6473</td>\n",
       "      <td>SHOX</td>\n",
       "      <td>NM_006883.2</td>\n",
       "      <td>+</td>\n",
       "      <td>4</td>\n",
       "      <td>3</td>\n",
       "      <td>4</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3761</th>\n",
       "      <td>7339</td>\n",
       "      <td>551733</td>\n",
       "      <td>551822</td>\n",
       "      <td>chrY</td>\n",
       "      <td>6473</td>\n",
       "      <td>SHOX</td>\n",
       "      <td>NM_006883.2</td>\n",
       "      <td>+</td>\n",
       "      <td>5</td>\n",
       "      <td>2</td>\n",
       "      <td>5</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3762</th>\n",
       "      <td>7339</td>\n",
       "      <td>569519</td>\n",
       "      <td>570146</td>\n",
       "      <td>chrY</td>\n",
       "      <td>6473</td>\n",
       "      <td>SHOX</td>\n",
       "      <td>NM_006883.2</td>\n",
       "      <td>+</td>\n",
       "      <td>6</td>\n",
       "      <td>1</td>\n",
       "      <td>6</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4615</th>\n",
       "      <td>18161</td>\n",
       "      <td>2654895</td>\n",
       "      <td>2655782</td>\n",
       "      <td>chrY</td>\n",
       "      <td>6736</td>\n",
       "      <td>SRY</td>\n",
       "      <td>NM_003140.2</td>\n",
       "      <td>-</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        row    start      end chrom  GeneID name2         name strand  +  -  \\\n",
       "3759   7339   545352   545561  chrY    6473  SHOX  NM_006883.2      +  3  4   \n",
       "3760   7339   551555   551613  chrY    6473  SHOX  NM_006883.2      +  4  3   \n",
       "3761   7339   551733   551822  chrY    6473  SHOX  NM_006883.2      +  5  2   \n",
       "3762   7339   569519   570146  chrY    6473  SHOX  NM_006883.2      +  6  1   \n",
       "4615  18161  2654895  2655782  chrY    6736   SRY  NM_003140.2      -  1  1   \n",
       "\n",
       "      exon  last_exon  \n",
       "3759     3      False  \n",
       "3760     4      False  \n",
       "3761     5      False  \n",
       "3762     6       True  \n",
       "4615     1       True  "
      ]
     },
     "execution_count": 49,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "exon.tail()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [],
   "source": [
    "exon = exon[\n",
    "    ['chrom', 'start', 'end', 'GeneID', 'name2', 'name', 'exon', 'last_exon']\n",
    "].sort_values(['chrom', 'start', 'end'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>GeneID</th>\n",
       "      <th>name2</th>\n",
       "      <th>name</th>\n",
       "      <th>exon</th>\n",
       "      <th>last_exon</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>126</th>\n",
       "      <td>chr1</td>\n",
       "      <td>6845383</td>\n",
       "      <td>6845635</td>\n",
       "      <td>23261</td>\n",
       "      <td>CAMTA1</td>\n",
       "      <td>NM_015215.3</td>\n",
       "      <td>1</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>127</th>\n",
       "      <td>chr1</td>\n",
       "      <td>6880240</td>\n",
       "      <td>6880310</td>\n",
       "      <td>23261</td>\n",
       "      <td>CAMTA1</td>\n",
       "      <td>NM_015215.3</td>\n",
       "      <td>2</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>128</th>\n",
       "      <td>chr1</td>\n",
       "      <td>6885151</td>\n",
       "      <td>6885270</td>\n",
       "      <td>23261</td>\n",
       "      <td>CAMTA1</td>\n",
       "      <td>NM_015215.3</td>\n",
       "      <td>3</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>129</th>\n",
       "      <td>chr1</td>\n",
       "      <td>7151363</td>\n",
       "      <td>7151431</td>\n",
       "      <td>23261</td>\n",
       "      <td>CAMTA1</td>\n",
       "      <td>NM_015215.3</td>\n",
       "      <td>4</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>130</th>\n",
       "      <td>chr1</td>\n",
       "      <td>7309550</td>\n",
       "      <td>7309686</td>\n",
       "      <td>23261</td>\n",
       "      <td>CAMTA1</td>\n",
       "      <td>NM_015215.3</td>\n",
       "      <td>5</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    chrom    start      end  GeneID   name2         name  exon  last_exon\n",
       "126  chr1  6845383  6845635   23261  CAMTA1  NM_015215.3     1      False\n",
       "127  chr1  6880240  6880310   23261  CAMTA1  NM_015215.3     2      False\n",
       "128  chr1  6885151  6885270   23261  CAMTA1  NM_015215.3     3      False\n",
       "129  chr1  7151363  7151431   23261  CAMTA1  NM_015215.3     4      False\n",
       "130  chr1  7309550  7309686   23261  CAMTA1  NM_015215.3     5      False"
      ]
     },
     "execution_count": 51,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "exon.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [],
   "source": [
    "exon.rename(columns={\n",
    "    'chrom': '#chrom',\n",
    "    'GeneID': 'gene_id', 'name2': 'symbol', 'name': 'transcript'\n",
    "}).to_csv('data/hi-exon-20200109.sorted.bed', index=False, sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [],
   "source": [
    "!bgzip -cf data/hi-exon-20200109.sorted.bed > ../data/hi-exon-20200109.sorted.bed.gz"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [],
   "source": [
    "!tabix -fp bed ../data/hi-exon-20200109.sorted.bed.gz"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {},
   "outputs": [],
   "source": [
    "last_exon = exon[exon['last_exon']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>GeneID</th>\n",
       "      <th>name2</th>\n",
       "      <th>name</th>\n",
       "      <th>exon</th>\n",
       "      <th>last_exon</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>148</th>\n",
       "      <td>chr1</td>\n",
       "      <td>7826518</td>\n",
       "      <td>7829766</td>\n",
       "      <td>23261</td>\n",
       "      <td>CAMTA1</td>\n",
       "      <td>NM_015215.3</td>\n",
       "      <td>23</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3443</th>\n",
       "      <td>chr1</td>\n",
       "      <td>17345216</td>\n",
       "      <td>17345453</td>\n",
       "      <td>6390</td>\n",
       "      <td>SDHB</td>\n",
       "      <td>NM_003000.2</td>\n",
       "      <td>8</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2226</th>\n",
       "      <td>chr1</td>\n",
       "      <td>27105513</td>\n",
       "      <td>27108601</td>\n",
       "      <td>8289</td>\n",
       "      <td>ARID1A</td>\n",
       "      <td>NM_006015.4</td>\n",
       "      <td>20</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4530</th>\n",
       "      <td>chr1</td>\n",
       "      <td>27860755</td>\n",
       "      <td>27861427</td>\n",
       "      <td>27245</td>\n",
       "      <td>AHDC1</td>\n",
       "      <td>NM_001029882.3</td>\n",
       "      <td>7</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3646</th>\n",
       "      <td>chr1</td>\n",
       "      <td>43391045</td>\n",
       "      <td>43392912</td>\n",
       "      <td>6513</td>\n",
       "      <td>SLC2A1</td>\n",
       "      <td>NM_006516.2</td>\n",
       "      <td>10</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     chrom     start       end  GeneID   name2            name  exon  \\\n",
       "148   chr1   7826518   7829766   23261  CAMTA1     NM_015215.3    23   \n",
       "3443  chr1  17345216  17345453    6390    SDHB     NM_003000.2     8   \n",
       "2226  chr1  27105513  27108601    8289  ARID1A     NM_006015.4    20   \n",
       "4530  chr1  27860755  27861427   27245   AHDC1  NM_001029882.3     7   \n",
       "3646  chr1  43391045  43392912    6513  SLC2A1     NM_006516.2    10   \n",
       "\n",
       "      last_exon  \n",
       "148        True  \n",
       "3443       True  \n",
       "2226       True  \n",
       "4530       True  \n",
       "3646       True  "
      ]
     },
     "execution_count": 56,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "last_exon.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {},
   "outputs": [],
   "source": [
    "last_exon_region = last_exon['chrom'] + ':' + last_exon['start'].astype(str) + '-' + last_exon['end'].astype(str)\n",
    "last_exon_region = last_exon_region.str.replace('chr', '')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "148       1:7826518-7829766\n",
       "3443    1:17345216-17345453\n",
       "2226    1:27105513-27108601\n",
       "4530    1:27860755-27861427\n",
       "3646    1:43391045-43392912\n",
       "dtype: object"
      ]
     },
     "execution_count": 58,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "last_exon_region.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {},
   "outputs": [],
   "source": [
    "need_fields = [\n",
    "    'variants/CHROM', 'variants/POS', 'variants/REF', 'variants/ALT',\n",
    "    'variants/AF_ESP', 'variants/AF_EXAC', 'variants/AF_TGP', 'variants/CLNSIG'\n",
    "]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open('data/clinvar_pathogenic_20191125.sorted.vcf', 'w') as f:\n",
    "    headers = allel.read_vcf_headers('data/clinvar_20191125.vcf.gz')\n",
    "    f.write(''.join(headers.headers))\n",
    "    \n",
    "    def fetch_variants(region):\n",
    "        fields, samples, headers, it = allel.iter_vcf_chunks(\n",
    "            'data/clinvar_20191125.vcf.gz', fields=need_fields, alt_number=1, region=region\n",
    "        )\n",
    "        for variants, *_ in it:\n",
    "            esp_filter = np.isnan(variants['variants/AF_ESP'])\n",
    "            esp_filter[~esp_filter] |= variants['variants/AF_ESP'][~esp_filter] < 0.01\n",
    "\n",
    "            exac_filter = np.isnan(variants['variants/AF_EXAC'])\n",
    "            exac_filter[~exac_filter] |= variants['variants/AF_EXAC'][~exac_filter] < 0.01\n",
    "\n",
    "            tgp_filter = np.isnan(variants['variants/AF_TGP'])\n",
    "            tgp_filter[~tgp_filter] |= variants['variants/AF_TGP'][~tgp_filter] < 0.01\n",
    "\n",
    "            pathogenic_filter = np.isin(\n",
    "                variants['variants/CLNSIG'], ['Likely_pathogenic', 'Pathogenic', 'Pathogenic/Likely_pathogenic']\n",
    "            )\n",
    "\n",
    "            af_filter = esp_filter & exac_filter & tgp_filter & pathogenic_filter\n",
    "\n",
    "            filtered_variants = {k: v[af_filter] for k, v in variants.items()}\n",
    "\n",
    "            filtered_variants['variants/CHROM'] = 'chr' + filtered_variants['variants/CHROM']\n",
    "\n",
    "            return allel.normalize_callset(filtered_variants)\n",
    "    \n",
    "    with Pool(processes=7) as pool:\n",
    "        variants = pool.map(fetch_variants, last_exon_region)\n",
    "    \n",
    "    for names, callset in filter(lambda x: x is not None, variants):\n",
    "        allel.write_vcf_data(f, names, callset, None, {field: np.nan for field in need_fields})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {},
   "outputs": [],
   "source": [
    "!sed -i 's/AF_.\\+=nan;//' data/clinvar_pathogenic_20191125.sorted.vcf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "metadata": {},
   "outputs": [],
   "source": [
    "!bgzip -cf data/clinvar_pathogenic_20191125.sorted.vcf > ../data/clinvar-pathogenic-20191125.sorted.vcf.gz"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "metadata": {},
   "outputs": [],
   "source": [
    "!tabix -fp vcf ../data/clinvar-pathogenic-20191125.sorted.vcf.gz"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "metadata": {},
   "outputs": [],
   "source": [
    "uhi_genes = set(\n",
    "    curation_gene.loc[curation_gene['Haploinsufficiency Score'] == '40', '#Gene Symbol']\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "metadata": {},
   "outputs": [],
   "source": [
    "uhi_gene = refgene_info.loc[\n",
    "    refgene_info['name2'].isin(uhi_genes),\n",
    "    ['chrom', 'txStart', 'txEnd', 'GeneID', 'name2', 'name', 'strand']\n",
    "].sort_values(['chrom', 'txStart', 'txEnd'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>txStart</th>\n",
       "      <th>txEnd</th>\n",
       "      <th>GeneID</th>\n",
       "      <th>name2</th>\n",
       "      <th>name</th>\n",
       "      <th>strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>8064</th>\n",
       "      <td>chr1</td>\n",
       "      <td>55505148</td>\n",
       "      <td>55530526</td>\n",
       "      <td>255738</td>\n",
       "      <td>PCSK9</td>\n",
       "      <td>NM_174936.3</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18085</th>\n",
       "      <td>chr13</td>\n",
       "      <td>20761601</td>\n",
       "      <td>20767114</td>\n",
       "      <td>2706</td>\n",
       "      <td>GJB2</td>\n",
       "      <td>NM_004004.5</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       chrom   txStart     txEnd  GeneID  name2         name strand\n",
       "8064    chr1  55505148  55530526  255738  PCSK9  NM_174936.3      +\n",
       "18085  chr13  20761601  20767114    2706   GJB2  NM_004004.5      -"
      ]
     },
     "execution_count": 66,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "uhi_gene.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "metadata": {},
   "outputs": [],
   "source": [
    "uhi_gene.rename(columns={\n",
    "    'chrom': '#chrom', 'txStart': 'start', 'txEnd': 'end',\n",
    "    'GeneID': 'gene_id', 'name2': 'symbol', 'name': 'transcript'\n",
    "}).to_csv('data/uhi-gene-20200109.sorted.bed', index=False, sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "metadata": {},
   "outputs": [],
   "source": [
    "!bgzip -cf data/uhi-gene-20200109.sorted.bed > ../data/uhi-gene-20200109.sorted.bed.gz"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "metadata": {},
   "outputs": [],
   "source": [
    "!tabix -fp bed ../data/uhi-gene-20200109.sorted.bed.gz"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "metadata": {},
   "outputs": [],
   "source": [
    "region = pd.read_csv('data/ClinGen_region_curation_list_GRCh37.tsv', sep='\\t', skiprows=5, dtype=str)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>#ISCA ID</th>\n",
       "      <th>ISCA Region Name</th>\n",
       "      <th>cytoBand</th>\n",
       "      <th>Genomic Location</th>\n",
       "      <th>Haploinsufficiency Score</th>\n",
       "      <th>Haploinsufficiency Description</th>\n",
       "      <th>Haploinsufficiency PMID1</th>\n",
       "      <th>Haploinsufficiency PMID2</th>\n",
       "      <th>Haploinsufficiency PMID3</th>\n",
       "      <th>Triplosensitivity Score</th>\n",
       "      <th>Triplosensitivity Description</th>\n",
       "      <th>Triplosensitivity PMID1</th>\n",
       "      <th>Triplosensitivity PMID2</th>\n",
       "      <th>Triplosensitivity PMID3</th>\n",
       "      <th>Date Last Evaluated</th>\n",
       "      <th>Loss phenotype OMIM ID</th>\n",
       "      <th>Triplosensitive phenotype OMIM ID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ISCA-46300</td>\n",
       "      <td>15q24 recurrent region (C-D) (includes SIN3A)</td>\n",
       "      <td>NaN</td>\n",
       "      <td>chr15: 75631787-75972909</td>\n",
       "      <td>3</td>\n",
       "      <td>Sufficient evidence for dosage pathogenicity</td>\n",
       "      <td>27399968</td>\n",
       "      <td>22180641</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>No evidence available</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2020-01-13</td>\n",
       "      <td>613406</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ISCA-46299</td>\n",
       "      <td>Xp11.22 region (includes HUWE1)</td>\n",
       "      <td>Xp11.22</td>\n",
       "      <td>chrX:53363456-53793054</td>\n",
       "      <td>0</td>\n",
       "      <td>No evidence available</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3</td>\n",
       "      <td>Sufficient evidence for dosage pathogenicity</td>\n",
       "      <td>22840365</td>\n",
       "      <td>20655035</td>\n",
       "      <td>26692240</td>\n",
       "      <td>2018-11-19</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ISCA-46298</td>\n",
       "      <td>13q12.12 recurrent region (includes SACS and S...</td>\n",
       "      <td>13q12.12</td>\n",
       "      <td>chr13:23544582-24881361</td>\n",
       "      <td>30</td>\n",
       "      <td>Gene associated with autosomal recessive pheno...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>No evidence available</td>\n",
       "      <td>23239789</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2019-11-03</td>\n",
       "      <td>270550,253700</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ISCA-46297</td>\n",
       "      <td>16p12.2 recurrent region (includes OTOA) (dist...</td>\n",
       "      <td>16p12.2</td>\n",
       "      <td>chr16:21570113-21740423</td>\n",
       "      <td>30</td>\n",
       "      <td>Gene associated with autosomal recessive pheno...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>40</td>\n",
       "      <td>Dosage sensitivity unlikely</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2019-11-03</td>\n",
       "      <td>607039</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ISCA-46296</td>\n",
       "      <td>15q24 recurrent region (A-C)</td>\n",
       "      <td>15q24.1-q24.2</td>\n",
       "      <td>chr15:72963715-75508312</td>\n",
       "      <td>3</td>\n",
       "      <td>Sufficient evidence for dosage pathogenicity</td>\n",
       "      <td>22180641</td>\n",
       "      <td>19921647</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>Little evidence for dosage pathogenicity</td>\n",
       "      <td>19557438</td>\n",
       "      <td>20860070</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2020-01-03</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     #ISCA ID                                   ISCA Region Name  \\\n",
       "0  ISCA-46300      15q24 recurrent region (C-D) (includes SIN3A)   \n",
       "1  ISCA-46299                    Xp11.22 region (includes HUWE1)   \n",
       "2  ISCA-46298  13q12.12 recurrent region (includes SACS and S...   \n",
       "3  ISCA-46297  16p12.2 recurrent region (includes OTOA) (dist...   \n",
       "4  ISCA-46296                       15q24 recurrent region (A-C)   \n",
       "\n",
       "        cytoBand          Genomic Location Haploinsufficiency Score  \\\n",
       "0            NaN  chr15: 75631787-75972909                        3   \n",
       "1        Xp11.22    chrX:53363456-53793054                        0   \n",
       "2       13q12.12   chr13:23544582-24881361                       30   \n",
       "3        16p12.2   chr16:21570113-21740423                       30   \n",
       "4  15q24.1-q24.2   chr15:72963715-75508312                        3   \n",
       "\n",
       "                      Haploinsufficiency Description Haploinsufficiency PMID1  \\\n",
       "0       Sufficient evidence for dosage pathogenicity                 27399968   \n",
       "1                              No evidence available                      NaN   \n",
       "2  Gene associated with autosomal recessive pheno...                      NaN   \n",
       "3  Gene associated with autosomal recessive pheno...                      NaN   \n",
       "4       Sufficient evidence for dosage pathogenicity                 22180641   \n",
       "\n",
       "  Haploinsufficiency PMID2 Haploinsufficiency PMID3 Triplosensitivity Score  \\\n",
       "0                 22180641                      NaN                       0   \n",
       "1                      NaN                      NaN                       3   \n",
       "2                      NaN                      NaN                       0   \n",
       "3                      NaN                      NaN                      40   \n",
       "4                 19921647                      NaN                       1   \n",
       "\n",
       "                  Triplosensitivity Description Triplosensitivity PMID1  \\\n",
       "0                         No evidence available                     NaN   \n",
       "1  Sufficient evidence for dosage pathogenicity                22840365   \n",
       "2                         No evidence available                23239789   \n",
       "3                   Dosage sensitivity unlikely                     NaN   \n",
       "4      Little evidence for dosage pathogenicity                19557438   \n",
       "\n",
       "  Triplosensitivity PMID2 Triplosensitivity PMID3 Date Last Evaluated  \\\n",
       "0                     NaN                     NaN          2020-01-13   \n",
       "1                20655035                26692240          2018-11-19   \n",
       "2                     NaN                     NaN          2019-11-03   \n",
       "3                     NaN                     NaN          2019-11-03   \n",
       "4                20860070                     NaN          2020-01-03   \n",
       "\n",
       "  Loss phenotype OMIM ID Triplosensitive phenotype OMIM ID  \n",
       "0                 613406                               NaN  \n",
       "1                    NaN                               NaN  \n",
       "2          270550,253700                               NaN  \n",
       "3                 607039                               NaN  \n",
       "4                    NaN                               NaN  "
      ]
     },
     "execution_count": 71,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "region.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "metadata": {},
   "outputs": [],
   "source": [
    "position = region['Genomic Location'].str.extract(r'(?P<chrom>chr\\w+)\\s*:\\s*(?P<start>\\d+)\\s*-\\s*(?P<end>\\d+)')\n",
    "position['start'] = position['start'].astype(int)\n",
    "position['end'] = position['end'].astype(int)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr15</td>\n",
       "      <td>75631787</td>\n",
       "      <td>75972909</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chrX</td>\n",
       "      <td>53363456</td>\n",
       "      <td>53793054</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr13</td>\n",
       "      <td>23544582</td>\n",
       "      <td>24881361</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr16</td>\n",
       "      <td>21570113</td>\n",
       "      <td>21740423</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr15</td>\n",
       "      <td>72963715</td>\n",
       "      <td>75508312</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   chrom     start       end\n",
       "0  chr15  75631787  75972909\n",
       "1   chrX  53363456  53793054\n",
       "2  chr13  23544582  24881361\n",
       "3  chr16  21570113  21740423\n",
       "4  chr15  72963715  75508312"
      ]
     },
     "execution_count": 73,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "position.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "metadata": {},
   "outputs": [],
   "source": [
    "region_pos = region.merge(position, how='left', left_index=True, right_index=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>#ISCA ID</th>\n",
       "      <th>ISCA Region Name</th>\n",
       "      <th>cytoBand</th>\n",
       "      <th>Genomic Location</th>\n",
       "      <th>Haploinsufficiency Score</th>\n",
       "      <th>Haploinsufficiency Description</th>\n",
       "      <th>Haploinsufficiency PMID1</th>\n",
       "      <th>Haploinsufficiency PMID2</th>\n",
       "      <th>Haploinsufficiency PMID3</th>\n",
       "      <th>Triplosensitivity Score</th>\n",
       "      <th>Triplosensitivity Description</th>\n",
       "      <th>Triplosensitivity PMID1</th>\n",
       "      <th>Triplosensitivity PMID2</th>\n",
       "      <th>Triplosensitivity PMID3</th>\n",
       "      <th>Date Last Evaluated</th>\n",
       "      <th>Loss phenotype OMIM ID</th>\n",
       "      <th>Triplosensitive phenotype OMIM ID</th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ISCA-46300</td>\n",
       "      <td>15q24 recurrent region (C-D) (includes SIN3A)</td>\n",
       "      <td>NaN</td>\n",
       "      <td>chr15: 75631787-75972909</td>\n",
       "      <td>3</td>\n",
       "      <td>Sufficient evidence for dosage pathogenicity</td>\n",
       "      <td>27399968</td>\n",
       "      <td>22180641</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>No evidence available</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2020-01-13</td>\n",
       "      <td>613406</td>\n",
       "      <td>NaN</td>\n",
       "      <td>chr15</td>\n",
       "      <td>75631787</td>\n",
       "      <td>75972909</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ISCA-46299</td>\n",
       "      <td>Xp11.22 region (includes HUWE1)</td>\n",
       "      <td>Xp11.22</td>\n",
       "      <td>chrX:53363456-53793054</td>\n",
       "      <td>0</td>\n",
       "      <td>No evidence available</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3</td>\n",
       "      <td>Sufficient evidence for dosage pathogenicity</td>\n",
       "      <td>22840365</td>\n",
       "      <td>20655035</td>\n",
       "      <td>26692240</td>\n",
       "      <td>2018-11-19</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>chrX</td>\n",
       "      <td>53363456</td>\n",
       "      <td>53793054</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ISCA-46298</td>\n",
       "      <td>13q12.12 recurrent region (includes SACS and S...</td>\n",
       "      <td>13q12.12</td>\n",
       "      <td>chr13:23544582-24881361</td>\n",
       "      <td>30</td>\n",
       "      <td>Gene associated with autosomal recessive pheno...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>No evidence available</td>\n",
       "      <td>23239789</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2019-11-03</td>\n",
       "      <td>270550,253700</td>\n",
       "      <td>NaN</td>\n",
       "      <td>chr13</td>\n",
       "      <td>23544582</td>\n",
       "      <td>24881361</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ISCA-46297</td>\n",
       "      <td>16p12.2 recurrent region (includes OTOA) (dist...</td>\n",
       "      <td>16p12.2</td>\n",
       "      <td>chr16:21570113-21740423</td>\n",
       "      <td>30</td>\n",
       "      <td>Gene associated with autosomal recessive pheno...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>40</td>\n",
       "      <td>Dosage sensitivity unlikely</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2019-11-03</td>\n",
       "      <td>607039</td>\n",
       "      <td>NaN</td>\n",
       "      <td>chr16</td>\n",
       "      <td>21570113</td>\n",
       "      <td>21740423</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ISCA-46296</td>\n",
       "      <td>15q24 recurrent region (A-C)</td>\n",
       "      <td>15q24.1-q24.2</td>\n",
       "      <td>chr15:72963715-75508312</td>\n",
       "      <td>3</td>\n",
       "      <td>Sufficient evidence for dosage pathogenicity</td>\n",
       "      <td>22180641</td>\n",
       "      <td>19921647</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>Little evidence for dosage pathogenicity</td>\n",
       "      <td>19557438</td>\n",
       "      <td>20860070</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2020-01-03</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>chr15</td>\n",
       "      <td>72963715</td>\n",
       "      <td>75508312</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     #ISCA ID                                   ISCA Region Name  \\\n",
       "0  ISCA-46300      15q24 recurrent region (C-D) (includes SIN3A)   \n",
       "1  ISCA-46299                    Xp11.22 region (includes HUWE1)   \n",
       "2  ISCA-46298  13q12.12 recurrent region (includes SACS and S...   \n",
       "3  ISCA-46297  16p12.2 recurrent region (includes OTOA) (dist...   \n",
       "4  ISCA-46296                       15q24 recurrent region (A-C)   \n",
       "\n",
       "        cytoBand          Genomic Location Haploinsufficiency Score  \\\n",
       "0            NaN  chr15: 75631787-75972909                        3   \n",
       "1        Xp11.22    chrX:53363456-53793054                        0   \n",
       "2       13q12.12   chr13:23544582-24881361                       30   \n",
       "3        16p12.2   chr16:21570113-21740423                       30   \n",
       "4  15q24.1-q24.2   chr15:72963715-75508312                        3   \n",
       "\n",
       "                      Haploinsufficiency Description Haploinsufficiency PMID1  \\\n",
       "0       Sufficient evidence for dosage pathogenicity                 27399968   \n",
       "1                              No evidence available                      NaN   \n",
       "2  Gene associated with autosomal recessive pheno...                      NaN   \n",
       "3  Gene associated with autosomal recessive pheno...                      NaN   \n",
       "4       Sufficient evidence for dosage pathogenicity                 22180641   \n",
       "\n",
       "  Haploinsufficiency PMID2 Haploinsufficiency PMID3 Triplosensitivity Score  \\\n",
       "0                 22180641                      NaN                       0   \n",
       "1                      NaN                      NaN                       3   \n",
       "2                      NaN                      NaN                       0   \n",
       "3                      NaN                      NaN                      40   \n",
       "4                 19921647                      NaN                       1   \n",
       "\n",
       "                  Triplosensitivity Description Triplosensitivity PMID1  \\\n",
       "0                         No evidence available                     NaN   \n",
       "1  Sufficient evidence for dosage pathogenicity                22840365   \n",
       "2                         No evidence available                23239789   \n",
       "3                   Dosage sensitivity unlikely                     NaN   \n",
       "4      Little evidence for dosage pathogenicity                19557438   \n",
       "\n",
       "  Triplosensitivity PMID2 Triplosensitivity PMID3 Date Last Evaluated  \\\n",
       "0                     NaN                     NaN          2020-01-13   \n",
       "1                20655035                26692240          2018-11-19   \n",
       "2                     NaN                     NaN          2019-11-03   \n",
       "3                     NaN                     NaN          2019-11-03   \n",
       "4                20860070                     NaN          2020-01-03   \n",
       "\n",
       "  Loss phenotype OMIM ID Triplosensitive phenotype OMIM ID  chrom     start  \\\n",
       "0                 613406                               NaN  chr15  75631787   \n",
       "1                    NaN                               NaN   chrX  53363456   \n",
       "2          270550,253700                               NaN  chr13  23544582   \n",
       "3                 607039                               NaN  chr16  21570113   \n",
       "4                    NaN                               NaN  chr15  72963715   \n",
       "\n",
       "        end  \n",
       "0  75972909  \n",
       "1  53793054  \n",
       "2  24881361  \n",
       "3  21740423  \n",
       "4  75508312  "
      ]
     },
     "execution_count": 75,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "region_pos.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "metadata": {},
   "outputs": [],
   "source": [
    "hi_region = region_pos.loc[\n",
    "    region_pos['Haploinsufficiency Score'] == '3',\n",
    "    ['chrom', 'start', 'end', '#ISCA ID', 'ISCA Region Name']\n",
    "].sort_values(['chrom', 'start', 'end'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>#ISCA ID</th>\n",
       "      <th>ISCA Region Name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>43</th>\n",
       "      <td>chr1</td>\n",
       "      <td>834083</td>\n",
       "      <td>6289973</td>\n",
       "      <td>ISCA-37434</td>\n",
       "      <td>1p36 terminal region (includes GABRD)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>52</th>\n",
       "      <td>chr1</td>\n",
       "      <td>146577486</td>\n",
       "      <td>147394506</td>\n",
       "      <td>ISCA-37421</td>\n",
       "      <td>1q21.1 recurrent region (BP3-BP4, distal) (inc...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>chr1</td>\n",
       "      <td>243287730</td>\n",
       "      <td>245318287</td>\n",
       "      <td>ISCA-37493</td>\n",
       "      <td>1q43q44 terminal region (includes AKT3)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>51</th>\n",
       "      <td>chr10</td>\n",
       "      <td>81682843</td>\n",
       "      <td>88739388</td>\n",
       "      <td>ISCA-37424</td>\n",
       "      <td>10q22.3q23.2 recurrent region (LCR-3/4-flanked...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>63</th>\n",
       "      <td>chr11</td>\n",
       "      <td>31803509</td>\n",
       "      <td>32510988</td>\n",
       "      <td>ISCA-37401</td>\n",
       "      <td>11p13 (WAGR syndrome) region</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    chrom      start        end    #ISCA ID  \\\n",
       "43   chr1     834083    6289973  ISCA-37434   \n",
       "52   chr1  146577486  147394506  ISCA-37421   \n",
       "17   chr1  243287730  245318287  ISCA-37493   \n",
       "51  chr10   81682843   88739388  ISCA-37424   \n",
       "63  chr11   31803509   32510988  ISCA-37401   \n",
       "\n",
       "                                     ISCA Region Name  \n",
       "43              1p36 terminal region (includes GABRD)  \n",
       "52  1q21.1 recurrent region (BP3-BP4, distal) (inc...  \n",
       "17            1q43q44 terminal region (includes AKT3)  \n",
       "51  10q22.3q23.2 recurrent region (LCR-3/4-flanked...  \n",
       "63                       11p13 (WAGR syndrome) region  "
      ]
     },
     "execution_count": 77,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "hi_region.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "metadata": {},
   "outputs": [],
   "source": [
    "hi_region['omim_genes'] = hi_region.apply(\n",
    "    lambda row: ','.join(omim_gene.loc[\n",
    "        (omim_gene['chrom'] == row['chrom']) & (omim_gene['txStart'] >= row['start']) & (omim_gene['txEnd'] <= row['end']), 'name2'\n",
    "    ]),\n",
    "    axis=1\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>#ISCA ID</th>\n",
       "      <th>ISCA Region Name</th>\n",
       "      <th>omim_genes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>43</th>\n",
       "      <td>chr1</td>\n",
       "      <td>834083</td>\n",
       "      <td>6289973</td>\n",
       "      <td>ISCA-37434</td>\n",
       "      <td>1p36 terminal region (includes GABRD)</td>\n",
       "      <td>ISG15,AGRN,TNFRSF4,B3GALT6,DVL1,ATAD3A,TMEM240...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>52</th>\n",
       "      <td>chr1</td>\n",
       "      <td>146577486</td>\n",
       "      <td>147394506</td>\n",
       "      <td>ISCA-37421</td>\n",
       "      <td>1q21.1 recurrent region (BP3-BP4, distal) (inc...</td>\n",
       "      <td>GJA5,GJA8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>chr1</td>\n",
       "      <td>243287730</td>\n",
       "      <td>245318287</td>\n",
       "      <td>ISCA-37493</td>\n",
       "      <td>1q43q44 terminal region (includes AKT3)</td>\n",
       "      <td>SDCCAG8,AKT3,ZBTB18,COX20,HNRNPU</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>51</th>\n",
       "      <td>chr10</td>\n",
       "      <td>81682843</td>\n",
       "      <td>88739388</td>\n",
       "      <td>ISCA-37424</td>\n",
       "      <td>10q22.3q23.2 recurrent region (LCR-3/4-flanked...</td>\n",
       "      <td>ANXA11,MAT1A,CDHR1,RGR,LDB3,BMPR1A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>63</th>\n",
       "      <td>chr11</td>\n",
       "      <td>31803509</td>\n",
       "      <td>32510988</td>\n",
       "      <td>ISCA-37401</td>\n",
       "      <td>11p13 (WAGR syndrome) region</td>\n",
       "      <td>PAX6,WT1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    chrom      start        end    #ISCA ID  \\\n",
       "43   chr1     834083    6289973  ISCA-37434   \n",
       "52   chr1  146577486  147394506  ISCA-37421   \n",
       "17   chr1  243287730  245318287  ISCA-37493   \n",
       "51  chr10   81682843   88739388  ISCA-37424   \n",
       "63  chr11   31803509   32510988  ISCA-37401   \n",
       "\n",
       "                                     ISCA Region Name  \\\n",
       "43              1p36 terminal region (includes GABRD)   \n",
       "52  1q21.1 recurrent region (BP3-BP4, distal) (inc...   \n",
       "17            1q43q44 terminal region (includes AKT3)   \n",
       "51  10q22.3q23.2 recurrent region (LCR-3/4-flanked...   \n",
       "63                       11p13 (WAGR syndrome) region   \n",
       "\n",
       "                                           omim_genes  \n",
       "43  ISG15,AGRN,TNFRSF4,B3GALT6,DVL1,ATAD3A,TMEM240...  \n",
       "52                                          GJA5,GJA8  \n",
       "17                   SDCCAG8,AKT3,ZBTB18,COX20,HNRNPU  \n",
       "51                 ANXA11,MAT1A,CDHR1,RGR,LDB3,BMPR1A  \n",
       "63                                           PAX6,WT1  "
      ]
     },
     "execution_count": 79,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "hi_region.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "metadata": {},
   "outputs": [],
   "source": [
    "hi_region.rename(columns={\n",
    "    'chrom': '#chrom',\n",
    "    '#ISCA ID': 'isca_id', 'ISCA Region Name': 'name'\n",
    "}).to_csv('data/hi-region-20200109.sorted.bed', sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "metadata": {},
   "outputs": [],
   "source": [
    "!bgzip -cf data/hi-region-20200109.sorted.bed > ../data/hi-region-20200109.sorted.bed.gz"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "metadata": {},
   "outputs": [],
   "source": [
    "!tabix -fp bed ../data/hi-region-20200109.sorted.bed.gz"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "metadata": {},
   "outputs": [],
   "source": [
    "uhi_region = region_pos.loc[\n",
    "    region_pos['Haploinsufficiency Score'] == '40',\n",
    "    ['chrom', 'start', 'end', '#ISCA ID', 'ISCA Region Name']\n",
    "].sort_values(['chrom', 'start', 'end'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>#ISCA ID</th>\n",
       "      <th>ISCA Region Name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>chr1</td>\n",
       "      <td>148867551</td>\n",
       "      <td>149768855</td>\n",
       "      <td>ISCA-37469</td>\n",
       "      <td>1q21.2 region (polymorphic region)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>chr14</td>\n",
       "      <td>22111109</td>\n",
       "      <td>23021097</td>\n",
       "      <td>ISCA-37476</td>\n",
       "      <td>14q11.2 region (TCRA region)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>chr14</td>\n",
       "      <td>106050000</td>\n",
       "      <td>107289540</td>\n",
       "      <td>ISCA-37477</td>\n",
       "      <td>1 copy: 14q telomere; 3 copies: 14q telomere</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>chr16</td>\n",
       "      <td>34202088</td>\n",
       "      <td>35147508</td>\n",
       "      <td>ISCA-37481</td>\n",
       "      <td>3 copies: 16p centromere</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>chr19</td>\n",
       "      <td>43242796</td>\n",
       "      <td>43741310</td>\n",
       "      <td>ISCA-37483</td>\n",
       "      <td>19q13.3 region (PSG gene cluster)</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    chrom      start        end    #ISCA ID  \\\n",
       "29   chr1  148867551  149768855  ISCA-37469   \n",
       "25  chr14   22111109   23021097  ISCA-37476   \n",
       "24  chr14  106050000  107289540  ISCA-37477   \n",
       "21  chr16   34202088   35147508  ISCA-37481   \n",
       "20  chr19   43242796   43741310  ISCA-37483   \n",
       "\n",
       "                                ISCA Region Name  \n",
       "29            1q21.2 region (polymorphic region)  \n",
       "25                  14q11.2 region (TCRA region)  \n",
       "24  1 copy: 14q telomere; 3 copies: 14q telomere  \n",
       "21                      3 copies: 16p centromere  \n",
       "20             19q13.3 region (PSG gene cluster)  "
      ]
     },
     "execution_count": 84,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "uhi_region.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "metadata": {},
   "outputs": [],
   "source": [
    "uhi_region['genes'] = uhi_region.apply(\n",
    "    lambda row: ','.join(gene.loc[\n",
    "        (gene['chrom'] == row['chrom']) & (gene['txStart'] >= row['start']) & (gene['txEnd'] <= row['end']), 'name2'\n",
    "    ]),\n",
    "    axis=1\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>#ISCA ID</th>\n",
       "      <th>ISCA Region Name</th>\n",
       "      <th>genes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>chr1</td>\n",
       "      <td>148867551</td>\n",
       "      <td>149768855</td>\n",
       "      <td>ISCA-37469</td>\n",
       "      <td>1q21.2 region (polymorphic region)</td>\n",
       "      <td>FAM72C,PPIAL4C,FCGR1A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>chr14</td>\n",
       "      <td>22111109</td>\n",
       "      <td>23021097</td>\n",
       "      <td>ISCA-37476</td>\n",
       "      <td>14q11.2 region (TCRA region)</td>\n",
       "      <td>OR4E2,OR4E1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>chr14</td>\n",
       "      <td>106050000</td>\n",
       "      <td>107289540</td>\n",
       "      <td>ISCA-37477</td>\n",
       "      <td>1 copy: 14q telomere; 3 copies: 14q telomere</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>chr16</td>\n",
       "      <td>34202088</td>\n",
       "      <td>35147508</td>\n",
       "      <td>ISCA-37481</td>\n",
       "      <td>3 copies: 16p centromere</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>chr19</td>\n",
       "      <td>43242796</td>\n",
       "      <td>43741310</td>\n",
       "      <td>ISCA-37483</td>\n",
       "      <td>19q13.3 region (PSG gene cluster)</td>\n",
       "      <td>PSG8,PSG1,PSG6,PSG7,PSG11,PSG2,PSG5,PSG4</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    chrom      start        end    #ISCA ID  \\\n",
       "29   chr1  148867551  149768855  ISCA-37469   \n",
       "25  chr14   22111109   23021097  ISCA-37476   \n",
       "24  chr14  106050000  107289540  ISCA-37477   \n",
       "21  chr16   34202088   35147508  ISCA-37481   \n",
       "20  chr19   43242796   43741310  ISCA-37483   \n",
       "\n",
       "                                ISCA Region Name  \\\n",
       "29            1q21.2 region (polymorphic region)   \n",
       "25                  14q11.2 region (TCRA region)   \n",
       "24  1 copy: 14q telomere; 3 copies: 14q telomere   \n",
       "21                      3 copies: 16p centromere   \n",
       "20             19q13.3 region (PSG gene cluster)   \n",
       "\n",
       "                                       genes  \n",
       "29                     FAM72C,PPIAL4C,FCGR1A  \n",
       "25                               OR4E2,OR4E1  \n",
       "24                                            \n",
       "21                                            \n",
       "20  PSG8,PSG1,PSG6,PSG7,PSG11,PSG2,PSG5,PSG4  "
      ]
     },
     "execution_count": 86,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "uhi_region.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 87,
   "metadata": {},
   "outputs": [],
   "source": [
    "uhi_region.rename(columns={\n",
    "    'chrom': '#chrom',\n",
    "    '#ISCA ID': 'isca_id', 'ISCA Region Name': 'name'\n",
    "}).to_csv('data/uhi-region-20200109.sorted.bed', sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 88,
   "metadata": {},
   "outputs": [],
   "source": [
    "!bgzip -cf data/uhi-region-20200109.sorted.bed > ../data/uhi-region-20200109.sorted.bed.gz"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 89,
   "metadata": {},
   "outputs": [],
   "source": [
    "!tabix -fp bed ../data/uhi-region-20200109.sorted.bed.gz"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 90,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open('data/gene-table.json') as f:\n",
    "    data = json.load(f)\n",
    "    decipher = pd.DataFrame.from_records(data, columns=['name', 'hi_score', 'p_li']).rename(columns={'name': 'symbol'})\n",
    "    decipher = decipher.merge(gene, left_on='symbol', right_on='name2')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 91,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>symbol</th>\n",
       "      <th>hi_score</th>\n",
       "      <th>p_li</th>\n",
       "      <th>chrom</th>\n",
       "      <th>txStart</th>\n",
       "      <th>txEnd</th>\n",
       "      <th>GeneID</th>\n",
       "      <th>name2</th>\n",
       "      <th>name</th>\n",
       "      <th>strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A2ML1</td>\n",
       "      <td>0.011664</td>\n",
       "      <td>1.329022e-22</td>\n",
       "      <td>chr12</td>\n",
       "      <td>8975067</td>\n",
       "      <td>9029379</td>\n",
       "      <td>144568</td>\n",
       "      <td>A2ML1</td>\n",
       "      <td>NM_144670.5</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AARS2</td>\n",
       "      <td>0.089486</td>\n",
       "      <td>1.648580e-07</td>\n",
       "      <td>chr6</td>\n",
       "      <td>44266462</td>\n",
       "      <td>44281063</td>\n",
       "      <td>57505</td>\n",
       "      <td>AARS2</td>\n",
       "      <td>NM_020745.3</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AASS</td>\n",
       "      <td>0.132071</td>\n",
       "      <td>1.355315e-06</td>\n",
       "      <td>chr7</td>\n",
       "      <td>121713597</td>\n",
       "      <td>121784344</td>\n",
       "      <td>10157</td>\n",
       "      <td>AASS</td>\n",
       "      <td>NM_005763.3</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ABCA10</td>\n",
       "      <td>0.002180</td>\n",
       "      <td>1.206886e-27</td>\n",
       "      <td>chr17</td>\n",
       "      <td>67144147</td>\n",
       "      <td>67240956</td>\n",
       "      <td>10349</td>\n",
       "      <td>ABCA10</td>\n",
       "      <td>NM_080282.3</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ABCA13</td>\n",
       "      <td>0.047218</td>\n",
       "      <td>3.307328e-78</td>\n",
       "      <td>chr7</td>\n",
       "      <td>48211056</td>\n",
       "      <td>48687091</td>\n",
       "      <td>154664</td>\n",
       "      <td>ABCA13</td>\n",
       "      <td>NM_152701.4</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   symbol  hi_score          p_li  chrom    txStart      txEnd  GeneID  \\\n",
       "0   A2ML1  0.011664  1.329022e-22  chr12    8975067    9029379  144568   \n",
       "1   AARS2  0.089486  1.648580e-07   chr6   44266462   44281063   57505   \n",
       "2    AASS  0.132071  1.355315e-06   chr7  121713597  121784344   10157   \n",
       "3  ABCA10  0.002180  1.206886e-27  chr17   67144147   67240956   10349   \n",
       "4  ABCA13  0.047218  3.307328e-78   chr7   48211056   48687091  154664   \n",
       "\n",
       "    name2         name strand  \n",
       "0   A2ML1  NM_144670.5      +  \n",
       "1   AARS2  NM_020745.3      -  \n",
       "2    AASS  NM_005763.3      -  \n",
       "3  ABCA10  NM_080282.3      -  \n",
       "4  ABCA13  NM_152701.4      +  "
      ]
     },
     "execution_count": 91,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "decipher.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 92,
   "metadata": {},
   "outputs": [],
   "source": [
    "gnomad = pd.read_csv('data/gnomad.v2.1.1.lof_metrics.by_gene.txt', sep='\\t', index_col=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 93,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>transcript</th>\n",
       "      <th>obs_mis</th>\n",
       "      <th>exp_mis</th>\n",
       "      <th>oe_mis</th>\n",
       "      <th>mu_mis</th>\n",
       "      <th>possible_mis</th>\n",
       "      <th>obs_mis_pphen</th>\n",
       "      <th>exp_mis_pphen</th>\n",
       "      <th>oe_mis_pphen</th>\n",
       "      <th>possible_mis_pphen</th>\n",
       "      <th>obs_syn</th>\n",
       "      <th>exp_syn</th>\n",
       "      <th>oe_syn</th>\n",
       "      <th>mu_syn</th>\n",
       "      <th>possible_syn</th>\n",
       "      <th>obs_lof</th>\n",
       "      <th>mu_lof</th>\n",
       "      <th>possible_lof</th>\n",
       "      <th>exp_lof</th>\n",
       "      <th>pLI</th>\n",
       "      <th>pNull</th>\n",
       "      <th>pRec</th>\n",
       "      <th>oe_lof</th>\n",
       "      <th>oe_syn_lower</th>\n",
       "      <th>oe_syn_upper</th>\n",
       "      <th>oe_mis_lower</th>\n",
       "      <th>oe_mis_upper</th>\n",
       "      <th>oe_lof_lower</th>\n",
       "      <th>oe_lof_upper</th>\n",
       "      <th>constraint_flag</th>\n",
       "      <th>syn_z</th>\n",
       "      <th>mis_z</th>\n",
       "      <th>lof_z</th>\n",
       "      <th>oe_lof_upper_rank</th>\n",
       "      <th>oe_lof_upper_bin</th>\n",
       "      <th>oe_lof_upper_bin_6</th>\n",
       "      <th>n_sites</th>\n",
       "      <th>classic_caf</th>\n",
       "      <th>max_af</th>\n",
       "      <th>no_lofs</th>\n",
       "      <th>obs_het_lof</th>\n",
       "      <th>obs_hom_lof</th>\n",
       "      <th>defined</th>\n",
       "      <th>p</th>\n",
       "      <th>exp_hom_lof</th>\n",
       "      <th>classic_caf_afr</th>\n",
       "      <th>classic_caf_amr</th>\n",
       "      <th>classic_caf_asj</th>\n",
       "      <th>classic_caf_eas</th>\n",
       "      <th>classic_caf_fin</th>\n",
       "      <th>classic_caf_nfe</th>\n",
       "      <th>classic_caf_oth</th>\n",
       "      <th>classic_caf_sas</th>\n",
       "      <th>p_afr</th>\n",
       "      <th>p_amr</th>\n",
       "      <th>p_asj</th>\n",
       "      <th>p_eas</th>\n",
       "      <th>p_fin</th>\n",
       "      <th>p_nfe</th>\n",
       "      <th>p_oth</th>\n",
       "      <th>p_sas</th>\n",
       "      <th>transcript_type</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>transcript_level</th>\n",
       "      <th>cds_length</th>\n",
       "      <th>num_coding_exons</th>\n",
       "      <th>gene_type</th>\n",
       "      <th>gene_length</th>\n",
       "      <th>exac_pLI</th>\n",
       "      <th>exac_obs_lof</th>\n",
       "      <th>exac_exp_lof</th>\n",
       "      <th>exac_oe_lof</th>\n",
       "      <th>brain_expression</th>\n",
       "      <th>chromosome</th>\n",
       "      <th>start_position</th>\n",
       "      <th>end_position</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>MED13</th>\n",
       "      <td>ENST00000397786</td>\n",
       "      <td>871</td>\n",
       "      <td>1117.80</td>\n",
       "      <td>0.77921</td>\n",
       "      <td>0.000056</td>\n",
       "      <td>14195</td>\n",
       "      <td>314.0</td>\n",
       "      <td>529.75</td>\n",
       "      <td>0.59273</td>\n",
       "      <td>6708.0</td>\n",
       "      <td>422</td>\n",
       "      <td>387.53</td>\n",
       "      <td>1.0890</td>\n",
       "      <td>0.000019</td>\n",
       "      <td>4248</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000005</td>\n",
       "      <td>1257.0</td>\n",
       "      <td>98.429</td>\n",
       "      <td>1.0</td>\n",
       "      <td>8.943600e-40</td>\n",
       "      <td>1.838300e-16</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.005</td>\n",
       "      <td>1.180</td>\n",
       "      <td>0.736</td>\n",
       "      <td>0.824</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.030</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-1.376500</td>\n",
       "      <td>2.6232</td>\n",
       "      <td>9.1935</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.000012</td>\n",
       "      <td>0.000008</td>\n",
       "      <td>124782.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>124785.0</td>\n",
       "      <td>0.000012</td>\n",
       "      <td>0.000018</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000093</td>\n",
       "      <td>0.000009</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000093</td>\n",
       "      <td>0.000009</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>ENSG00000108510</td>\n",
       "      <td>2</td>\n",
       "      <td>6522</td>\n",
       "      <td>30</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>122678</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>64.393</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>17</td>\n",
       "      <td>60019966</td>\n",
       "      <td>60142643</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NIPBL</th>\n",
       "      <td>ENST00000282516</td>\n",
       "      <td>846</td>\n",
       "      <td>1441.50</td>\n",
       "      <td>0.58688</td>\n",
       "      <td>0.000074</td>\n",
       "      <td>18540</td>\n",
       "      <td>158.0</td>\n",
       "      <td>543.10</td>\n",
       "      <td>0.29092</td>\n",
       "      <td>7135.0</td>\n",
       "      <td>496</td>\n",
       "      <td>495.01</td>\n",
       "      <td>1.0020</td>\n",
       "      <td>0.000025</td>\n",
       "      <td>5211</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.000009</td>\n",
       "      <td>1781.0</td>\n",
       "      <td>150.320</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.977300e-59</td>\n",
       "      <td>3.572400e-24</td>\n",
       "      <td>0.006653</td>\n",
       "      <td>0.930</td>\n",
       "      <td>1.079</td>\n",
       "      <td>0.554</td>\n",
       "      <td>0.621</td>\n",
       "      <td>0.001</td>\n",
       "      <td>0.032</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.035119</td>\n",
       "      <td>5.5737</td>\n",
       "      <td>11.2860</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.000012</td>\n",
       "      <td>0.000008</td>\n",
       "      <td>125693.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>125696.0</td>\n",
       "      <td>0.000012</td>\n",
       "      <td>0.000018</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000099</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000065</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000099</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000065</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>ENSG00000164190</td>\n",
       "      <td>2</td>\n",
       "      <td>8412</td>\n",
       "      <td>46</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>189655</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>110.570</td>\n",
       "      <td>0.009044</td>\n",
       "      <td>NaN</td>\n",
       "      <td>5</td>\n",
       "      <td>36876861</td>\n",
       "      <td>37066515</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SMC3</th>\n",
       "      <td>ENST00000361804</td>\n",
       "      <td>178</td>\n",
       "      <td>630.07</td>\n",
       "      <td>0.28251</td>\n",
       "      <td>0.000032</td>\n",
       "      <td>8109</td>\n",
       "      <td>21.0</td>\n",
       "      <td>182.52</td>\n",
       "      <td>0.11506</td>\n",
       "      <td>2197.0</td>\n",
       "      <td>215</td>\n",
       "      <td>203.25</td>\n",
       "      <td>1.0578</td>\n",
       "      <td>0.000010</td>\n",
       "      <td>2091</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000005</td>\n",
       "      <td>937.0</td>\n",
       "      <td>79.490</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.785300e-32</td>\n",
       "      <td>2.191400e-13</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.946</td>\n",
       "      <td>1.184</td>\n",
       "      <td>0.249</td>\n",
       "      <td>0.320</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.037</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.647760</td>\n",
       "      <td>6.3999</td>\n",
       "      <td>8.2618</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>0.000032</td>\n",
       "      <td>0.000004</td>\n",
       "      <td>125731.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>125739.0</td>\n",
       "      <td>0.000032</td>\n",
       "      <td>0.000127</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000099</td>\n",
       "      <td>0.000054</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000044</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000033</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000099</td>\n",
       "      <td>0.000054</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000044</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000033</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>ENSG00000108055</td>\n",
       "      <td>2</td>\n",
       "      <td>3651</td>\n",
       "      <td>29</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>36946</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>58.523</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>10</td>\n",
       "      <td>112327449</td>\n",
       "      <td>112364394</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CNOT1</th>\n",
       "      <td>ENST00000317147</td>\n",
       "      <td>561</td>\n",
       "      <td>1295.90</td>\n",
       "      <td>0.43290</td>\n",
       "      <td>0.000069</td>\n",
       "      <td>15670</td>\n",
       "      <td>51.0</td>\n",
       "      <td>290.68</td>\n",
       "      <td>0.17545</td>\n",
       "      <td>3560.0</td>\n",
       "      <td>470</td>\n",
       "      <td>456.03</td>\n",
       "      <td>1.0306</td>\n",
       "      <td>0.000024</td>\n",
       "      <td>4564</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.000007</td>\n",
       "      <td>1440.0</td>\n",
       "      <td>125.030</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.992400e-49</td>\n",
       "      <td>4.562900e-20</td>\n",
       "      <td>0.007998</td>\n",
       "      <td>0.955</td>\n",
       "      <td>1.112</td>\n",
       "      <td>0.403</td>\n",
       "      <td>0.464</td>\n",
       "      <td>0.002</td>\n",
       "      <td>0.038</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.514100</td>\n",
       "      <td>7.2546</td>\n",
       "      <td>10.2790</td>\n",
       "      <td>3.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>0.000020</td>\n",
       "      <td>0.000004</td>\n",
       "      <td>125740.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>125744.0</td>\n",
       "      <td>0.000016</td>\n",
       "      <td>0.000032</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000029</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000055</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000026</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000029</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000054</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000018</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>ENSG00000125107</td>\n",
       "      <td>2</td>\n",
       "      <td>7128</td>\n",
       "      <td>48</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>109936</td>\n",
       "      <td>1.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>90.130</td>\n",
       "      <td>0.033285</td>\n",
       "      <td>NaN</td>\n",
       "      <td>16</td>\n",
       "      <td>58553855</td>\n",
       "      <td>58663790</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RLF</th>\n",
       "      <td>ENST00000372771</td>\n",
       "      <td>669</td>\n",
       "      <td>972.87</td>\n",
       "      <td>0.68766</td>\n",
       "      <td>0.000047</td>\n",
       "      <td>12682</td>\n",
       "      <td>107.0</td>\n",
       "      <td>321.14</td>\n",
       "      <td>0.33319</td>\n",
       "      <td>4151.0</td>\n",
       "      <td>358</td>\n",
       "      <td>352.62</td>\n",
       "      <td>1.0153</td>\n",
       "      <td>0.000017</td>\n",
       "      <td>3482</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000004</td>\n",
       "      <td>1024.0</td>\n",
       "      <td>73.222</td>\n",
       "      <td>1.0</td>\n",
       "      <td>8.405500e-30</td>\n",
       "      <td>2.284200e-12</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.930</td>\n",
       "      <td>1.108</td>\n",
       "      <td>0.645</td>\n",
       "      <td>0.733</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.040</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.225180</td>\n",
       "      <td>3.4620</td>\n",
       "      <td>7.9294</td>\n",
       "      <td>4.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.000004</td>\n",
       "      <td>0.000004</td>\n",
       "      <td>125122.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>125123.0</td>\n",
       "      <td>0.000004</td>\n",
       "      <td>0.000002</td>\n",
       "      <td>0.000062</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000062</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>ENSG00000117000</td>\n",
       "      <td>2</td>\n",
       "      <td>5742</td>\n",
       "      <td>8</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>79549</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>43.607</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>40627045</td>\n",
       "      <td>40706593</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            transcript  obs_mis  exp_mis   oe_mis    mu_mis  possible_mis  \\\n",
       "gene                                                                        \n",
       "MED13  ENST00000397786      871  1117.80  0.77921  0.000056         14195   \n",
       "NIPBL  ENST00000282516      846  1441.50  0.58688  0.000074         18540   \n",
       "SMC3   ENST00000361804      178   630.07  0.28251  0.000032          8109   \n",
       "CNOT1  ENST00000317147      561  1295.90  0.43290  0.000069         15670   \n",
       "RLF    ENST00000372771      669   972.87  0.68766  0.000047         12682   \n",
       "\n",
       "       obs_mis_pphen  exp_mis_pphen  oe_mis_pphen  possible_mis_pphen  \\\n",
       "gene                                                                    \n",
       "MED13          314.0         529.75       0.59273              6708.0   \n",
       "NIPBL          158.0         543.10       0.29092              7135.0   \n",
       "SMC3            21.0         182.52       0.11506              2197.0   \n",
       "CNOT1           51.0         290.68       0.17545              3560.0   \n",
       "RLF            107.0         321.14       0.33319              4151.0   \n",
       "\n",
       "       obs_syn  exp_syn  oe_syn    mu_syn  possible_syn  obs_lof    mu_lof  \\\n",
       "gene                                                                         \n",
       "MED13      422   387.53  1.0890  0.000019          4248      0.0  0.000005   \n",
       "NIPBL      496   495.01  1.0020  0.000025          5211      1.0  0.000009   \n",
       "SMC3       215   203.25  1.0578  0.000010          2091      0.0  0.000005   \n",
       "CNOT1      470   456.03  1.0306  0.000024          4564      1.0  0.000007   \n",
       "RLF        358   352.62  1.0153  0.000017          3482      0.0  0.000004   \n",
       "\n",
       "       possible_lof  exp_lof  pLI         pNull          pRec    oe_lof  \\\n",
       "gene                                                                      \n",
       "MED13        1257.0   98.429  1.0  8.943600e-40  1.838300e-16  0.000000   \n",
       "NIPBL        1781.0  150.320  1.0  2.977300e-59  3.572400e-24  0.006653   \n",
       "SMC3          937.0   79.490  1.0  2.785300e-32  2.191400e-13  0.000000   \n",
       "CNOT1        1440.0  125.030  1.0  2.992400e-49  4.562900e-20  0.007998   \n",
       "RLF          1024.0   73.222  1.0  8.405500e-30  2.284200e-12  0.000000   \n",
       "\n",
       "       oe_syn_lower  oe_syn_upper  oe_mis_lower  oe_mis_upper  oe_lof_lower  \\\n",
       "gene                                                                          \n",
       "MED13         1.005         1.180         0.736         0.824         0.000   \n",
       "NIPBL         0.930         1.079         0.554         0.621         0.001   \n",
       "SMC3          0.946         1.184         0.249         0.320         0.000   \n",
       "CNOT1         0.955         1.112         0.403         0.464         0.002   \n",
       "RLF           0.930         1.108         0.645         0.733         0.000   \n",
       "\n",
       "       oe_lof_upper constraint_flag     syn_z   mis_z    lof_z  \\\n",
       "gene                                                             \n",
       "MED13         0.030             NaN -1.376500  2.6232   9.1935   \n",
       "NIPBL         0.032             NaN -0.035119  5.5737  11.2860   \n",
       "SMC3          0.037             NaN -0.647760  6.3999   8.2618   \n",
       "CNOT1         0.038             NaN -0.514100  7.2546  10.2790   \n",
       "RLF           0.040             NaN -0.225180  3.4620   7.9294   \n",
       "\n",
       "       oe_lof_upper_rank  oe_lof_upper_bin  oe_lof_upper_bin_6  n_sites  \\\n",
       "gene                                                                      \n",
       "MED13                0.0               0.0                 0.0      2.0   \n",
       "NIPBL                1.0               0.0                 0.0      2.0   \n",
       "SMC3                 2.0               0.0                 0.0      8.0   \n",
       "CNOT1                3.0               0.0                 0.0      5.0   \n",
       "RLF                  4.0               0.0                 0.0      1.0   \n",
       "\n",
       "       classic_caf    max_af   no_lofs  obs_het_lof  obs_hom_lof   defined  \\\n",
       "gene                                                                         \n",
       "MED13     0.000012  0.000008  124782.0          3.0          0.0  124785.0   \n",
       "NIPBL     0.000012  0.000008  125693.0          3.0          0.0  125696.0   \n",
       "SMC3      0.000032  0.000004  125731.0          8.0          0.0  125739.0   \n",
       "CNOT1     0.000020  0.000004  125740.0          4.0          0.0  125744.0   \n",
       "RLF       0.000004  0.000004  125122.0          1.0          0.0  125123.0   \n",
       "\n",
       "              p  exp_hom_lof  classic_caf_afr  classic_caf_amr  \\\n",
       "gene                                                             \n",
       "MED13  0.000012     0.000018         0.000000         0.000000   \n",
       "NIPBL  0.000012     0.000018         0.000000         0.000000   \n",
       "SMC3   0.000032     0.000127         0.000000         0.000000   \n",
       "CNOT1  0.000016     0.000032         0.000000         0.000029   \n",
       "RLF    0.000004     0.000002         0.000062         0.000000   \n",
       "\n",
       "       classic_caf_asj  classic_caf_eas  classic_caf_fin  classic_caf_nfe  \\\n",
       "gene                                                                        \n",
       "MED13         0.000000         0.000000         0.000093         0.000009   \n",
       "NIPBL         0.000099         0.000000         0.000000         0.000000   \n",
       "SMC3          0.000099         0.000054         0.000000         0.000044   \n",
       "CNOT1         0.000000         0.000055         0.000000         0.000026   \n",
       "RLF           0.000000         0.000000         0.000000         0.000000   \n",
       "\n",
       "       classic_caf_oth  classic_caf_sas     p_afr     p_amr     p_asj  \\\n",
       "gene                                                                    \n",
       "MED13              0.0         0.000000  0.000000  0.000000  0.000000   \n",
       "NIPBL              0.0         0.000065  0.000000  0.000000  0.000099   \n",
       "SMC3               0.0         0.000033  0.000000  0.000000  0.000099   \n",
       "CNOT1              0.0         0.000000  0.000000  0.000029  0.000000   \n",
       "RLF                0.0         0.000000  0.000062  0.000000  0.000000   \n",
       "\n",
       "          p_eas     p_fin     p_nfe  p_oth     p_sas transcript_type  \\\n",
       "gene                                                                   \n",
       "MED13  0.000000  0.000093  0.000009    0.0  0.000000  protein_coding   \n",
       "NIPBL  0.000000  0.000000  0.000000    0.0  0.000065  protein_coding   \n",
       "SMC3   0.000054  0.000000  0.000044    0.0  0.000033  protein_coding   \n",
       "CNOT1  0.000054  0.000000  0.000018    0.0  0.000000  protein_coding   \n",
       "RLF    0.000000  0.000000  0.000000    0.0  0.000000  protein_coding   \n",
       "\n",
       "               gene_id  transcript_level  cds_length  num_coding_exons  \\\n",
       "gene                                                                     \n",
       "MED13  ENSG00000108510                 2        6522                30   \n",
       "NIPBL  ENSG00000164190                 2        8412                46   \n",
       "SMC3   ENSG00000108055                 2        3651                29   \n",
       "CNOT1  ENSG00000125107                 2        7128                48   \n",
       "RLF    ENSG00000117000                 2        5742                 8   \n",
       "\n",
       "            gene_type  gene_length  exac_pLI  exac_obs_lof  exac_exp_lof  \\\n",
       "gene                                                                       \n",
       "MED13  protein_coding       122678       1.0           0.0        64.393   \n",
       "NIPBL  protein_coding       189655       1.0           1.0       110.570   \n",
       "SMC3   protein_coding        36946       1.0           0.0        58.523   \n",
       "CNOT1  protein_coding       109936       1.0           3.0        90.130   \n",
       "RLF    protein_coding        79549       1.0           0.0        43.607   \n",
       "\n",
       "       exac_oe_lof  brain_expression chromosome  start_position  end_position  \n",
       "gene                                                                           \n",
       "MED13     0.000000               NaN         17        60019966      60142643  \n",
       "NIPBL     0.009044               NaN          5        36876861      37066515  \n",
       "SMC3      0.000000               NaN         10       112327449     112364394  \n",
       "CNOT1     0.033285               NaN         16        58553855      58663790  \n",
       "RLF       0.000000               NaN          1        40627045      40706593  "
      ]
     },
     "execution_count": 93,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gnomad.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 94,
   "metadata": {},
   "outputs": [],
   "source": [
    "decipher = decipher.join(gnomad['pLI'], on='name2')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 95,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>symbol</th>\n",
       "      <th>hi_score</th>\n",
       "      <th>p_li</th>\n",
       "      <th>chrom</th>\n",
       "      <th>txStart</th>\n",
       "      <th>txEnd</th>\n",
       "      <th>GeneID</th>\n",
       "      <th>name2</th>\n",
       "      <th>name</th>\n",
       "      <th>strand</th>\n",
       "      <th>pLI</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A2ML1</td>\n",
       "      <td>0.011664</td>\n",
       "      <td>1.329022e-22</td>\n",
       "      <td>chr12</td>\n",
       "      <td>8975067</td>\n",
       "      <td>9029379</td>\n",
       "      <td>144568</td>\n",
       "      <td>A2ML1</td>\n",
       "      <td>NM_144670.5</td>\n",
       "      <td>+</td>\n",
       "      <td>1.610900e-34</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AARS2</td>\n",
       "      <td>0.089486</td>\n",
       "      <td>1.648580e-07</td>\n",
       "      <td>chr6</td>\n",
       "      <td>44266462</td>\n",
       "      <td>44281063</td>\n",
       "      <td>57505</td>\n",
       "      <td>AARS2</td>\n",
       "      <td>NM_020745.3</td>\n",
       "      <td>-</td>\n",
       "      <td>4.977700e-17</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AASS</td>\n",
       "      <td>0.132071</td>\n",
       "      <td>1.355315e-06</td>\n",
       "      <td>chr7</td>\n",
       "      <td>121713597</td>\n",
       "      <td>121784344</td>\n",
       "      <td>10157</td>\n",
       "      <td>AASS</td>\n",
       "      <td>NM_005763.3</td>\n",
       "      <td>-</td>\n",
       "      <td>2.310100e-12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ABCA10</td>\n",
       "      <td>0.002180</td>\n",
       "      <td>1.206886e-27</td>\n",
       "      <td>chr17</td>\n",
       "      <td>67144147</td>\n",
       "      <td>67240956</td>\n",
       "      <td>10349</td>\n",
       "      <td>ABCA10</td>\n",
       "      <td>NM_080282.3</td>\n",
       "      <td>-</td>\n",
       "      <td>7.659300e-34</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ABCA13</td>\n",
       "      <td>0.047218</td>\n",
       "      <td>3.307328e-78</td>\n",
       "      <td>chr7</td>\n",
       "      <td>48211056</td>\n",
       "      <td>48687091</td>\n",
       "      <td>154664</td>\n",
       "      <td>ABCA13</td>\n",
       "      <td>NM_152701.4</td>\n",
       "      <td>+</td>\n",
       "      <td>8.775300e-114</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   symbol  hi_score          p_li  chrom    txStart      txEnd  GeneID  \\\n",
       "0   A2ML1  0.011664  1.329022e-22  chr12    8975067    9029379  144568   \n",
       "1   AARS2  0.089486  1.648580e-07   chr6   44266462   44281063   57505   \n",
       "2    AASS  0.132071  1.355315e-06   chr7  121713597  121784344   10157   \n",
       "3  ABCA10  0.002180  1.206886e-27  chr17   67144147   67240956   10349   \n",
       "4  ABCA13  0.047218  3.307328e-78   chr7   48211056   48687091  154664   \n",
       "\n",
       "    name2         name strand            pLI  \n",
       "0   A2ML1  NM_144670.5      +   1.610900e-34  \n",
       "1   AARS2  NM_020745.3      -   4.977700e-17  \n",
       "2    AASS  NM_005763.3      -   2.310100e-12  \n",
       "3  ABCA10  NM_080282.3      -   7.659300e-34  \n",
       "4  ABCA13  NM_152701.4      +  8.775300e-114  "
      ]
     },
     "execution_count": 95,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "decipher.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 96,
   "metadata": {},
   "outputs": [],
   "source": [
    "decipher = decipher.loc[\n",
    "    (decipher['pLI'] >= 0.9) & (decipher['hi_score'] < 0.1),\n",
    "    ['chrom', 'txStart', 'txEnd', 'GeneID', 'name2', 'name', 'pLI', 'hi_score']\n",
    "].sort_values(['chrom', 'txStart', 'txEnd'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 97,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>txStart</th>\n",
       "      <th>txEnd</th>\n",
       "      <th>GeneID</th>\n",
       "      <th>name2</th>\n",
       "      <th>name</th>\n",
       "      <th>pLI</th>\n",
       "      <th>hi_score</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2112</th>\n",
       "      <td>chr1</td>\n",
       "      <td>2439969</td>\n",
       "      <td>2458067</td>\n",
       "      <td>55229</td>\n",
       "      <td>PANK4</td>\n",
       "      <td>NM_018216.2</td>\n",
       "      <td>0.98325</td>\n",
       "      <td>0.098966</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2587</th>\n",
       "      <td>chr1</td>\n",
       "      <td>155290639</td>\n",
       "      <td>155300909</td>\n",
       "      <td>23623</td>\n",
       "      <td>RUSC1</td>\n",
       "      <td>NM_001105203.1</td>\n",
       "      <td>0.94130</td>\n",
       "      <td>0.035285</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2705</th>\n",
       "      <td>chr1</td>\n",
       "      <td>232533710</td>\n",
       "      <td>232651354</td>\n",
       "      <td>57568</td>\n",
       "      <td>SIPA1L2</td>\n",
       "      <td>NM_020808.4</td>\n",
       "      <td>0.99215</td>\n",
       "      <td>0.079950</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1608</th>\n",
       "      <td>chr10</td>\n",
       "      <td>852853</td>\n",
       "      <td>977645</td>\n",
       "      <td>23185</td>\n",
       "      <td>LARP4B</td>\n",
       "      <td>NM_015155.2</td>\n",
       "      <td>0.99998</td>\n",
       "      <td>0.071477</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>199</th>\n",
       "      <td>chr10</td>\n",
       "      <td>24872537</td>\n",
       "      <td>25012597</td>\n",
       "      <td>57584</td>\n",
       "      <td>ARHGAP21</td>\n",
       "      <td>NM_020824.3</td>\n",
       "      <td>1.00000</td>\n",
       "      <td>0.073197</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      chrom    txStart      txEnd  GeneID     name2            name      pLI  \\\n",
       "2112   chr1    2439969    2458067   55229     PANK4     NM_018216.2  0.98325   \n",
       "2587   chr1  155290639  155300909   23623     RUSC1  NM_001105203.1  0.94130   \n",
       "2705   chr1  232533710  232651354   57568   SIPA1L2     NM_020808.4  0.99215   \n",
       "1608  chr10     852853     977645   23185    LARP4B     NM_015155.2  0.99998   \n",
       "199   chr10   24872537   25012597   57584  ARHGAP21     NM_020824.3  1.00000   \n",
       "\n",
       "      hi_score  \n",
       "2112  0.098966  \n",
       "2587  0.035285  \n",
       "2705  0.079950  \n",
       "1608  0.071477  \n",
       "199   0.073197  "
      ]
     },
     "execution_count": 97,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "decipher.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 98,
   "metadata": {},
   "outputs": [],
   "source": [
    "decipher.rename(columns={\n",
    "    'chrom': '#chrom', 'txStart': 'start', 'txEnd': 'end',\n",
    "    'GeneID': 'gene_id', 'name2': 'symbol', 'name': 'transcript'\n",
    "}).to_csv('data/decipher-gene-20200109.sorted.bed', sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 99,
   "metadata": {},
   "outputs": [],
   "source": [
    "!bgzip -cf data/decipher-gene-20200109.sorted.bed > ../data/decipher-gene-20200109.sorted.bed.gz"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 100,
   "metadata": {},
   "outputs": [],
   "source": [
    "!tabix -fp bed ../data/decipher-gene-20200109.sorted.bed.gz"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 101,
   "metadata": {},
   "outputs": [],
   "source": [
    "ts_genes = set(\n",
    "    curation_gene.loc[\n",
    "        curation_gene['Triplosensitivity Score'] == '3', '#Gene Symbol'\n",
    "    ]\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 102,
   "metadata": {},
   "outputs": [],
   "source": [
    "ts_gene = refgene_info.loc[\n",
    "    refgene_info['name2'].isin(ts_genes),\n",
    "    ['chrom', 'txStart', 'txEnd', 'GeneID', 'name2', 'name', 'strand']\n",
    "].sort_values(['chrom', 'txStart', 'txEnd'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 103,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>txStart</th>\n",
       "      <th>txEnd</th>\n",
       "      <th>GeneID</th>\n",
       "      <th>name2</th>\n",
       "      <th>name</th>\n",
       "      <th>strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>4186</th>\n",
       "      <td>chr5</td>\n",
       "      <td>126112314</td>\n",
       "      <td>126172712</td>\n",
       "      <td>4001</td>\n",
       "      <td>LMNB1</td>\n",
       "      <td>NM_005573.3</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10499</th>\n",
       "      <td>chrX</td>\n",
       "      <td>103031433</td>\n",
       "      <td>103047548</td>\n",
       "      <td>5354</td>\n",
       "      <td>PLP1</td>\n",
       "      <td>NM_001128834.2</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3662</th>\n",
       "      <td>chrX</td>\n",
       "      <td>153295684</td>\n",
       "      <td>153363188</td>\n",
       "      <td>4204</td>\n",
       "      <td>MECP2</td>\n",
       "      <td>NM_001110792.1</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      chrom    txStart      txEnd  GeneID  name2            name strand\n",
       "4186   chr5  126112314  126172712    4001  LMNB1     NM_005573.3      +\n",
       "10499  chrX  103031433  103047548    5354   PLP1  NM_001128834.2      +\n",
       "3662   chrX  153295684  153363188    4204  MECP2  NM_001110792.1      -"
      ]
     },
     "execution_count": 103,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ts_gene.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 104,
   "metadata": {},
   "outputs": [],
   "source": [
    "ts_gene.rename(columns={\n",
    "    'chrom': '#chrom', 'txStart': 'start', 'txEnd': 'end',\n",
    "    'GeneID': 'gene_id', 'name2': 'symbol', 'name': 'transcript'\n",
    "}).to_csv('data/ts-gene-20200109.sorted.bed', index=False, sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 105,
   "metadata": {},
   "outputs": [],
   "source": [
    "!bgzip -cf data/ts-gene-20200109.sorted.bed > ../data/ts-gene-20200109.sorted.bed.gz"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 106,
   "metadata": {},
   "outputs": [],
   "source": [
    "!tabix -fp bed ../data/ts-gene-20200109.sorted.bed.gz"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 107,
   "metadata": {},
   "outputs": [],
   "source": [
    "uts_genes = set(\n",
    "    curation_gene.loc[curation_gene['Triplosensitivity Score'] == '40', '#Gene Symbol']\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 108,
   "metadata": {},
   "outputs": [],
   "source": [
    "uts_gene = refgene_info.loc[\n",
    "    refgene_info['name2'].isin(uts_genes),\n",
    "    ['chrom', 'txStart', 'txEnd', 'GeneID', 'name2', 'name', 'strand']\n",
    "].sort_values(['chrom', 'txStart', 'txEnd'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 109,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>txStart</th>\n",
       "      <th>txEnd</th>\n",
       "      <th>GeneID</th>\n",
       "      <th>name2</th>\n",
       "      <th>name</th>\n",
       "      <th>strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>17987</th>\n",
       "      <td>chrX</td>\n",
       "      <td>6451658</td>\n",
       "      <td>6453159</td>\n",
       "      <td>51481</td>\n",
       "      <td>VCX3A</td>\n",
       "      <td>NM_016379.3</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      chrom  txStart    txEnd  GeneID  name2         name strand\n",
       "17987  chrX  6451658  6453159   51481  VCX3A  NM_016379.3      -"
      ]
     },
     "execution_count": 109,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "uts_gene.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 110,
   "metadata": {},
   "outputs": [],
   "source": [
    "uts_gene.rename(columns={\n",
    "    'chrom': '#chrom', 'txStart': 'start', 'txEnd': 'end',\n",
    "    'GeneID': 'gene_id', 'name2': 'symbol', 'name': 'transcript'\n",
    "}).to_csv('data/uts-gene-20200109.sorted.bed', index=False, sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 111,
   "metadata": {},
   "outputs": [],
   "source": [
    "!bgzip -cf data/uts-gene-20200109.sorted.bed > ../data/uts-gene-20200109.sorted.bed.gz"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 112,
   "metadata": {},
   "outputs": [],
   "source": [
    "!tabix -fp bed ../data/uts-gene-20200109.sorted.bed.gz"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 113,
   "metadata": {},
   "outputs": [],
   "source": [
    "ts_region = region_pos.loc[\n",
    "    region_pos['Triplosensitivity Score'] == '3',\n",
    "    ['chrom', 'start', 'end', '#ISCA ID', 'ISCA Region Name']\n",
    "].sort_values(['chrom', 'start', 'end'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 114,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>#ISCA ID</th>\n",
       "      <th>ISCA Region Name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>52</th>\n",
       "      <td>chr1</td>\n",
       "      <td>146577486</td>\n",
       "      <td>147394506</td>\n",
       "      <td>ISCA-37421</td>\n",
       "      <td>1q21.1 recurrent region (BP3-BP4, distal) (inc...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>62</th>\n",
       "      <td>chr15</td>\n",
       "      <td>22832519</td>\n",
       "      <td>28379874</td>\n",
       "      <td>ISCA-37404</td>\n",
       "      <td>15q11q13 recurrent (PWS/AS) region (BP1-BP3, C...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>chr15</td>\n",
       "      <td>23747996</td>\n",
       "      <td>28379874</td>\n",
       "      <td>ISCA-37478</td>\n",
       "      <td>15q11q13 recurrent (PWS/AS) region (BP2-BP3, C...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>64</th>\n",
       "      <td>chr16</td>\n",
       "      <td>29649997</td>\n",
       "      <td>30199852</td>\n",
       "      <td>ISCA-37400</td>\n",
       "      <td>16p11.2 recurrent region (proximal, BP4-BP5) (...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>47</th>\n",
       "      <td>chr17</td>\n",
       "      <td>1247833</td>\n",
       "      <td>2588909</td>\n",
       "      <td>ISCA-37430</td>\n",
       "      <td>17p13.3 (Miller-Dieker syndrome) region (inclu...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    chrom      start        end    #ISCA ID  \\\n",
       "52   chr1  146577486  147394506  ISCA-37421   \n",
       "62  chr15   22832519   28379874  ISCA-37404   \n",
       "23  chr15   23747996   28379874  ISCA-37478   \n",
       "64  chr16   29649997   30199852  ISCA-37400   \n",
       "47  chr17    1247833    2588909  ISCA-37430   \n",
       "\n",
       "                                     ISCA Region Name  \n",
       "52  1q21.1 recurrent region (BP3-BP4, distal) (inc...  \n",
       "62  15q11q13 recurrent (PWS/AS) region (BP1-BP3, C...  \n",
       "23  15q11q13 recurrent (PWS/AS) region (BP2-BP3, C...  \n",
       "64  16p11.2 recurrent region (proximal, BP4-BP5) (...  \n",
       "47  17p13.3 (Miller-Dieker syndrome) region (inclu...  "
      ]
     },
     "execution_count": 114,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ts_region.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 115,
   "metadata": {},
   "outputs": [],
   "source": [
    "ts_region['omim_genes'] = ts_region.apply(\n",
    "    lambda row: ','.join(omim_gene.loc[\n",
    "        (omim_gene['chrom'] == row['chrom']) & (omim_gene['txStart'] >= row['start']) & (omim_gene['txEnd'] <= row['end']), 'name2'\n",
    "    ]),\n",
    "    axis=1\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 116,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>#ISCA ID</th>\n",
       "      <th>ISCA Region Name</th>\n",
       "      <th>omim_genes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>52</th>\n",
       "      <td>chr1</td>\n",
       "      <td>146577486</td>\n",
       "      <td>147394506</td>\n",
       "      <td>ISCA-37421</td>\n",
       "      <td>1q21.1 recurrent region (BP3-BP4, distal) (inc...</td>\n",
       "      <td>GJA5,GJA8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>62</th>\n",
       "      <td>chr15</td>\n",
       "      <td>22832519</td>\n",
       "      <td>28379874</td>\n",
       "      <td>ISCA-37404</td>\n",
       "      <td>15q11q13 recurrent (PWS/AS) region (BP1-BP3, C...</td>\n",
       "      <td>NIPA1,MKRN3,MAGEL2,NDN,SNRPN,UBE3A,GABRB3,GABR...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>chr15</td>\n",
       "      <td>23747996</td>\n",
       "      <td>28379874</td>\n",
       "      <td>ISCA-37478</td>\n",
       "      <td>15q11q13 recurrent (PWS/AS) region (BP2-BP3, C...</td>\n",
       "      <td>MKRN3,MAGEL2,NDN,SNRPN,UBE3A,GABRB3,GABRA5,OCA2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>64</th>\n",
       "      <td>chr16</td>\n",
       "      <td>29649997</td>\n",
       "      <td>30199852</td>\n",
       "      <td>ISCA-37400</td>\n",
       "      <td>16p11.2 recurrent region (proximal, BP4-BP5) (...</td>\n",
       "      <td>KIF22,PRRT2,ALDOA,TBX6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>47</th>\n",
       "      <td>chr17</td>\n",
       "      <td>1247833</td>\n",
       "      <td>2588909</td>\n",
       "      <td>ISCA-37430</td>\n",
       "      <td>17p13.3 (Miller-Dieker syndrome) region (inclu...</td>\n",
       "      <td>INPP5K,PRPF8,WDR81,SERPINF2,SERPINF1,DPH1,PAFA...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    chrom      start        end    #ISCA ID  \\\n",
       "52   chr1  146577486  147394506  ISCA-37421   \n",
       "62  chr15   22832519   28379874  ISCA-37404   \n",
       "23  chr15   23747996   28379874  ISCA-37478   \n",
       "64  chr16   29649997   30199852  ISCA-37400   \n",
       "47  chr17    1247833    2588909  ISCA-37430   \n",
       "\n",
       "                                     ISCA Region Name  \\\n",
       "52  1q21.1 recurrent region (BP3-BP4, distal) (inc...   \n",
       "62  15q11q13 recurrent (PWS/AS) region (BP1-BP3, C...   \n",
       "23  15q11q13 recurrent (PWS/AS) region (BP2-BP3, C...   \n",
       "64  16p11.2 recurrent region (proximal, BP4-BP5) (...   \n",
       "47  17p13.3 (Miller-Dieker syndrome) region (inclu...   \n",
       "\n",
       "                                           omim_genes  \n",
       "52                                          GJA5,GJA8  \n",
       "62  NIPA1,MKRN3,MAGEL2,NDN,SNRPN,UBE3A,GABRB3,GABR...  \n",
       "23    MKRN3,MAGEL2,NDN,SNRPN,UBE3A,GABRB3,GABRA5,OCA2  \n",
       "64                             KIF22,PRRT2,ALDOA,TBX6  \n",
       "47  INPP5K,PRPF8,WDR81,SERPINF2,SERPINF1,DPH1,PAFA...  "
      ]
     },
     "execution_count": 116,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ts_region.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 117,
   "metadata": {},
   "outputs": [],
   "source": [
    "ts_region.rename(columns={\n",
    "    'chrom': '#chrom',\n",
    "    '#ISCA ID': 'isca_id', 'ISCA Region Name': 'name'\n",
    "}).to_csv('data/ts-region-20200109.sorted.bed', sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 118,
   "metadata": {},
   "outputs": [],
   "source": [
    "!bgzip -cf data/ts-region-20200109.sorted.bed > ../data/ts-region-20200109.sorted.bed.gz"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 119,
   "metadata": {},
   "outputs": [],
   "source": [
    "!tabix -fp bed ../data/ts-region-20200109.sorted.bed.gz"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 120,
   "metadata": {},
   "outputs": [],
   "source": [
    "uts_region = region_pos.loc[\n",
    "    region_pos['Triplosensitivity Score'] == '40',\n",
    "    ['chrom', 'start', 'end', '#ISCA ID', 'ISCA Region Name']\n",
    "].sort_values(['chrom', 'start', 'end'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 121,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>#ISCA ID</th>\n",
       "      <th>ISCA Region Name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>chr1</td>\n",
       "      <td>148867551</td>\n",
       "      <td>149768855</td>\n",
       "      <td>ISCA-37469</td>\n",
       "      <td>1q21.2 region (polymorphic region)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>chr14</td>\n",
       "      <td>22111109</td>\n",
       "      <td>23021097</td>\n",
       "      <td>ISCA-37476</td>\n",
       "      <td>14q11.2 region (TCRA region)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>chr14</td>\n",
       "      <td>106050000</td>\n",
       "      <td>107289540</td>\n",
       "      <td>ISCA-37477</td>\n",
       "      <td>1 copy: 14q telomere; 3 copies: 14q telomere</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>chr15</td>\n",
       "      <td>22832519</td>\n",
       "      <td>23090897</td>\n",
       "      <td>ISCA-37448</td>\n",
       "      <td>15q11.2 recurrent region (BP1-BP2) (includes N...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>chr15</td>\n",
       "      <td>32019621</td>\n",
       "      <td>32445405</td>\n",
       "      <td>ISCA-46295</td>\n",
       "      <td>15q13.3 recurrent region (D-CHRNA7 to BP5) (in...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    chrom      start        end    #ISCA ID  \\\n",
       "29   chr1  148867551  149768855  ISCA-37469   \n",
       "25  chr14   22111109   23021097  ISCA-37476   \n",
       "24  chr14  106050000  107289540  ISCA-37477   \n",
       "33  chr15   22832519   23090897  ISCA-37448   \n",
       "5   chr15   32019621   32445405  ISCA-46295   \n",
       "\n",
       "                                     ISCA Region Name  \n",
       "29                 1q21.2 region (polymorphic region)  \n",
       "25                       14q11.2 region (TCRA region)  \n",
       "24       1 copy: 14q telomere; 3 copies: 14q telomere  \n",
       "33  15q11.2 recurrent region (BP1-BP2) (includes N...  \n",
       "5   15q13.3 recurrent region (D-CHRNA7 to BP5) (in...  "
      ]
     },
     "execution_count": 121,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "uts_region.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 122,
   "metadata": {},
   "outputs": [],
   "source": [
    "uts_region['genes'] = uts_region.apply(\n",
    "    lambda row: ','.join(gene.loc[\n",
    "        (gene['chrom'] == row['chrom']) & (gene['txStart'] >= row['start']) & (gene['txEnd'] <= row['end']), 'name2'\n",
    "    ]),\n",
    "    axis=1\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 123,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>#ISCA ID</th>\n",
       "      <th>ISCA Region Name</th>\n",
       "      <th>genes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>chr1</td>\n",
       "      <td>148867551</td>\n",
       "      <td>149768855</td>\n",
       "      <td>ISCA-37469</td>\n",
       "      <td>1q21.2 region (polymorphic region)</td>\n",
       "      <td>FAM72C,PPIAL4C,FCGR1A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>chr14</td>\n",
       "      <td>22111109</td>\n",
       "      <td>23021097</td>\n",
       "      <td>ISCA-37476</td>\n",
       "      <td>14q11.2 region (TCRA region)</td>\n",
       "      <td>OR4E2,OR4E1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>chr14</td>\n",
       "      <td>106050000</td>\n",
       "      <td>107289540</td>\n",
       "      <td>ISCA-37477</td>\n",
       "      <td>1 copy: 14q telomere; 3 copies: 14q telomere</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>chr15</td>\n",
       "      <td>22832519</td>\n",
       "      <td>23090897</td>\n",
       "      <td>ISCA-37448</td>\n",
       "      <td>15q11.2 recurrent region (BP1-BP2) (includes N...</td>\n",
       "      <td>TUBGCP5,CYFIP1,NIPA2,NIPA1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>chr15</td>\n",
       "      <td>32019621</td>\n",
       "      <td>32445405</td>\n",
       "      <td>ISCA-46295</td>\n",
       "      <td>15q13.3 recurrent region (D-CHRNA7 to BP5) (in...</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    chrom      start        end    #ISCA ID  \\\n",
       "29   chr1  148867551  149768855  ISCA-37469   \n",
       "25  chr14   22111109   23021097  ISCA-37476   \n",
       "24  chr14  106050000  107289540  ISCA-37477   \n",
       "33  chr15   22832519   23090897  ISCA-37448   \n",
       "5   chr15   32019621   32445405  ISCA-46295   \n",
       "\n",
       "                                     ISCA Region Name  \\\n",
       "29                 1q21.2 region (polymorphic region)   \n",
       "25                       14q11.2 region (TCRA region)   \n",
       "24       1 copy: 14q telomere; 3 copies: 14q telomere   \n",
       "33  15q11.2 recurrent region (BP1-BP2) (includes N...   \n",
       "5   15q13.3 recurrent region (D-CHRNA7 to BP5) (in...   \n",
       "\n",
       "                         genes  \n",
       "29       FAM72C,PPIAL4C,FCGR1A  \n",
       "25                 OR4E2,OR4E1  \n",
       "24                              \n",
       "33  TUBGCP5,CYFIP1,NIPA2,NIPA1  \n",
       "5                               "
      ]
     },
     "execution_count": 123,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "uts_region.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 124,
   "metadata": {},
   "outputs": [],
   "source": [
    "uts_region.rename(columns={\n",
    "    'chrom': '#chrom',\n",
    "    '#ISCA ID': 'isca_id', 'ISCA Region Name': 'name'\n",
    "}).to_csv('data/uts-region-20200109.sorted.bed', sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 125,
   "metadata": {},
   "outputs": [],
   "source": [
    "!bgzip -cf data/uts-region-20200109.sorted.bed > ../data/uts-region-20200109.sorted.bed.gz"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 126,
   "metadata": {},
   "outputs": [],
   "source": [
    "!tabix -fp bed ../data/uts-region-20200109.sorted.bed.gz"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 127,
   "metadata": {},
   "outputs": [],
   "source": [
    "dgv = pd.read_csv(\n",
    "    'data/DGV.GS.March2016.50percent.GainLossSep.Final.hg19.gff3', sep='\\t', names=['chrom', 'info'], usecols=[0, 8]\n",
    ").drop_duplicates('info')\n",
    "info = dgv['info'].str.extract(\n",
    "    r'ID=(?P<id>[^;]+).*variant_sub_type=(?P<type>[^;]+).*outer_start=(?P<start>[^;]+).*outer_end=(?P<end>[^;]+).*num_samples=(?P<sample>[^;]+).*Frequency=(?P<freq>\\S+?)%;'\n",
    ").astype({'start': int, 'end': int, 'sample': int, 'freq': float})\n",
    "dgv = dgv.merge(info, left_index=True, right_index=True)\n",
    "dgv = dgv[(dgv['freq'] >= 1) & (dgv['sample'] >= 10)].sort_values(['chrom', 'start', 'end'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 128,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>info</th>\n",
       "      <th>id</th>\n",
       "      <th>type</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>sample</th>\n",
       "      <th>freq</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>chr1</td>\n",
       "      <td>ID=gssvG19;Name=gssvG19;variant_type=CNV;varia...</td>\n",
       "      <td>gssvG19</td>\n",
       "      <td>Gain</td>\n",
       "      <td>10001</td>\n",
       "      <td>82189</td>\n",
       "      <td>13</td>\n",
       "      <td>19.40</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>chr1</td>\n",
       "      <td>ID=gssvG13;Name=gssvG13;variant_type=CNV;varia...</td>\n",
       "      <td>gssvG13</td>\n",
       "      <td>Gain</td>\n",
       "      <td>10001</td>\n",
       "      <td>127330</td>\n",
       "      <td>30</td>\n",
       "      <td>44.78</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>chr1</td>\n",
       "      <td>ID=gssvG9;Name=gssvG9;variant_type=CNV;variant...</td>\n",
       "      <td>gssvG9</td>\n",
       "      <td>Gain</td>\n",
       "      <td>10499</td>\n",
       "      <td>235938</td>\n",
       "      <td>13</td>\n",
       "      <td>1.13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>93</th>\n",
       "      <td>chr1</td>\n",
       "      <td>ID=gssvG53;Name=gssvG53;variant_type=CNV;varia...</td>\n",
       "      <td>gssvG53</td>\n",
       "      <td>Gain</td>\n",
       "      <td>21064</td>\n",
       "      <td>35983</td>\n",
       "      <td>12</td>\n",
       "      <td>30.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>90</th>\n",
       "      <td>chr1</td>\n",
       "      <td>ID=gssvG50;Name=gssvG50;variant_type=CNV;varia...</td>\n",
       "      <td>gssvG50</td>\n",
       "      <td>Gain</td>\n",
       "      <td>70007</td>\n",
       "      <td>88084</td>\n",
       "      <td>11</td>\n",
       "      <td>27.50</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   chrom                                               info       id  type  \\\n",
       "33  chr1  ID=gssvG19;Name=gssvG19;variant_type=CNV;varia...  gssvG19  Gain   \n",
       "24  chr1  ID=gssvG13;Name=gssvG13;variant_type=CNV;varia...  gssvG13  Gain   \n",
       "12  chr1  ID=gssvG9;Name=gssvG9;variant_type=CNV;variant...   gssvG9  Gain   \n",
       "93  chr1  ID=gssvG53;Name=gssvG53;variant_type=CNV;varia...  gssvG53  Gain   \n",
       "90  chr1  ID=gssvG50;Name=gssvG50;variant_type=CNV;varia...  gssvG50  Gain   \n",
       "\n",
       "    start     end  sample   freq  \n",
       "33  10001   82189      13  19.40  \n",
       "24  10001  127330      30  44.78  \n",
       "12  10499  235938      13   1.13  \n",
       "93  21064   35983      12  30.00  \n",
       "90  70007   88084      11  27.50  "
      ]
     },
     "execution_count": 128,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dgv.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 129,
   "metadata": {},
   "outputs": [],
   "source": [
    "def fetch_gene(gene, chrom, start, end):\n",
    "    return ','.join(gene.loc[\n",
    "        (gene['chrom'] == chrom) & (gene['txStart'] >= start) & (gene['txEnd'] <= end), 'name2'\n",
    "    ])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 130,
   "metadata": {},
   "outputs": [],
   "source": [
    "with Pool(processes=7) as pool:\n",
    "    dgv['genes'] = pool.starmap(fetch_gene, (\n",
    "        (gene, row['chrom'], row['start'], row['end']) for _, row in dgv.iterrows()\n",
    "    ), chunksize=70)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 131,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>info</th>\n",
       "      <th>id</th>\n",
       "      <th>type</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>sample</th>\n",
       "      <th>freq</th>\n",
       "      <th>genes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>chr1</td>\n",
       "      <td>ID=gssvG19;Name=gssvG19;variant_type=CNV;varia...</td>\n",
       "      <td>gssvG19</td>\n",
       "      <td>Gain</td>\n",
       "      <td>10001</td>\n",
       "      <td>82189</td>\n",
       "      <td>13</td>\n",
       "      <td>19.40</td>\n",
       "      <td>OR4F5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>chr1</td>\n",
       "      <td>ID=gssvG13;Name=gssvG13;variant_type=CNV;varia...</td>\n",
       "      <td>gssvG13</td>\n",
       "      <td>Gain</td>\n",
       "      <td>10001</td>\n",
       "      <td>127330</td>\n",
       "      <td>30</td>\n",
       "      <td>44.78</td>\n",
       "      <td>OR4F5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>chr1</td>\n",
       "      <td>ID=gssvG9;Name=gssvG9;variant_type=CNV;variant...</td>\n",
       "      <td>gssvG9</td>\n",
       "      <td>Gain</td>\n",
       "      <td>10499</td>\n",
       "      <td>235938</td>\n",
       "      <td>13</td>\n",
       "      <td>1.13</td>\n",
       "      <td>OR4F5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>93</th>\n",
       "      <td>chr1</td>\n",
       "      <td>ID=gssvG53;Name=gssvG53;variant_type=CNV;varia...</td>\n",
       "      <td>gssvG53</td>\n",
       "      <td>Gain</td>\n",
       "      <td>21064</td>\n",
       "      <td>35983</td>\n",
       "      <td>12</td>\n",
       "      <td>30.00</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>90</th>\n",
       "      <td>chr1</td>\n",
       "      <td>ID=gssvG50;Name=gssvG50;variant_type=CNV;varia...</td>\n",
       "      <td>gssvG50</td>\n",
       "      <td>Gain</td>\n",
       "      <td>70007</td>\n",
       "      <td>88084</td>\n",
       "      <td>11</td>\n",
       "      <td>27.50</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   chrom                                               info       id  type  \\\n",
       "33  chr1  ID=gssvG19;Name=gssvG19;variant_type=CNV;varia...  gssvG19  Gain   \n",
       "24  chr1  ID=gssvG13;Name=gssvG13;variant_type=CNV;varia...  gssvG13  Gain   \n",
       "12  chr1  ID=gssvG9;Name=gssvG9;variant_type=CNV;variant...   gssvG9  Gain   \n",
       "93  chr1  ID=gssvG53;Name=gssvG53;variant_type=CNV;varia...  gssvG53  Gain   \n",
       "90  chr1  ID=gssvG50;Name=gssvG50;variant_type=CNV;varia...  gssvG50  Gain   \n",
       "\n",
       "    start     end  sample   freq  genes  \n",
       "33  10001   82189      13  19.40  OR4F5  \n",
       "24  10001  127330      30  44.78  OR4F5  \n",
       "12  10499  235938      13   1.13  OR4F5  \n",
       "93  21064   35983      12  30.00         \n",
       "90  70007   88084      11  27.50         "
      ]
     },
     "execution_count": 131,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dgv.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 132,
   "metadata": {},
   "outputs": [],
   "source": [
    "dgv.loc[dgv['type'] == 'Gain', ['chrom', 'start', 'end', 'id', 'genes']\n",
    "].rename(columns={'chrom': '#chrom'}).to_csv('data/dgv-gain.sorted.bed', sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 133,
   "metadata": {},
   "outputs": [],
   "source": [
    "!bgzip -cf data/dgv-gain.sorted.bed > ../data/dgv-gain.sorted.bed.gz"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 134,
   "metadata": {},
   "outputs": [],
   "source": [
    "!tabix -fp bed ../data/dgv-gain.sorted.bed.gz"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 135,
   "metadata": {},
   "outputs": [],
   "source": [
    "dgv.loc[dgv['type'] == 'Loss', ['chrom', 'start', 'end', 'id', 'genes']\n",
    "].rename(columns={'chrom': '#chrom'}).to_csv('data/dgv-loss.sorted.bed', sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 136,
   "metadata": {},
   "outputs": [],
   "source": [
    "!bgzip -cf data/dgv-loss.sorted.bed > ../data/dgv-loss.sorted.bed.gz"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 137,
   "metadata": {},
   "outputs": [],
   "source": [
    "!tabix -fp bed ../data/dgv-loss.sorted.bed.gz"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 138,
   "metadata": {},
   "outputs": [],
   "source": [
    "gnomad = pd.read_csv(\n",
    "    'data/gnomad_v2.1_sv.controls_only.sites.bed.gz', sep='\\t', usecols=[0, 1, 2, 3, 4, 241], dtype=str\n",
    ")\n",
    "gnomad = gnomad[(gnomad['FILTER'] == 'PASS') & gnomad['svtype'].isin(['DEL', 'DUP'])]\n",
    "gnomad['#chrom'] = 'chr' + gnomad['#chrom']\n",
    "gnomad['start'] = gnomad['start'].astype(int)\n",
    "gnomad['end'] = gnomad['end'].astype(int)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 139,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>#chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>name</th>\n",
       "      <th>svtype</th>\n",
       "      <th>FILTER</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>chr1</td>\n",
       "      <td>62398</td>\n",
       "      <td>62489</td>\n",
       "      <td>gnomAD-SV_v2.1_DEL_1_3</td>\n",
       "      <td>DEL</td>\n",
       "      <td>PASS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>chr1</td>\n",
       "      <td>66349</td>\n",
       "      <td>66427</td>\n",
       "      <td>gnomAD-SV_v2.1_DEL_1_4</td>\n",
       "      <td>DEL</td>\n",
       "      <td>PASS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>chr1</td>\n",
       "      <td>136999</td>\n",
       "      <td>155500</td>\n",
       "      <td>gnomAD-SV_v2.1_DUP_1_3</td>\n",
       "      <td>DUP</td>\n",
       "      <td>PASS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>chr1</td>\n",
       "      <td>141474</td>\n",
       "      <td>155000</td>\n",
       "      <td>gnomAD-SV_v2.1_DUP_1_4</td>\n",
       "      <td>DUP</td>\n",
       "      <td>PASS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>chr1</td>\n",
       "      <td>319999</td>\n",
       "      <td>327650</td>\n",
       "      <td>gnomAD-SV_v2.1_DEL_1_9</td>\n",
       "      <td>DEL</td>\n",
       "      <td>PASS</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   #chrom   start     end                    name svtype FILTER\n",
       "6    chr1   62398   62489  gnomAD-SV_v2.1_DEL_1_3    DEL   PASS\n",
       "7    chr1   66349   66427  gnomAD-SV_v2.1_DEL_1_4    DEL   PASS\n",
       "15   chr1  136999  155500  gnomAD-SV_v2.1_DUP_1_3    DUP   PASS\n",
       "16   chr1  141474  155000  gnomAD-SV_v2.1_DUP_1_4    DUP   PASS\n",
       "24   chr1  319999  327650  gnomAD-SV_v2.1_DEL_1_9    DEL   PASS"
      ]
     },
     "execution_count": 139,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gnomad.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 140,
   "metadata": {},
   "outputs": [],
   "source": [
    "with Pool(processes=7) as pool:\n",
    "    gnomad['genes'] = pool.starmap(fetch_gene, (\n",
    "        (gene, row['#chrom'], row['start'], row['end']) for _, row in gnomad.iterrows()\n",
    "    ), chunksize=70)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 141,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>#chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>name</th>\n",
       "      <th>svtype</th>\n",
       "      <th>FILTER</th>\n",
       "      <th>genes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>chr1</td>\n",
       "      <td>62398</td>\n",
       "      <td>62489</td>\n",
       "      <td>gnomAD-SV_v2.1_DEL_1_3</td>\n",
       "      <td>DEL</td>\n",
       "      <td>PASS</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>chr1</td>\n",
       "      <td>66349</td>\n",
       "      <td>66427</td>\n",
       "      <td>gnomAD-SV_v2.1_DEL_1_4</td>\n",
       "      <td>DEL</td>\n",
       "      <td>PASS</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>chr1</td>\n",
       "      <td>136999</td>\n",
       "      <td>155500</td>\n",
       "      <td>gnomAD-SV_v2.1_DUP_1_3</td>\n",
       "      <td>DUP</td>\n",
       "      <td>PASS</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>chr1</td>\n",
       "      <td>141474</td>\n",
       "      <td>155000</td>\n",
       "      <td>gnomAD-SV_v2.1_DUP_1_4</td>\n",
       "      <td>DUP</td>\n",
       "      <td>PASS</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>chr1</td>\n",
       "      <td>319999</td>\n",
       "      <td>327650</td>\n",
       "      <td>gnomAD-SV_v2.1_DEL_1_9</td>\n",
       "      <td>DEL</td>\n",
       "      <td>PASS</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   #chrom   start     end                    name svtype FILTER genes\n",
       "6    chr1   62398   62489  gnomAD-SV_v2.1_DEL_1_3    DEL   PASS      \n",
       "7    chr1   66349   66427  gnomAD-SV_v2.1_DEL_1_4    DEL   PASS      \n",
       "15   chr1  136999  155500  gnomAD-SV_v2.1_DUP_1_3    DUP   PASS      \n",
       "16   chr1  141474  155000  gnomAD-SV_v2.1_DUP_1_4    DUP   PASS      \n",
       "24   chr1  319999  327650  gnomAD-SV_v2.1_DEL_1_9    DEL   PASS      "
      ]
     },
     "execution_count": 141,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gnomad.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 142,
   "metadata": {},
   "outputs": [],
   "source": [
    "gnomad.loc[gnomad['svtype'] == 'DEL', ['#chrom', 'start', 'end', 'name', 'genes']\n",
    "].to_csv('data/gnomad-del.sorted.bed', sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 143,
   "metadata": {},
   "outputs": [],
   "source": [
    "!bgzip -cf data/gnomad-del.sorted.bed > ../data/gnomad-del.sorted.bed.gz"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 144,
   "metadata": {},
   "outputs": [],
   "source": [
    "!tabix -fp bed ../data/gnomad-del.sorted.bed.gz"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 145,
   "metadata": {},
   "outputs": [],
   "source": [
    "gnomad.loc[gnomad['svtype'] == 'DUP', ['#chrom', 'start', 'end', 'name', 'genes']\n",
    "].to_csv('data/gnomad-dup.sorted.bed', sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 146,
   "metadata": {},
   "outputs": [],
   "source": [
    "!bgzip -cf data/gnomad-dup.sorted.bed > ../data/gnomad-dup.sorted.bed.gz"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 147,
   "metadata": {},
   "outputs": [],
   "source": [
    "!tabix -fp bed ../data/gnomad-dup.sorted.bed.gz"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "jupyter",
   "language": "python",
   "name": "jupyter"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
