{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import allel\n",
    "import json"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "pd.set_option('display.max_columns', None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "refgene = pd.read_csv('data/refseq-20200109.txt', sep='\\t')\n",
    "refgene['length'] = refgene['txEnd'] - refgene['txStart']\n",
    "refgene = refgene.sort_values('length')\n",
    "refgene = refgene.drop_duplicates('name2', keep='last')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>#bin</th>\n",
       "      <th>name</th>\n",
       "      <th>chrom</th>\n",
       "      <th>strand</th>\n",
       "      <th>txStart</th>\n",
       "      <th>txEnd</th>\n",
       "      <th>cdsStart</th>\n",
       "      <th>cdsEnd</th>\n",
       "      <th>exonCount</th>\n",
       "      <th>exonStarts</th>\n",
       "      <th>exonEnds</th>\n",
       "      <th>score</th>\n",
       "      <th>name2</th>\n",
       "      <th>cdsStartStat</th>\n",
       "      <th>cdsEndStat</th>\n",
       "      <th>exonFrames</th>\n",
       "      <th>length</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>43333</th>\n",
       "      <td>779</td>\n",
       "      <td>NR_003499.1</td>\n",
       "      <td>chr15</td>\n",
       "      <td>+</td>\n",
       "      <td>25513663</td>\n",
       "      <td>25513696</td>\n",
       "      <td>25513696</td>\n",
       "      <td>25513696</td>\n",
       "      <td>1</td>\n",
       "      <td>25513663,</td>\n",
       "      <td>25513696,</td>\n",
       "      <td>0</td>\n",
       "      <td>SNORD115-47</td>\n",
       "      <td>none</td>\n",
       "      <td>none</td>\n",
       "      <td>-1,</td>\n",
       "      <td>33</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3416</th>\n",
       "      <td>1311</td>\n",
       "      <td>NR_039620.1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>-</td>\n",
       "      <td>95211415</td>\n",
       "      <td>95211456</td>\n",
       "      <td>95211456</td>\n",
       "      <td>95211456</td>\n",
       "      <td>1</td>\n",
       "      <td>95211415,</td>\n",
       "      <td>95211456,</td>\n",
       "      <td>0</td>\n",
       "      <td>MIR378G</td>\n",
       "      <td>none</td>\n",
       "      <td>none</td>\n",
       "      <td>-1,</td>\n",
       "      <td>41</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16216</th>\n",
       "      <td>589</td>\n",
       "      <td>NR_039661.1</td>\n",
       "      <td>chr5</td>\n",
       "      <td>-</td>\n",
       "      <td>535954</td>\n",
       "      <td>535997</td>\n",
       "      <td>535997</td>\n",
       "      <td>535997</td>\n",
       "      <td>1</td>\n",
       "      <td>535954,</td>\n",
       "      <td>535997,</td>\n",
       "      <td>0</td>\n",
       "      <td>MIR4456</td>\n",
       "      <td>none</td>\n",
       "      <td>none</td>\n",
       "      <td>-1,</td>\n",
       "      <td>43</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4887</th>\n",
       "      <td>1911</td>\n",
       "      <td>NR_003938.1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>-</td>\n",
       "      <td>173833312</td>\n",
       "      <td>173833355</td>\n",
       "      <td>173833355</td>\n",
       "      <td>173833355</td>\n",
       "      <td>1</td>\n",
       "      <td>173833312,</td>\n",
       "      <td>173833355,</td>\n",
       "      <td>0</td>\n",
       "      <td>SNORD81</td>\n",
       "      <td>none</td>\n",
       "      <td>none</td>\n",
       "      <td>-1,</td>\n",
       "      <td>43</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14863</th>\n",
       "      <td>1479</td>\n",
       "      <td>NR_031737.1</td>\n",
       "      <td>chr4</td>\n",
       "      <td>+</td>\n",
       "      <td>117220880</td>\n",
       "      <td>117220924</td>\n",
       "      <td>117220924</td>\n",
       "      <td>117220924</td>\n",
       "      <td>1</td>\n",
       "      <td>117220880,</td>\n",
       "      <td>117220924,</td>\n",
       "      <td>0</td>\n",
       "      <td>MIR1973</td>\n",
       "      <td>none</td>\n",
       "      <td>none</td>\n",
       "      <td>-1,</td>\n",
       "      <td>44</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       #bin         name  chrom strand    txStart      txEnd   cdsStart  \\\n",
       "43333   779  NR_003499.1  chr15      +   25513663   25513696   25513696   \n",
       "3416   1311  NR_039620.1   chr1      -   95211415   95211456   95211456   \n",
       "16216   589  NR_039661.1   chr5      -     535954     535997     535997   \n",
       "4887   1911  NR_003938.1   chr1      -  173833312  173833355  173833355   \n",
       "14863  1479  NR_031737.1   chr4      +  117220880  117220924  117220924   \n",
       "\n",
       "          cdsEnd  exonCount  exonStarts    exonEnds  score        name2  \\\n",
       "43333   25513696          1   25513663,   25513696,      0  SNORD115-47   \n",
       "3416    95211456          1   95211415,   95211456,      0      MIR378G   \n",
       "16216     535997          1     535954,     535997,      0      MIR4456   \n",
       "4887   173833355          1  173833312,  173833355,      0      SNORD81   \n",
       "14863  117220924          1  117220880,  117220924,      0      MIR1973   \n",
       "\n",
       "      cdsStartStat cdsEndStat exonFrames  length  \n",
       "43333         none       none        -1,      33  \n",
       "3416          none       none        -1,      41  \n",
       "16216         none       none        -1,      43  \n",
       "4887          none       none        -1,      43  \n",
       "14863         none       none        -1,      44  "
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "refgene.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_info = pd.read_csv('data/Homo_sapiens.gene_info.gz', sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>#tax_id</th>\n",
       "      <th>GeneID</th>\n",
       "      <th>Symbol</th>\n",
       "      <th>LocusTag</th>\n",
       "      <th>Synonyms</th>\n",
       "      <th>dbXrefs</th>\n",
       "      <th>chromosome</th>\n",
       "      <th>map_location</th>\n",
       "      <th>description</th>\n",
       "      <th>type_of_gene</th>\n",
       "      <th>Symbol_from_nomenclature_authority</th>\n",
       "      <th>Full_name_from_nomenclature_authority</th>\n",
       "      <th>Nomenclature_status</th>\n",
       "      <th>Other_designations</th>\n",
       "      <th>Modification_date</th>\n",
       "      <th>Feature_type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>9606</td>\n",
       "      <td>1</td>\n",
       "      <td>A1BG</td>\n",
       "      <td>-</td>\n",
       "      <td>A1B|ABG|GAB|HYST2477</td>\n",
       "      <td>MIM:138670|HGNC:HGNC:5|Ensembl:ENSG00000121410</td>\n",
       "      <td>19</td>\n",
       "      <td>19q13.43</td>\n",
       "      <td>alpha-1-B glycoprotein</td>\n",
       "      <td>protein-coding</td>\n",
       "      <td>A1BG</td>\n",
       "      <td>alpha-1-B glycoprotein</td>\n",
       "      <td>O</td>\n",
       "      <td>alpha-1B-glycoprotein|HEL-S-163pA|epididymis s...</td>\n",
       "      <td>20191220</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>9606</td>\n",
       "      <td>2</td>\n",
       "      <td>A2M</td>\n",
       "      <td>-</td>\n",
       "      <td>A2MD|CPAMD5|FWP007|S863-7</td>\n",
       "      <td>MIM:103950|HGNC:HGNC:7|Ensembl:ENSG00000175899</td>\n",
       "      <td>12</td>\n",
       "      <td>12p13.31</td>\n",
       "      <td>alpha-2-macroglobulin</td>\n",
       "      <td>protein-coding</td>\n",
       "      <td>A2M</td>\n",
       "      <td>alpha-2-macroglobulin</td>\n",
       "      <td>O</td>\n",
       "      <td>alpha-2-macroglobulin|C3 and PZP-like alpha-2-...</td>\n",
       "      <td>20200113</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>9606</td>\n",
       "      <td>3</td>\n",
       "      <td>A2MP1</td>\n",
       "      <td>-</td>\n",
       "      <td>A2MP</td>\n",
       "      <td>HGNC:HGNC:8|Ensembl:ENSG00000256069</td>\n",
       "      <td>12</td>\n",
       "      <td>12p13.31</td>\n",
       "      <td>alpha-2-macroglobulin pseudogene 1</td>\n",
       "      <td>pseudo</td>\n",
       "      <td>A2MP1</td>\n",
       "      <td>alpha-2-macroglobulin pseudogene 1</td>\n",
       "      <td>O</td>\n",
       "      <td>pregnancy-zone protein pseudogene</td>\n",
       "      <td>20191221</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>9606</td>\n",
       "      <td>9</td>\n",
       "      <td>NAT1</td>\n",
       "      <td>-</td>\n",
       "      <td>AAC1|MNAT|NAT-1|NATI</td>\n",
       "      <td>MIM:108345|HGNC:HGNC:7645|Ensembl:ENSG00000171428</td>\n",
       "      <td>8</td>\n",
       "      <td>8p22</td>\n",
       "      <td>N-acetyltransferase 1</td>\n",
       "      <td>protein-coding</td>\n",
       "      <td>NAT1</td>\n",
       "      <td>N-acetyltransferase 1</td>\n",
       "      <td>O</td>\n",
       "      <td>arylamine N-acetyltransferase 1|N-acetyltransf...</td>\n",
       "      <td>20200209</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>9606</td>\n",
       "      <td>10</td>\n",
       "      <td>NAT2</td>\n",
       "      <td>-</td>\n",
       "      <td>AAC2|NAT-2|PNAT</td>\n",
       "      <td>MIM:612182|HGNC:HGNC:7646|Ensembl:ENSG00000156006</td>\n",
       "      <td>8</td>\n",
       "      <td>8p22</td>\n",
       "      <td>N-acetyltransferase 2</td>\n",
       "      <td>protein-coding</td>\n",
       "      <td>NAT2</td>\n",
       "      <td>N-acetyltransferase 2</td>\n",
       "      <td>O</td>\n",
       "      <td>arylamine N-acetyltransferase 2|N-acetyltransf...</td>\n",
       "      <td>20200121</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   #tax_id  GeneID Symbol LocusTag                   Synonyms  \\\n",
       "0     9606       1   A1BG        -       A1B|ABG|GAB|HYST2477   \n",
       "1     9606       2    A2M        -  A2MD|CPAMD5|FWP007|S863-7   \n",
       "2     9606       3  A2MP1        -                       A2MP   \n",
       "3     9606       9   NAT1        -       AAC1|MNAT|NAT-1|NATI   \n",
       "4     9606      10   NAT2        -            AAC2|NAT-2|PNAT   \n",
       "\n",
       "                                             dbXrefs chromosome map_location  \\\n",
       "0     MIM:138670|HGNC:HGNC:5|Ensembl:ENSG00000121410         19     19q13.43   \n",
       "1     MIM:103950|HGNC:HGNC:7|Ensembl:ENSG00000175899         12     12p13.31   \n",
       "2                HGNC:HGNC:8|Ensembl:ENSG00000256069         12     12p13.31   \n",
       "3  MIM:108345|HGNC:HGNC:7645|Ensembl:ENSG00000171428          8         8p22   \n",
       "4  MIM:612182|HGNC:HGNC:7646|Ensembl:ENSG00000156006          8         8p22   \n",
       "\n",
       "                          description    type_of_gene  \\\n",
       "0              alpha-1-B glycoprotein  protein-coding   \n",
       "1               alpha-2-macroglobulin  protein-coding   \n",
       "2  alpha-2-macroglobulin pseudogene 1          pseudo   \n",
       "3               N-acetyltransferase 1  protein-coding   \n",
       "4               N-acetyltransferase 2  protein-coding   \n",
       "\n",
       "  Symbol_from_nomenclature_authority Full_name_from_nomenclature_authority  \\\n",
       "0                               A1BG                alpha-1-B glycoprotein   \n",
       "1                                A2M                 alpha-2-macroglobulin   \n",
       "2                              A2MP1    alpha-2-macroglobulin pseudogene 1   \n",
       "3                               NAT1                 N-acetyltransferase 1   \n",
       "4                               NAT2                 N-acetyltransferase 2   \n",
       "\n",
       "  Nomenclature_status                                 Other_designations  \\\n",
       "0                   O  alpha-1B-glycoprotein|HEL-S-163pA|epididymis s...   \n",
       "1                   O  alpha-2-macroglobulin|C3 and PZP-like alpha-2-...   \n",
       "2                   O                  pregnancy-zone protein pseudogene   \n",
       "3                   O  arylamine N-acetyltransferase 1|N-acetyltransf...   \n",
       "4                   O  arylamine N-acetyltransferase 2|N-acetyltransf...   \n",
       "\n",
       "   Modification_date Feature_type  \n",
       "0           20191220            -  \n",
       "1           20200113            -  \n",
       "2           20191221            -  \n",
       "3           20200209            -  \n",
       "4           20200121            -  "
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_info.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "refgene_info = refgene.merge(gene_info, left_on='name2', right_on='Symbol')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>#bin</th>\n",
       "      <th>name</th>\n",
       "      <th>chrom</th>\n",
       "      <th>strand</th>\n",
       "      <th>txStart</th>\n",
       "      <th>txEnd</th>\n",
       "      <th>cdsStart</th>\n",
       "      <th>cdsEnd</th>\n",
       "      <th>exonCount</th>\n",
       "      <th>exonStarts</th>\n",
       "      <th>exonEnds</th>\n",
       "      <th>score</th>\n",
       "      <th>name2</th>\n",
       "      <th>cdsStartStat</th>\n",
       "      <th>cdsEndStat</th>\n",
       "      <th>exonFrames</th>\n",
       "      <th>length</th>\n",
       "      <th>#tax_id</th>\n",
       "      <th>GeneID</th>\n",
       "      <th>Symbol</th>\n",
       "      <th>LocusTag</th>\n",
       "      <th>Synonyms</th>\n",
       "      <th>dbXrefs</th>\n",
       "      <th>chromosome</th>\n",
       "      <th>map_location</th>\n",
       "      <th>description</th>\n",
       "      <th>type_of_gene</th>\n",
       "      <th>Symbol_from_nomenclature_authority</th>\n",
       "      <th>Full_name_from_nomenclature_authority</th>\n",
       "      <th>Nomenclature_status</th>\n",
       "      <th>Other_designations</th>\n",
       "      <th>Modification_date</th>\n",
       "      <th>Feature_type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>779</td>\n",
       "      <td>NR_003499.1</td>\n",
       "      <td>chr15</td>\n",
       "      <td>+</td>\n",
       "      <td>25513663</td>\n",
       "      <td>25513696</td>\n",
       "      <td>25513696</td>\n",
       "      <td>25513696</td>\n",
       "      <td>1</td>\n",
       "      <td>25513663,</td>\n",
       "      <td>25513696,</td>\n",
       "      <td>0</td>\n",
       "      <td>SNORD115-47</td>\n",
       "      <td>none</td>\n",
       "      <td>none</td>\n",
       "      <td>-1,</td>\n",
       "      <td>33</td>\n",
       "      <td>9606</td>\n",
       "      <td>100036567</td>\n",
       "      <td>SNORD115-47</td>\n",
       "      <td>-</td>\n",
       "      <td>HBII-52-46|HBII-52-47</td>\n",
       "      <td>HGNC:HGNC:33065</td>\n",
       "      <td>15</td>\n",
       "      <td>15q11.2</td>\n",
       "      <td>small nucleolar RNA, C/D box 115-47</td>\n",
       "      <td>snoRNA</td>\n",
       "      <td>SNORD115-47</td>\n",
       "      <td>small nucleolar RNA, C/D box 115-47</td>\n",
       "      <td>O</td>\n",
       "      <td>HBII-52-46 snoRNA</td>\n",
       "      <td>20191221</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1311</td>\n",
       "      <td>NR_039620.1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>-</td>\n",
       "      <td>95211415</td>\n",
       "      <td>95211456</td>\n",
       "      <td>95211456</td>\n",
       "      <td>95211456</td>\n",
       "      <td>1</td>\n",
       "      <td>95211415,</td>\n",
       "      <td>95211456,</td>\n",
       "      <td>0</td>\n",
       "      <td>MIR378G</td>\n",
       "      <td>none</td>\n",
       "      <td>none</td>\n",
       "      <td>-1,</td>\n",
       "      <td>41</td>\n",
       "      <td>9606</td>\n",
       "      <td>100616321</td>\n",
       "      <td>MIR378G</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>HGNC:HGNC:41785|Ensembl:ENSG00000263526|miRBas...</td>\n",
       "      <td>1</td>\n",
       "      <td>1p21.3</td>\n",
       "      <td>microRNA 378g</td>\n",
       "      <td>ncRNA</td>\n",
       "      <td>MIR378G</td>\n",
       "      <td>microRNA 378g</td>\n",
       "      <td>O</td>\n",
       "      <td>hsa-mir-378g</td>\n",
       "      <td>20191221</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>589</td>\n",
       "      <td>NR_039661.1</td>\n",
       "      <td>chr5</td>\n",
       "      <td>-</td>\n",
       "      <td>535954</td>\n",
       "      <td>535997</td>\n",
       "      <td>535997</td>\n",
       "      <td>535997</td>\n",
       "      <td>1</td>\n",
       "      <td>535954,</td>\n",
       "      <td>535997,</td>\n",
       "      <td>0</td>\n",
       "      <td>MIR4456</td>\n",
       "      <td>none</td>\n",
       "      <td>none</td>\n",
       "      <td>-1,</td>\n",
       "      <td>43</td>\n",
       "      <td>9606</td>\n",
       "      <td>100616381</td>\n",
       "      <td>MIR4456</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>HGNC:HGNC:41606|Ensembl:ENSG00000264233|miRBas...</td>\n",
       "      <td>5</td>\n",
       "      <td>5p15.33</td>\n",
       "      <td>microRNA 4456</td>\n",
       "      <td>ncRNA</td>\n",
       "      <td>MIR4456</td>\n",
       "      <td>microRNA 4456</td>\n",
       "      <td>O</td>\n",
       "      <td>hsa-mir-4456</td>\n",
       "      <td>20191221</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1911</td>\n",
       "      <td>NR_003938.1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>-</td>\n",
       "      <td>173833312</td>\n",
       "      <td>173833355</td>\n",
       "      <td>173833355</td>\n",
       "      <td>173833355</td>\n",
       "      <td>1</td>\n",
       "      <td>173833312,</td>\n",
       "      <td>173833355,</td>\n",
       "      <td>0</td>\n",
       "      <td>SNORD81</td>\n",
       "      <td>none</td>\n",
       "      <td>none</td>\n",
       "      <td>-1,</td>\n",
       "      <td>43</td>\n",
       "      <td>9606</td>\n",
       "      <td>26769</td>\n",
       "      <td>SNORD81</td>\n",
       "      <td>-</td>\n",
       "      <td>RNU104|U81|Z23</td>\n",
       "      <td>HGNC:HGNC:10101</td>\n",
       "      <td>1</td>\n",
       "      <td>1q25.1</td>\n",
       "      <td>small nucleolar RNA, C/D box 81</td>\n",
       "      <td>snoRNA</td>\n",
       "      <td>SNORD81</td>\n",
       "      <td>small nucleolar RNA, C/D box 81</td>\n",
       "      <td>O</td>\n",
       "      <td>RNA, U104 small nucleolar</td>\n",
       "      <td>20191221</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1479</td>\n",
       "      <td>NR_031737.1</td>\n",
       "      <td>chr4</td>\n",
       "      <td>+</td>\n",
       "      <td>117220880</td>\n",
       "      <td>117220924</td>\n",
       "      <td>117220924</td>\n",
       "      <td>117220924</td>\n",
       "      <td>1</td>\n",
       "      <td>117220880,</td>\n",
       "      <td>117220924,</td>\n",
       "      <td>0</td>\n",
       "      <td>MIR1973</td>\n",
       "      <td>none</td>\n",
       "      <td>none</td>\n",
       "      <td>-1,</td>\n",
       "      <td>44</td>\n",
       "      <td>9606</td>\n",
       "      <td>100302290</td>\n",
       "      <td>MIR1973</td>\n",
       "      <td>-</td>\n",
       "      <td>hsa-mir-1973|mir-1973</td>\n",
       "      <td>HGNC:HGNC:37061|Ensembl:ENSG00000284253|miRBas...</td>\n",
       "      <td>4</td>\n",
       "      <td>4q26</td>\n",
       "      <td>microRNA 1973</td>\n",
       "      <td>ncRNA</td>\n",
       "      <td>MIR1973</td>\n",
       "      <td>microRNA 1973</td>\n",
       "      <td>O</td>\n",
       "      <td>-</td>\n",
       "      <td>20191221</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   #bin         name  chrom strand    txStart      txEnd   cdsStart  \\\n",
       "0   779  NR_003499.1  chr15      +   25513663   25513696   25513696   \n",
       "1  1311  NR_039620.1   chr1      -   95211415   95211456   95211456   \n",
       "2   589  NR_039661.1   chr5      -     535954     535997     535997   \n",
       "3  1911  NR_003938.1   chr1      -  173833312  173833355  173833355   \n",
       "4  1479  NR_031737.1   chr4      +  117220880  117220924  117220924   \n",
       "\n",
       "      cdsEnd  exonCount  exonStarts    exonEnds  score        name2  \\\n",
       "0   25513696          1   25513663,   25513696,      0  SNORD115-47   \n",
       "1   95211456          1   95211415,   95211456,      0      MIR378G   \n",
       "2     535997          1     535954,     535997,      0      MIR4456   \n",
       "3  173833355          1  173833312,  173833355,      0      SNORD81   \n",
       "4  117220924          1  117220880,  117220924,      0      MIR1973   \n",
       "\n",
       "  cdsStartStat cdsEndStat exonFrames  length  #tax_id     GeneID       Symbol  \\\n",
       "0         none       none        -1,      33     9606  100036567  SNORD115-47   \n",
       "1         none       none        -1,      41     9606  100616321      MIR378G   \n",
       "2         none       none        -1,      43     9606  100616381      MIR4456   \n",
       "3         none       none        -1,      43     9606      26769      SNORD81   \n",
       "4         none       none        -1,      44     9606  100302290      MIR1973   \n",
       "\n",
       "  LocusTag               Synonyms  \\\n",
       "0        -  HBII-52-46|HBII-52-47   \n",
       "1        -                      -   \n",
       "2        -                      -   \n",
       "3        -         RNU104|U81|Z23   \n",
       "4        -  hsa-mir-1973|mir-1973   \n",
       "\n",
       "                                             dbXrefs chromosome map_location  \\\n",
       "0                                    HGNC:HGNC:33065         15      15q11.2   \n",
       "1  HGNC:HGNC:41785|Ensembl:ENSG00000263526|miRBas...          1       1p21.3   \n",
       "2  HGNC:HGNC:41606|Ensembl:ENSG00000264233|miRBas...          5      5p15.33   \n",
       "3                                    HGNC:HGNC:10101          1       1q25.1   \n",
       "4  HGNC:HGNC:37061|Ensembl:ENSG00000284253|miRBas...          4         4q26   \n",
       "\n",
       "                           description type_of_gene  \\\n",
       "0  small nucleolar RNA, C/D box 115-47       snoRNA   \n",
       "1                        microRNA 378g        ncRNA   \n",
       "2                        microRNA 4456        ncRNA   \n",
       "3      small nucleolar RNA, C/D box 81       snoRNA   \n",
       "4                        microRNA 1973        ncRNA   \n",
       "\n",
       "  Symbol_from_nomenclature_authority Full_name_from_nomenclature_authority  \\\n",
       "0                        SNORD115-47   small nucleolar RNA, C/D box 115-47   \n",
       "1                            MIR378G                         microRNA 378g   \n",
       "2                            MIR4456                         microRNA 4456   \n",
       "3                            SNORD81       small nucleolar RNA, C/D box 81   \n",
       "4                            MIR1973                         microRNA 1973   \n",
       "\n",
       "  Nomenclature_status         Other_designations  Modification_date  \\\n",
       "0                   O          HBII-52-46 snoRNA           20191221   \n",
       "1                   O               hsa-mir-378g           20191221   \n",
       "2                   O               hsa-mir-4456           20191221   \n",
       "3                   O  RNA, U104 small nucleolar           20191221   \n",
       "4                   O                          -           20191221   \n",
       "\n",
       "  Feature_type  \n",
       "0            -  \n",
       "1            -  \n",
       "2            -  \n",
       "3            -  \n",
       "4            -  "
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "refgene_info.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "gene = refgene_info.loc[\n",
    "    refgene_info['type_of_gene'] == 'protein-coding',\n",
    "    ['chrom', 'txStart', 'txEnd', 'GeneID', 'name2', 'name', 'strand']\n",
    "].sort_values(['chrom', 'txStart', 'txEnd'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>txStart</th>\n",
       "      <th>txEnd</th>\n",
       "      <th>GeneID</th>\n",
       "      <th>name2</th>\n",
       "      <th>name</th>\n",
       "      <th>strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2719</th>\n",
       "      <td>chr1</td>\n",
       "      <td>69090</td>\n",
       "      <td>70008</td>\n",
       "      <td>79501</td>\n",
       "      <td>OR4F5</td>\n",
       "      <td>NM_001005484.1</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2832</th>\n",
       "      <td>chr1</td>\n",
       "      <td>367658</td>\n",
       "      <td>368597</td>\n",
       "      <td>729759</td>\n",
       "      <td>OR4F29</td>\n",
       "      <td>NM_001005221.2</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2855</th>\n",
       "      <td>chr1</td>\n",
       "      <td>621095</td>\n",
       "      <td>622034</td>\n",
       "      <td>81399</td>\n",
       "      <td>OR4F16</td>\n",
       "      <td>NM_001005277.1</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13665</th>\n",
       "      <td>chr1</td>\n",
       "      <td>861120</td>\n",
       "      <td>879961</td>\n",
       "      <td>148398</td>\n",
       "      <td>SAMD11</td>\n",
       "      <td>NM_152486.2</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12356</th>\n",
       "      <td>chr1</td>\n",
       "      <td>879582</td>\n",
       "      <td>894679</td>\n",
       "      <td>26155</td>\n",
       "      <td>NOC2L</td>\n",
       "      <td>NM_015658.3</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      chrom  txStart   txEnd  GeneID   name2            name strand\n",
       "2719   chr1    69090   70008   79501   OR4F5  NM_001005484.1      +\n",
       "2832   chr1   367658  368597  729759  OR4F29  NM_001005221.2      +\n",
       "2855   chr1   621095  622034   81399  OR4F16  NM_001005277.1      -\n",
       "13665  chr1   861120  879961  148398  SAMD11     NM_152486.2      +\n",
       "12356  chr1   879582  894679   26155   NOC2L     NM_015658.3      -"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "gene.rename(columns={\n",
    "    'chrom': '#chrom', 'txStart': 'start', 'txEnd': 'end',\n",
    "    'GeneID': 'gene_id', 'name2': 'symbol', 'name': 'transcript'\n",
    "}).to_csv('data/gene-20200109.sorted.bed', index=False, sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "!bgzip -cf data/gene-20200109.sorted.bed > ../data/gene-20200109.sorted.bed.gz"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "!tabix -fp bed ../data/gene-20200109.sorted.bed.gz"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "curation_gene = pd.read_csv('data/ClinGen_gene_curation_list_GRCh37.tsv', sep='\\t', dtype=str, skiprows=5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>#Gene Symbol</th>\n",
       "      <th>Gene ID</th>\n",
       "      <th>cytoBand</th>\n",
       "      <th>Genomic Location</th>\n",
       "      <th>Haploinsufficiency Score</th>\n",
       "      <th>Haploinsufficiency Description</th>\n",
       "      <th>Haploinsufficiency PMID1</th>\n",
       "      <th>Haploinsufficiency PMID2</th>\n",
       "      <th>Haploinsufficiency PMID3</th>\n",
       "      <th>Triplosensitivity Score</th>\n",
       "      <th>Triplosensitivity Description</th>\n",
       "      <th>Triplosensitivity PMID1</th>\n",
       "      <th>Triplosensitivity PMID2</th>\n",
       "      <th>Triplosensitivity PMID3</th>\n",
       "      <th>Date Last Evaluated</th>\n",
       "      <th>Loss phenotype OMIM ID</th>\n",
       "      <th>Triplosensitive phenotype OMIM ID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A4GALT</td>\n",
       "      <td>53947</td>\n",
       "      <td>22q13.2</td>\n",
       "      <td>chr22:43088121-43117307</td>\n",
       "      <td>30</td>\n",
       "      <td>Gene associated with autosomal recessive pheno...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>No evidence available</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2014-12-11</td>\n",
       "      <td>111400</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AAGAB</td>\n",
       "      <td>79719</td>\n",
       "      <td>15q23</td>\n",
       "      <td>chr15:67493013-67547536</td>\n",
       "      <td>3</td>\n",
       "      <td>Sufficient evidence for dosage pathogenicity</td>\n",
       "      <td>23064416</td>\n",
       "      <td>23000146</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>No evidence available</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2013-02-28</td>\n",
       "      <td>148600</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AARS</td>\n",
       "      <td>16</td>\n",
       "      <td>16q22.1</td>\n",
       "      <td>chr16:70286297-70323412</td>\n",
       "      <td>0</td>\n",
       "      <td>No evidence available</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>No evidence available</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2018-01-11</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AARS2</td>\n",
       "      <td>57505</td>\n",
       "      <td>6p21.1</td>\n",
       "      <td>chr6:44266463-44281063</td>\n",
       "      <td>30</td>\n",
       "      <td>Gene associated with autosomal recessive pheno...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Not yet evaluated</td>\n",
       "      <td>Not yet evaluated</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2016-08-22</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AASS</td>\n",
       "      <td>10157</td>\n",
       "      <td>7q31.32</td>\n",
       "      <td>chr7:121713598-121784344</td>\n",
       "      <td>30</td>\n",
       "      <td>Gene associated with autosomal recessive pheno...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Not yet evaluated</td>\n",
       "      <td>Not yet evaluated</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2016-08-22</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  #Gene Symbol Gene ID cytoBand          Genomic Location  \\\n",
       "0       A4GALT   53947  22q13.2   chr22:43088121-43117307   \n",
       "1        AAGAB   79719    15q23   chr15:67493013-67547536   \n",
       "2         AARS      16  16q22.1   chr16:70286297-70323412   \n",
       "3        AARS2   57505   6p21.1    chr6:44266463-44281063   \n",
       "4         AASS   10157  7q31.32  chr7:121713598-121784344   \n",
       "\n",
       "  Haploinsufficiency Score                     Haploinsufficiency Description  \\\n",
       "0                       30  Gene associated with autosomal recessive pheno...   \n",
       "1                        3       Sufficient evidence for dosage pathogenicity   \n",
       "2                        0                              No evidence available   \n",
       "3                       30  Gene associated with autosomal recessive pheno...   \n",
       "4                       30  Gene associated with autosomal recessive pheno...   \n",
       "\n",
       "  Haploinsufficiency PMID1 Haploinsufficiency PMID2 Haploinsufficiency PMID3  \\\n",
       "0                      NaN                      NaN                      NaN   \n",
       "1                 23064416                 23000146                      NaN   \n",
       "2                      NaN                      NaN                      NaN   \n",
       "3                      NaN                      NaN                      NaN   \n",
       "4                      NaN                      NaN                      NaN   \n",
       "\n",
       "  Triplosensitivity Score Triplosensitivity Description  \\\n",
       "0                       0         No evidence available   \n",
       "1                       0         No evidence available   \n",
       "2                       0         No evidence available   \n",
       "3       Not yet evaluated             Not yet evaluated   \n",
       "4       Not yet evaluated             Not yet evaluated   \n",
       "\n",
       "  Triplosensitivity PMID1 Triplosensitivity PMID2 Triplosensitivity PMID3  \\\n",
       "0                     NaN                     NaN                     NaN   \n",
       "1                     NaN                     NaN                     NaN   \n",
       "2                     NaN                     NaN                     NaN   \n",
       "3                     NaN                     NaN                     NaN   \n",
       "4                     NaN                     NaN                     NaN   \n",
       "\n",
       "  Date Last Evaluated Loss phenotype OMIM ID Triplosensitive phenotype OMIM ID  \n",
       "0          2014-12-11                 111400                               NaN  \n",
       "1          2013-02-28                 148600                               NaN  \n",
       "2          2018-01-11                    NaN                               NaN  \n",
       "3          2016-08-22                    NaN                               NaN  \n",
       "4          2016-08-22                    NaN                               NaN  "
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "curation_gene.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "hi_genes = set(\n",
    "    curation_gene.loc[\n",
    "        curation_gene['Haploinsufficiency Score'] == '3', '#Gene Symbol'\n",
    "    ]\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "hi_gene = gene[gene['name2'].isin(hi_genes)].copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>txStart</th>\n",
       "      <th>txEnd</th>\n",
       "      <th>GeneID</th>\n",
       "      <th>name2</th>\n",
       "      <th>name</th>\n",
       "      <th>strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>25636</th>\n",
       "      <td>chr1</td>\n",
       "      <td>6845383</td>\n",
       "      <td>7829766</td>\n",
       "      <td>23261</td>\n",
       "      <td>CAMTA1</td>\n",
       "      <td>NM_015215.3</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17434</th>\n",
       "      <td>chr1</td>\n",
       "      <td>17345216</td>\n",
       "      <td>17380665</td>\n",
       "      <td>6390</td>\n",
       "      <td>SDHB</td>\n",
       "      <td>NM_003000.2</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21801</th>\n",
       "      <td>chr1</td>\n",
       "      <td>27022521</td>\n",
       "      <td>27108601</td>\n",
       "      <td>8289</td>\n",
       "      <td>ARID1A</td>\n",
       "      <td>NM_139135.2</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20929</th>\n",
       "      <td>chr1</td>\n",
       "      <td>27860755</td>\n",
       "      <td>27930005</td>\n",
       "      <td>27245</td>\n",
       "      <td>AHDC1</td>\n",
       "      <td>NM_001029882.3</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17173</th>\n",
       "      <td>chr1</td>\n",
       "      <td>43391045</td>\n",
       "      <td>43424847</td>\n",
       "      <td>6513</td>\n",
       "      <td>SLC2A1</td>\n",
       "      <td>NM_006516.2</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      chrom   txStart     txEnd  GeneID   name2            name strand\n",
       "25636  chr1   6845383   7829766   23261  CAMTA1     NM_015215.3      +\n",
       "17434  chr1  17345216  17380665    6390    SDHB     NM_003000.2      -\n",
       "21801  chr1  27022521  27108601    8289  ARID1A     NM_139135.2      +\n",
       "20929  chr1  27860755  27930005   27245   AHDC1  NM_001029882.3      -\n",
       "17173  chr1  43391045  43424847    6513  SLC2A1     NM_006516.2      -"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "hi_gene.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "hi_gene.rename(columns={\n",
    "    'chrom': '#chrom', 'txStart': 'start', 'txEnd': 'end',\n",
    "    'GeneID': 'gene_id', 'name2': 'symbol', 'name': 'transcript'\n",
    "}).to_csv('data/hi-gene-20200109.sorted.bed', index=False, sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "!bgzip -cf data/hi-gene-20200109.sorted.bed > ../data/hi-gene-20200109.sorted.bed.gz"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "!tabix -fp bed ../data/hi-gene-20200109.sorted.bed.gz"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "hi_cds = refgene_info.loc[\n",
    "    refgene_info['name2'].isin(hi_genes), ['chrom', 'cdsStart', 'cdsEnd', 'GeneID', 'name2', 'name']\n",
    "].copy()\n",
    "hi_cds = hi_cds.sort_values(['chrom', 'cdsStart', 'cdsEnd'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>cdsStart</th>\n",
       "      <th>cdsEnd</th>\n",
       "      <th>GeneID</th>\n",
       "      <th>name2</th>\n",
       "      <th>name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>25636</th>\n",
       "      <td>chr1</td>\n",
       "      <td>6845590</td>\n",
       "      <td>7826551</td>\n",
       "      <td>23261</td>\n",
       "      <td>CAMTA1</td>\n",
       "      <td>NM_015215.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17434</th>\n",
       "      <td>chr1</td>\n",
       "      <td>17345375</td>\n",
       "      <td>17380514</td>\n",
       "      <td>6390</td>\n",
       "      <td>SDHB</td>\n",
       "      <td>NM_003000.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21801</th>\n",
       "      <td>chr1</td>\n",
       "      <td>27022894</td>\n",
       "      <td>27107247</td>\n",
       "      <td>8289</td>\n",
       "      <td>ARID1A</td>\n",
       "      <td>NM_139135.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20929</th>\n",
       "      <td>chr1</td>\n",
       "      <td>27873814</td>\n",
       "      <td>27878626</td>\n",
       "      <td>27245</td>\n",
       "      <td>AHDC1</td>\n",
       "      <td>NM_001029882.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17173</th>\n",
       "      <td>chr1</td>\n",
       "      <td>43392711</td>\n",
       "      <td>43424322</td>\n",
       "      <td>6513</td>\n",
       "      <td>SLC2A1</td>\n",
       "      <td>NM_006516.2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      chrom  cdsStart    cdsEnd  GeneID   name2            name\n",
       "25636  chr1   6845590   7826551   23261  CAMTA1     NM_015215.3\n",
       "17434  chr1  17345375  17380514    6390    SDHB     NM_003000.2\n",
       "21801  chr1  27022894  27107247    8289  ARID1A     NM_139135.2\n",
       "20929  chr1  27873814  27878626   27245   AHDC1  NM_001029882.3\n",
       "17173  chr1  43392711  43424322    6513  SLC2A1     NM_006516.2"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "hi_cds.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "hi_cds.rename(columns={\n",
    "    'chrom': '#chrom', 'cdsStart': 'start', 'cdsEnd': 'end',\n",
    "    'GeneID': 'gene_id', 'name2': 'symbol', 'name': 'transcript'\n",
    "}).to_csv('data/hi-cds-20200109.sorted.bed', index=False, sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "!bgzip -cf data/hi-cds-20200109.sorted.bed > ../data/hi-cds-20200109.sorted.bed.gz"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "!tabix -fp bed ../data/hi-cds-20200109.sorted.bed.gz"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [],
   "source": [
    "hi_exons = refgene_info.loc[\n",
    "    refgene_info['name2'].isin(hi_genes), ['chrom', 'exonStarts', 'exonEnds', 'GeneID', 'name2', 'name', 'strand']\n",
    "].copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>exonStarts</th>\n",
       "      <th>exonEnds</th>\n",
       "      <th>GeneID</th>\n",
       "      <th>name2</th>\n",
       "      <th>name</th>\n",
       "      <th>strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2693</th>\n",
       "      <td>chrY</td>\n",
       "      <td>2654895,</td>\n",
       "      <td>2655782,</td>\n",
       "      <td>6736</td>\n",
       "      <td>SRY</td>\n",
       "      <td>NM_003140.2</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3929</th>\n",
       "      <td>chr16</td>\n",
       "      <td>86600856,</td>\n",
       "      <td>86602539,</td>\n",
       "      <td>2303</td>\n",
       "      <td>FOXC2</td>\n",
       "      <td>NM_005251.2</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4495</th>\n",
       "      <td>chr7</td>\n",
       "      <td>19155090,19156293,</td>\n",
       "      <td>19155754,19157295,</td>\n",
       "      <td>7291</td>\n",
       "      <td>TWIST1</td>\n",
       "      <td>NM_000474.3</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4610</th>\n",
       "      <td>chr12</td>\n",
       "      <td>56435685,56436208,56437146,56437902,</td>\n",
       "      <td>56435953,56436386,56437277,56438007,</td>\n",
       "      <td>6231</td>\n",
       "      <td>RPS26</td>\n",
       "      <td>NM_001029.3</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4822</th>\n",
       "      <td>chr3</td>\n",
       "      <td>181429711,</td>\n",
       "      <td>181432224,</td>\n",
       "      <td>6657</td>\n",
       "      <td>SOX2</td>\n",
       "      <td>NM_003106.3</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      chrom                            exonStarts  \\\n",
       "2693   chrY                              2654895,   \n",
       "3929  chr16                             86600856,   \n",
       "4495   chr7                    19155090,19156293,   \n",
       "4610  chr12  56435685,56436208,56437146,56437902,   \n",
       "4822   chr3                            181429711,   \n",
       "\n",
       "                                  exonEnds  GeneID   name2         name strand  \n",
       "2693                              2655782,    6736     SRY  NM_003140.2      -  \n",
       "3929                             86602539,    2303   FOXC2  NM_005251.2      +  \n",
       "4495                    19155754,19157295,    7291  TWIST1  NM_000474.3      -  \n",
       "4610  56435953,56436386,56437277,56438007,    6231   RPS26  NM_001029.3      +  \n",
       "4822                            181432224,    6657    SOX2  NM_003106.3      +  "
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "hi_exons.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [],
   "source": [
    "hi_exons['exonStarts'] = hi_exons['exonStarts'].str.replace(r',$', '')\n",
    "hi_exons['exonEnds'] = hi_exons['exonEnds'].str.replace(r',$', '')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>exonStarts</th>\n",
       "      <th>exonEnds</th>\n",
       "      <th>GeneID</th>\n",
       "      <th>name2</th>\n",
       "      <th>name</th>\n",
       "      <th>strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2693</th>\n",
       "      <td>chrY</td>\n",
       "      <td>2654895</td>\n",
       "      <td>2655782</td>\n",
       "      <td>6736</td>\n",
       "      <td>SRY</td>\n",
       "      <td>NM_003140.2</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3929</th>\n",
       "      <td>chr16</td>\n",
       "      <td>86600856</td>\n",
       "      <td>86602539</td>\n",
       "      <td>2303</td>\n",
       "      <td>FOXC2</td>\n",
       "      <td>NM_005251.2</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4495</th>\n",
       "      <td>chr7</td>\n",
       "      <td>19155090,19156293</td>\n",
       "      <td>19155754,19157295</td>\n",
       "      <td>7291</td>\n",
       "      <td>TWIST1</td>\n",
       "      <td>NM_000474.3</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4610</th>\n",
       "      <td>chr12</td>\n",
       "      <td>56435685,56436208,56437146,56437902</td>\n",
       "      <td>56435953,56436386,56437277,56438007</td>\n",
       "      <td>6231</td>\n",
       "      <td>RPS26</td>\n",
       "      <td>NM_001029.3</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4822</th>\n",
       "      <td>chr3</td>\n",
       "      <td>181429711</td>\n",
       "      <td>181432224</td>\n",
       "      <td>6657</td>\n",
       "      <td>SOX2</td>\n",
       "      <td>NM_003106.3</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      chrom                           exonStarts  \\\n",
       "2693   chrY                              2654895   \n",
       "3929  chr16                             86600856   \n",
       "4495   chr7                    19155090,19156293   \n",
       "4610  chr12  56435685,56436208,56437146,56437902   \n",
       "4822   chr3                            181429711   \n",
       "\n",
       "                                 exonEnds  GeneID   name2         name strand  \n",
       "2693                              2655782    6736     SRY  NM_003140.2      -  \n",
       "3929                             86602539    2303   FOXC2  NM_005251.2      +  \n",
       "4495                    19155754,19157295    7291  TWIST1  NM_000474.3      -  \n",
       "4610  56435953,56436386,56437277,56438007    6231   RPS26  NM_001029.3      +  \n",
       "4822                            181432224    6657    SOX2  NM_003106.3      +  "
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "hi_exons.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [],
   "source": [
    "start = hi_exons['exonStarts'].str.split(',').apply(pd.Series).stack().reset_index()\n",
    "start = start.rename(columns={'level_0': 'row', 0: 'start'})[['row', 'start']]\n",
    "start['start'] = start['start'].astype(int)\n",
    "end = hi_exons['exonEnds'].str.split(',').apply(pd.Series).stack().reset_index()\n",
    "end = end.rename(columns={0: 'end'})['end'].astype(int)\n",
    "position = start.join(end)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>row</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>2693</td>\n",
       "      <td>2654895</td>\n",
       "      <td>2655782</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>3929</td>\n",
       "      <td>86600856</td>\n",
       "      <td>86602539</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>4495</td>\n",
       "      <td>19155090</td>\n",
       "      <td>19155754</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>4495</td>\n",
       "      <td>19156293</td>\n",
       "      <td>19157295</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4610</td>\n",
       "      <td>56435685</td>\n",
       "      <td>56435953</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    row     start       end\n",
       "0  2693   2654895   2655782\n",
       "1  3929  86600856  86602539\n",
       "2  4495  19155090  19155754\n",
       "3  4495  19156293  19157295\n",
       "4  4610  56435685  56435953"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "position.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [],
   "source": [
    "exon = position.merge(\n",
    "    hi_exons[['chrom', 'GeneID', 'name2', 'name', 'strand']], how='left', left_on='row', right_index=True\n",
    ")\n",
    "exon = exon.sort_values(['chrom', 'start', 'end'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>row</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>chrom</th>\n",
       "      <th>GeneID</th>\n",
       "      <th>name2</th>\n",
       "      <th>name</th>\n",
       "      <th>strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>4527</th>\n",
       "      <td>25636</td>\n",
       "      <td>6845383</td>\n",
       "      <td>6845635</td>\n",
       "      <td>chr1</td>\n",
       "      <td>23261</td>\n",
       "      <td>CAMTA1</td>\n",
       "      <td>NM_015215.3</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4528</th>\n",
       "      <td>25636</td>\n",
       "      <td>6880240</td>\n",
       "      <td>6880310</td>\n",
       "      <td>chr1</td>\n",
       "      <td>23261</td>\n",
       "      <td>CAMTA1</td>\n",
       "      <td>NM_015215.3</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4529</th>\n",
       "      <td>25636</td>\n",
       "      <td>6885151</td>\n",
       "      <td>6885270</td>\n",
       "      <td>chr1</td>\n",
       "      <td>23261</td>\n",
       "      <td>CAMTA1</td>\n",
       "      <td>NM_015215.3</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4530</th>\n",
       "      <td>25636</td>\n",
       "      <td>7151363</td>\n",
       "      <td>7151431</td>\n",
       "      <td>chr1</td>\n",
       "      <td>23261</td>\n",
       "      <td>CAMTA1</td>\n",
       "      <td>NM_015215.3</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4531</th>\n",
       "      <td>25636</td>\n",
       "      <td>7309550</td>\n",
       "      <td>7309686</td>\n",
       "      <td>chr1</td>\n",
       "      <td>23261</td>\n",
       "      <td>CAMTA1</td>\n",
       "      <td>NM_015215.3</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        row    start      end chrom  GeneID   name2         name strand\n",
       "4527  25636  6845383  6845635  chr1   23261  CAMTA1  NM_015215.3      +\n",
       "4528  25636  6880240  6880310  chr1   23261  CAMTA1  NM_015215.3      +\n",
       "4529  25636  6885151  6885270  chr1   23261  CAMTA1  NM_015215.3      +\n",
       "4530  25636  7151363  7151431  chr1   23261  CAMTA1  NM_015215.3      +\n",
       "4531  25636  7309550  7309686  chr1   23261  CAMTA1  NM_015215.3      +"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "exon.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [],
   "source": [
    "exon['+'] = exon.groupby(['name2', 'name'])['start'].rank('first', ascending=True).astype(int)\n",
    "exon['-'] = exon.groupby(['name2', 'name'])['start'].rank('first', ascending=False).astype(int)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>row</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>chrom</th>\n",
       "      <th>GeneID</th>\n",
       "      <th>name2</th>\n",
       "      <th>name</th>\n",
       "      <th>strand</th>\n",
       "      <th>+</th>\n",
       "      <th>-</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>4527</th>\n",
       "      <td>25636</td>\n",
       "      <td>6845383</td>\n",
       "      <td>6845635</td>\n",
       "      <td>chr1</td>\n",
       "      <td>23261</td>\n",
       "      <td>CAMTA1</td>\n",
       "      <td>NM_015215.3</td>\n",
       "      <td>+</td>\n",
       "      <td>1</td>\n",
       "      <td>23</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4528</th>\n",
       "      <td>25636</td>\n",
       "      <td>6880240</td>\n",
       "      <td>6880310</td>\n",
       "      <td>chr1</td>\n",
       "      <td>23261</td>\n",
       "      <td>CAMTA1</td>\n",
       "      <td>NM_015215.3</td>\n",
       "      <td>+</td>\n",
       "      <td>2</td>\n",
       "      <td>22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4529</th>\n",
       "      <td>25636</td>\n",
       "      <td>6885151</td>\n",
       "      <td>6885270</td>\n",
       "      <td>chr1</td>\n",
       "      <td>23261</td>\n",
       "      <td>CAMTA1</td>\n",
       "      <td>NM_015215.3</td>\n",
       "      <td>+</td>\n",
       "      <td>3</td>\n",
       "      <td>21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4530</th>\n",
       "      <td>25636</td>\n",
       "      <td>7151363</td>\n",
       "      <td>7151431</td>\n",
       "      <td>chr1</td>\n",
       "      <td>23261</td>\n",
       "      <td>CAMTA1</td>\n",
       "      <td>NM_015215.3</td>\n",
       "      <td>+</td>\n",
       "      <td>4</td>\n",
       "      <td>20</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4531</th>\n",
       "      <td>25636</td>\n",
       "      <td>7309550</td>\n",
       "      <td>7309686</td>\n",
       "      <td>chr1</td>\n",
       "      <td>23261</td>\n",
       "      <td>CAMTA1</td>\n",
       "      <td>NM_015215.3</td>\n",
       "      <td>+</td>\n",
       "      <td>5</td>\n",
       "      <td>19</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        row    start      end chrom  GeneID   name2         name strand  +   -\n",
       "4527  25636  6845383  6845635  chr1   23261  CAMTA1  NM_015215.3      +  1  23\n",
       "4528  25636  6880240  6880310  chr1   23261  CAMTA1  NM_015215.3      +  2  22\n",
       "4529  25636  6885151  6885270  chr1   23261  CAMTA1  NM_015215.3      +  3  21\n",
       "4530  25636  7151363  7151431  chr1   23261  CAMTA1  NM_015215.3      +  4  20\n",
       "4531  25636  7309550  7309686  chr1   23261  CAMTA1  NM_015215.3      +  5  19"
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "exon.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>row</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>chrom</th>\n",
       "      <th>GeneID</th>\n",
       "      <th>name2</th>\n",
       "      <th>name</th>\n",
       "      <th>strand</th>\n",
       "      <th>+</th>\n",
       "      <th>-</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>3204</th>\n",
       "      <td>24186</td>\n",
       "      <td>154227753</td>\n",
       "      <td>154227875</td>\n",
       "      <td>chrX</td>\n",
       "      <td>2157</td>\n",
       "      <td>F8</td>\n",
       "      <td>NM_000132.3</td>\n",
       "      <td>-</td>\n",
       "      <td>25</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3205</th>\n",
       "      <td>24186</td>\n",
       "      <td>154250684</td>\n",
       "      <td>154250998</td>\n",
       "      <td>chrX</td>\n",
       "      <td>2157</td>\n",
       "      <td>F8</td>\n",
       "      <td>NM_000132.3</td>\n",
       "      <td>-</td>\n",
       "      <td>26</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>63</th>\n",
       "      <td>7924</td>\n",
       "      <td>154487518</td>\n",
       "      <td>154490514</td>\n",
       "      <td>chrX</td>\n",
       "      <td>116442</td>\n",
       "      <td>RAB39B</td>\n",
       "      <td>NM_171998.3</td>\n",
       "      <td>-</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>64</th>\n",
       "      <td>7924</td>\n",
       "      <td>154493358</td>\n",
       "      <td>154493874</td>\n",
       "      <td>chrX</td>\n",
       "      <td>116442</td>\n",
       "      <td>RAB39B</td>\n",
       "      <td>NM_171998.3</td>\n",
       "      <td>-</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>2693</td>\n",
       "      <td>2654895</td>\n",
       "      <td>2655782</td>\n",
       "      <td>chrY</td>\n",
       "      <td>6736</td>\n",
       "      <td>SRY</td>\n",
       "      <td>NM_003140.2</td>\n",
       "      <td>-</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        row      start        end chrom  GeneID   name2         name strand  \\\n",
       "3204  24186  154227753  154227875  chrX    2157      F8  NM_000132.3      -   \n",
       "3205  24186  154250684  154250998  chrX    2157      F8  NM_000132.3      -   \n",
       "63     7924  154487518  154490514  chrX  116442  RAB39B  NM_171998.3      -   \n",
       "64     7924  154493358  154493874  chrX  116442  RAB39B  NM_171998.3      -   \n",
       "0      2693    2654895    2655782  chrY    6736     SRY  NM_003140.2      -   \n",
       "\n",
       "       +  -  \n",
       "3204  25  2  \n",
       "3205  26  1  \n",
       "63     1  2  \n",
       "64     2  1  \n",
       "0      1  1  "
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "exon.tail()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [],
   "source": [
    "exon['exon'] = pd.concat([exon.loc[exon['strand'] == '+', '+'], exon.loc[exon['strand'] == '-', '-']])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>row</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>chrom</th>\n",
       "      <th>GeneID</th>\n",
       "      <th>name2</th>\n",
       "      <th>name</th>\n",
       "      <th>strand</th>\n",
       "      <th>+</th>\n",
       "      <th>-</th>\n",
       "      <th>exon</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>4527</th>\n",
       "      <td>25636</td>\n",
       "      <td>6845383</td>\n",
       "      <td>6845635</td>\n",
       "      <td>chr1</td>\n",
       "      <td>23261</td>\n",
       "      <td>CAMTA1</td>\n",
       "      <td>NM_015215.3</td>\n",
       "      <td>+</td>\n",
       "      <td>1</td>\n",
       "      <td>23</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4528</th>\n",
       "      <td>25636</td>\n",
       "      <td>6880240</td>\n",
       "      <td>6880310</td>\n",
       "      <td>chr1</td>\n",
       "      <td>23261</td>\n",
       "      <td>CAMTA1</td>\n",
       "      <td>NM_015215.3</td>\n",
       "      <td>+</td>\n",
       "      <td>2</td>\n",
       "      <td>22</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4529</th>\n",
       "      <td>25636</td>\n",
       "      <td>6885151</td>\n",
       "      <td>6885270</td>\n",
       "      <td>chr1</td>\n",
       "      <td>23261</td>\n",
       "      <td>CAMTA1</td>\n",
       "      <td>NM_015215.3</td>\n",
       "      <td>+</td>\n",
       "      <td>3</td>\n",
       "      <td>21</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4530</th>\n",
       "      <td>25636</td>\n",
       "      <td>7151363</td>\n",
       "      <td>7151431</td>\n",
       "      <td>chr1</td>\n",
       "      <td>23261</td>\n",
       "      <td>CAMTA1</td>\n",
       "      <td>NM_015215.3</td>\n",
       "      <td>+</td>\n",
       "      <td>4</td>\n",
       "      <td>20</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4531</th>\n",
       "      <td>25636</td>\n",
       "      <td>7309550</td>\n",
       "      <td>7309686</td>\n",
       "      <td>chr1</td>\n",
       "      <td>23261</td>\n",
       "      <td>CAMTA1</td>\n",
       "      <td>NM_015215.3</td>\n",
       "      <td>+</td>\n",
       "      <td>5</td>\n",
       "      <td>19</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        row    start      end chrom  GeneID   name2         name strand  +  \\\n",
       "4527  25636  6845383  6845635  chr1   23261  CAMTA1  NM_015215.3      +  1   \n",
       "4528  25636  6880240  6880310  chr1   23261  CAMTA1  NM_015215.3      +  2   \n",
       "4529  25636  6885151  6885270  chr1   23261  CAMTA1  NM_015215.3      +  3   \n",
       "4530  25636  7151363  7151431  chr1   23261  CAMTA1  NM_015215.3      +  4   \n",
       "4531  25636  7309550  7309686  chr1   23261  CAMTA1  NM_015215.3      +  5   \n",
       "\n",
       "       -  exon  \n",
       "4527  23     1  \n",
       "4528  22     2  \n",
       "4529  21     3  \n",
       "4530  20     4  \n",
       "4531  19     5  "
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "exon.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>row</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>chrom</th>\n",
       "      <th>GeneID</th>\n",
       "      <th>name2</th>\n",
       "      <th>name</th>\n",
       "      <th>strand</th>\n",
       "      <th>+</th>\n",
       "      <th>-</th>\n",
       "      <th>exon</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>3204</th>\n",
       "      <td>24186</td>\n",
       "      <td>154227753</td>\n",
       "      <td>154227875</td>\n",
       "      <td>chrX</td>\n",
       "      <td>2157</td>\n",
       "      <td>F8</td>\n",
       "      <td>NM_000132.3</td>\n",
       "      <td>-</td>\n",
       "      <td>25</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3205</th>\n",
       "      <td>24186</td>\n",
       "      <td>154250684</td>\n",
       "      <td>154250998</td>\n",
       "      <td>chrX</td>\n",
       "      <td>2157</td>\n",
       "      <td>F8</td>\n",
       "      <td>NM_000132.3</td>\n",
       "      <td>-</td>\n",
       "      <td>26</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>63</th>\n",
       "      <td>7924</td>\n",
       "      <td>154487518</td>\n",
       "      <td>154490514</td>\n",
       "      <td>chrX</td>\n",
       "      <td>116442</td>\n",
       "      <td>RAB39B</td>\n",
       "      <td>NM_171998.3</td>\n",
       "      <td>-</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>64</th>\n",
       "      <td>7924</td>\n",
       "      <td>154493358</td>\n",
       "      <td>154493874</td>\n",
       "      <td>chrX</td>\n",
       "      <td>116442</td>\n",
       "      <td>RAB39B</td>\n",
       "      <td>NM_171998.3</td>\n",
       "      <td>-</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>2693</td>\n",
       "      <td>2654895</td>\n",
       "      <td>2655782</td>\n",
       "      <td>chrY</td>\n",
       "      <td>6736</td>\n",
       "      <td>SRY</td>\n",
       "      <td>NM_003140.2</td>\n",
       "      <td>-</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        row      start        end chrom  GeneID   name2         name strand  \\\n",
       "3204  24186  154227753  154227875  chrX    2157      F8  NM_000132.3      -   \n",
       "3205  24186  154250684  154250998  chrX    2157      F8  NM_000132.3      -   \n",
       "63     7924  154487518  154490514  chrX  116442  RAB39B  NM_171998.3      -   \n",
       "64     7924  154493358  154493874  chrX  116442  RAB39B  NM_171998.3      -   \n",
       "0      2693    2654895    2655782  chrY    6736     SRY  NM_003140.2      -   \n",
       "\n",
       "       +  -  exon  \n",
       "3204  25  2     2  \n",
       "3205  26  1     1  \n",
       "63     1  2     2  \n",
       "64     2  1     1  \n",
       "0      1  1     1  "
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "exon.tail()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [],
   "source": [
    "exon['last_exon'] = exon.groupby(['name2', 'name'])['exon'].transform('max') == exon['exon']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>row</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>chrom</th>\n",
       "      <th>GeneID</th>\n",
       "      <th>name2</th>\n",
       "      <th>name</th>\n",
       "      <th>strand</th>\n",
       "      <th>+</th>\n",
       "      <th>-</th>\n",
       "      <th>exon</th>\n",
       "      <th>last_exon</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>4527</th>\n",
       "      <td>25636</td>\n",
       "      <td>6845383</td>\n",
       "      <td>6845635</td>\n",
       "      <td>chr1</td>\n",
       "      <td>23261</td>\n",
       "      <td>CAMTA1</td>\n",
       "      <td>NM_015215.3</td>\n",
       "      <td>+</td>\n",
       "      <td>1</td>\n",
       "      <td>23</td>\n",
       "      <td>1</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4528</th>\n",
       "      <td>25636</td>\n",
       "      <td>6880240</td>\n",
       "      <td>6880310</td>\n",
       "      <td>chr1</td>\n",
       "      <td>23261</td>\n",
       "      <td>CAMTA1</td>\n",
       "      <td>NM_015215.3</td>\n",
       "      <td>+</td>\n",
       "      <td>2</td>\n",
       "      <td>22</td>\n",
       "      <td>2</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4529</th>\n",
       "      <td>25636</td>\n",
       "      <td>6885151</td>\n",
       "      <td>6885270</td>\n",
       "      <td>chr1</td>\n",
       "      <td>23261</td>\n",
       "      <td>CAMTA1</td>\n",
       "      <td>NM_015215.3</td>\n",
       "      <td>+</td>\n",
       "      <td>3</td>\n",
       "      <td>21</td>\n",
       "      <td>3</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4530</th>\n",
       "      <td>25636</td>\n",
       "      <td>7151363</td>\n",
       "      <td>7151431</td>\n",
       "      <td>chr1</td>\n",
       "      <td>23261</td>\n",
       "      <td>CAMTA1</td>\n",
       "      <td>NM_015215.3</td>\n",
       "      <td>+</td>\n",
       "      <td>4</td>\n",
       "      <td>20</td>\n",
       "      <td>4</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4531</th>\n",
       "      <td>25636</td>\n",
       "      <td>7309550</td>\n",
       "      <td>7309686</td>\n",
       "      <td>chr1</td>\n",
       "      <td>23261</td>\n",
       "      <td>CAMTA1</td>\n",
       "      <td>NM_015215.3</td>\n",
       "      <td>+</td>\n",
       "      <td>5</td>\n",
       "      <td>19</td>\n",
       "      <td>5</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        row    start      end chrom  GeneID   name2         name strand  +  \\\n",
       "4527  25636  6845383  6845635  chr1   23261  CAMTA1  NM_015215.3      +  1   \n",
       "4528  25636  6880240  6880310  chr1   23261  CAMTA1  NM_015215.3      +  2   \n",
       "4529  25636  6885151  6885270  chr1   23261  CAMTA1  NM_015215.3      +  3   \n",
       "4530  25636  7151363  7151431  chr1   23261  CAMTA1  NM_015215.3      +  4   \n",
       "4531  25636  7309550  7309686  chr1   23261  CAMTA1  NM_015215.3      +  5   \n",
       "\n",
       "       -  exon  last_exon  \n",
       "4527  23     1      False  \n",
       "4528  22     2      False  \n",
       "4529  21     3      False  \n",
       "4530  20     4      False  \n",
       "4531  19     5      False  "
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "exon.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>row</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>chrom</th>\n",
       "      <th>GeneID</th>\n",
       "      <th>name2</th>\n",
       "      <th>name</th>\n",
       "      <th>strand</th>\n",
       "      <th>+</th>\n",
       "      <th>-</th>\n",
       "      <th>exon</th>\n",
       "      <th>last_exon</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>3204</th>\n",
       "      <td>24186</td>\n",
       "      <td>154227753</td>\n",
       "      <td>154227875</td>\n",
       "      <td>chrX</td>\n",
       "      <td>2157</td>\n",
       "      <td>F8</td>\n",
       "      <td>NM_000132.3</td>\n",
       "      <td>-</td>\n",
       "      <td>25</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3205</th>\n",
       "      <td>24186</td>\n",
       "      <td>154250684</td>\n",
       "      <td>154250998</td>\n",
       "      <td>chrX</td>\n",
       "      <td>2157</td>\n",
       "      <td>F8</td>\n",
       "      <td>NM_000132.3</td>\n",
       "      <td>-</td>\n",
       "      <td>26</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>63</th>\n",
       "      <td>7924</td>\n",
       "      <td>154487518</td>\n",
       "      <td>154490514</td>\n",
       "      <td>chrX</td>\n",
       "      <td>116442</td>\n",
       "      <td>RAB39B</td>\n",
       "      <td>NM_171998.3</td>\n",
       "      <td>-</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>64</th>\n",
       "      <td>7924</td>\n",
       "      <td>154493358</td>\n",
       "      <td>154493874</td>\n",
       "      <td>chrX</td>\n",
       "      <td>116442</td>\n",
       "      <td>RAB39B</td>\n",
       "      <td>NM_171998.3</td>\n",
       "      <td>-</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>2693</td>\n",
       "      <td>2654895</td>\n",
       "      <td>2655782</td>\n",
       "      <td>chrY</td>\n",
       "      <td>6736</td>\n",
       "      <td>SRY</td>\n",
       "      <td>NM_003140.2</td>\n",
       "      <td>-</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        row      start        end chrom  GeneID   name2         name strand  \\\n",
       "3204  24186  154227753  154227875  chrX    2157      F8  NM_000132.3      -   \n",
       "3205  24186  154250684  154250998  chrX    2157      F8  NM_000132.3      -   \n",
       "63     7924  154487518  154490514  chrX  116442  RAB39B  NM_171998.3      -   \n",
       "64     7924  154493358  154493874  chrX  116442  RAB39B  NM_171998.3      -   \n",
       "0      2693    2654895    2655782  chrY    6736     SRY  NM_003140.2      -   \n",
       "\n",
       "       +  -  exon  last_exon  \n",
       "3204  25  2     2      False  \n",
       "3205  26  1     1      False  \n",
       "63     1  2     2       True  \n",
       "64     2  1     1      False  \n",
       "0      1  1     1       True  "
      ]
     },
     "execution_count": 43,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "exon.tail()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [],
   "source": [
    "exon = exon[\n",
    "    ['chrom', 'start', 'end', 'GeneID', 'name2', 'name', 'exon', 'last_exon']\n",
    "].sort_values(['chrom', 'start', 'end'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>GeneID</th>\n",
       "      <th>name2</th>\n",
       "      <th>name</th>\n",
       "      <th>exon</th>\n",
       "      <th>last_exon</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>4527</th>\n",
       "      <td>chr1</td>\n",
       "      <td>6845383</td>\n",
       "      <td>6845635</td>\n",
       "      <td>23261</td>\n",
       "      <td>CAMTA1</td>\n",
       "      <td>NM_015215.3</td>\n",
       "      <td>1</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4528</th>\n",
       "      <td>chr1</td>\n",
       "      <td>6880240</td>\n",
       "      <td>6880310</td>\n",
       "      <td>23261</td>\n",
       "      <td>CAMTA1</td>\n",
       "      <td>NM_015215.3</td>\n",
       "      <td>2</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4529</th>\n",
       "      <td>chr1</td>\n",
       "      <td>6885151</td>\n",
       "      <td>6885270</td>\n",
       "      <td>23261</td>\n",
       "      <td>CAMTA1</td>\n",
       "      <td>NM_015215.3</td>\n",
       "      <td>3</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4530</th>\n",
       "      <td>chr1</td>\n",
       "      <td>7151363</td>\n",
       "      <td>7151431</td>\n",
       "      <td>23261</td>\n",
       "      <td>CAMTA1</td>\n",
       "      <td>NM_015215.3</td>\n",
       "      <td>4</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4531</th>\n",
       "      <td>chr1</td>\n",
       "      <td>7309550</td>\n",
       "      <td>7309686</td>\n",
       "      <td>23261</td>\n",
       "      <td>CAMTA1</td>\n",
       "      <td>NM_015215.3</td>\n",
       "      <td>5</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     chrom    start      end  GeneID   name2         name  exon  last_exon\n",
       "4527  chr1  6845383  6845635   23261  CAMTA1  NM_015215.3     1      False\n",
       "4528  chr1  6880240  6880310   23261  CAMTA1  NM_015215.3     2      False\n",
       "4529  chr1  6885151  6885270   23261  CAMTA1  NM_015215.3     3      False\n",
       "4530  chr1  7151363  7151431   23261  CAMTA1  NM_015215.3     4      False\n",
       "4531  chr1  7309550  7309686   23261  CAMTA1  NM_015215.3     5      False"
      ]
     },
     "execution_count": 45,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "exon.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [],
   "source": [
    "exon.rename(columns={\n",
    "    'chrom': '#chrom',\n",
    "    'GeneID': 'gene_id', 'name2': 'symbol', 'name': 'transcript'\n",
    "}).to_csv('data/hi-exon-20200109.sorted.bed', index=False, sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [],
   "source": [
    "!bgzip -cf data/hi-exon-20200109.sorted.bed > ../data/hi-exon-20200109.sorted.bed.gz"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [],
   "source": [
    "!tabix -fp bed ../data/hi-exon-20200109.sorted.bed.gz"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [],
   "source": [
    "last_exon = exon[exon['last_exon']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>GeneID</th>\n",
       "      <th>name2</th>\n",
       "      <th>name</th>\n",
       "      <th>exon</th>\n",
       "      <th>last_exon</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>4549</th>\n",
       "      <td>chr1</td>\n",
       "      <td>7826518</td>\n",
       "      <td>7829766</td>\n",
       "      <td>23261</td>\n",
       "      <td>CAMTA1</td>\n",
       "      <td>NM_015215.3</td>\n",
       "      <td>23</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>948</th>\n",
       "      <td>chr1</td>\n",
       "      <td>17345216</td>\n",
       "      <td>17345453</td>\n",
       "      <td>6390</td>\n",
       "      <td>SDHB</td>\n",
       "      <td>NM_003000.2</td>\n",
       "      <td>8</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2185</th>\n",
       "      <td>chr1</td>\n",
       "      <td>27105513</td>\n",
       "      <td>27108601</td>\n",
       "      <td>8289</td>\n",
       "      <td>ARID1A</td>\n",
       "      <td>NM_139135.2</td>\n",
       "      <td>20</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1875</th>\n",
       "      <td>chr1</td>\n",
       "      <td>27860755</td>\n",
       "      <td>27861427</td>\n",
       "      <td>27245</td>\n",
       "      <td>AHDC1</td>\n",
       "      <td>NM_001029882.3</td>\n",
       "      <td>7</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>872</th>\n",
       "      <td>chr1</td>\n",
       "      <td>43391045</td>\n",
       "      <td>43392912</td>\n",
       "      <td>6513</td>\n",
       "      <td>SLC2A1</td>\n",
       "      <td>NM_006516.2</td>\n",
       "      <td>10</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     chrom     start       end  GeneID   name2            name  exon  \\\n",
       "4549  chr1   7826518   7829766   23261  CAMTA1     NM_015215.3    23   \n",
       "948   chr1  17345216  17345453    6390    SDHB     NM_003000.2     8   \n",
       "2185  chr1  27105513  27108601    8289  ARID1A     NM_139135.2    20   \n",
       "1875  chr1  27860755  27861427   27245   AHDC1  NM_001029882.3     7   \n",
       "872   chr1  43391045  43392912    6513  SLC2A1     NM_006516.2    10   \n",
       "\n",
       "      last_exon  \n",
       "4549       True  \n",
       "948        True  \n",
       "2185       True  \n",
       "1875       True  \n",
       "872        True  "
      ]
     },
     "execution_count": 50,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "last_exon.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [],
   "source": [
    "last_exon_region = last_exon['chrom'] + ':' + last_exon['start'].astype(str) + '-' + last_exon['end'].astype(str)\n",
    "last_exon_region = last_exon_region.str.replace('chr', '')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "4549      1:7826518-7829766\n",
       "948     1:17345216-17345453\n",
       "2185    1:27105513-27108601\n",
       "1875    1:27860755-27861427\n",
       "872     1:43391045-43392912\n",
       "dtype: object"
      ]
     },
     "execution_count": 52,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "last_exon_region.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [],
   "source": [
    "need_fields = [\n",
    "    'variants/CHROM', 'variants/POS', 'variants/REF', 'variants/ALT',\n",
    "    'variants/AF_ESP', 'variants/AF_EXAC', 'variants/AF_TGP', 'variants/CLNSIG'\n",
    "]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open('data/clinvar_pathogenic_20191125.sorted.vcf', 'w') as f:\n",
    "    headers = allel.read_vcf_headers('data/clinvar_20191125.vcf.gz')\n",
    "    f.write(''.join(headers.headers))\n",
    "    \n",
    "    for region in last_exon_region:\n",
    "        fields, samples, headers, it = allel.iter_vcf_chunks(\n",
    "            'data/clinvar_20191125.vcf.gz', fields=need_fields, alt_number=1, region=region\n",
    "        )\n",
    "        for variants, *_ in it:\n",
    "            esp_filter = np.isnan(variants['variants/AF_ESP'])\n",
    "            esp_filter[~esp_filter] |= variants['variants/AF_ESP'][~esp_filter] < 0.01\n",
    "\n",
    "            exac_filter = np.isnan(variants['variants/AF_EXAC'])\n",
    "            exac_filter[~exac_filter] |= variants['variants/AF_EXAC'][~exac_filter] < 0.01\n",
    "\n",
    "            tgp_filter = np.isnan(variants['variants/AF_TGP'])\n",
    "            tgp_filter[~tgp_filter] |= variants['variants/AF_TGP'][~tgp_filter] < 0.01\n",
    "\n",
    "            pathogenic_filter = np.isin(\n",
    "                variants['variants/CLNSIG'], ['Likely_pathogenic', 'Pathogenic', 'Pathogenic/Likely_pathogenic']\n",
    "            )\n",
    "\n",
    "            af_filter = esp_filter & exac_filter & tgp_filter & pathogenic_filter\n",
    "\n",
    "            filtered_variants = {k: v[af_filter] for k, v in variants.items()}\n",
    "\n",
    "            filtered_variants['variants/CHROM'] = 'chr' + filtered_variants['variants/CHROM']\n",
    "\n",
    "            names, callset = allel.normalize_callset(filtered_variants)\n",
    "            allel.write_vcf_data(f, names, callset, None, {field: np.nan for field in need_fields})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {},
   "outputs": [],
   "source": [
    "!sed -i 's/AF_.\\+=nan;//' data/clinvar_pathogenic_20191125.sorted.vcf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {},
   "outputs": [],
   "source": [
    "!bgzip -cf data/clinvar_pathogenic_20191125.sorted.vcf > ../data/clinvar-pathogenic-20191125.sorted.vcf.gz"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {},
   "outputs": [],
   "source": [
    "!tabix -fp vcf ../data/clinvar-pathogenic-20191125.sorted.vcf.gz"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {},
   "outputs": [],
   "source": [
    "uhi_genes = set(\n",
    "    curation_gene.loc[curation_gene['Haploinsufficiency Score'] == '40', '#Gene Symbol']\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {},
   "outputs": [],
   "source": [
    "uhi_gene = gene[gene['name2'].isin(uhi_genes)].copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>txStart</th>\n",
       "      <th>txEnd</th>\n",
       "      <th>GeneID</th>\n",
       "      <th>name2</th>\n",
       "      <th>name</th>\n",
       "      <th>strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>15408</th>\n",
       "      <td>chr1</td>\n",
       "      <td>55505148</td>\n",
       "      <td>55530526</td>\n",
       "      <td>255738</td>\n",
       "      <td>PCSK9</td>\n",
       "      <td>NM_174936.3</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7326</th>\n",
       "      <td>chr13</td>\n",
       "      <td>20761601</td>\n",
       "      <td>20767114</td>\n",
       "      <td>2706</td>\n",
       "      <td>GJB2</td>\n",
       "      <td>NM_004004.5</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       chrom   txStart     txEnd  GeneID  name2         name strand\n",
       "15408   chr1  55505148  55530526  255738  PCSK9  NM_174936.3      +\n",
       "7326   chr13  20761601  20767114    2706   GJB2  NM_004004.5      -"
      ]
     },
     "execution_count": 60,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "uhi_gene.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {},
   "outputs": [],
   "source": [
    "uhi_gene.rename(columns={\n",
    "    'chrom': '#chrom', 'txStart': 'start', 'txEnd': 'end',\n",
    "    'GeneID': 'gene_id', 'name2': 'symbol', 'name': 'transcript'\n",
    "}).to_csv('data/uhi-gene-20200109.sorted.bed', index=False, sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "metadata": {},
   "outputs": [],
   "source": [
    "!bgzip -cf data/uhi-gene-20200109.sorted.bed > ../data/uhi-gene-20200109.sorted.bed.gz"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "metadata": {},
   "outputs": [],
   "source": [
    "!tabix -fp bed ../data/uhi-gene-20200109.sorted.bed.gz"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "metadata": {},
   "outputs": [],
   "source": [
    "region = pd.read_csv('data/ClinGen_region_curation_list_GRCh37.tsv', sep='\\t', skiprows=5, dtype=str)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>#ISCA ID</th>\n",
       "      <th>ISCA Region Name</th>\n",
       "      <th>cytoBand</th>\n",
       "      <th>Genomic Location</th>\n",
       "      <th>Haploinsufficiency Score</th>\n",
       "      <th>Haploinsufficiency Description</th>\n",
       "      <th>Haploinsufficiency PMID1</th>\n",
       "      <th>Haploinsufficiency PMID2</th>\n",
       "      <th>Haploinsufficiency PMID3</th>\n",
       "      <th>Triplosensitivity Score</th>\n",
       "      <th>Triplosensitivity Description</th>\n",
       "      <th>Triplosensitivity PMID1</th>\n",
       "      <th>Triplosensitivity PMID2</th>\n",
       "      <th>Triplosensitivity PMID3</th>\n",
       "      <th>Date Last Evaluated</th>\n",
       "      <th>Loss phenotype OMIM ID</th>\n",
       "      <th>Triplosensitive phenotype OMIM ID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ISCA-46300</td>\n",
       "      <td>15q24 recurrent region (C-D) (includes SIN3A)</td>\n",
       "      <td>NaN</td>\n",
       "      <td>chr15: 75631787-75972909</td>\n",
       "      <td>3</td>\n",
       "      <td>Sufficient evidence for dosage pathogenicity</td>\n",
       "      <td>27399968</td>\n",
       "      <td>22180641</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>No evidence available</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2020-01-13</td>\n",
       "      <td>613406</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ISCA-46299</td>\n",
       "      <td>Xp11.22 region (includes HUWE1)</td>\n",
       "      <td>Xp11.22</td>\n",
       "      <td>chrX:53363456-53793054</td>\n",
       "      <td>0</td>\n",
       "      <td>No evidence available</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3</td>\n",
       "      <td>Sufficient evidence for dosage pathogenicity</td>\n",
       "      <td>22840365</td>\n",
       "      <td>20655035</td>\n",
       "      <td>26692240</td>\n",
       "      <td>2018-11-19</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ISCA-46298</td>\n",
       "      <td>13q12.12 recurrent region (includes SACS and S...</td>\n",
       "      <td>13q12.12</td>\n",
       "      <td>chr13:23544582-24881361</td>\n",
       "      <td>30</td>\n",
       "      <td>Gene associated with autosomal recessive pheno...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>No evidence available</td>\n",
       "      <td>23239789</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2019-11-03</td>\n",
       "      <td>270550,253700</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ISCA-46297</td>\n",
       "      <td>16p12.2 recurrent region (includes OTOA) (dist...</td>\n",
       "      <td>16p12.2</td>\n",
       "      <td>chr16:21570113-21740423</td>\n",
       "      <td>30</td>\n",
       "      <td>Gene associated with autosomal recessive pheno...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>40</td>\n",
       "      <td>Dosage sensitivity unlikely</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2019-11-03</td>\n",
       "      <td>607039</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ISCA-46296</td>\n",
       "      <td>15q24 recurrent region (A-C)</td>\n",
       "      <td>15q24.1-q24.2</td>\n",
       "      <td>chr15:72963715-75508312</td>\n",
       "      <td>3</td>\n",
       "      <td>Sufficient evidence for dosage pathogenicity</td>\n",
       "      <td>22180641</td>\n",
       "      <td>19921647</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>Little evidence for dosage pathogenicity</td>\n",
       "      <td>19557438</td>\n",
       "      <td>20860070</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2020-01-03</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     #ISCA ID                                   ISCA Region Name  \\\n",
       "0  ISCA-46300      15q24 recurrent region (C-D) (includes SIN3A)   \n",
       "1  ISCA-46299                    Xp11.22 region (includes HUWE1)   \n",
       "2  ISCA-46298  13q12.12 recurrent region (includes SACS and S...   \n",
       "3  ISCA-46297  16p12.2 recurrent region (includes OTOA) (dist...   \n",
       "4  ISCA-46296                       15q24 recurrent region (A-C)   \n",
       "\n",
       "        cytoBand          Genomic Location Haploinsufficiency Score  \\\n",
       "0            NaN  chr15: 75631787-75972909                        3   \n",
       "1        Xp11.22    chrX:53363456-53793054                        0   \n",
       "2       13q12.12   chr13:23544582-24881361                       30   \n",
       "3        16p12.2   chr16:21570113-21740423                       30   \n",
       "4  15q24.1-q24.2   chr15:72963715-75508312                        3   \n",
       "\n",
       "                      Haploinsufficiency Description Haploinsufficiency PMID1  \\\n",
       "0       Sufficient evidence for dosage pathogenicity                 27399968   \n",
       "1                              No evidence available                      NaN   \n",
       "2  Gene associated with autosomal recessive pheno...                      NaN   \n",
       "3  Gene associated with autosomal recessive pheno...                      NaN   \n",
       "4       Sufficient evidence for dosage pathogenicity                 22180641   \n",
       "\n",
       "  Haploinsufficiency PMID2 Haploinsufficiency PMID3 Triplosensitivity Score  \\\n",
       "0                 22180641                      NaN                       0   \n",
       "1                      NaN                      NaN                       3   \n",
       "2                      NaN                      NaN                       0   \n",
       "3                      NaN                      NaN                      40   \n",
       "4                 19921647                      NaN                       1   \n",
       "\n",
       "                  Triplosensitivity Description Triplosensitivity PMID1  \\\n",
       "0                         No evidence available                     NaN   \n",
       "1  Sufficient evidence for dosage pathogenicity                22840365   \n",
       "2                         No evidence available                23239789   \n",
       "3                   Dosage sensitivity unlikely                     NaN   \n",
       "4      Little evidence for dosage pathogenicity                19557438   \n",
       "\n",
       "  Triplosensitivity PMID2 Triplosensitivity PMID3 Date Last Evaluated  \\\n",
       "0                     NaN                     NaN          2020-01-13   \n",
       "1                20655035                26692240          2018-11-19   \n",
       "2                     NaN                     NaN          2019-11-03   \n",
       "3                     NaN                     NaN          2019-11-03   \n",
       "4                20860070                     NaN          2020-01-03   \n",
       "\n",
       "  Loss phenotype OMIM ID Triplosensitive phenotype OMIM ID  \n",
       "0                 613406                               NaN  \n",
       "1                    NaN                               NaN  \n",
       "2          270550,253700                               NaN  \n",
       "3                 607039                               NaN  \n",
       "4                    NaN                               NaN  "
      ]
     },
     "execution_count": 65,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "region.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "metadata": {},
   "outputs": [],
   "source": [
    "position = region['Genomic Location'].str.extract(r'(?P<chrom>chr\\w+)\\s*:\\s*(?P<start>\\d+)\\s*-\\s*(?P<end>\\d+)')\n",
    "position['start'] = position['start'].astype(int)\n",
    "position['end'] = position['end'].astype(int)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr15</td>\n",
       "      <td>75631787</td>\n",
       "      <td>75972909</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chrX</td>\n",
       "      <td>53363456</td>\n",
       "      <td>53793054</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr13</td>\n",
       "      <td>23544582</td>\n",
       "      <td>24881361</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr16</td>\n",
       "      <td>21570113</td>\n",
       "      <td>21740423</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr15</td>\n",
       "      <td>72963715</td>\n",
       "      <td>75508312</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   chrom     start       end\n",
       "0  chr15  75631787  75972909\n",
       "1   chrX  53363456  53793054\n",
       "2  chr13  23544582  24881361\n",
       "3  chr16  21570113  21740423\n",
       "4  chr15  72963715  75508312"
      ]
     },
     "execution_count": 67,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "position.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "metadata": {},
   "outputs": [],
   "source": [
    "region_pos = region.merge(position, how='left', left_index=True, right_index=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>#ISCA ID</th>\n",
       "      <th>ISCA Region Name</th>\n",
       "      <th>cytoBand</th>\n",
       "      <th>Genomic Location</th>\n",
       "      <th>Haploinsufficiency Score</th>\n",
       "      <th>Haploinsufficiency Description</th>\n",
       "      <th>Haploinsufficiency PMID1</th>\n",
       "      <th>Haploinsufficiency PMID2</th>\n",
       "      <th>Haploinsufficiency PMID3</th>\n",
       "      <th>Triplosensitivity Score</th>\n",
       "      <th>Triplosensitivity Description</th>\n",
       "      <th>Triplosensitivity PMID1</th>\n",
       "      <th>Triplosensitivity PMID2</th>\n",
       "      <th>Triplosensitivity PMID3</th>\n",
       "      <th>Date Last Evaluated</th>\n",
       "      <th>Loss phenotype OMIM ID</th>\n",
       "      <th>Triplosensitive phenotype OMIM ID</th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ISCA-46300</td>\n",
       "      <td>15q24 recurrent region (C-D) (includes SIN3A)</td>\n",
       "      <td>NaN</td>\n",
       "      <td>chr15: 75631787-75972909</td>\n",
       "      <td>3</td>\n",
       "      <td>Sufficient evidence for dosage pathogenicity</td>\n",
       "      <td>27399968</td>\n",
       "      <td>22180641</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>No evidence available</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2020-01-13</td>\n",
       "      <td>613406</td>\n",
       "      <td>NaN</td>\n",
       "      <td>chr15</td>\n",
       "      <td>75631787</td>\n",
       "      <td>75972909</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ISCA-46299</td>\n",
       "      <td>Xp11.22 region (includes HUWE1)</td>\n",
       "      <td>Xp11.22</td>\n",
       "      <td>chrX:53363456-53793054</td>\n",
       "      <td>0</td>\n",
       "      <td>No evidence available</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3</td>\n",
       "      <td>Sufficient evidence for dosage pathogenicity</td>\n",
       "      <td>22840365</td>\n",
       "      <td>20655035</td>\n",
       "      <td>26692240</td>\n",
       "      <td>2018-11-19</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>chrX</td>\n",
       "      <td>53363456</td>\n",
       "      <td>53793054</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ISCA-46298</td>\n",
       "      <td>13q12.12 recurrent region (includes SACS and S...</td>\n",
       "      <td>13q12.12</td>\n",
       "      <td>chr13:23544582-24881361</td>\n",
       "      <td>30</td>\n",
       "      <td>Gene associated with autosomal recessive pheno...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>No evidence available</td>\n",
       "      <td>23239789</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2019-11-03</td>\n",
       "      <td>270550,253700</td>\n",
       "      <td>NaN</td>\n",
       "      <td>chr13</td>\n",
       "      <td>23544582</td>\n",
       "      <td>24881361</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ISCA-46297</td>\n",
       "      <td>16p12.2 recurrent region (includes OTOA) (dist...</td>\n",
       "      <td>16p12.2</td>\n",
       "      <td>chr16:21570113-21740423</td>\n",
       "      <td>30</td>\n",
       "      <td>Gene associated with autosomal recessive pheno...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>40</td>\n",
       "      <td>Dosage sensitivity unlikely</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2019-11-03</td>\n",
       "      <td>607039</td>\n",
       "      <td>NaN</td>\n",
       "      <td>chr16</td>\n",
       "      <td>21570113</td>\n",
       "      <td>21740423</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ISCA-46296</td>\n",
       "      <td>15q24 recurrent region (A-C)</td>\n",
       "      <td>15q24.1-q24.2</td>\n",
       "      <td>chr15:72963715-75508312</td>\n",
       "      <td>3</td>\n",
       "      <td>Sufficient evidence for dosage pathogenicity</td>\n",
       "      <td>22180641</td>\n",
       "      <td>19921647</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>Little evidence for dosage pathogenicity</td>\n",
       "      <td>19557438</td>\n",
       "      <td>20860070</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2020-01-03</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>chr15</td>\n",
       "      <td>72963715</td>\n",
       "      <td>75508312</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     #ISCA ID                                   ISCA Region Name  \\\n",
       "0  ISCA-46300      15q24 recurrent region (C-D) (includes SIN3A)   \n",
       "1  ISCA-46299                    Xp11.22 region (includes HUWE1)   \n",
       "2  ISCA-46298  13q12.12 recurrent region (includes SACS and S...   \n",
       "3  ISCA-46297  16p12.2 recurrent region (includes OTOA) (dist...   \n",
       "4  ISCA-46296                       15q24 recurrent region (A-C)   \n",
       "\n",
       "        cytoBand          Genomic Location Haploinsufficiency Score  \\\n",
       "0            NaN  chr15: 75631787-75972909                        3   \n",
       "1        Xp11.22    chrX:53363456-53793054                        0   \n",
       "2       13q12.12   chr13:23544582-24881361                       30   \n",
       "3        16p12.2   chr16:21570113-21740423                       30   \n",
       "4  15q24.1-q24.2   chr15:72963715-75508312                        3   \n",
       "\n",
       "                      Haploinsufficiency Description Haploinsufficiency PMID1  \\\n",
       "0       Sufficient evidence for dosage pathogenicity                 27399968   \n",
       "1                              No evidence available                      NaN   \n",
       "2  Gene associated with autosomal recessive pheno...                      NaN   \n",
       "3  Gene associated with autosomal recessive pheno...                      NaN   \n",
       "4       Sufficient evidence for dosage pathogenicity                 22180641   \n",
       "\n",
       "  Haploinsufficiency PMID2 Haploinsufficiency PMID3 Triplosensitivity Score  \\\n",
       "0                 22180641                      NaN                       0   \n",
       "1                      NaN                      NaN                       3   \n",
       "2                      NaN                      NaN                       0   \n",
       "3                      NaN                      NaN                      40   \n",
       "4                 19921647                      NaN                       1   \n",
       "\n",
       "                  Triplosensitivity Description Triplosensitivity PMID1  \\\n",
       "0                         No evidence available                     NaN   \n",
       "1  Sufficient evidence for dosage pathogenicity                22840365   \n",
       "2                         No evidence available                23239789   \n",
       "3                   Dosage sensitivity unlikely                     NaN   \n",
       "4      Little evidence for dosage pathogenicity                19557438   \n",
       "\n",
       "  Triplosensitivity PMID2 Triplosensitivity PMID3 Date Last Evaluated  \\\n",
       "0                     NaN                     NaN          2020-01-13   \n",
       "1                20655035                26692240          2018-11-19   \n",
       "2                     NaN                     NaN          2019-11-03   \n",
       "3                     NaN                     NaN          2019-11-03   \n",
       "4                20860070                     NaN          2020-01-03   \n",
       "\n",
       "  Loss phenotype OMIM ID Triplosensitive phenotype OMIM ID  chrom     start  \\\n",
       "0                 613406                               NaN  chr15  75631787   \n",
       "1                    NaN                               NaN   chrX  53363456   \n",
       "2          270550,253700                               NaN  chr13  23544582   \n",
       "3                 607039                               NaN  chr16  21570113   \n",
       "4                    NaN                               NaN  chr15  72963715   \n",
       "\n",
       "        end  \n",
       "0  75972909  \n",
       "1  53793054  \n",
       "2  24881361  \n",
       "3  21740423  \n",
       "4  75508312  "
      ]
     },
     "execution_count": 69,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "region_pos.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "metadata": {},
   "outputs": [],
   "source": [
    "hi_region = region_pos.loc[\n",
    "    region_pos['Haploinsufficiency Score'] == '3',\n",
    "    ['chrom', 'start', 'end', '#ISCA ID', 'ISCA Region Name']\n",
    "].sort_values(['chrom', 'start', 'end'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>#ISCA ID</th>\n",
       "      <th>ISCA Region Name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>43</th>\n",
       "      <td>chr1</td>\n",
       "      <td>834083</td>\n",
       "      <td>6289973</td>\n",
       "      <td>ISCA-37434</td>\n",
       "      <td>1p36 terminal region (includes GABRD)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>52</th>\n",
       "      <td>chr1</td>\n",
       "      <td>146577486</td>\n",
       "      <td>147394506</td>\n",
       "      <td>ISCA-37421</td>\n",
       "      <td>1q21.1 recurrent region (BP3-BP4, distal) (inc...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>chr1</td>\n",
       "      <td>243287730</td>\n",
       "      <td>245318287</td>\n",
       "      <td>ISCA-37493</td>\n",
       "      <td>1q43q44 terminal region (includes AKT3)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>51</th>\n",
       "      <td>chr10</td>\n",
       "      <td>81682843</td>\n",
       "      <td>88739388</td>\n",
       "      <td>ISCA-37424</td>\n",
       "      <td>10q22.3q23.2 recurrent region (LCR-3/4-flanked...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>63</th>\n",
       "      <td>chr11</td>\n",
       "      <td>31803509</td>\n",
       "      <td>32510988</td>\n",
       "      <td>ISCA-37401</td>\n",
       "      <td>11p13 (WAGR syndrome) region</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    chrom      start        end    #ISCA ID  \\\n",
       "43   chr1     834083    6289973  ISCA-37434   \n",
       "52   chr1  146577486  147394506  ISCA-37421   \n",
       "17   chr1  243287730  245318287  ISCA-37493   \n",
       "51  chr10   81682843   88739388  ISCA-37424   \n",
       "63  chr11   31803509   32510988  ISCA-37401   \n",
       "\n",
       "                                     ISCA Region Name  \n",
       "43              1p36 terminal region (includes GABRD)  \n",
       "52  1q21.1 recurrent region (BP3-BP4, distal) (inc...  \n",
       "17            1q43q44 terminal region (includes AKT3)  \n",
       "51  10q22.3q23.2 recurrent region (LCR-3/4-flanked...  \n",
       "63                       11p13 (WAGR syndrome) region  "
      ]
     },
     "execution_count": 71,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "hi_region.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "metadata": {},
   "outputs": [],
   "source": [
    "hi_region.rename(columns={\n",
    "    'chrom': '#chrom',\n",
    "    '#ISCA ID': 'isca_id', 'ISCA Region Name': 'name'\n",
    "}).to_csv('data/hi-region-20200109.sorted.bed', sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "metadata": {},
   "outputs": [],
   "source": [
    "!bgzip -cf data/hi-region-20200109.sorted.bed > ../data/hi-region-20200109.sorted.bed.gz"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "metadata": {},
   "outputs": [],
   "source": [
    "!tabix -fp bed ../data/hi-region-20200109.sorted.bed.gz"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "metadata": {},
   "outputs": [],
   "source": [
    "uhi_region = region_pos.loc[\n",
    "    region_pos['Haploinsufficiency Score'] == '40',\n",
    "    ['chrom', 'start', 'end', '#ISCA ID', 'ISCA Region Name']\n",
    "].sort_values(['chrom', 'start', 'end'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>#ISCA ID</th>\n",
       "      <th>ISCA Region Name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>chr1</td>\n",
       "      <td>148867551</td>\n",
       "      <td>149768855</td>\n",
       "      <td>ISCA-37469</td>\n",
       "      <td>1q21.2 region (polymorphic region)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>chr14</td>\n",
       "      <td>22111109</td>\n",
       "      <td>23021097</td>\n",
       "      <td>ISCA-37476</td>\n",
       "      <td>14q11.2 region (TCRA region)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>chr14</td>\n",
       "      <td>106050000</td>\n",
       "      <td>107289540</td>\n",
       "      <td>ISCA-37477</td>\n",
       "      <td>1 copy: 14q telomere; 3 copies: 14q telomere</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>chr16</td>\n",
       "      <td>34202088</td>\n",
       "      <td>35147508</td>\n",
       "      <td>ISCA-37481</td>\n",
       "      <td>3 copies: 16p centromere</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>chr19</td>\n",
       "      <td>43242796</td>\n",
       "      <td>43741310</td>\n",
       "      <td>ISCA-37483</td>\n",
       "      <td>19q13.3 region (PSG gene cluster)</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    chrom      start        end    #ISCA ID  \\\n",
       "29   chr1  148867551  149768855  ISCA-37469   \n",
       "25  chr14   22111109   23021097  ISCA-37476   \n",
       "24  chr14  106050000  107289540  ISCA-37477   \n",
       "21  chr16   34202088   35147508  ISCA-37481   \n",
       "20  chr19   43242796   43741310  ISCA-37483   \n",
       "\n",
       "                                ISCA Region Name  \n",
       "29            1q21.2 region (polymorphic region)  \n",
       "25                  14q11.2 region (TCRA region)  \n",
       "24  1 copy: 14q telomere; 3 copies: 14q telomere  \n",
       "21                      3 copies: 16p centromere  \n",
       "20             19q13.3 region (PSG gene cluster)  "
      ]
     },
     "execution_count": 76,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "uhi_region.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "metadata": {},
   "outputs": [],
   "source": [
    "uhi_region['genes'] = uhi_region.apply(\n",
    "    lambda row: ','.join(gene.loc[\n",
    "        (gene['chrom'] == row['chrom']) & (gene['txStart'] >= row['start']) & (gene['txEnd'] <= row['end']), 'name2'\n",
    "    ]),\n",
    "    axis=1\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>#ISCA ID</th>\n",
       "      <th>ISCA Region Name</th>\n",
       "      <th>genes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>chr1</td>\n",
       "      <td>148867551</td>\n",
       "      <td>149768855</td>\n",
       "      <td>ISCA-37469</td>\n",
       "      <td>1q21.2 region (polymorphic region)</td>\n",
       "      <td>FAM72C,PPIAL4C,FCGR1A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>chr14</td>\n",
       "      <td>22111109</td>\n",
       "      <td>23021097</td>\n",
       "      <td>ISCA-37476</td>\n",
       "      <td>14q11.2 region (TCRA region)</td>\n",
       "      <td>OR4E2,OR4E1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>chr14</td>\n",
       "      <td>106050000</td>\n",
       "      <td>107289540</td>\n",
       "      <td>ISCA-37477</td>\n",
       "      <td>1 copy: 14q telomere; 3 copies: 14q telomere</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>chr16</td>\n",
       "      <td>34202088</td>\n",
       "      <td>35147508</td>\n",
       "      <td>ISCA-37481</td>\n",
       "      <td>3 copies: 16p centromere</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>chr19</td>\n",
       "      <td>43242796</td>\n",
       "      <td>43741310</td>\n",
       "      <td>ISCA-37483</td>\n",
       "      <td>19q13.3 region (PSG gene cluster)</td>\n",
       "      <td>PSG8,PSG1,PSG6,PSG7,PSG11,PSG2,PSG5,PSG4</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    chrom      start        end    #ISCA ID  \\\n",
       "29   chr1  148867551  149768855  ISCA-37469   \n",
       "25  chr14   22111109   23021097  ISCA-37476   \n",
       "24  chr14  106050000  107289540  ISCA-37477   \n",
       "21  chr16   34202088   35147508  ISCA-37481   \n",
       "20  chr19   43242796   43741310  ISCA-37483   \n",
       "\n",
       "                                ISCA Region Name  \\\n",
       "29            1q21.2 region (polymorphic region)   \n",
       "25                  14q11.2 region (TCRA region)   \n",
       "24  1 copy: 14q telomere; 3 copies: 14q telomere   \n",
       "21                      3 copies: 16p centromere   \n",
       "20             19q13.3 region (PSG gene cluster)   \n",
       "\n",
       "                                       genes  \n",
       "29                     FAM72C,PPIAL4C,FCGR1A  \n",
       "25                               OR4E2,OR4E1  \n",
       "24                                            \n",
       "21                                            \n",
       "20  PSG8,PSG1,PSG6,PSG7,PSG11,PSG2,PSG5,PSG4  "
      ]
     },
     "execution_count": 78,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "uhi_region.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "metadata": {},
   "outputs": [],
   "source": [
    "uhi_region.rename(columns={\n",
    "    'chrom': '#chrom',\n",
    "    '#ISCA ID': 'isca_id', 'ISCA Region Name': 'name'\n",
    "}).to_csv('data/uhi-region-20200109.sorted.bed', sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "metadata": {},
   "outputs": [],
   "source": [
    "!bgzip -cf data/uhi-region-20200109.sorted.bed > ../data/uhi-region-20200109.sorted.bed.gz"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "metadata": {},
   "outputs": [],
   "source": [
    "!tabix -fp bed ../data/uhi-region-20200109.sorted.bed.gz"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open('data/gene-table.json') as f:\n",
    "    data = json.load(f)\n",
    "    decipher = pd.DataFrame.from_records(data, columns=['name', 'hi_score', 'p_li']).rename(columns={'name': 'symbol'})\n",
    "    decipher = decipher.merge(gene, left_on='symbol', right_on='name2')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "metadata": {},
   "outputs": [],
   "source": [
    "decipher = decipher.loc[\n",
    "    (decipher['p_li'] >= 0.9) & (decipher['hi_score'] < 0.1),\n",
    "    ['chrom', 'txStart', 'txEnd', 'GeneID', 'name2', 'name', 'p_li', 'hi_score']\n",
    "].sort_values(['chrom', 'txStart', 'txEnd'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>txStart</th>\n",
       "      <th>txEnd</th>\n",
       "      <th>GeneID</th>\n",
       "      <th>name2</th>\n",
       "      <th>name</th>\n",
       "      <th>p_li</th>\n",
       "      <th>hi_score</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2112</th>\n",
       "      <td>chr1</td>\n",
       "      <td>2439969</td>\n",
       "      <td>2458067</td>\n",
       "      <td>55229</td>\n",
       "      <td>PANK4</td>\n",
       "      <td>NM_018216.2</td>\n",
       "      <td>0.945333</td>\n",
       "      <td>0.098966</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2587</th>\n",
       "      <td>chr1</td>\n",
       "      <td>155290639</td>\n",
       "      <td>155300909</td>\n",
       "      <td>23623</td>\n",
       "      <td>RUSC1</td>\n",
       "      <td>NM_001105203.1</td>\n",
       "      <td>0.949429</td>\n",
       "      <td>0.035285</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2705</th>\n",
       "      <td>chr1</td>\n",
       "      <td>232533710</td>\n",
       "      <td>232651354</td>\n",
       "      <td>57568</td>\n",
       "      <td>SIPA1L2</td>\n",
       "      <td>NM_020808.4</td>\n",
       "      <td>0.999220</td>\n",
       "      <td>0.079950</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1608</th>\n",
       "      <td>chr10</td>\n",
       "      <td>852853</td>\n",
       "      <td>977645</td>\n",
       "      <td>23185</td>\n",
       "      <td>LARP4B</td>\n",
       "      <td>NM_015155.2</td>\n",
       "      <td>0.999886</td>\n",
       "      <td>0.071477</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>199</th>\n",
       "      <td>chr10</td>\n",
       "      <td>24872537</td>\n",
       "      <td>25012597</td>\n",
       "      <td>57584</td>\n",
       "      <td>ARHGAP21</td>\n",
       "      <td>NM_020824.3</td>\n",
       "      <td>0.999947</td>\n",
       "      <td>0.073197</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      chrom    txStart      txEnd  GeneID     name2            name      p_li  \\\n",
       "2112   chr1    2439969    2458067   55229     PANK4     NM_018216.2  0.945333   \n",
       "2587   chr1  155290639  155300909   23623     RUSC1  NM_001105203.1  0.949429   \n",
       "2705   chr1  232533710  232651354   57568   SIPA1L2     NM_020808.4  0.999220   \n",
       "1608  chr10     852853     977645   23185    LARP4B     NM_015155.2  0.999886   \n",
       "199   chr10   24872537   25012597   57584  ARHGAP21     NM_020824.3  0.999947   \n",
       "\n",
       "      hi_score  \n",
       "2112  0.098966  \n",
       "2587  0.035285  \n",
       "2705  0.079950  \n",
       "1608  0.071477  \n",
       "199   0.073197  "
      ]
     },
     "execution_count": 84,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "decipher.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "metadata": {},
   "outputs": [],
   "source": [
    "decipher.rename(columns={\n",
    "    'chrom': '#chrom', 'txStart': 'start', 'txEnd': 'end',\n",
    "    'GeneID': 'gene_id', 'name2': 'symbol', 'name': 'transcript'\n",
    "}).to_csv('data/decipher-gene-20200109.sorted.bed', sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "metadata": {},
   "outputs": [],
   "source": [
    "!bgzip -cf data/decipher-gene-20200109.sorted.bed > ../data/decipher-gene-20200109.sorted.bed.gz"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 87,
   "metadata": {},
   "outputs": [],
   "source": [
    "!tabix -fp bed ../data/decipher-gene-20200109.sorted.bed.gz"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 88,
   "metadata": {},
   "outputs": [],
   "source": [
    "ts_genes = set(\n",
    "    curation_gene.loc[\n",
    "        curation_gene['Triplosensitivity Score'] == '3', '#Gene Symbol'\n",
    "    ]\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 89,
   "metadata": {},
   "outputs": [],
   "source": [
    "ts_gene = gene[gene['name2'].isin(ts_genes)].copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 90,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>txStart</th>\n",
       "      <th>txEnd</th>\n",
       "      <th>GeneID</th>\n",
       "      <th>name2</th>\n",
       "      <th>name</th>\n",
       "      <th>strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>20275</th>\n",
       "      <td>chr5</td>\n",
       "      <td>126112314</td>\n",
       "      <td>126172712</td>\n",
       "      <td>4001</td>\n",
       "      <td>LMNB1</td>\n",
       "      <td>NM_005573.3</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12737</th>\n",
       "      <td>chrX</td>\n",
       "      <td>103031433</td>\n",
       "      <td>103047548</td>\n",
       "      <td>5354</td>\n",
       "      <td>PLP1</td>\n",
       "      <td>NM_001128834.2</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21287</th>\n",
       "      <td>chrX</td>\n",
       "      <td>153287263</td>\n",
       "      <td>153363188</td>\n",
       "      <td>4204</td>\n",
       "      <td>MECP2</td>\n",
       "      <td>NM_004992.3</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      chrom    txStart      txEnd  GeneID  name2            name strand\n",
       "20275  chr5  126112314  126172712    4001  LMNB1     NM_005573.3      +\n",
       "12737  chrX  103031433  103047548    5354   PLP1  NM_001128834.2      +\n",
       "21287  chrX  153287263  153363188    4204  MECP2     NM_004992.3      -"
      ]
     },
     "execution_count": 90,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ts_gene.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 91,
   "metadata": {},
   "outputs": [],
   "source": [
    "ts_gene.rename(columns={\n",
    "    'chrom': '#chrom', 'txStart': 'start', 'txEnd': 'end',\n",
    "    'GeneID': 'gene_id', 'name2': 'symbol', 'name': 'transcript'\n",
    "}).to_csv('data/ts-gene-20200109.sorted.bed', index=False, sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 92,
   "metadata": {},
   "outputs": [],
   "source": [
    "!bgzip -cf data/ts-gene-20200109.sorted.bed > ../data/ts-gene-20200109.sorted.bed.gz"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 93,
   "metadata": {},
   "outputs": [],
   "source": [
    "!tabix -fp bed ../data/ts-gene-20200109.sorted.bed.gz"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 94,
   "metadata": {},
   "outputs": [],
   "source": [
    "uts_genes = set(\n",
    "    curation_gene.loc[curation_gene['Triplosensitivity Score'] == '40', '#Gene Symbol']\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 95,
   "metadata": {},
   "outputs": [],
   "source": [
    "uts_gene = gene[gene['name2'].isin(uts_genes)].copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 96,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>txStart</th>\n",
       "      <th>txEnd</th>\n",
       "      <th>GeneID</th>\n",
       "      <th>name2</th>\n",
       "      <th>name</th>\n",
       "      <th>strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>3727</th>\n",
       "      <td>chrX</td>\n",
       "      <td>6451658</td>\n",
       "      <td>6453159</td>\n",
       "      <td>51481</td>\n",
       "      <td>VCX3A</td>\n",
       "      <td>NM_016379.3</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     chrom  txStart    txEnd  GeneID  name2         name strand\n",
       "3727  chrX  6451658  6453159   51481  VCX3A  NM_016379.3      -"
      ]
     },
     "execution_count": 96,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "uts_gene.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 97,
   "metadata": {},
   "outputs": [],
   "source": [
    "uts_gene.rename(columns={\n",
    "    'chrom': '#chrom', 'txStart': 'start', 'txEnd': 'end',\n",
    "    'GeneID': 'gene_id', 'name2': 'symbol', 'name': 'transcript'\n",
    "}).to_csv('data/uts-gene-20200109.sorted.bed', index=False, sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 98,
   "metadata": {},
   "outputs": [],
   "source": [
    "!bgzip -cf data/uts-gene-20200109.sorted.bed > ../data/uts-gene-20200109.sorted.bed.gz"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 99,
   "metadata": {},
   "outputs": [],
   "source": [
    "!tabix -fp bed ../data/uts-gene-20200109.sorted.bed.gz"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 100,
   "metadata": {},
   "outputs": [],
   "source": [
    "ts_region = region_pos.loc[\n",
    "    region_pos['Triplosensitivity Score'] == '3',\n",
    "    ['chrom', 'start', 'end', '#ISCA ID', 'ISCA Region Name']\n",
    "].sort_values(['chrom', 'start', 'end'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 101,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>#ISCA ID</th>\n",
       "      <th>ISCA Region Name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>52</th>\n",
       "      <td>chr1</td>\n",
       "      <td>146577486</td>\n",
       "      <td>147394506</td>\n",
       "      <td>ISCA-37421</td>\n",
       "      <td>1q21.1 recurrent region (BP3-BP4, distal) (inc...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>62</th>\n",
       "      <td>chr15</td>\n",
       "      <td>22832519</td>\n",
       "      <td>28379874</td>\n",
       "      <td>ISCA-37404</td>\n",
       "      <td>15q11q13 recurrent (PWS/AS) region (BP1-BP3, C...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>chr15</td>\n",
       "      <td>23747996</td>\n",
       "      <td>28379874</td>\n",
       "      <td>ISCA-37478</td>\n",
       "      <td>15q11q13 recurrent (PWS/AS) region (BP2-BP3, C...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>64</th>\n",
       "      <td>chr16</td>\n",
       "      <td>29649997</td>\n",
       "      <td>30199852</td>\n",
       "      <td>ISCA-37400</td>\n",
       "      <td>16p11.2 recurrent region (proximal, BP4-BP5) (...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>47</th>\n",
       "      <td>chr17</td>\n",
       "      <td>1247833</td>\n",
       "      <td>2588909</td>\n",
       "      <td>ISCA-37430</td>\n",
       "      <td>17p13.3 (Miller-Dieker syndrome) region (inclu...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    chrom      start        end    #ISCA ID  \\\n",
       "52   chr1  146577486  147394506  ISCA-37421   \n",
       "62  chr15   22832519   28379874  ISCA-37404   \n",
       "23  chr15   23747996   28379874  ISCA-37478   \n",
       "64  chr16   29649997   30199852  ISCA-37400   \n",
       "47  chr17    1247833    2588909  ISCA-37430   \n",
       "\n",
       "                                     ISCA Region Name  \n",
       "52  1q21.1 recurrent region (BP3-BP4, distal) (inc...  \n",
       "62  15q11q13 recurrent (PWS/AS) region (BP1-BP3, C...  \n",
       "23  15q11q13 recurrent (PWS/AS) region (BP2-BP3, C...  \n",
       "64  16p11.2 recurrent region (proximal, BP4-BP5) (...  \n",
       "47  17p13.3 (Miller-Dieker syndrome) region (inclu...  "
      ]
     },
     "execution_count": 101,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ts_region.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 102,
   "metadata": {},
   "outputs": [],
   "source": [
    "ts_region.rename(columns={\n",
    "    'chrom': '#chrom',\n",
    "    '#ISCA ID': 'isca_id', 'ISCA Region Name': 'name'\n",
    "}).to_csv('data/ts-region-20200109.sorted.bed', sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 103,
   "metadata": {},
   "outputs": [],
   "source": [
    "!bgzip -cf data/ts-region-20200109.sorted.bed > ../data/ts-region-20200109.sorted.bed.gz"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 104,
   "metadata": {},
   "outputs": [],
   "source": [
    "!tabix -fp bed ../data/ts-region-20200109.sorted.bed.gz"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 105,
   "metadata": {},
   "outputs": [],
   "source": [
    "uts_region = region_pos.loc[\n",
    "    region_pos['Triplosensitivity Score'] == '40',\n",
    "    ['chrom', 'start', 'end', '#ISCA ID', 'ISCA Region Name']\n",
    "].sort_values(['chrom', 'start', 'end'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 106,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>#ISCA ID</th>\n",
       "      <th>ISCA Region Name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>chr1</td>\n",
       "      <td>148867551</td>\n",
       "      <td>149768855</td>\n",
       "      <td>ISCA-37469</td>\n",
       "      <td>1q21.2 region (polymorphic region)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>chr14</td>\n",
       "      <td>22111109</td>\n",
       "      <td>23021097</td>\n",
       "      <td>ISCA-37476</td>\n",
       "      <td>14q11.2 region (TCRA region)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>chr14</td>\n",
       "      <td>106050000</td>\n",
       "      <td>107289540</td>\n",
       "      <td>ISCA-37477</td>\n",
       "      <td>1 copy: 14q telomere; 3 copies: 14q telomere</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>chr15</td>\n",
       "      <td>22832519</td>\n",
       "      <td>23090897</td>\n",
       "      <td>ISCA-37448</td>\n",
       "      <td>15q11.2 recurrent region (BP1-BP2) (includes N...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>chr15</td>\n",
       "      <td>32019621</td>\n",
       "      <td>32445405</td>\n",
       "      <td>ISCA-46295</td>\n",
       "      <td>15q13.3 recurrent region (D-CHRNA7 to BP5) (in...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    chrom      start        end    #ISCA ID  \\\n",
       "29   chr1  148867551  149768855  ISCA-37469   \n",
       "25  chr14   22111109   23021097  ISCA-37476   \n",
       "24  chr14  106050000  107289540  ISCA-37477   \n",
       "33  chr15   22832519   23090897  ISCA-37448   \n",
       "5   chr15   32019621   32445405  ISCA-46295   \n",
       "\n",
       "                                     ISCA Region Name  \n",
       "29                 1q21.2 region (polymorphic region)  \n",
       "25                       14q11.2 region (TCRA region)  \n",
       "24       1 copy: 14q telomere; 3 copies: 14q telomere  \n",
       "33  15q11.2 recurrent region (BP1-BP2) (includes N...  \n",
       "5   15q13.3 recurrent region (D-CHRNA7 to BP5) (in...  "
      ]
     },
     "execution_count": 106,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "uts_region.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 107,
   "metadata": {},
   "outputs": [],
   "source": [
    "uts_region['genes'] = uts_region.apply(\n",
    "    lambda row: ','.join(gene.loc[\n",
    "        (gene['chrom'] == row['chrom']) & (gene['txStart'] >= row['start']) & (gene['txEnd'] <= row['end']), 'name2'\n",
    "    ]),\n",
    "    axis=1\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 108,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>#ISCA ID</th>\n",
       "      <th>ISCA Region Name</th>\n",
       "      <th>genes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>chr1</td>\n",
       "      <td>148867551</td>\n",
       "      <td>149768855</td>\n",
       "      <td>ISCA-37469</td>\n",
       "      <td>1q21.2 region (polymorphic region)</td>\n",
       "      <td>FAM72C,PPIAL4C,FCGR1A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>chr14</td>\n",
       "      <td>22111109</td>\n",
       "      <td>23021097</td>\n",
       "      <td>ISCA-37476</td>\n",
       "      <td>14q11.2 region (TCRA region)</td>\n",
       "      <td>OR4E2,OR4E1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>chr14</td>\n",
       "      <td>106050000</td>\n",
       "      <td>107289540</td>\n",
       "      <td>ISCA-37477</td>\n",
       "      <td>1 copy: 14q telomere; 3 copies: 14q telomere</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>chr15</td>\n",
       "      <td>22832519</td>\n",
       "      <td>23090897</td>\n",
       "      <td>ISCA-37448</td>\n",
       "      <td>15q11.2 recurrent region (BP1-BP2) (includes N...</td>\n",
       "      <td>TUBGCP5,CYFIP1,NIPA2,NIPA1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>chr15</td>\n",
       "      <td>32019621</td>\n",
       "      <td>32445405</td>\n",
       "      <td>ISCA-46295</td>\n",
       "      <td>15q13.3 recurrent region (D-CHRNA7 to BP5) (in...</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    chrom      start        end    #ISCA ID  \\\n",
       "29   chr1  148867551  149768855  ISCA-37469   \n",
       "25  chr14   22111109   23021097  ISCA-37476   \n",
       "24  chr14  106050000  107289540  ISCA-37477   \n",
       "33  chr15   22832519   23090897  ISCA-37448   \n",
       "5   chr15   32019621   32445405  ISCA-46295   \n",
       "\n",
       "                                     ISCA Region Name  \\\n",
       "29                 1q21.2 region (polymorphic region)   \n",
       "25                       14q11.2 region (TCRA region)   \n",
       "24       1 copy: 14q telomere; 3 copies: 14q telomere   \n",
       "33  15q11.2 recurrent region (BP1-BP2) (includes N...   \n",
       "5   15q13.3 recurrent region (D-CHRNA7 to BP5) (in...   \n",
       "\n",
       "                         genes  \n",
       "29       FAM72C,PPIAL4C,FCGR1A  \n",
       "25                 OR4E2,OR4E1  \n",
       "24                              \n",
       "33  TUBGCP5,CYFIP1,NIPA2,NIPA1  \n",
       "5                               "
      ]
     },
     "execution_count": 108,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "uts_region.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 109,
   "metadata": {},
   "outputs": [],
   "source": [
    "uts_region.rename(columns={\n",
    "    'chrom': '#chrom',\n",
    "    '#ISCA ID': 'isca_id', 'ISCA Region Name': 'name'\n",
    "}).to_csv('data/uts-region-20200109.sorted.bed', sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 110,
   "metadata": {},
   "outputs": [],
   "source": [
    "!bgzip -cf data/uts-region-20200109.sorted.bed > ../data/uts-region-20200109.sorted.bed.gz"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 111,
   "metadata": {},
   "outputs": [],
   "source": [
    "!tabix -fp bed ../data/uts-region-20200109.sorted.bed.gz"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "jupyter",
   "language": "python",
   "name": "jupyter"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
